{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Complete functions\n",
    "def revComp(inputSeq):\n",
    "  \"\"\"\n",
    "  This function takes an input sequence and returns the reverse complement.\n",
    "\n",
    "  Input: inputSeq in str format\n",
    "  Output: revComp in str format\n",
    "\n",
    "  \"\"\"\n",
    "  complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "  \n",
    "  revComp = \"\"\n",
    "  for base in inputSeq[::-1]:\n",
    "    revComp += complement[(base.upper())]\n",
    "\n",
    "  return revComp\n",
    "\n",
    "def make_homology_arm_fragments(tfsDF):\n",
    "    \"\"\"\n",
    "    Takes in the dataframe of information about start/stop codon regions, \n",
    "    and appends with columns for the 225 bp upstream and downstream of\n",
    "    this site (the homlogy arm fragments).\n",
    "\n",
    "    input: tfsDF - as defined in previous function and on GitHub README, with Reference_Seq 1600bp either side of the start/stop.\n",
    "    output: tfsDF appended with upstreamHA and downstreamHA\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    tfsDF[\"upstreamHA\"] = tfsDF.Reference_Seq.str[1375:1600]\n",
    "    tfsDF[\"downstreamHA\"] = tfsDF.Reference_Seq.str[1604:1829]\n",
    "\n",
    "    return tfsDF\n",
    "\n",
    "def refSeq():\n",
    "    \"\"\"\n",
    "    Creates a Bio SeqIO element from the Drosophila melanogaster reference genome.\n",
    "\n",
    "    returns: refSeqPerChromosome - a dictionary of seq ID: sequence (in SeqIO fasta format) for the full reference genome.\n",
    "    \"\"\"\n",
    "    from Bio import SeqIO\n",
    "\n",
    "    refSeqPerChromosome = {}\n",
    "    for seq in SeqIO.parse(open(\"inputfiles/dmel-all-chromosome-r6.48.fasta\"), 'fasta'): #This is the FASTA file of the reference genome sequence\n",
    "        refSeqPerChromosome[seq.id] = seq.seq \n",
    "\n",
    "    return refSeqPerChromosome\n",
    "\n",
    "def make_dataframe_from_TFs_list(TF_list, refSeqPerChromosome, annotation = \"inputfiles/dmel-all-r6.48.gtf\"):\n",
    "    '''\n",
    "    Creating a dataframe from sequence information and genes of interest. Depends on functions revComp() and make_homology_arm_fragments(). \n",
    "\n",
    "    params:\n",
    "        TF_list: xlsx file listing the genes of interest\n",
    "        refSeqPerChromosome: reference chromosome stored in SeqIO sequence format, indexed by refSeqPerChromosome[seq.id] = seq.seq\n",
    "        annotation: gtf file containing the gene annotation information for the reference genome\n",
    "\n",
    "    returns: dataframe of information about each start/stop site, in format:\n",
    "        'Gene_ID': 'FBgn0004652',\n",
    "        'Transcript_ID': 'FBtr0083651', \n",
    "        'Chromosome': '3R', \n",
    "        'Gene_Region': 'stop_codon', \n",
    "        'Start': 18426145, \n",
    "        'Stop': 18426147, \n",
    "        'Strand': '-', \n",
    "        'Reference_Seq': 'TTGATCGTAGGACAC', \n",
    "        'upstreamHA': 'ATGCCTG', \n",
    "        'downstreamHA': 'CTGGATC'\n",
    "\n",
    "    '''\n",
    "    import pandas as pd\n",
    "    \n",
    "    #This is the input file containing the TFs we want to query\n",
    "    queryTFsdf = pd.read_excel(TF_list)\n",
    "\n",
    "    #This is the .gtf file with annotations for each gene on the reference genome\n",
    "    refAnnotationsHeaders = [\"Chromosome\", \"Source\", \"Gene_Region\", \"Start\", \"Stop\", \"Score\", \"Strand\", \"Frame\", \"Attribute\"]\n",
    "    refGenomeAnnotation = pd.read_csv(annotation, sep = \"\\t\", header = None, index_col = False, names = refAnnotationsHeaders)\n",
    "    \n",
    "    #This is to reformat the \"Attribute\" category in refGenomeAnnotation, to extract Gene_ID, Gene_Symbol, and Transcript ID\n",
    "    index = 0\n",
    "\n",
    "    #Add new categories to the dataframe\n",
    "    refGenomeAnnotation = refGenomeAnnotation.assign(Gene_ID = \"\", Gene_Symbol = \"\", Transcript_ID = \"\")\n",
    "\n",
    "    #For each attribute value, extract the gene ID and symbol and add this to the new categories\n",
    "    for attribute in refGenomeAnnotation['Attribute']:\n",
    "        fullatt = (refGenomeAnnotation.loc[index][\"Attribute\"]).replace(\";\", \"\")\n",
    "        fullatt = fullatt.replace('\"', \"\")\n",
    "        fullattsplit = fullatt.split(\" \")\n",
    "        refGenomeAnnotation.at[index,\"Gene_ID\"] = fullattsplit[1]\n",
    "        refGenomeAnnotation.at[index,\"Gene_Symbol\"] = fullattsplit[3]\n",
    "        if len(fullattsplit) > 4:\n",
    "            refGenomeAnnotation.at[index,\"Transcript_ID\"] = fullattsplit[5]\n",
    "            refGenomeAnnotation.at[index,\"Transcript_Symbol\"] = fullattsplit[7]\n",
    "        index+=1\n",
    "\n",
    "    #Delete Attributes category\n",
    "    del refGenomeAnnotation[\"Attribute\"]\n",
    "\n",
    "    #Select only rows that TFs are in, and keep only the start and stop codon gene regions\n",
    "\n",
    "    refGenomeAnnotation = refGenomeAnnotation.loc[refGenomeAnnotation[\"Gene_Region\"].isin([\"start_codon\", \"stop_codon\"])]\n",
    "\n",
    "    TFsdf = refGenomeAnnotation[[\"Gene_ID\", \"Transcript_ID\", \"Chromosome\", \"Gene_Region\", \"Start\", \"Stop\", \"Strand\"]].loc[refGenomeAnnotation[\"Gene_ID\"].isin(queryTFsdf[\"Flybase_ID\"])]\n",
    "\n",
    "    #Add reference genome sequence per gene region\n",
    "    #This will correspond to 1.6kb upstream and downstream of ATG/stop codon \n",
    "    TFsdf = TFsdf.assign(Reference_Seq = \"\")\n",
    "\n",
    "    for index, rowcontents in TFsdf.iterrows():\n",
    "            #Define 3.2kb gene region\n",
    "            regionStart = rowcontents[\"Start\"] - 1601\n",
    "            regionStop = rowcontents[\"Stop\"] + 1600\n",
    "\n",
    "            #Add reference sequence\n",
    "            refPosStrandSeq = str(refSeqPerChromosome[rowcontents[\"Chromosome\"]][regionStart:regionStop]) #This is the + strand seq, so goes from end to beginning\n",
    "            TFsdf.at[index,\"Reference_Seq\"] = revComp(refPosStrandSeq)\n",
    "\n",
    "    #Create fragments for the HDR-arm\n",
    "    TFsdf = make_homology_arm_fragments(TFsdf)\n",
    "\n",
    "    #re-index\n",
    "    TFsdf = TFsdf.reset_index()\n",
    "    del TFsdf[\"index\"]\n",
    "\n",
    "    return TFsdf\n",
    "\n",
    "def find_synonymous_codons(query_codon, base_to_change, codon_table_excel = \"inputfiles/codon_table.xlsx\"):\n",
    "\n",
    "    '''\n",
    "    Uses the amino acids table to select codons that encode for the same amino acid as the query codon. Will ensure the specified base has been mutated.\n",
    "\n",
    "    Params:\n",
    "        query_codon: string, codon to select synonymous codons for\n",
    "        base_to_change: the base within the codon that needs to change. Numeric value from 1-3.\n",
    "        codon_table_excel: string, path to an excel file that lists per codon which amino acid that codon encodes.\n",
    "    \n",
    "    Returns:\n",
    "        synonymous_codons: list of strings, each string is a codon that encodes for the same amino acid as the query codon.\n",
    "    \n",
    "    '''\n",
    "\n",
    "    import pandas as pd\n",
    "\n",
    "    codon_table = pd.read_excel(codon_table_excel)\n",
    "\n",
    "    #Extract amino acid given codon\n",
    "    amino_acid_query = codon_table[codon_table['codon'] == query_codon].iloc[0][\"amino_acid\"]\n",
    "\n",
    "    #Subset df for other rows corresponding to this amino acid\n",
    "    same_aa_df=codon_table[codon_table[\"amino_acid\"] == amino_acid_query]\n",
    "\n",
    "    #Convert the codons to a list\n",
    "    codon_list = same_aa_df[\"codon\"].values.tolist()\n",
    "\n",
    "    #Keep only codons where the indicated base has changed (this will also remove original codon)\n",
    "    changedCodons = []\n",
    "    for codon in codon_list:\n",
    "        if codon[base_to_change-1] != query_codon[base_to_change-1]:\n",
    "            changedCodons.append(codon)\n",
    "            \n",
    "    return changedCodons\n",
    "\n",
    "def filter_gRNA(gRNA_file, tfSingleRow, refSeqPerChromosome):\n",
    "    \"\"\"\n",
    "    Selects gRNAs within 20pb of the start/stop site and returns these in a dataframe with information about their coordinates and the start/stop site.\n",
    "    params:\n",
    "        gRNA_file: gff file\n",
    "        tfSingleRow: a pandas series for one row of the original tfsDF with the following format\n",
    "            Gene_ID                                                    FBgn0000022\n",
    "            Transcript_ID                                              FBtr0070072\n",
    "            Chromosome                                                           X\n",
    "            Gene_Region                                                start_codon\n",
    "            Start                                                           370094\n",
    "            Stop                                                            370096\n",
    "            Strand                                                               +\n",
    "            Reference_Seq        CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...\n",
    "            upstreamHA           TTTGCTCAGTTTTTTATTGGCGCCGGGACCAATTCCCCGGCGACCA...\n",
    "            downstreamHA         TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...\n",
    "        \n",
    "    returns: filtered_gRNAs - a dataframe of sgRNAs for one start/stop site of the format:\n",
    "            fmin\tfmax\t#chr\tstrand\tsgRNA_sequence\t        Gene_ID\t    Transcript_ID\tChromosome\tGene_Region\tStart\t...\tdownstreamHA\n",
    "        0\t370082\t370104\tX\t    +\t    CTATCTCTTAAAATGGCTTTGGG\tFBgn0000022\tFBtr0070072\t    X\t        start_codon\t370094\t...\tTTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...\n",
    "        1\t370693\t370715\tX\t    +\t    CCTGTAAAAAAACAGATCAAATC\tFBgn0000022\tFBtr0070072\t    X\t        start_codon\t370694\t...\tTTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...\n",
    "\n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "\n",
    "    # create a dataframe from the gRNA file\n",
    "    gRNAFileAnnotation = pd.read_csv(gRNA_file, sep = \"\\t\", index_col = False)\n",
    "\n",
    "    # add a new category to the dataframe that provides information about whether the sequence deviates from the transgenic strain \n",
    "    gRNAFileAnnotation = gRNAFileAnnotation.assign(target_site_variation= \"\")\n",
    "\n",
    "    # reformat the \"Attribute\" category in refGenomeAnnotation, to extract Gene_ID, Gene_Symbol, and Transcript ID\n",
    "    index = 0 #TODO@Marina improve this code\n",
    "\n",
    "    # for each attribute value, extract the gene ID and symbol and add this to the new categories\n",
    "    for attribute in gRNAFileAnnotation['attributes']:\n",
    "\n",
    "        fullatt = (attribute).split(\";\")\n",
    "        gRNAFileAnnotation.at[index,\"target_site_variation\"] = fullatt[8]\n",
    "        index+=1\n",
    "    \n",
    "    # shorten file to essential information\n",
    "    GenomeCoordinates = gRNAFileAnnotation[[\"target_site_variation\", \"fmin\", \"fmax\", \"#chr\", \"strand\"]]\n",
    "\n",
    "    #Filter for the appropriate conditions:\n",
    "        # check whether the sgRNAs match the transgeneic genome, whether the sgRNAs match to the same chromosome as the transcription factor\n",
    "        # select sgRNA that are located maximally 20 bp upstream of the start/stop codon of the transcription factors\n",
    "        # select sgRNA that are located maximally 20 bp downstream of the start/stop codon of the transcription factors\n",
    "    filtered_gRNAs = GenomeCoordinates[(GenomeCoordinates['target_site_variation'] == \"target_site_variation=none observed\") & \n",
    "                                                  (GenomeCoordinates['#chr'] == tfSingleRow[\"Chromosome\"]) & \n",
    "                                                  (GenomeCoordinates['fmin']-1 >= tfSingleRow[\"Start\"] - 20) & \n",
    "                                                  (GenomeCoordinates['fmax'] <= tfSingleRow[\"Stop\"] + 20)]\n",
    "\n",
    "    #add sgRNA sequence using coordinates\n",
    "    filtered_gRNAs = filtered_gRNAs.assign(sgRNA_sequence= \"\")\n",
    "    for index, row in filtered_gRNAs.iterrows():\n",
    "        filtered_gRNAs.loc[index,'sgRNA_sequence'] = str(refSeqPerChromosome[row['#chr']][int(row[\"fmin\"])-1:int(row[\"fmax\"])])\n",
    "    \n",
    "    #Add columns from tfSingleRow containing start/stop site info\n",
    "    for (columnName, columnData) in tfSingleRow.items():\n",
    "        filtered_gRNAs = filtered_gRNAs.assign(**{columnName: columnData})\n",
    "\n",
    "    #drop target site variation\n",
    "    filtered_gRNAs = filtered_gRNAs.drop(columns=[\"target_site_variation\"])\n",
    "\n",
    "    #reset index\n",
    "    filtered_gRNAs = filtered_gRNAs.reset_index(drop = True)\n",
    "\n",
    "    return filtered_gRNAs\n",
    "\n",
    "def gRNA_stringencyIterator(tfSingleRow, refSeqPerChromosome):\n",
    "    \"\"\"\n",
    "    Iterates trough gRNA stringency files until at least one gRNA is returned from filter_gRNAs function.\n",
    "\n",
    "    params:\n",
    "        tfSingleRow: a pandas Series for one row of the tfsDF dataframe produced by make_dataframe_from_TFs_list\n",
    "        refSeqPerChromosome: a Bio SeqIO dictionary for the D. melanogaster genome\n",
    "\n",
    "    returns:\n",
    "        filtered_gRNAs: a df of gRNAs that meet the conditions at this site\n",
    "    \"\"\"\n",
    "\n",
    "    gRNAfiles = [\"inputfiles/Hu.2019.8.28.sgRNA_designs/NoOffTarget_high_stringency.gff\", \"inputfiles/Hu.2019.8.28.sgRNA_designs/NoOffTarget_med_stringency.gff\", \"inputfiles/Hu.2019.8.28.sgRNA_designs/NoOffTarget_low_stringency.gff\", \"inputfiles/Hu.2019.8.28.sgRNA_designs/1to3NonCdsOffTarget_low_stringency.gff\", \"inputfiles/Hu.2019.8.28.sgRNA_designs/ManyOffTarget_low_stringency.gff\"]\n",
    "\n",
    "    #Set up sgRNA variable\n",
    "    sgRNA = \"\"\n",
    "    file_ind = 0\n",
    "\n",
    "    #Loop through files and retain the filtered sgRNA df if at least one row is present\n",
    "    while len(sgRNA) == 0:\n",
    "        sgRNA = filter_gRNA(gRNAfiles[file_ind], tfSingleRow, refSeqPerChromosome)\n",
    "        print(f\"stringency {file_ind}\")\n",
    "        file_ind +=1\n",
    "        if file_ind == 4: #If we've tried all files and sgRNA is still 0, break out of loop\n",
    "            break\n",
    "    \n",
    "    return sgRNA\n",
    "\n",
    "def sgRNApositionCheck(sgRNAdf):\n",
    "    \"\"\"\n",
    "    Given filtered sgRNAs for a start/stop site in the 'filtered_gRNAs' format, will create dataframe containing positional information and condition checks\n",
    "    for each sgRNA.\n",
    "\n",
    "    params: df: a pandas series of sgRNAs for one start/stop site:\n",
    "        fmin                                                            370082\n",
    "        fmax                                                            370104\n",
    "        #chr                                                                 X\n",
    "        strand                                                               +\n",
    "        sgRNA_sequence                                 CTATCTCTTAAAATGGCTTTGGG\n",
    "        Gene_ID                                                    FBgn0000022\n",
    "        Transcript_ID                                              FBtr0070072\n",
    "        Chromosome                                                           X\n",
    "        Gene_Region                                                start_codon\n",
    "        Start                                                           370094\n",
    "        Stop                                                            370096\n",
    "        Strand                                                               +\n",
    "        Reference_Seq        CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...\n",
    "        upstreamHA           TTTGCTCAGTTTTTTATTGGCGCCGGGACCAATTCCCCGGCGACCA...\n",
    "        downstreamHA         TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...\n",
    "        \n",
    "    output: dataframe with columns:\n",
    "        fmin                                                            370082\n",
    "        fmax                                                            370104\n",
    "        #chr                                                                 X\n",
    "        strand                                                               +\n",
    "        sgRNA_sequence                                 CTATCTCTTAAAATGGCTTTGGG\n",
    "        Gene_ID                                                    FBgn0000022\n",
    "        Transcript_ID                                              FBtr0070072\n",
    "        Chromosome                                                           X\n",
    "        Gene_Region                                                start_codon\n",
    "        Start                                                           370094\n",
    "        Stop                                                            370096\n",
    "        Strand                                                               +\n",
    "        Reference_Seq        CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...\n",
    "        upstreamHA           TTTGCTCAGTTTTTTATTGGCGCCGGGACCAATTCCCCGGCGACCA...\n",
    "        downstreamHA         TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...\n",
    "        positionScore                                                        8\n",
    "        PAM_in_start/stop                                                False\n",
    "        <15_bp3’_overhang                                                 True\n",
    "        PAM_in_CDS                                                        True\n",
    "        PAM_outside_CDS                                                  False\n",
    "        CDS_boundary                                                       >22\n",
    "        lastG                                                             31.0\n",
    "        cutSite                                                           26.0\n",
    "    \n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "\n",
    "    #Adding position scores (fmax - stop)\n",
    "    sgRNAdf[\"positionScore\"] = sgRNAdf[\"fmax\"] - sgRNAdf[\"Stop\"]\n",
    "    \n",
    "    #Dataframe containing parameter ranges to interpret the positon score, based on gene strand, sgRNA strand, and start/stop\n",
    "    positionScoreParameters = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\")\n",
    "\n",
    "    #Per parameter, append the sgRNAdf with a TRUE/FALSE value per condition.\n",
    "    #Also add positional information about the CDS boundary, position of the last G/C, and the cut site from the conditions file.\n",
    "\n",
    "    #Start by adding columns for these to sgRNAdf\n",
    "    booleanColumns = [\"PAM_in_start/stop\", \"<15_bp3’_overhang\", \"PAM_in_CDS\", \"PAM_outside_CDS\"]\n",
    "    positionColumns = [\"CDS_boundary\", \"lastG\", \"cutSite\"]\n",
    "    sgRNAdf = sgRNAdf.reindex(columns = sgRNAdf.columns.tolist() + booleanColumns + positionColumns)\n",
    "    \n",
    "    #Iterating through sgRNAdf, extract the appropriate row of the conditions file and add new information to sgRNAdf.\n",
    "    for ind, sgRNA in sgRNAdf.iterrows():\n",
    "\n",
    "        #Extract the appropriate parameter row per sgRNA\n",
    "        conditions = positionScoreParameters.loc[(positionScoreParameters[\"start/stop\"] == sgRNA['Gene_Region']) & (positionScoreParameters[\"strand_type\"] == sgRNA['Strand']) & (positionScoreParameters[\"sgRNA_strand\"] == sgRNA['strand'])]\n",
    "        conditions = conditions.reset_index(drop = True)\n",
    "        \n",
    "        #Per column, input true/false as to whether the position score meets that condition\n",
    "        for col in booleanColumns:\n",
    "            colValue = conditions.at[0,col] #extract parameter range values from dataframe\n",
    "            #Process the value into a range (in format [min, max])\n",
    "            #If the values should be 'more than' or 'less than', 25 is used as a max or -25 as min because distances cannot be more than 20\n",
    "            if \">\" in colValue: #could simplify this further by just defining all as ranges in excel\n",
    "                minMax = [int(colValue[1:])+1, 25]\n",
    "            elif \"<\" in colValue:\n",
    "                minMax = [-25,int(colValue[1:])]\n",
    "            elif \":\" in colValue:\n",
    "                min, max = colValue.split(\":\")\n",
    "                minMax = [int(min), int(max)]\n",
    "            else:\n",
    "                print(\"Incorrect format of range value. Verify inputs.\")\n",
    "\n",
    "            #Into the output dataframe, print true/false as to whether the positionScore has met the condition for that column\n",
    "            sgRNAdf.at[ind, col] = bool(sgRNA[\"positionScore\"] in range(minMax[0], minMax[1]))\n",
    "       \n",
    "        #Calculate PAM last G or C position using positionScore. Add this to the sgRNA catalogue.\n",
    "        sgRNAdf.at[ind, \"lastG\"] = sgRNAdf.at[ind, \"positionScore\"] + int(conditions.at[0, \"lastG\"])\n",
    "\n",
    "        #Calculate cutsite coordinate using positionScore. Add this to the sgRNA catalogue.\n",
    "        sgRNAdf.at[ind, \"cutSite\"] = sgRNAdf.at[ind, \"positionScore\"] + int(conditions.at[0, \"cutSite\"])\n",
    "\n",
    "        #Keep CDS boundary position in output df\n",
    "        sgRNAdf.at[ind, \"CDS_boundary\"] = conditions.at[0, \"CDS_boundary\"]\n",
    "    \n",
    "    return sgRNAdf\n",
    "\n",
    "def checkCDSCutandOrder(sgRNAdf):\n",
    "    \"\"\"\n",
    "    Given an sgRNAdf, will calculate firstly rows where the sgRNA cuts inside CDS. If multiple, selects that which cuts closest to start/stop.\n",
    "    If none, selects closest cut sgRNA that is outside CDS.\n",
    "    \n",
    "    params: sgRNAdf in format:\n",
    "        fmin                                                            370082\n",
    "        fmax                                                            370104\n",
    "        #chr                                                                 X\n",
    "        strand                                                               +\n",
    "        sgRNA_sequence                                 CTATCTCTTAAAATGGCTTTGGG\n",
    "        Gene_ID                                                    FBgn0000022\n",
    "        Transcript_ID                                              FBtr0070072\n",
    "        Chromosome                                                           X\n",
    "        Gene_Region                                                start_codon\n",
    "        Start                                                           370094\n",
    "        Stop                                                            370096\n",
    "        Strand                                                               +\n",
    "        Reference_Seq        CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...\n",
    "        upstreamHA           TTTGCTCAGTTTTTTATTGGCGCCGGGACCAATTCCCCGGCGACCA...\n",
    "        downstreamHA         TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...\n",
    "        positionScore                                                        8\n",
    "        PAM_in_start/stop                                                False\n",
    "        <15_bp3’_overhang                                                 True\n",
    "        PAM_in_CDS                                                        True\n",
    "        PAM_outside_CDS                                                  False\n",
    "        CDS_boundary                                                       >22\n",
    "        lastG                                                             31.0\n",
    "        cutSite                                                           26.0\n",
    "\n",
    "\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    #extract CDS boundary value\n",
    "    sgRNAdf.reset_index(inplace= True, drop= True)\n",
    "    CDSBoundary = sgRNAdf.at[0, 'CDS_boundary']\n",
    "\n",
    "    #Calculate sgRNA/cut site difference\n",
    "    sgRNAdf[\"cutsite-CDSbound\"] = abs(sgRNAdf[\"cutSite\"] - 20)\n",
    "    \n",
    "    #C. Check cut site in CDS\n",
    "    \n",
    "    #Define CDS boundary range\n",
    "    if \">\" in CDSBoundary: #could simplify this further by just defining all as ranges in excel\n",
    "        CDSBoundaryRange = [int(CDSBoundary[1:])+1, 43]\n",
    "    elif \"<\" in CDSBoundary:\n",
    "        CDSBoundaryRange = [0,int(CDSBoundary[1:])]\n",
    "    else:\n",
    "        print(\"CDS boundary value not found/invalid\")\n",
    "\n",
    "    #filter for only sgRNAs that cut in CDS\n",
    "    conditionC = sgRNAdf[sgRNAdf[\"cutSite\"].between(CDSBoundaryRange[0], CDSBoundaryRange[1])]\n",
    "\n",
    "    #if there are sgRNAs that cut in CDS, sort by distance of cutsite and CDS start. Select the one that cuts closest.\n",
    "    if len(conditionC) > 0: #cuts in CDS, closest cut (C1, C2)\n",
    "        conditionCclosestCut = conditionC.sort_values('cutsite-CDSbound')\n",
    "        conditionCclosestCut = conditionCclosestCut.reset_index(drop = True) #reset index\n",
    "        winnersgRNA = conditionCclosestCut.at[0, \"sgRNA_sequence\"]\n",
    "    else: #no sgRNAs cut in CDS, select closest that still met condition B (C3)\n",
    "        nonCDSclosestCut = sgRNAdf.sort_values('cutsite-CDSbound')\n",
    "        winnersgRNA = nonCDSclosestCut.at[0, \"sgRNA_sequence\"]\n",
    "\n",
    "    return winnersgRNA\n",
    "\n",
    "def find_best_gRNA(sgRNAdf):\n",
    "    \"\"\"\n",
    "    Select the best guide RNA from a dataframe of guideRNAs that are ±20bp from the start/stop site.\n",
    "\n",
    "    params:\n",
    "        df: the dataframe of potential guideRNAs\n",
    "    \n",
    "    output:\n",
    "        winnersgRNA: the sequence of the best guideRNA (str)\n",
    "        mutatinoNeeded: True/False of whether this sgRNA needs to be mutated.\n",
    "\n",
    "    \"\"\"\n",
    "    #Set up the winning guide RNA\n",
    "    winnerFound = False\n",
    "    mutationNeeded = False\n",
    "    winnersgRNA = \"\"\n",
    "\n",
    "    #A. Ideal condition - PAM in start/stop\n",
    "    conditionA = sgRNAdf[sgRNAdf[\"PAM_in_start/stop\"] == True] #This is the subset df for which PAM is in the start/stop\n",
    "    if len(conditionA) > 0: #if there is one or more sgRNAs for this condition, select the first as the winner\n",
    "        conditionA = conditionA.reset_index(drop = True) #reset index\n",
    "        winnersgRNA = conditionA.at[0, \"sgRNA_sequence\"]\n",
    "        winnerFound = True\n",
    "\n",
    "    #B. sgRNA overhang is less than 15bp\n",
    "    if winnerFound == False:\n",
    "        conditionB = sgRNAdf[sgRNAdf[\"<15_bp3’_overhang\"] == True]\n",
    "        if len(conditionB) == 1:\n",
    "            conditionB = conditionB.reset_index(drop = True)\n",
    "            winnersgRNA = conditionB.at[0, \"sgRNA_sequence\"]\n",
    "            winnerFound = True\n",
    "        elif len(conditionB) > 1: #if more than one, check condition C (select for cut site in CDS preferencially, and the sgRNA where cut site is closest to CDS)\n",
    "            winnersgRNA = checkCDSCutandOrder(conditionB) \n",
    "            winnerFound = True   \n",
    "\n",
    "    #D. sgRNA overhang is more than 15bp, need to mutate\n",
    "    if winnerFound == False:\n",
    "        if len(sgRNAdf) == 1: #if just 1 sgRNA, select this one\n",
    "            winnersgRNA = sgRNAdf.at[0, \"sgRNA_sequence\"]\n",
    "        else: #more than 1 sgRNA\n",
    "            winnersgRNA = checkCDSCutandOrder(sgRNAdf) #Again, iterate through condition C to select an sgRNA\n",
    "            winnerFound = True         \n",
    "\n",
    "        mutationNeeded = True\n",
    "\n",
    "    return winnersgRNA, mutationNeeded\n",
    "\n",
    "def codonFragmenter(winnerdf):\n",
    "    \"\"\"\n",
    "    For the start/stop site, will create a list of codons in the appropriate range where sgRNAs might be found (start/stop ±20 bp).\n",
    "    If gene is on - strand, these will be revComp codons.\n",
    "\n",
    "    params:\n",
    "        winnerdf: pandas Series for the row of the sgRNA dataframe containing information about the winner sgRNA.capitalize\n",
    "\n",
    "    returns:\n",
    "        codonList: codons from the ±21bp region around the start/stop site in mutable format (revComp if gene is -)\n",
    "    \"\"\"\n",
    "    \n",
    "    #Extract HA sequences - just the 21bp near start/stop\n",
    "    HAL = winnerdf[\"upstreamHA\"][-21:]\n",
    "    HAR = winnerdf[\"downstreamHA\"][0:21]\n",
    "\n",
    "    #Start codon list\n",
    "    codonList = []\n",
    "\n",
    "    #Add HAL codons\n",
    "    for codonBase1 in range(0, len(HAL), 3):\n",
    "        codonList.append(HAL[codonBase1:codonBase1+3])\n",
    "\n",
    "    #add start/stop - this is added as 'ATG' (even if stop) but will never be mutated. As HAL and HAR only will be replaced into tfsDF,\n",
    "    #this will not affect original sequences.\n",
    "    codonList.append('ATG')\n",
    "\n",
    "    #Add HAR codons\n",
    "    for codonBase1 in range(0, len(HAR), 3):\n",
    "        codonList.append(HAR[codonBase1:codonBase1+3])\n",
    "    \n",
    "    #Define gene strand:\n",
    "    if winnerdf[\"Strand\"] == '-': #If on the minus strand, take revComp per codon\n",
    "        for ind, codon in enumerate(codonList):\n",
    "            codonList[ind] = revComp(codon)\n",
    "\n",
    "    return codonList\n",
    "\n",
    "def codonReverseFragmenter(codonsList, winnerdf):\n",
    "    \"\"\"\n",
    "    Will take a fragmented list of codons of the start/stop site region and replace the mutated HAL/HAR arms in tfsDF.\n",
    "    If gene is on - strand, revComp codons will be reversed back to the + strand sequences.\n",
    "    \"\"\"\n",
    "    HAL = \"\"\n",
    "    HAR = \"\"\n",
    "\n",
    "    #return to + strand if the gene is on -\n",
    "    if winnerdf[\"Strand\"] == '-':\n",
    "        for ind, codon in enumerate(codonsList):\n",
    "            codonsList[ind] = revComp(codon)\n",
    "    \n",
    "    #replace HAL\n",
    "    for codon in range(0, 7):\n",
    "        HAL += codonsList[codon]\n",
    "\n",
    "    #replace HAR\n",
    "    for codon in range(8, 15):\n",
    "        HAR += codonsList[codon]\n",
    "\n",
    "    #replace mutated HAL and HAR\n",
    "    winnerdf.at[\"upstreamHA\"] = winnerdf[\"upstreamHA\"][:-21] + HAL\n",
    "    winnerdf.at[\"downstreamHA\"] = HAR + winnerdf[\"downstreamHA\"][21:]\n",
    "\n",
    "    return winnerdf\n",
    "\n",
    "def mutator(winnerdf):\n",
    "    \"\"\"\n",
    "    In the case where a fragment needs to be mutated, will mutate in CDS (preferably PAM, if not then in the sgRNA). \n",
    "    If not possible, will mutate PAM outside of CDS to NTG/CTN.\n",
    "    \n",
    "    params:\n",
    "        sequenceToMutate: the homology arm or primer fragment that will be mutated.\n",
    "        winner_sgRNACatalogue: sgRNACatalogue in format as above, with only the row for the winner sgRNA selected.\n",
    "    \"\"\"\n",
    "\n",
    "    import pandas as pd\n",
    "\n",
    "    #Mutated starts as false\n",
    "    mutated = False\n",
    "\n",
    "    #Codon fragmenter - codons are now in mutable format in a list, from last 21bp of HAL, the start/stop site, then the first 21bp of HAR. RevComp if gene is -\n",
    "    mutableCodons = codonFragmenter(winnerdf)\n",
    "\n",
    "    ## Coordinate information per codon of mutableCodons. This accounts for gene strand.\n",
    "    if winnerdf[\"Strand\"] == '+':\n",
    "        codonCoordinates = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\", sheet_name=\"CodonCoordinatePlus\", index_col= 0)\n",
    "    elif winnerdf[\"Strand\"] == '-':\n",
    "        codonCoordinates = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\", sheet_name=\"CodonCoordinateMinus\", index_col= 0)\n",
    "\n",
    "    #Define CDS boundary range\n",
    "    if \">\" in winnerdf[\"CDS_boundary\"]:\n",
    "        CDSboundary = [int(winnerdf[\"CDS_boundary\"][1:])+1, 43]\n",
    "    elif \"<\" in winnerdf[\"CDS_boundary\"]:\n",
    "        CDSboundary = [0,int(winnerdf[\"CDS_boundary\"][1:])]\n",
    "\n",
    "    #PAM last G coordinate\n",
    "    lastG = winnerdf['lastG'] #this is a C if on minus strand\n",
    "\n",
    "    #Coordinates for sgRNA bases sgRNA1, sgRNA2, sgRNA3, sgRNA4, sgRNA5, sgRNA6\n",
    "    #Any of these coordinates that are outside of the CDS will be removed, leaving only sgRNA coordinates in CDS.\n",
    "    #We also define the second PAM base based on sgRNA +/- here - secondG (again, might be c)\n",
    "    sgRNAcoordinates = []\n",
    "    if winnerdf['strand'] == '+': #sgRNA on + strand\n",
    "        secondG = lastG-1\n",
    "        for shift in range(3, 9):\n",
    "            sgRNAcoordinates.append(lastG - shift)\n",
    "    else:\n",
    "        secondG = lastG+1\n",
    "        for shift in range(3, 9): #sgRNA on - strand\n",
    "            sgRNAcoordinates.append(lastG + shift)\n",
    "\n",
    "    sgRNAcoordinates = [x for x in sgRNAcoordinates if x in range(CDSboundary[0], CDSboundary[1])]\n",
    "\n",
    "    #A. Mutate PAM in CDS (1 mutation)\n",
    "    if winnerdf[\"PAM_in_CDS\"] == True: #check condition\n",
    "        #mutate last G if possible using relative coordinates\n",
    "        codonNumber = codonCoordinates.at[lastG, 'codon'] #this is the codon number (as an index in mutableCodons)\n",
    "        codon = mutableCodons[codonNumber] #This is the codon we want to mutate\n",
    "        base = codonCoordinates.at[lastG, 'base'] #This is the base within that codon (1-3)\n",
    "        potentialCodons = find_synonymous_codons(query_codon =codon, base_to_change= base)\n",
    "        if len(potentialCodons) >0:\n",
    "        #check for NGA and remove this\n",
    "            for cod in potentialCodons:\n",
    "                if cod[:-1] == \"A\":\n",
    "                    potentialCodons.remove(cod)\n",
    "            #If we still have replacement codons left, swap the first one in:\n",
    "            if len(potentialCodons) >0:\n",
    "                mutableCodons[codonNumber] = potentialCodons[0]\n",
    "                mutated = True\n",
    "        else: #if no mutation is found, try the second G.\n",
    "            codonNumber = codonCoordinates.at[secondG, 'codon']\n",
    "            codon = mutableCodons[codonNumber]\n",
    "            base = codonCoordinates.at[secondG, 'base']\n",
    "            potentialCodons = find_synonymous_codons(query_codon = codon, base_to_change = base)\n",
    "             #If we have replacement codons, swap the first one in:\n",
    "            if len(potentialCodons) >0:\n",
    "                mutableCodons[codonNumber] = potentialCodons[0]\n",
    "                mutated = True\n",
    "    \n",
    "    #B. Mutate sgRNA in CDS (up to 2 mutations)\n",
    "    #Check if mutation was successful in A, then check if there are mutable bases of the sgRNA in CDS \n",
    "    if mutated == False and len(sgRNAcoordinates) != 0:\n",
    "        print(\"Condition B\")\n",
    "        sgMutatedBases = 0 #count of bases mutated\n",
    "        excludeCodons = [] #this is the codon number of codons we've already mutated. This ensures that we won't re-mutate already mutated codons.\n",
    "        index = 0 #this is the index of the sgRNA coordinate we are trying to mutate\n",
    "        while sgMutatedBases < 2 and index <= len(sgRNAcoordinates): #while 2 mutations have not yet been made and the we've not uet used all coordinates in the sgRNA coordinate list\n",
    "            print(f\"mutating base {index}, mutations so far are {sgMutatedBases}\")\n",
    "            coordinate = sgRNAcoordinates[index] #This is the coordinate of the sgRNA base\n",
    "            codonNumber = codonCoordinates.at[coordinate, 'codon']\n",
    "            codon = mutableCodons[codonNumber] #this is the codon number that this coordinate is in\n",
    "            base = codonCoordinates.at[coordinate, 'base'] #this is the base within that codon that the coordinate corresponds to (1-3)\n",
    "            \n",
    "            #check if we've already mutated this codon\n",
    "            if codonNumber not in excludeCodons: #if not, attempt to mutate this base\n",
    "                potentialCodons = find_synonymous_codons(query_codon = codon, base_to_change = base)\n",
    "                if len(potentialCodons) > 0: #if mutation is possible, accept this mutation\n",
    "                    mutableCodons[codonNumber] = potentialCodons[0] #replace this codon in the fragmenter list\n",
    "                    excludeCodons.append(codonNumber) #add this to already mutated codons\n",
    "                    sgMutatedBases +=1 #add 1 to count of mutated bases\n",
    "            \n",
    "            index += 1 #move to next sgRNA base\n",
    "\n",
    "        if sgMutatedBases > 0: #if we've mutated at least 1 sgRNA base, set mutated to True\n",
    "            mutated = True\n",
    "\n",
    "    #C. Mutate PAM outside CDS - to NTG or CTN\n",
    "    if mutated == False:\n",
    "        codonNumber = codonCoordinates.at[secondG, 'codon']\n",
    "        codon = mutableCodons[codonNumber]\n",
    "        base = codonCoordinates.at[secondG, 'base']\n",
    "        newCodon = codon[:base-1] + 'T' + codon[base:] #add a 'T' where the old base was in this codon\n",
    "        mutableCodons[codonNumber] = newCodon\n",
    "        mutated = True\n",
    "\n",
    "    #Replace the potentially mutated homology arms back in the upstreamHA and downstreamHA of our winnerdf\n",
    "    winnerdfmutated = codonReverseFragmenter(mutableCodons, winnerdf)\n",
    "\n",
    "    #Indicate if mutation has occurred within winner df\n",
    "    winnerdfmutated['mutated'] = mutated\n",
    "\n",
    "    return winnerdfmutated\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 478,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sgRNArunner():\n",
    "    \"\"\"\n",
    "    Check functions so far work with the proposed changes.\n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "    #set up reference sequence Bio SeqIO element\n",
    "    refSeqPerChromosome = refSeq()\n",
    "\n",
    "    #Make the dataframe for all transcription factor start/stop site infp\n",
    "    TFsdf = make_dataframe_from_TFs_list(\"inputfiles/TFs.xlsx\", refSeqPerChromosome)\n",
    "    display(TFsdf)\n",
    "    \n",
    "    #set up the winner sgRNA variable. If we want more information about this sgRNA to be returned, \n",
    "    # we can use the winnersgRNA sequence as the index to extract that dataframe row and return the full information. e.g. sgRNAdf[sgRNAdf[\"sgRNA_sequence\"] == winnersgRNA]\n",
    "    TFsdfWinnersandMutated = pd.DataFrame(columns=[\"fmin\", \"fmax\", \"#chr\", \"strand\", \"sgRNA_sequence\", \"Gene_ID\",\n",
    "                                                   \"Transcript_ID\", \"Chromosome\", \"Gene_Region\", \"Start\", \"Stop\",\n",
    "                                                   \"Strand\", \"Reference_Seq\", \"upstreamHA\", \"downstreamHA\", \"positionScore\",\n",
    "                                                    \"PAM_in_start/stop\", \"<15_bp3’_overhang\", \"PAM_in_CDS\", \"PAM_outside_CDS\",\n",
    "                                                    \"CDS_boundary\", \"lastG\", \"cutSite\", \"mutated\"])\n",
    "    \n",
    "    display(TFsdfWinnersandMutated)\n",
    "\n",
    "    #Per row of this dataframe, select a guideRNA\n",
    "    for ind, row in TFsdf.iterrows():\n",
    "        print(ind)\n",
    "\n",
    "        filtered_sgRNA = gRNA_stringencyIterator(row, refSeqPerChromosome)\n",
    "\n",
    "        #Score the sgRNAs for this site\n",
    "        sgRNAdf = sgRNApositionCheck(filtered_sgRNA)\n",
    "\n",
    "        display(sgRNAdf)\n",
    "\n",
    "        if len(filtered_sgRNA) == 0:\n",
    "            print(f\"No sgRNAs found at all stringencies for {ind}.\")\n",
    "            print(filtered_sgRNA)\n",
    "\n",
    "            for col in list(row.index):\n",
    "                TFsdfWinnersandMutated.at[ind, col] = row[col]\n",
    "                \n",
    "        else:\n",
    "            #Select winner\n",
    "            winnersgRNA, mutationNeeded = find_best_gRNA(sgRNAdf)\n",
    "            winnerdf = sgRNAdf[sgRNAdf[\"sgRNA_sequence\"] == winnersgRNA] #This is the Pandas Series for the winning sgRNA from the original dataframe\n",
    "\n",
    "            if mutationNeeded == True:\n",
    "                winnerdf = mutator(winnerdf)\n",
    "        \n",
    "            TFsdfWinnersandMutated.at[ind] = winnerdf\n",
    "        \n",
    "        display(TFsdfWinnersandMutated)\n",
    "            \n",
    "    return TFsdfWinnersandMutated\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sgRNArunner()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Free variables\n",
    "\n",
    "#refSeqPerChromosome\n",
    "from Bio import SeqIO\n",
    "\n",
    "refSeqPerChromosome = {}\n",
    "\n",
    "for seq in SeqIO.parse(open(\"inputfiles/dmel-all-chromosome-r6.48.fasta\"), 'fasta'):\n",
    "    refSeqPerChromosome[seq.id] = seq.seq \n",
    "\n",
    "#TFsdf\n",
    "TFsdf = make_dataframe_from_TFs_list(\"inputfiles/TFs.xlsx\", refSeqPerChromosome)\n",
    "display(TFsdf)\n",
    "\n",
    "\"\"\"\n",
    "#singletfsDF = pandas series for one row extracted from tfsDF\n",
    "for index, row in TFsdf.iterrows():\n",
    "    if index == 0:\n",
    "        singletfsDF = row\n",
    "        print(singletfsDF)\n",
    "        \n",
    "#gRNAFileAnnotation = dataframe of gRNAs as extracted from original gff file\n",
    "import pandas as pd\n",
    "gRNAFileAnnotation = pd.read_csv(\"inputfiles/Hu.2019.8.28.sgRNA_designs/1to3NonCdsOffTarget_low_stringency.gff\", sep = \"\\t\", index_col = False)\n",
    "gRNAFileAnnotation = gRNAFileAnnotation.assign(target_site_variation= \"\")\n",
    "index = 0 #TODO@Marina improve this code\n",
    "for attribute in gRNAFileAnnotation['attributes']:\n",
    "    fullatt = (attribute).split(\";\")\n",
    "    gRNAFileAnnotation.at[index,\"target_site_variation\"] = fullatt[8]\n",
    "    index+=1\n",
    "\n",
    "print(type(gRNAFileAnnotation))\n",
    "\n",
    "#GenomeCoordinatesFiltered = filtered gRNAs for 20bp region, no target site variation, and with start/stop info from tfsDF\n",
    "GenomeCoordinatesFiltered = gRNAFileAnnotation[(gRNAFileAnnotation['target_site_variation'] == \"target_site_variation=none observed\") & \n",
    "                                                (gRNAFileAnnotation['#chr'] == singletfsDF[\"Chromosome\"]) & \n",
    "                                                (gRNAFileAnnotation['fmin']-1 >= singletfsDF[\"Start\"] - 20)]\n",
    "for (columnName, columnData) in singletfsDF.items():\n",
    "       GenomeCoordinatesFiltered = GenomeCoordinatesFiltered.assign(**{columnName: columnData})\n",
    "display(GenomeCoordinatesFiltered)\n",
    "\"\"\"\n",
    "#filteredgRNAs - theoretical sgRNAs for one site\n",
    "filteredgRNAs = pd.read_excel(\"inputfiles/mockMaterials/filtered_gRNAs.xlsx\", index_col= 0)\n",
    "\n",
    "display(filteredgRNAs)\n",
    "\n",
    "sgRNAdf = sgRNApositionCheck(filteredgRNAs)\n",
    "\n",
    "display(sgRNAdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>upstreamHA</th>\n",
       "      <th>downstreamHA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "      <td>TTTGCTCAGTTTTTTATTGGCGCCGGGACCAATTCCCCGGCGACCA...</td>\n",
       "      <td>TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>370697</td>\n",
       "      <td>370699</td>\n",
       "      <td>+</td>\n",
       "      <td>GACATTGTGTCGTTCGTATGTCGCCCATTGAGACCGCCAATGGAGG...</td>\n",
       "      <td>TCCTCAATGTATGCTATAAAAAAAAACTTAAATATAAATTGTCTTC...</td>\n",
       "      <td>AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FBgn0004170</td>\n",
       "      <td>FBtr0070073</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>396177</td>\n",
       "      <td>396179</td>\n",
       "      <td>+</td>\n",
       "      <td>ACCTGCGATAATTTGACATTCTTAGAAACTACCTGAAGAAATAGGA...</td>\n",
       "      <td>TCTGGTCAGTGCCATACCCCTTGGTGTATACTTGCGAGTCTTAATT...</td>\n",
       "      <td>AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FBgn0004170</td>\n",
       "      <td>FBtr0070073</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>397212</td>\n",
       "      <td>397214</td>\n",
       "      <td>+</td>\n",
       "      <td>TCGCGCCTGCGGTGTAAGAATAAATGCTTCAAGTAACAACTTCGTA...</td>\n",
       "      <td>GCAAAAAAATATGTTTTCGTTTCCATGCTATAGTTATTTCATAAAT...</td>\n",
       "      <td>TGCTCCTGCCATAGAGAGATGTAGTCCAAAATCTCCTCGTCGTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FBgn0002561</td>\n",
       "      <td>FBtr0070074</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>409796</td>\n",
       "      <td>409798</td>\n",
       "      <td>+</td>\n",
       "      <td>CGACGTCGTCCAGCTTGATGAAAAACTTTGTGGCGGTGTGCGCCAG...</td>\n",
       "      <td>CATGTTATTGTAGTTGAACTTCTTCTTCTGGGAGGACAACATGGGT...</td>\n",
       "      <td>GTAATCCTTGCGAGAGTTTTCTAAGATTTAGTTTACAGATGTTGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4367</th>\n",
       "      <td>FBgn0031375</td>\n",
       "      <td>FBtr0330201</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1950927</td>\n",
       "      <td>1950929</td>\n",
       "      <td>-</td>\n",
       "      <td>TGACTCCGCTTTCAAGAAGTATGTGCCGGGTCCGTGTTCGGGCGCC...</td>\n",
       "      <td>TTCTTGCAACTCCAGCCGCAGCAGCCACATCCGCAAAGCCAGCAAC...</td>\n",
       "      <td>TGCAGTTTGTGGCAGCGGCGGCCGCTGCGAGCAGCTCGAATTCCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4368</th>\n",
       "      <td>FBgn0031375</td>\n",
       "      <td>FBtr0330202</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1954323</td>\n",
       "      <td>1954325</td>\n",
       "      <td>-</td>\n",
       "      <td>ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...</td>\n",
       "      <td>TCCGCTTCGGTCGCTGATCCTGTGCTGAACGCGTCCTGTTCCTCGC...</td>\n",
       "      <td>TTTACTTTGTACGTATTCGATACATAAAAACTTAATTTATTACTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4369</th>\n",
       "      <td>FBgn0031375</td>\n",
       "      <td>FBtr0330202</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>1950795</td>\n",
       "      <td>1950797</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTGCTGTACTATGCACCCAATCAGCTGATGTGCGCCGCTGCTGCC...</td>\n",
       "      <td>ACGGGGTCGACCTCGACGACCAGCTCCACGGTGACCACCACCACCA...</td>\n",
       "      <td>CAGAAGGACCCCTGTCCACCTCCACCCACCCCGCAGAACCTCAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4370</th>\n",
       "      <td>FBgn0031086</td>\n",
       "      <td>FBtr0070032</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>20092383</td>\n",
       "      <td>20092385</td>\n",
       "      <td>-</td>\n",
       "      <td>CTGTGTCTTCAAAATTCTGGCGGCTGCGAGACGCGACAAAGGAAGT...</td>\n",
       "      <td>GACTCCCGTTCCCGTAGATGTGGATCAATGAATGAATCCGAAGGCT...</td>\n",
       "      <td>ATACGACACCAATCTTCCAGTCCAGCTTCTCCATCCGATCGCTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4371</th>\n",
       "      <td>FBgn0031086</td>\n",
       "      <td>FBtr0070032</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>20092383</td>\n",
       "      <td>20092385</td>\n",
       "      <td>-</td>\n",
       "      <td>CTGTGTCTTCAAAATTCTGGCGGCTGCGAGACGCGACAAAGGAAGT...</td>\n",
       "      <td>GACTCCCGTTCCCGTAGATGTGGATCAATGAATGAATCCGAAGGCT...</td>\n",
       "      <td>ATACGACACCAATCTTCCAGTCCAGCTTCTCCATCCGATCGCTGCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4370 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene_ID Transcript_ID Chromosome  Gene_Region     Start      Stop  \\\n",
       "2     FBgn0000022   FBtr0070072          X  start_codon    370094    370096   \n",
       "3     FBgn0000022   FBtr0070072          X   stop_codon    370697    370699   \n",
       "4     FBgn0004170   FBtr0070073          X  start_codon    396177    396179   \n",
       "5     FBgn0004170   FBtr0070073          X   stop_codon    397212    397214   \n",
       "6     FBgn0002561   FBtr0070074          X  start_codon    409796    409798   \n",
       "...           ...           ...        ...          ...       ...       ...   \n",
       "4367  FBgn0031375   FBtr0330201         2L   stop_codon   1950927   1950929   \n",
       "4368  FBgn0031375   FBtr0330202         2L  start_codon   1954323   1954325   \n",
       "4369  FBgn0031375   FBtr0330202         2L   stop_codon   1950795   1950797   \n",
       "4370  FBgn0031086   FBtr0070032          X  start_codon  20092383  20092385   \n",
       "4371  FBgn0031086   FBtr0070032          X  start_codon  20092383  20092385   \n",
       "\n",
       "     Strand                                      Reference_Seq  \\\n",
       "2         +  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...   \n",
       "3         +  GACATTGTGTCGTTCGTATGTCGCCCATTGAGACCGCCAATGGAGG...   \n",
       "4         +  ACCTGCGATAATTTGACATTCTTAGAAACTACCTGAAGAAATAGGA...   \n",
       "5         +  TCGCGCCTGCGGTGTAAGAATAAATGCTTCAAGTAACAACTTCGTA...   \n",
       "6         +  CGACGTCGTCCAGCTTGATGAAAAACTTTGTGGCGGTGTGCGCCAG...   \n",
       "...     ...                                                ...   \n",
       "4367      -  TGACTCCGCTTTCAAGAAGTATGTGCCGGGTCCGTGTTCGGGCGCC...   \n",
       "4368      -  ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...   \n",
       "4369      -  CCTGCTGTACTATGCACCCAATCAGCTGATGTGCGCCGCTGCTGCC...   \n",
       "4370      -  CTGTGTCTTCAAAATTCTGGCGGCTGCGAGACGCGACAAAGGAAGT...   \n",
       "4371      -  CTGTGTCTTCAAAATTCTGGCGGCTGCGAGACGCGACAAAGGAAGT...   \n",
       "\n",
       "                                             upstreamHA  \\\n",
       "2     TTTGCTCAGTTTTTTATTGGCGCCGGGACCAATTCCCCGGCGACCA...   \n",
       "3     TCCTCAATGTATGCTATAAAAAAAAACTTAAATATAAATTGTCTTC...   \n",
       "4     TCTGGTCAGTGCCATACCCCTTGGTGTATACTTGCGAGTCTTAATT...   \n",
       "5     GCAAAAAAATATGTTTTCGTTTCCATGCTATAGTTATTTCATAAAT...   \n",
       "6     CATGTTATTGTAGTTGAACTTCTTCTTCTGGGAGGACAACATGGGT...   \n",
       "...                                                 ...   \n",
       "4367  TTCTTGCAACTCCAGCCGCAGCAGCCACATCCGCAAAGCCAGCAAC...   \n",
       "4368  TCCGCTTCGGTCGCTGATCCTGTGCTGAACGCGTCCTGTTCCTCGC...   \n",
       "4369  ACGGGGTCGACCTCGACGACCAGCTCCACGGTGACCACCACCACCA...   \n",
       "4370  GACTCCCGTTCCCGTAGATGTGGATCAATGAATGAATCCGAAGGCT...   \n",
       "4371  GACTCCCGTTCCCGTAGATGTGGATCAATGAATGAATCCGAAGGCT...   \n",
       "\n",
       "                                           downstreamHA  \n",
       "2     TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...  \n",
       "3     AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...  \n",
       "4     AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...  \n",
       "5     TGCTCCTGCCATAGAGAGATGTAGTCCAAAATCTCCTCGTCGTCGG...  \n",
       "6     GTAATCCTTGCGAGAGTTTTCTAAGATTTAGTTTACAGATGTTGAC...  \n",
       "...                                                 ...  \n",
       "4367  TGCAGTTTGTGGCAGCGGCGGCCGCTGCGAGCAGCTCGAATTCCGG...  \n",
       "4368  TTTACTTTGTACGTATTCGATACATAAAAACTTAATTTATTACTCT...  \n",
       "4369  CAGAAGGACCCCTGTCCACCTCCACCCACCCCGCAGAACCTCAGCA...  \n",
       "4370  ATACGACACCAATCTTCCAGTCCAGCTTCTCCATCCGATCGCTGCT...  \n",
       "4371  ATACGACACCAATCTTCCAGTCCAGCTTCTCCATCCGATCGCTGCT...  \n",
       "\n",
       "[4370 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "refSeqPerChrom = refSeqPerChromosome\n",
    "TFsDF = TFsdf.loc[2:, :]\n",
    "display(TFsDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def workingsgRNArunner(refSeqPerChromosome = refSeqPerChrom, TFsdf = TFsDF):\n",
    "    \"\"\"\n",
    "    Check functions so far work with the proposed changes.\n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "    \n",
    "    #set up the winner sgRNA variable. If we want more information about this sgRNA to be returned, \n",
    "    # we can use the winnersgRNA sequence as the index to extract that dataframe row and return the full information. e.g. sgRNAdf[sgRNAdf[\"sgRNA_sequence\"] == winnersgRNA]\n",
    "    TFsdfWinnersandMutated = pd.DataFrame(columns=[\"fmin\", \"fmax\", \"#chr\", \"strand\", \"sgRNA_sequence\", \"Gene_ID\",\n",
    "                                                   \"Transcript_ID\", \"Chromosome\", \"Gene_Region\", \"Start\", \"Stop\",\n",
    "                                                   \"Strand\", \"Reference_Seq\", \"upstreamHA\", \"downstreamHA\", \"positionScore\",\n",
    "                                                    \"PAM_in_start/stop\", \"<15_bp3’_overhang\", \"PAM_in_CDS\", \"PAM_outside_CDS\",\n",
    "                                                    \"CDS_boundary\", \"lastG\", \"cutSite\", \"mutated\"])\n",
    "    \n",
    "    #Per row of this dataframe, select a guideRNA\n",
    "    for ind, row in TFsdf.iterrows():\n",
    "        print(f\"Selecting guide RNA for TFsdf row {ind}\")\n",
    "\n",
    "        filtered_sgRNA = gRNA_stringencyIterator(row, refSeqPerChromosome)\n",
    "\n",
    "        #Score the sgRNAs for this site\n",
    "        sgRNAdf = sgRNApositionCheck(filtered_sgRNA)\n",
    "\n",
    "        #Add a column to establish whether mutation has occurred. This will be set to 'True' in the mutator function and starts as False by default.\n",
    "        sgRNAdf['mutated'] = False\n",
    "\n",
    "        #If no sgRNAs were found at any stringency\n",
    "        if len(sgRNAdf) == 0:\n",
    "            print(f\"No sgRNAs found at all stringencies for {ind}.\")\n",
    "            print(sgRNAdf)\n",
    "\n",
    "            #Just input the information about this site into the final DF. The columns about the guideRNA will be filled with 'NaN', indicating no guideRNA could be found.\n",
    "            for col in list(row.index):\n",
    "                TFsdfWinnersandMutated.at[ind, col] = row[col]\n",
    "            display(TFsdfWinnersandMutated)\n",
    "        else: #If we have at least one sgRNA\n",
    "            #Select winner\n",
    "            winnersgRNA, mutationNeeded = find_best_gRNA(sgRNAdf) #winnersgRNA is a string of the winning sequence, mutationNeeded is a bool variable\n",
    "            winnerdf = sgRNAdf[sgRNAdf[\"sgRNA_sequence\"] == winnersgRNA] #This is the dataframe row for the winning sgRNA from the original sgRNA dataframe\n",
    "            winnerdf = winnerdf.loc[0] #This is the pandas series for the same information\n",
    "            display(winnerdf)\n",
    "            if mutationNeeded == True: #run mutator if indicated\n",
    "                winnerdf = mutator(winnerdf) #this will mutate HAL or HAR as needed and return the original DF with mutated sequences, and the mutated column set to True\n",
    "        \n",
    "            #Add the winning sgRNA into the output df for this TF start/stop site\n",
    "            TFsdfWinnersandMutated.loc[ind] = winnerdf\n",
    "            display(TFsdfWinnersandMutated)\n",
    "    \n",
    "        #Return dataframe as an excel file (This should be unindented one, but while testing I'd like to see the output file updated after each row of the TFsdf)\n",
    "        TFsdfWinnersandMutated.to_excel(\"outputFiles/winningsgRNAs.xlsx\")\n",
    "\n",
    "    return TFsdfWinnersandMutated\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting guide RNA for TFsdf row 2\n",
      "stringency 0\n",
      "stringency 1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "fmin                                                            370082\n",
       "fmax                                                            370104\n",
       "#chr                                                                 X\n",
       "strand                                                               +\n",
       "sgRNA_sequence                                 CTATCTCTTAAAATGGCTTTGGG\n",
       "Gene_ID                                                    FBgn0000022\n",
       "Transcript_ID                                              FBtr0070072\n",
       "Chromosome                                                           X\n",
       "Gene_Region                                                start_codon\n",
       "Start                                                           370094\n",
       "Stop                                                            370096\n",
       "Strand                                                               +\n",
       "Reference_Seq        CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...\n",
       "upstreamHA           TTTGCTCAGTTTTTTATTGGCGCCGGGACCAATTCCCCGGCGACCA...\n",
       "downstreamHA         TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...\n",
       "positionScore                                                        8\n",
       "PAM_in_start/stop                                                False\n",
       "<15_bp3’_overhang                                                 True\n",
       "PAM_in_CDS                                                        True\n",
       "PAM_outside_CDS                                                  False\n",
       "CDS_boundary                                                       >22\n",
       "lastG                                                             31.0\n",
       "cutSite                                                           26.0\n",
       "mutated                                                          False\n",
       "Name: 0, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>downstreamHA</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>&lt;15_bp3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>lastG</th>\n",
       "      <th>cutSite</th>\n",
       "      <th>mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>...</td>\n",
       "      <td>TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>31.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "2  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  \\\n",
       "2   FBtr0070072          X  start_codon  370094  ...   \n",
       "\n",
       "                                        downstreamHA positionScore  \\\n",
       "2  TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...             8   \n",
       "\n",
       "  PAM_in_start/stop <15_bp3’_overhang PAM_in_CDS PAM_outside_CDS CDS_boundary  \\\n",
       "2             False              True       True           False          >22   \n",
       "\n",
       "  lastG cutSite mutated  \n",
       "2  31.0    26.0   False  \n",
       "\n",
       "[1 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting guide RNA for TFsdf row 3\n",
      "stringency 0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "fmin                                                            370693\n",
       "fmax                                                            370715\n",
       "#chr                                                                 X\n",
       "strand                                                               -\n",
       "sgRNA_sequence                                 CCTGTAAAAAAACAGATCAAATC\n",
       "Gene_ID                                                    FBgn0000022\n",
       "Transcript_ID                                              FBtr0070072\n",
       "Chromosome                                                           X\n",
       "Gene_Region                                                 stop_codon\n",
       "Start                                                           370697\n",
       "Stop                                                            370699\n",
       "Strand                                                               +\n",
       "Reference_Seq        GACATTGTGTCGTTCGTATGTCGCCCATTGAGACCGCCAATGGAGG...\n",
       "upstreamHA           TCCTCAATGTATGCTATAAAAAAAAACTTAAATATAAATTGTCTTC...\n",
       "downstreamHA         AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...\n",
       "positionScore                                                       16\n",
       "PAM_in_start/stop                                                False\n",
       "<15_bp3’_overhang                                                False\n",
       "PAM_in_CDS                                                        True\n",
       "PAM_outside_CDS                                                  False\n",
       "CDS_boundary                                                       <20\n",
       "lastG                                                             16.0\n",
       "cutSite                                                           21.0\n",
       "mutated                                                          False\n",
       "Name: 0, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/5l/nmfpq6250w56y2dw2_l6jh4w0000gp/T/ipykernel_9575/2015968293.py:46: FutureWarning: In a future version, object-dtype columns with all-bool values will not be included in reductions with bool_only=True. Explicitly cast to bool dtype instead.\n",
      "  TFsdfWinnersandMutated.loc[ind] = winnerdf\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>downstreamHA</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>&lt;15_bp3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>lastG</th>\n",
       "      <th>cutSite</th>\n",
       "      <th>mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>...</td>\n",
       "      <td>TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>31.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>370693</td>\n",
       "      <td>370715</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTGTAAAAAAACAGATCAAATC</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>370697</td>\n",
       "      <td>...</td>\n",
       "      <td>AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;20</td>\n",
       "      <td>16.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "2  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "3  370693  370715    X      -  CCTGTAAAAAAACAGATCAAATC  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  \\\n",
       "2   FBtr0070072          X  start_codon  370094  ...   \n",
       "3   FBtr0070072          X   stop_codon  370697  ...   \n",
       "\n",
       "                                        downstreamHA positionScore  \\\n",
       "2  TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...             8   \n",
       "3  AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...            16   \n",
       "\n",
       "  PAM_in_start/stop <15_bp3’_overhang PAM_in_CDS PAM_outside_CDS CDS_boundary  \\\n",
       "2             False              True       True           False          >22   \n",
       "3             False             False       True           False          <20   \n",
       "\n",
       "  lastG cutSite mutated  \n",
       "2  31.0    26.0   False  \n",
       "3  16.0    21.0    True  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting guide RNA for TFsdf row 4\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n",
      "stringency 3\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "fmin                                                            396165\n",
       "fmax                                                            396187\n",
       "#chr                                                                 X\n",
       "strand                                                               -\n",
       "sgRNA_sequence                                 CCGAGTGTGTTAATGAAAAATAA\n",
       "Gene_ID                                                    FBgn0004170\n",
       "Transcript_ID                                              FBtr0070073\n",
       "Chromosome                                                           X\n",
       "Gene_Region                                                start_codon\n",
       "Start                                                           396177\n",
       "Stop                                                            396179\n",
       "Strand                                                               +\n",
       "Reference_Seq        ACCTGCGATAATTTGACATTCTTAGAAACTACCTGAAGAAATAGGA...\n",
       "upstreamHA           TCTGGTCAGTGCCATACCCCTTGGTGTATACTTGCGAGTCTTAATT...\n",
       "downstreamHA         AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...\n",
       "positionScore                                                        8\n",
       "PAM_in_start/stop                                                False\n",
       "<15_bp3’_overhang                                                 True\n",
       "PAM_in_CDS                                                       False\n",
       "PAM_outside_CDS                                                   True\n",
       "CDS_boundary                                                       >22\n",
       "lastG                                                              8.0\n",
       "cutSite                                                           13.0\n",
       "mutated                                                          False\n",
       "Name: 0, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/5l/nmfpq6250w56y2dw2_l6jh4w0000gp/T/ipykernel_9575/2015968293.py:46: FutureWarning: In a future version, object-dtype columns with all-bool values will not be included in reductions with bool_only=True. Explicitly cast to bool dtype instead.\n",
      "  TFsdfWinnersandMutated.loc[ind] = winnerdf\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>downstreamHA</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>&lt;15_bp3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>lastG</th>\n",
       "      <th>cutSite</th>\n",
       "      <th>mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>...</td>\n",
       "      <td>TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>31.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>370693</td>\n",
       "      <td>370715</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTGTAAAAAAACAGATCAAATC</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>370697</td>\n",
       "      <td>...</td>\n",
       "      <td>AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;20</td>\n",
       "      <td>16.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>396165</td>\n",
       "      <td>396187</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGAGTGTGTTAATGAAAAATAA</td>\n",
       "      <td>FBgn0004170</td>\n",
       "      <td>FBtr0070073</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>396177</td>\n",
       "      <td>...</td>\n",
       "      <td>AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>8.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "2  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "3  370693  370715    X      -  CCTGTAAAAAAACAGATCAAATC  FBgn0000022   \n",
       "4  396165  396187    X      -  CCGAGTGTGTTAATGAAAAATAA  FBgn0004170   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  \\\n",
       "2   FBtr0070072          X  start_codon  370094  ...   \n",
       "3   FBtr0070072          X   stop_codon  370697  ...   \n",
       "4   FBtr0070073          X  start_codon  396177  ...   \n",
       "\n",
       "                                        downstreamHA positionScore  \\\n",
       "2  TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...             8   \n",
       "3  AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...            16   \n",
       "4  AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...             8   \n",
       "\n",
       "  PAM_in_start/stop <15_bp3’_overhang PAM_in_CDS PAM_outside_CDS CDS_boundary  \\\n",
       "2             False              True       True           False          >22   \n",
       "3             False             False       True           False          <20   \n",
       "4             False              True      False            True          >22   \n",
       "\n",
       "  lastG cutSite mutated  \n",
       "2  31.0    26.0   False  \n",
       "3  16.0    21.0    True  \n",
       "4   8.0    13.0   False  \n",
       "\n",
       "[3 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting guide RNA for TFsdf row 5\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n",
      "stringency 3\n",
      "No sgRNAs found at all stringencies for 5.\n",
      "Empty DataFrame\n",
      "Columns: [fmin, fmax, #chr, strand, sgRNA_sequence, Gene_ID, Transcript_ID, Chromosome, Gene_Region, Start, Stop, Strand, Reference_Seq, upstreamHA, downstreamHA, positionScore, PAM_in_start/stop, <15_bp3’_overhang, PAM_in_CDS, PAM_outside_CDS, CDS_boundary, lastG, cutSite, mutated]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 24 columns]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>downstreamHA</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>&lt;15_bp3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>lastG</th>\n",
       "      <th>cutSite</th>\n",
       "      <th>mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>...</td>\n",
       "      <td>TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>31.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>370693</td>\n",
       "      <td>370715</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTGTAAAAAAACAGATCAAATC</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>370697</td>\n",
       "      <td>...</td>\n",
       "      <td>AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;20</td>\n",
       "      <td>16.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>396165</td>\n",
       "      <td>396187</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGAGTGTGTTAATGAAAAATAA</td>\n",
       "      <td>FBgn0004170</td>\n",
       "      <td>FBtr0070073</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>396177</td>\n",
       "      <td>...</td>\n",
       "      <td>AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>8.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBgn0004170</td>\n",
       "      <td>FBtr0070073</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>397212</td>\n",
       "      <td>...</td>\n",
       "      <td>TGCTCCTGCCATAGAGAGATGTAGTCCAAAATCTCCTCGTCGTCGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "2  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "3  370693  370715    X      -  CCTGTAAAAAAACAGATCAAATC  FBgn0000022   \n",
       "4  396165  396187    X      -  CCGAGTGTGTTAATGAAAAATAA  FBgn0004170   \n",
       "5     NaN     NaN  NaN    NaN                      NaN  FBgn0004170   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  \\\n",
       "2   FBtr0070072          X  start_codon  370094  ...   \n",
       "3   FBtr0070072          X   stop_codon  370697  ...   \n",
       "4   FBtr0070073          X  start_codon  396177  ...   \n",
       "5   FBtr0070073          X   stop_codon  397212  ...   \n",
       "\n",
       "                                        downstreamHA positionScore  \\\n",
       "2  TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...             8   \n",
       "3  AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...            16   \n",
       "4  AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...             8   \n",
       "5  TGCTCCTGCCATAGAGAGATGTAGTCCAAAATCTCCTCGTCGTCGG...           NaN   \n",
       "\n",
       "  PAM_in_start/stop <15_bp3’_overhang PAM_in_CDS PAM_outside_CDS CDS_boundary  \\\n",
       "2             False              True       True           False          >22   \n",
       "3             False             False       True           False          <20   \n",
       "4             False              True      False            True          >22   \n",
       "5               NaN               NaN        NaN             NaN          NaN   \n",
       "\n",
       "  lastG cutSite mutated  \n",
       "2  31.0    26.0   False  \n",
       "3  16.0    21.0    True  \n",
       "4   8.0    13.0   False  \n",
       "5   NaN     NaN     NaN  \n",
       "\n",
       "[4 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting guide RNA for TFsdf row 6\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "fmin                                                            409794\n",
       "fmax                                                            409816\n",
       "#chr                                                                 X\n",
       "strand                                                               -\n",
       "sgRNA_sequence                                 CCATGACGAGCATTTGCAGCAGC\n",
       "Gene_ID                                                    FBgn0002561\n",
       "Transcript_ID                                              FBtr0070074\n",
       "Chromosome                                                           X\n",
       "Gene_Region                                                start_codon\n",
       "Start                                                           409796\n",
       "Stop                                                            409798\n",
       "Strand                                                               +\n",
       "Reference_Seq        CGACGTCGTCCAGCTTGATGAAAAACTTTGTGGCGGTGTGCGCCAG...\n",
       "upstreamHA           CATGTTATTGTAGTTGAACTTCTTCTTCTGGGAGGACAACATGGGT...\n",
       "downstreamHA         GTAATCCTTGCGAGAGTTTTCTAAGATTTAGTTTACAGATGTTGAC...\n",
       "positionScore                                                       18\n",
       "PAM_in_start/stop                                                 True\n",
       "<15_bp3’_overhang                                                 True\n",
       "PAM_in_CDS                                                       False\n",
       "PAM_outside_CDS                                                  False\n",
       "CDS_boundary                                                       >22\n",
       "lastG                                                             18.0\n",
       "cutSite                                                           23.0\n",
       "mutated                                                          False\n",
       "Name: 0, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>downstreamHA</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>&lt;15_bp3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>lastG</th>\n",
       "      <th>cutSite</th>\n",
       "      <th>mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>...</td>\n",
       "      <td>TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>31.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>370693</td>\n",
       "      <td>370715</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTGTAAAAAAACAGATCAAATC</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>370697</td>\n",
       "      <td>...</td>\n",
       "      <td>AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;20</td>\n",
       "      <td>16.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>396165</td>\n",
       "      <td>396187</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGAGTGTGTTAATGAAAAATAA</td>\n",
       "      <td>FBgn0004170</td>\n",
       "      <td>FBtr0070073</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>396177</td>\n",
       "      <td>...</td>\n",
       "      <td>AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>8.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBgn0004170</td>\n",
       "      <td>FBtr0070073</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>397212</td>\n",
       "      <td>...</td>\n",
       "      <td>TGCTCCTGCCATAGAGAGATGTAGTCCAAAATCTCCTCGTCGTCGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>409794</td>\n",
       "      <td>409816</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATGACGAGCATTTGCAGCAGC</td>\n",
       "      <td>FBgn0002561</td>\n",
       "      <td>FBtr0070074</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>409796</td>\n",
       "      <td>...</td>\n",
       "      <td>GTAATCCTTGCGAGAGTTTTCTAAGATTTAGTTTACAGATGTTGAC...</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>18.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "2  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "3  370693  370715    X      -  CCTGTAAAAAAACAGATCAAATC  FBgn0000022   \n",
       "4  396165  396187    X      -  CCGAGTGTGTTAATGAAAAATAA  FBgn0004170   \n",
       "5     NaN     NaN  NaN    NaN                      NaN  FBgn0004170   \n",
       "6  409794  409816    X      -  CCATGACGAGCATTTGCAGCAGC  FBgn0002561   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  \\\n",
       "2   FBtr0070072          X  start_codon  370094  ...   \n",
       "3   FBtr0070072          X   stop_codon  370697  ...   \n",
       "4   FBtr0070073          X  start_codon  396177  ...   \n",
       "5   FBtr0070073          X   stop_codon  397212  ...   \n",
       "6   FBtr0070074          X  start_codon  409796  ...   \n",
       "\n",
       "                                        downstreamHA positionScore  \\\n",
       "2  TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...             8   \n",
       "3  AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...            16   \n",
       "4  AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...             8   \n",
       "5  TGCTCCTGCCATAGAGAGATGTAGTCCAAAATCTCCTCGTCGTCGG...           NaN   \n",
       "6  GTAATCCTTGCGAGAGTTTTCTAAGATTTAGTTTACAGATGTTGAC...            18   \n",
       "\n",
       "  PAM_in_start/stop <15_bp3’_overhang PAM_in_CDS PAM_outside_CDS CDS_boundary  \\\n",
       "2             False              True       True           False          >22   \n",
       "3             False             False       True           False          <20   \n",
       "4             False              True      False            True          >22   \n",
       "5               NaN               NaN        NaN             NaN          NaN   \n",
       "6              True              True      False           False          >22   \n",
       "\n",
       "  lastG cutSite mutated  \n",
       "2  31.0    26.0   False  \n",
       "3  16.0    21.0    True  \n",
       "4   8.0    13.0   False  \n",
       "5   NaN     NaN     NaN  \n",
       "6  18.0    23.0   False  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting guide RNA for TFsdf row 7\n",
      "stringency 0\n",
      "stringency 1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "fmin                                                            410550\n",
       "fmax                                                            410572\n",
       "#chr                                                                 X\n",
       "strand                                                               +\n",
       "sgRNA_sequence                                 CATCTTGGCAAGAGCAGTGAAGG\n",
       "Gene_ID                                                    FBgn0002561\n",
       "Transcript_ID                                              FBtr0070074\n",
       "Chromosome                                                           X\n",
       "Gene_Region                                                 stop_codon\n",
       "Start                                                           410567\n",
       "Stop                                                            410569\n",
       "Strand                                                               +\n",
       "Reference_Seq        TCCGGCCGAAAATGCTAAGCAAATACACACTTCCTGCAGCAAGGAC...\n",
       "upstreamHA           TTGGCTTAAAAGTAGTTAGTTTAGTGTCTTCCATGCTTATAACAAA...\n",
       "downstreamHA         TGCTCTTGCCAAGATGAAATATAATCGAGTAGCTCCTCGTCATCTG...\n",
       "positionScore                                                        3\n",
       "PAM_in_start/stop                                                False\n",
       "<15_bp3’_overhang                                                 True\n",
       "PAM_in_CDS                                                       False\n",
       "PAM_outside_CDS                                                   True\n",
       "CDS_boundary                                                       <20\n",
       "lastG                                                             26.0\n",
       "cutSite                                                           21.0\n",
       "mutated                                                          False\n",
       "Name: 0, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>downstreamHA</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>&lt;15_bp3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>lastG</th>\n",
       "      <th>cutSite</th>\n",
       "      <th>mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>...</td>\n",
       "      <td>TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>31.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>370693</td>\n",
       "      <td>370715</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTGTAAAAAAACAGATCAAATC</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>370697</td>\n",
       "      <td>...</td>\n",
       "      <td>AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;20</td>\n",
       "      <td>16.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>396165</td>\n",
       "      <td>396187</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGAGTGTGTTAATGAAAAATAA</td>\n",
       "      <td>FBgn0004170</td>\n",
       "      <td>FBtr0070073</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>396177</td>\n",
       "      <td>...</td>\n",
       "      <td>AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>8.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBgn0004170</td>\n",
       "      <td>FBtr0070073</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>397212</td>\n",
       "      <td>...</td>\n",
       "      <td>TGCTCCTGCCATAGAGAGATGTAGTCCAAAATCTCCTCGTCGTCGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>409794</td>\n",
       "      <td>409816</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATGACGAGCATTTGCAGCAGC</td>\n",
       "      <td>FBgn0002561</td>\n",
       "      <td>FBtr0070074</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>409796</td>\n",
       "      <td>...</td>\n",
       "      <td>GTAATCCTTGCGAGAGTTTTCTAAGATTTAGTTTACAGATGTTGAC...</td>\n",
       "      <td>18</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&gt;22</td>\n",
       "      <td>18.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>410550</td>\n",
       "      <td>410572</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CATCTTGGCAAGAGCAGTGAAGG</td>\n",
       "      <td>FBgn0002561</td>\n",
       "      <td>FBtr0070074</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>410567</td>\n",
       "      <td>...</td>\n",
       "      <td>TGCTCTTGCCAAGATGAAATATAATCGAGTAGCTCCTCGTCATCTG...</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>&lt;20</td>\n",
       "      <td>26.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "2  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "3  370693  370715    X      -  CCTGTAAAAAAACAGATCAAATC  FBgn0000022   \n",
       "4  396165  396187    X      -  CCGAGTGTGTTAATGAAAAATAA  FBgn0004170   \n",
       "5     NaN     NaN  NaN    NaN                      NaN  FBgn0004170   \n",
       "6  409794  409816    X      -  CCATGACGAGCATTTGCAGCAGC  FBgn0002561   \n",
       "7  410550  410572    X      +  CATCTTGGCAAGAGCAGTGAAGG  FBgn0002561   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  \\\n",
       "2   FBtr0070072          X  start_codon  370094  ...   \n",
       "3   FBtr0070072          X   stop_codon  370697  ...   \n",
       "4   FBtr0070073          X  start_codon  396177  ...   \n",
       "5   FBtr0070073          X   stop_codon  397212  ...   \n",
       "6   FBtr0070074          X  start_codon  409796  ...   \n",
       "7   FBtr0070074          X   stop_codon  410567  ...   \n",
       "\n",
       "                                        downstreamHA positionScore  \\\n",
       "2  TTAAGAGATAGTATAACGTTATTGTGTGACGATGCTCCTTGCTTCA...             8   \n",
       "3  AGGTCGTCCTGCCAGAGTGATATATAGTCGAGGATGTCCTCATCTT...            16   \n",
       "4  AACACACTCGGAGCTTTCTTTAACTTTCCGGATAACGATCAACAGA...             8   \n",
       "5  TGCTCCTGCCATAGAGAGATGTAGTCCAAAATCTCCTCGTCGTCGG...           NaN   \n",
       "6  GTAATCCTTGCGAGAGTTTTCTAAGATTTAGTTTACAGATGTTGAC...            18   \n",
       "7  TGCTCTTGCCAAGATGAAATATAATCGAGTAGCTCCTCGTCATCTG...             3   \n",
       "\n",
       "  PAM_in_start/stop <15_bp3’_overhang PAM_in_CDS PAM_outside_CDS CDS_boundary  \\\n",
       "2             False              True       True           False          >22   \n",
       "3             False             False       True           False          <20   \n",
       "4             False              True      False            True          >22   \n",
       "5               NaN               NaN        NaN             NaN          NaN   \n",
       "6              True              True      False           False          >22   \n",
       "7             False              True      False            True          <20   \n",
       "\n",
       "  lastG cutSite mutated  \n",
       "2  31.0    26.0   False  \n",
       "3  16.0    21.0    True  \n",
       "4   8.0    13.0   False  \n",
       "5   NaN     NaN     NaN  \n",
       "6  18.0    23.0   False  \n",
       "7  26.0    21.0   False  \n",
       "\n",
       "[6 rows x 24 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting guide RNA for TFsdf row 8\n",
      "stringency 0\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[65], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m workingsgRNArunner()\n",
      "Cell \u001b[0;32mIn[64], line 19\u001b[0m, in \u001b[0;36mworkingsgRNArunner\u001b[0;34m(refSeqPerChromosome, TFsdf)\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[39mfor\u001b[39;00m ind, row \u001b[39min\u001b[39;00m TFsdf\u001b[39m.\u001b[39miterrows():\n\u001b[1;32m     17\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mSelecting guide RNA for TFsdf row \u001b[39m\u001b[39m{\u001b[39;00mind\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[0;32m---> 19\u001b[0m     filtered_sgRNA \u001b[39m=\u001b[39m gRNA_stringencyIterator(row, refSeqPerChromosome)\n\u001b[1;32m     21\u001b[0m     \u001b[39m#Score the sgRNAs for this site\u001b[39;00m\n\u001b[1;32m     22\u001b[0m     sgRNAdf \u001b[39m=\u001b[39m sgRNApositionCheck(filtered_sgRNA)\n",
      "Cell \u001b[0;32mIn[23], line 267\u001b[0m, in \u001b[0;36mgRNA_stringencyIterator\u001b[0;34m(tfSingleRow, refSeqPerChromosome)\u001b[0m\n\u001b[1;32m    265\u001b[0m \u001b[39m#Loop through files and retain the filtered sgRNA df if at least one row is present\u001b[39;00m\n\u001b[1;32m    266\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mlen\u001b[39m(sgRNA) \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[0;32m--> 267\u001b[0m     sgRNA \u001b[39m=\u001b[39m filter_gRNA(gRNAfiles[file_ind], tfSingleRow, refSeqPerChromosome)\n\u001b[1;32m    268\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mstringency \u001b[39m\u001b[39m{\u001b[39;00mfile_ind\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    269\u001b[0m     file_ind \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m\n",
      "Cell \u001b[0;32mIn[23], line 215\u001b[0m, in \u001b[0;36mfilter_gRNA\u001b[0;34m(gRNA_file, tfSingleRow, refSeqPerChromosome)\u001b[0m\n\u001b[1;32m    212\u001b[0m \u001b[39mfor\u001b[39;00m attribute \u001b[39min\u001b[39;00m gRNAFileAnnotation[\u001b[39m'\u001b[39m\u001b[39mattributes\u001b[39m\u001b[39m'\u001b[39m]:\n\u001b[1;32m    214\u001b[0m     fullatt \u001b[39m=\u001b[39m (attribute)\u001b[39m.\u001b[39msplit(\u001b[39m\"\u001b[39m\u001b[39m;\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 215\u001b[0m     gRNAFileAnnotation\u001b[39m.\u001b[39;49mat[index,\u001b[39m\"\u001b[39;49m\u001b[39mtarget_site_variation\u001b[39;49m\u001b[39m\"\u001b[39;49m] \u001b[39m=\u001b[39m fullatt[\u001b[39m8\u001b[39m]\n\u001b[1;32m    216\u001b[0m     index\u001b[39m+\u001b[39m\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m\n\u001b[1;32m    218\u001b[0m \u001b[39m# shorten file to essential information\u001b[39;00m\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/indexing.py:2442\u001b[0m, in \u001b[0;36m_AtIndexer.__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   2439\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mobj\u001b[39m.\u001b[39mloc[key] \u001b[39m=\u001b[39m value\n\u001b[1;32m   2440\u001b[0m     \u001b[39mreturn\u001b[39;00m\n\u001b[0;32m-> 2442\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49m\u001b[39m__setitem__\u001b[39;49m(key, value)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/indexing.py:2397\u001b[0m, in \u001b[0;36m_ScalarAccessIndexer.__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   2394\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(key) \u001b[39m!=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mndim:\n\u001b[1;32m   2395\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mNot enough indexers for scalar access (setting)!\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m-> 2397\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mobj\u001b[39m.\u001b[39;49m_set_value(\u001b[39m*\u001b[39;49mkey, value\u001b[39m=\u001b[39;49mvalue, takeable\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_takeable)\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/frame.py:4211\u001b[0m, in \u001b[0;36mDataFrame._set_value\u001b[0;34m(self, index, col, value, takeable)\u001b[0m\n\u001b[1;32m   4209\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   4210\u001b[0m     icol \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mget_loc(col)\n\u001b[0;32m-> 4211\u001b[0m     iindex \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mindex\u001b[39m.\u001b[39;49mget_loc(index)\n\u001b[1;32m   4212\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_mgr\u001b[39m.\u001b[39mcolumn_setitem(icol, iindex, value, inplace\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[1;32m   4213\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_clear_item_cache()\n",
      "File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pandas/core/indexes/range.py:391\u001b[0m, in \u001b[0;36mRangeIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m    389\u001b[0m new_key \u001b[39m=\u001b[39m \u001b[39mint\u001b[39m(key)\n\u001b[1;32m    390\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 391\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_range\u001b[39m.\u001b[39mindex(new_key)\n\u001b[1;32m    392\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mValueError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m    393\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "TFsdfWinnersandMutated = workingsgRNArunner()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
