{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "window = 21"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CAT'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test function to form the reverse complement of a DNA sequence\n",
    "from sgRNAutils import revComp\n",
    "revComp('ATG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TGCTAGAGGG')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test function to extract the DNA sequence at particular genomic coordinates of a particular chromosome\n",
    "# to confirm look up the genome coordinates and the chromosome for drosophila melanogaster dm6 on the UCSC genome browser https://genome.ucsc.edu/cgi-bin/hgTracks?db=dm6\n",
    "# below the link to to verify the results of the query below\n",
    "# https://genome.ucsc.edu/cgi-bin/hgTracks?db=dm6&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chrX%3A20000080%2D20000090&hgsid=1877457346_JpaxsFWfPkr6i1rbo2jG4jhD73Fb\n",
    "from sgRNAutils import refSeq\n",
    "refSeqPerChromosome= refSeq()\n",
    "refSeqPerChromosome['X'][20000080:20000090]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>370697</td>\n",
       "      <td>370699</td>\n",
       "      <td>+</td>\n",
       "      <td>GACATTGTGTCGTTCGTATGTCGCCCATTGAGACCGCCAATGGAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>539954</td>\n",
       "      <td>539956</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0310027</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>491316</td>\n",
       "      <td>491318</td>\n",
       "      <td>+</td>\n",
       "      <td>CACTGATAAAAAATACCCTAAATTATCGCAACGCAAATGGTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0310027</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>539954</td>\n",
       "      <td>539956</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0329895</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0329895</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>540017</td>\n",
       "      <td>540019</td>\n",
       "      <td>+</td>\n",
       "      <td>TTTTTTAATGTTCTGCACTTGCTTAATTACCAGCATGTACCATCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0333012</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>525418</td>\n",
       "      <td>525420</td>\n",
       "      <td>+</td>\n",
       "      <td>ATCATTGGCCAAAAAAGTAGGAATTCTGGGAGGAGGGCATGCCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0333012</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>539954</td>\n",
       "      <td>539956</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0333013</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>491316</td>\n",
       "      <td>491318</td>\n",
       "      <td>+</td>\n",
       "      <td>CACTGATAAAAAATACCCTAAATTATCGCAACGCAAATGGTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0333013</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>539954</td>\n",
       "      <td>539956</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>9791982</td>\n",
       "      <td>9791984</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0331285</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9792792</td>\n",
       "      <td>9792794</td>\n",
       "      <td>-</td>\n",
       "      <td>TAAAATGCGGTGACCTGCAAAAGGAGCCATAAACTAAGCGTCCTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0331285</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>9791982</td>\n",
       "      <td>9791984</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0331286</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9792792</td>\n",
       "      <td>9792794</td>\n",
       "      <td>-</td>\n",
       "      <td>TAAAATGCGGTGACCTGCAAAAGGAGCCATAAACTAAGCGTCCTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0331286</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>9791982</td>\n",
       "      <td>9791984</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>26005306</td>\n",
       "      <td>26005308</td>\n",
       "      <td>-</td>\n",
       "      <td>TCCCCAAGCGAACCCAAAACCCCCAACCGAACTGCAGCTGATCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>30852549</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>30854024</td>\n",
       "      <td>30854026</td>\n",
       "      <td>+</td>\n",
       "      <td>ACAGAATGATGAGGATTAGGCCAACACACACATCCATCGAAAGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11279461</td>\n",
       "      <td>11279463</td>\n",
       "      <td>+</td>\n",
       "      <td>TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091680</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278113</td>\n",
       "      <td>11278115</td>\n",
       "      <td>+</td>\n",
       "      <td>CTGGCGATCCACAGACTCGTCCTCCACATCGCACAATTCATTTGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091680</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11279461</td>\n",
       "      <td>11279463</td>\n",
       "      <td>+</td>\n",
       "      <td>TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091681</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278370</td>\n",
       "      <td>11278372</td>\n",
       "      <td>+</td>\n",
       "      <td>CTGATCAAAATAAAAGTATGAAGTACAAATTACAGCAATGAGATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091681</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11279461</td>\n",
       "      <td>11279463</td>\n",
       "      <td>+</td>\n",
       "      <td>TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091682</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278113</td>\n",
       "      <td>11278115</td>\n",
       "      <td>+</td>\n",
       "      <td>CTGGCGATCCACAGACTCGTCCTCCACATCGCACAATTCATTTGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091682</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11279461</td>\n",
       "      <td>11279463</td>\n",
       "      <td>+</td>\n",
       "      <td>TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>8610886</td>\n",
       "      <td>8610888</td>\n",
       "      <td>+</td>\n",
       "      <td>AATATTTTTCTTTTAAAATGAAACTTTCCCAAACTTTTAGTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0339354</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0339354</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>8610886</td>\n",
       "      <td>8610888</td>\n",
       "      <td>+</td>\n",
       "      <td>AATATTTTTCTTTTAAAATGAAACTTTCCCAAACTTTTAGTTTTTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene_ID Transcript_ID Chromosome  Gene_Region     Start      Stop  \\\n",
       "0   FBgn0000022   FBtr0070072          X  start_codon    370094    370096   \n",
       "1   FBgn0000022   FBtr0070072          X   stop_codon    370697    370699   \n",
       "2   FBgn0003963   FBtr0078063         2L  start_codon    524038    524040   \n",
       "3   FBgn0003963   FBtr0078063         2L   stop_codon    539954    539956   \n",
       "4   FBgn0003963   FBtr0310027         2L  start_codon    491316    491318   \n",
       "5   FBgn0003963   FBtr0310027         2L   stop_codon    539954    539956   \n",
       "6   FBgn0003963   FBtr0329895         2L  start_codon    524038    524040   \n",
       "7   FBgn0003963   FBtr0329895         2L   stop_codon    540017    540019   \n",
       "8   FBgn0003963   FBtr0333012         2L  start_codon    525418    525420   \n",
       "9   FBgn0003963   FBtr0333012         2L   stop_codon    539954    539956   \n",
       "10  FBgn0003963   FBtr0333013         2L  start_codon    491316    491318   \n",
       "11  FBgn0003963   FBtr0333013         2L   stop_codon    539954    539956   \n",
       "12  FBgn0032150   FBtr0079917         2L  start_codon   9793004   9793006   \n",
       "13  FBgn0032150   FBtr0079917         2L   stop_codon   9791982   9791984   \n",
       "14  FBgn0032150   FBtr0331285         2L  start_codon   9792792   9792794   \n",
       "15  FBgn0032150   FBtr0331285         2L   stop_codon   9791982   9791984   \n",
       "16  FBgn0032150   FBtr0331286         2L  start_codon   9792792   9792794   \n",
       "17  FBgn0032150   FBtr0331286         2L   stop_codon   9791982   9791984   \n",
       "18  FBgn0002733   FBtr0084982         3R  start_codon  26005891  26005893   \n",
       "19  FBgn0002733   FBtr0084982         3R   stop_codon  26005306  26005308   \n",
       "20  FBgn0003720   FBtr0085709         3R  start_codon  30852547  30852549   \n",
       "21  FBgn0003720   FBtr0085709         3R   stop_codon  30854024  30854026   \n",
       "22  FBgn0052831   FBtr0091683         2L  start_codon  11278451  11278453   \n",
       "23  FBgn0052831   FBtr0091683         2L   stop_codon  11279461  11279463   \n",
       "24  FBgn0052831   FBtr0091680         2L  start_codon  11278113  11278115   \n",
       "25  FBgn0052831   FBtr0091680         2L   stop_codon  11279461  11279463   \n",
       "26  FBgn0052831   FBtr0091681         2L  start_codon  11278370  11278372   \n",
       "27  FBgn0052831   FBtr0091681         2L   stop_codon  11279461  11279463   \n",
       "28  FBgn0052831   FBtr0091682         2L  start_codon  11278113  11278115   \n",
       "29  FBgn0052831   FBtr0091682         2L   stop_codon  11279461  11279463   \n",
       "30  FBgn0267792   FBtr0088692         2R  start_codon   8600582   8600584   \n",
       "31  FBgn0267792   FBtr0088692         2R   stop_codon   8610886   8610888   \n",
       "32  FBgn0267792   FBtr0339354         2R  start_codon   8600582   8600584   \n",
       "33  FBgn0267792   FBtr0339354         2R   stop_codon   8610886   8610888   \n",
       "\n",
       "   Strand                                      Reference_Seq  \n",
       "0       +  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...  \n",
       "1       +  GACATTGTGTCGTTCGTATGTCGCCCATTGAGACCGCCAATGGAGG...  \n",
       "2       +  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...  \n",
       "3       +  CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...  \n",
       "4       +  CACTGATAAAAAATACCCTAAATTATCGCAACGCAAATGGTAAAGT...  \n",
       "5       +  CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...  \n",
       "6       +  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...  \n",
       "7       +  TTTTTTAATGTTCTGCACTTGCTTAATTACCAGCATGTACCATCTG...  \n",
       "8       +  ATCATTGGCCAAAAAAGTAGGAATTCTGGGAGGAGGGCATGCCAAA...  \n",
       "9       +  CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...  \n",
       "10      +  CACTGATAAAAAATACCCTAAATTATCGCAACGCAAATGGTAAAGT...  \n",
       "11      +  CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...  \n",
       "12      -  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...  \n",
       "13      -  AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...  \n",
       "14      -  TAAAATGCGGTGACCTGCAAAAGGAGCCATAAACTAAGCGTCCTAC...  \n",
       "15      -  AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...  \n",
       "16      -  TAAAATGCGGTGACCTGCAAAAGGAGCCATAAACTAAGCGTCCTAC...  \n",
       "17      -  AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...  \n",
       "18      -  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...  \n",
       "19      -  TCCCCAAGCGAACCCAAAACCCCCAACCGAACTGCAGCTGATCCCC...  \n",
       "20      +  TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...  \n",
       "21      +  ACAGAATGATGAGGATTAGGCCAACACACACATCCATCGAAAGGAC...  \n",
       "22      +  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...  \n",
       "23      +  TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...  \n",
       "24      +  CTGGCGATCCACAGACTCGTCCTCCACATCGCACAATTCATTTGTC...  \n",
       "25      +  TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...  \n",
       "26      +  CTGATCAAAATAAAAGTATGAAGTACAAATTACAGCAATGAGATTG...  \n",
       "27      +  TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...  \n",
       "28      +  CTGGCGATCCACAGACTCGTCCTCCACATCGCACAATTCATTTGTC...  \n",
       "29      +  TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...  \n",
       "30      +  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...  \n",
       "31      +  AATATTTTTCTTTTAAAATGAAACTTTCCCAAACTTTTAGTTTTTT...  \n",
       "32      +  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...  \n",
       "33      +  AATATTTTTCTTTTAAAATGAAACTTTCCCAAACTTTTAGTTTTTT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test function to identify known transcripts of transcription factors and extract their genomic information\n",
    "# this cell takes about 60s to run but you can avoid it by running the next cell and loading the output from this cell from MockMaterials\n",
    "# to confirm the results look up the Transcript_IDs below on the UCSC genome browser and double check the genomic information: https://genome.ucsc.edu/cgi-bin/hgTracks?db=dm6\n",
    "# here an example for FBtr0070072: https://genome.ucsc.edu/cgi-bin/hgTracks?db=dm6&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chrX%3A370031%2D370947&hgsid=1877489778_pLlolcHmOnq6afrqpZAa7WNNrw7r\n",
    "# you can also check if the different transcripts belong to the same gene in the UCSC genome browser\n",
    "from sgRNAutils import make_dataframe_from_TFs_list\n",
    "TFsdf = make_dataframe_from_TFs_list('inputfiles/mockMaterials/TFsTruncatedLong.xlsx', refSeqPerChromosome)\n",
    "display(TFsdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG 225 225\n",
      "ACGTTATACTATCTCTTAAAATGGCTTTGGGCAGCGAAAATCA\n",
      "ACGTTATACTATCTCTTAAAXXXGCTTTGGGCAGCGAAAATCA\n",
      "TAA 225 225\n",
      "CACTCTGGCAGGACGACCTGTAAAAAAACAGATCAAATCTTCA\n",
      "CACTCTGGCAGGACGACCTGXXXAAAAACAGATCAAATCTTCA\n",
      "ATG 225 225\n",
      "ATACACATATAGGGAAAACAATGCTTAGCAGCAATACAAGAGG\n",
      "ATACACATATAGGGAAAACAXXXCTTAGCAGCAATACAAGAGG\n",
      "TGA 225 225\n",
      "ACCTGCAGGAGGCGGCCATTTGAGAAAGCCAGCTGCGGTGGGA\n",
      "ACCTGCAGGAGGCGGCCATTXXXGAAAGCCAGCTGCGGTGGGA\n",
      "ATG 225 225\n",
      "GTCCAGCAGCTGGACTCGAAATGTCACGACGCAAACAGTCCAA\n",
      "GTCCAGCAGCTGGACTCGAAXXXTCACGACGCAAACAGTCCAA\n",
      "TGA 225 225\n",
      "ACCTGCAGGAGGCGGCCATTTGAGAAAGCCAGCTGCGGTGGGA\n",
      "ACCTGCAGGAGGCGGCCATTXXXGAAAGCCAGCTGCGGTGGGA\n",
      "ATG 225 225\n",
      "ATACACATATAGGGAAAACAATGCTTAGCAGCAATACAAGAGG\n",
      "ATACACATATAGGGAAAACAXXXCTTAGCAGCAATACAAGAGG\n",
      "TAA 225 225\n",
      "AGTGGTATTACACGTGCTTCTAAGCAGCAGCCCAACCCAATCG\n",
      "AGTGGTATTACACGTGCTTCXXXGCAGCAGCCCAACCCAATCG\n",
      "ATG 225 225\n",
      "GTTCCGATACCGCAGAAGAGATGACCGTCGACAGCAGAGGTAA\n",
      "GTTCCGATACCGCAGAAGAGXXXACCGTCGACAGCAGAGGTAA\n",
      "TGA 225 225\n",
      "ACCTGCAGGAGGCGGCCATTTGAGAAAGCCAGCTGCGGTGGGA\n",
      "ACCTGCAGGAGGCGGCCATTXXXGAAAGCCAGCTGCGGTGGGA\n",
      "ATG 225 225\n",
      "GTCCAGCAGCTGGACTCGAAATGTCACGACGCAAACAGTCCAA\n",
      "GTCCAGCAGCTGGACTCGAAXXXTCACGACGCAAACAGTCCAA\n",
      "TGA 225 225\n",
      "ACCTGCAGGAGGCGGCCATTTGAGAAAGCCAGCTGCGGTGGGA\n",
      "ACCTGCAGGAGGCGGCCATTXXXGAAAGCCAGCTGCGGTGGGA\n",
      "CAT 225 225\n",
      "CGGCACAGTTTACGCAGGTCCATGGGTTTTATTATTTAATTAA\n",
      "CGGCACAGTTTACGCAGGTCXXXGGGTTTTATTATTTAATTAA\n",
      "TTA 225 225\n",
      "ACTATGTAATATAATCTTTATTATGAGCGCGGATTATCCCGAT\n",
      "ACTATGTAATATAATCTTTAXXXTGAGCGCGGATTATCCCGAT\n",
      "CAT 225 225\n",
      "TCCAAGCAATCATTACAGATCATGGGCAGCTCCTCAGTAAGAT\n",
      "TCCAAGCAATCATTACAGATXXXGGGCAGCTCCTCAGTAAGAT\n",
      "TTA 225 225\n",
      "ACTATGTAATATAATCTTTATTATGAGCGCGGATTATCCCGAT\n",
      "ACTATGTAATATAATCTTTAXXXTGAGCGCGGATTATCCCGAT\n",
      "CAT 225 225\n",
      "TCCAAGCAATCATTACAGATCATGGGCAGCTCCTCAGTAAGAT\n",
      "TCCAAGCAATCATTACAGATXXXGGGCAGCTCCTCAGTAAGAT\n",
      "TTA 225 225\n",
      "ACTATGTAATATAATCTTTATTATGAGCGCGGATTATCCCGAT\n",
      "ACTATGTAATATAATCTTTAXXXTGAGCGCGGATTATCCCGAT\n",
      "CAT 225 225\n",
      "GACATCTCCATTTCCAGAACCATTTTGTTGTAGTTTGGTTTTG\n",
      "GACATCTCCATTTCCAGAACXXXTTTGTTGTAGTTTGGTTTTG\n",
      "TCA 225 225\n",
      "ATCATCCGGAGTCCAGTTGATCACCAGGGACGCCACATGGGGC\n",
      "ATCATCCGGAGTCCAGTTGAXXXCCAGGGACGCCACATGGGGC\n",
      "ATG 225 225\n",
      "AAATTACGCCCCACATCGGTATGCAGTCGTCGGAGGGTTCACC\n",
      "AAATTACGCCCCACATCGGTXXXCAGTCGTCGGAGGGTTCACC\n",
      "TGA 225 225\n",
      "TGTACAGTCAGCGCAAGATCTGAAAAGTATGTAGAGCCTAGAC\n",
      "TGTACAGTCAGCGCAAGATCXXXAAAGTATGTAGAGCCTAGAC\n",
      "ATG 225 225\n",
      "GCAGACTGGGCGATCTAACCATGCAGCTGCATCCCAATGCGGA\n",
      "GCAGACTGGGCGATCTAACCXXXCAGCTGCATCCCAATGCGGA\n",
      "TGA 225 225\n",
      "TCAGCGAGGAGGTGGCCGGGTGAGGTTCGAGCCCTCGGAACGG\n",
      "TCAGCGAGGAGGTGGCCGGGXXXGGTTCGAGCCCTCGGAACGG\n",
      "ATG 225 225\n",
      "GCGCCCCGCGAAAAAAGGCCATGAACTCGGCAATCAAGGTAGG\n",
      "GCGCCCCGCGAAAAAAGGCCXXXAACTCGGCAATCAAGGTAGG\n",
      "TGA 225 225\n",
      "TCAGCGAGGAGGTGGCCGGGTGAGGTTCGAGCCCTCGGAACGG\n",
      "TCAGCGAGGAGGTGGCCGGGXXXGGTTCGAGCCCTCGGAACGG\n",
      "ATG 225 225\n",
      "TAACAACAAACTGACTCCGGATGCAAATCCGCATGCACTTGCA\n",
      "TAACAACAAACTGACTCCGGXXXCAAATCCGCATGCACTTGCA\n",
      "TGA 225 225\n",
      "TCAGCGAGGAGGTGGCCGGGTGAGGTTCGAGCCCTCGGAACGG\n",
      "TCAGCGAGGAGGTGGCCGGGXXXGGTTCGAGCCCTCGGAACGG\n",
      "ATG 225 225\n",
      "GCGCCCCGCGAAAAAAGGCCATGAACTCGGCAATCAAGGTAGG\n",
      "GCGCCCCGCGAAAAAAGGCCXXXAACTCGGCAATCAAGGTAGG\n",
      "TGA 225 225\n",
      "TCAGCGAGGAGGTGGCCGGGTGAGGTTCGAGCCCTCGGAACGG\n",
      "TCAGCGAGGAGGTGGCCGGGXXXGGTTCGAGCCCTCGGAACGG\n",
      "ATG 225 225\n",
      "TCCAGTAAAGCGCAGCAGCAATGATGCCGCTGCAGGAGTCGCC\n",
      "TCCAGTAAAGCGCAGCAGCAXXXATGCCGCTGCAGGAGTCGCC\n",
      "TAG 225 225\n",
      "TGGGGCAAGCGGGCGGCAAGTAGAACCTTCAGCCACTCGAAGT\n",
      "TGGGGCAAGCGGGCGGCAAGXXXAACCTTCAGCCACTCGAAGT\n",
      "ATG 225 225\n",
      "TCCAGTAAAGCGCAGCAGCAATGATGCCGCTGCAGGAGTCGCC\n",
      "TCCAGTAAAGCGCAGCAGCAXXXATGCCGCTGCAGGAGTCGCC\n",
      "TAG 225 225\n",
      "TGGGGCAAGCGGGCGGCAAGTAGAACCTTCAGCCACTCGAAGT\n",
      "TGGGGCAAGCGGGCGGCAAGXXXAACCTTCAGCCACTCGAAGT\n"
     ]
    }
   ],
   "source": [
    "# test function to extract 225 bp of the genome left and right of the start/stop codon of a transcript as homology arms (HA)\n",
    "# to check whether the identified HAs are correct, you can align them to the dm6 drosophila melanogaster genome using the \n",
    "# UCSC BLAT Search: https://genome.ucsc.edu/cgi-bin/hgBlat?hgsid=1885485482_QUkcgOVglCJ2ybPDnj4fC1f6KdNR&command=start\n",
    "# here as an example the HA within window of FBtr0070072 aligned: https://genome.ucsc.edu/cgi-bin/hgTracks?db=dm6&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chrX%3A370074%2D370116&hgsid=1885485482_QUkcgOVglCJ2ybPDnj4fC1f6KdNR\n",
    "from sgRNAutils import make_homology_arm_fragments\n",
    "import pandas as pd\n",
    "TFsdf = pd.read_excel('inputfiles/mockMaterials/TFsdfTruncatedLong.xlsx')\n",
    "TFsdf = make_homology_arm_fragments(TFsdf, refSeqPerChromosome)\n",
    "TFsdf = TFsdf.reset_index()\n",
    "del TFsdf[\"index\"]\n",
    "\n",
    "# for every transcript in the dataframe above ...\n",
    "for index, row in TFsdf.iterrows():\n",
    "    # ... check that the HAs are 225 bp long and that the sequence between the HAs is a start or a stop codon in forward or reverse complement\n",
    "    # stop codons are TAA, TAG and TGA and their reverse complements are TTA, CTA and TCA\n",
    "    # start codons are ATG and its reverse complement is CAT\n",
    "    print(refSeqPerChromosome[row[\"Chromosome\"]][row['Start']-1:row['Stop']], len(TFsdf.iloc[0][\"HAL\"]), len(TFsdf.iloc[0][\"HAR\"]))\n",
    "    # ... compare the region around the start/stop codon from the genome file to the end of HAL and the beginning of HAR that this function extracts\n",
    "    # both sequences should be equal except for the \"XXX\" as placeholder for the start/stop codon in the HA sequence\n",
    "    print(refSeqPerChromosome[row[\"Chromosome\"]][row['Start']-21:row['Stop']+20])\n",
    "    print(TFsdf.iloc[index][\"HAL\"][-20:] + \"XXX\" + TFsdf.iloc[index][\"HAR\"][:20])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DO NOT RUN THE NEXT CELL without reading this!\n",
    "The next cell might take about 30 min to run. You can avoid it by running the cell after the next cell and loading the output from this cell from MockMaterials."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stringency 0\n",
      "stringency 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0070072          X  start_codon  370094  370096      +   \n",
       "\n",
       "                                       Reference_Seq  \n",
       "0  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[6], line 7\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mos\u001b[39;00m\n\u001b[0;32m      6\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m index, row \u001b[38;5;129;01min\u001b[39;00m TFsdf\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[1;32m----> 7\u001b[0m     result_df \u001b[38;5;241m=\u001b[39m gRNA_stringencyIterator(row, window, refSeqPerChromosome)\n\u001b[0;32m      8\u001b[0m     display(result_df)\n\u001b[0;32m      9\u001b[0m     \u001b[38;5;66;03m# test whether every sgRNA identified is within the window of 21 bp around the start/stop codon\u001b[39;00m\n\u001b[0;32m     10\u001b[0m     \u001b[38;5;66;03m# you can manually do this by comparing the genomic coordinates of the sgRNA (fmin and fmax) with the coordinates of the start/stop codon (Start and Stop)\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:225\u001b[0m, in \u001b[0;36mgRNA_stringencyIterator\u001b[1;34m(tfSingleRow, window, refSeqPerChromosome)\u001b[0m\n\u001b[0;32m    223\u001b[0m \u001b[38;5;66;03m#Loop through files and retain the filtered sgRNA df if at least one row is present\u001b[39;00m\n\u001b[0;32m    224\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m sgRNA\u001b[38;5;241m.\u001b[39mempty \u001b[38;5;129;01mand\u001b[39;00m file_ind \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m:\n\u001b[1;32m--> 225\u001b[0m     sgRNA \u001b[38;5;241m=\u001b[39m filter_gRNA(gRNAfiles[file_ind], window, tfSingleRow, refSeqPerChromosome)\n\u001b[0;32m    226\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstringency \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfile_ind\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    227\u001b[0m     file_ind \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m\n",
      "File \u001b[1;32mc:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:169\u001b[0m, in \u001b[0;36mfilter_gRNA\u001b[1;34m(gRNA_file, window, tfSingleRow, refSeqPerChromosome)\u001b[0m\n\u001b[0;32m    167\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m index, attribute \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(gRNAFileAnnotation[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mattributes\u001b[39m\u001b[38;5;124m'\u001b[39m]):\n\u001b[0;32m    168\u001b[0m     fullatt \u001b[38;5;241m=\u001b[39m (attribute)\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m;\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m--> 169\u001b[0m     gRNAFileAnnotation\u001b[38;5;241m.\u001b[39mat[index,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtarget_site_variation\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m fullatt[\u001b[38;5;241m8\u001b[39m]\n\u001b[0;32m    171\u001b[0m \u001b[38;5;66;03m# shorten file to essential information\u001b[39;00m\n\u001b[0;32m    172\u001b[0m GenomeCoordinates \u001b[38;5;241m=\u001b[39m gRNAFileAnnotation[[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtarget_site_variation\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfmin\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfmax\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m#chr\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstrand\u001b[39m\u001b[38;5;124m\"\u001b[39m]]\n",
      "File \u001b[1;32mc:\\ProgramData\\anaconda3\\envs\\AutoTagsEnv\\Lib\\site-packages\\pandas\\core\\indexing.py:2499\u001b[0m, in \u001b[0;36m_AtIndexer.__setitem__\u001b[1;34m(self, key, value)\u001b[0m\n\u001b[0;32m   2496\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39mloc[key] \u001b[38;5;241m=\u001b[39m value\n\u001b[0;32m   2497\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[1;32m-> 2499\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__setitem__\u001b[39m(key, value)\n",
      "File \u001b[1;32mc:\\ProgramData\\anaconda3\\envs\\AutoTagsEnv\\Lib\\site-packages\\pandas\\core\\indexing.py:2455\u001b[0m, in \u001b[0;36m_ScalarAccessIndexer.__setitem__\u001b[1;34m(self, key, value)\u001b[0m\n\u001b[0;32m   2452\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(key) \u001b[38;5;241m!=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mndim:\n\u001b[0;32m   2453\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNot enough indexers for scalar access (setting)!\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m-> 2455\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_set_value(\u001b[38;5;241m*\u001b[39mkey, value\u001b[38;5;241m=\u001b[39mvalue, takeable\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_takeable)\n",
      "File \u001b[1;32mc:\\ProgramData\\anaconda3\\envs\\AutoTagsEnv\\Lib\\site-packages\\pandas\\core\\frame.py:4339\u001b[0m, in \u001b[0;36mDataFrame._set_value\u001b[1;34m(self, index, col, value, takeable)\u001b[0m\n\u001b[0;32m   4337\u001b[0m         icol \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mget_loc(col)\n\u001b[0;32m   4338\u001b[0m         iindex \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindex\u001b[38;5;241m.\u001b[39mget_loc(index)\n\u001b[1;32m-> 4339\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mgr\u001b[38;5;241m.\u001b[39mcolumn_setitem(icol, iindex, value, inplace_only\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m   4340\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_clear_item_cache()\n\u001b[0;32m   4342\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mKeyError\u001b[39;00m, \u001b[38;5;167;01mTypeError\u001b[39;00m, \u001b[38;5;167;01mValueError\u001b[39;00m, LossySetitemError):\n\u001b[0;32m   4343\u001b[0m     \u001b[38;5;66;03m# get_loc might raise a KeyError for missing labels (falling back\u001b[39;00m\n\u001b[0;32m   4344\u001b[0m     \u001b[38;5;66;03m#  to (i)loc will do expansion of the index)\u001b[39;00m\n\u001b[0;32m   4345\u001b[0m     \u001b[38;5;66;03m# column_setitem will do validation that may raise TypeError,\u001b[39;00m\n\u001b[0;32m   4346\u001b[0m     \u001b[38;5;66;03m#  ValueError, or LossySetitemError\u001b[39;00m\n\u001b[0;32m   4347\u001b[0m     \u001b[38;5;66;03m# set using a non-recursive method & reset the cache\u001b[39;00m\n",
      "File \u001b[1;32mc:\\ProgramData\\anaconda3\\envs\\AutoTagsEnv\\Lib\\site-packages\\pandas\\core\\internals\\managers.py:1304\u001b[0m, in \u001b[0;36mBlockManager.column_setitem\u001b[1;34m(self, loc, idx, value, inplace_only)\u001b[0m\n\u001b[0;32m   1300\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_iset_split_block(blkno, [blk_loc], values)\n\u001b[0;32m   1302\u001b[0m \u001b[38;5;66;03m# this manager is only created temporarily to mutate the values in place\u001b[39;00m\n\u001b[0;32m   1303\u001b[0m \u001b[38;5;66;03m# so don't track references, otherwise the `setitem` would perform CoW again\u001b[39;00m\n\u001b[1;32m-> 1304\u001b[0m col_mgr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39miget(loc, track_ref\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m   1305\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m inplace_only:\n\u001b[0;32m   1306\u001b[0m     col_mgr\u001b[38;5;241m.\u001b[39msetitem_inplace(idx, value)\n",
      "File \u001b[1;32mc:\\ProgramData\\anaconda3\\envs\\AutoTagsEnv\\Lib\\site-packages\\pandas\\core\\internals\\managers.py:996\u001b[0m, in \u001b[0;36mBlockManager.iget\u001b[1;34m(self, i, track_ref)\u001b[0m\n\u001b[0;32m    992\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    993\u001b[0m \u001b[38;5;124;03mReturn the data as a SingleBlockManager.\u001b[39;00m\n\u001b[0;32m    994\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    995\u001b[0m block \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mblocks[\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mblknos[i]]\n\u001b[1;32m--> 996\u001b[0m values \u001b[38;5;241m=\u001b[39m block\u001b[38;5;241m.\u001b[39miget(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mblklocs[i])\n\u001b[0;32m    998\u001b[0m \u001b[38;5;66;03m# shortcut for select a single-dim from a 2-dim BM\u001b[39;00m\n\u001b[0;32m    999\u001b[0m bp \u001b[38;5;241m=\u001b[39m BlockPlacement(\u001b[38;5;28mslice\u001b[39m(\u001b[38;5;241m0\u001b[39m, \u001b[38;5;28mlen\u001b[39m(values)))\n",
      "File \u001b[1;32mc:\\ProgramData\\anaconda3\\envs\\AutoTagsEnv\\Lib\\site-packages\\pandas\\core\\internals\\blocks.py:1007\u001b[0m, in \u001b[0;36mBlock.iget\u001b[1;34m(self, i)\u001b[0m\n\u001b[0;32m   1003\u001b[0m \u001b[38;5;129m@property\u001b[39m\n\u001b[0;32m   1004\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mshape\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Shape:\n\u001b[0;32m   1005\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalues\u001b[38;5;241m.\u001b[39mshape\n\u001b[1;32m-> 1007\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21miget\u001b[39m(\u001b[38;5;28mself\u001b[39m, i: \u001b[38;5;28mint\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28mtuple\u001b[39m[\u001b[38;5;28mint\u001b[39m, \u001b[38;5;28mint\u001b[39m] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28mtuple\u001b[39m[\u001b[38;5;28mslice\u001b[39m, \u001b[38;5;28mint\u001b[39m]) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m np\u001b[38;5;241m.\u001b[39mndarray:\n\u001b[0;32m   1008\u001b[0m     \u001b[38;5;66;03m# In the case where we have a tuple[slice, int], the slice will always\u001b[39;00m\n\u001b[0;32m   1009\u001b[0m     \u001b[38;5;66;03m#  be slice(None)\u001b[39;00m\n\u001b[0;32m   1010\u001b[0m     \u001b[38;5;66;03m# Note: only reached with self.ndim == 2\u001b[39;00m\n\u001b[0;32m   1011\u001b[0m     \u001b[38;5;66;03m# Invalid index type \"Union[int, Tuple[int, int], Tuple[slice, int]]\"\u001b[39;00m\n\u001b[0;32m   1012\u001b[0m     \u001b[38;5;66;03m# for \"Union[ndarray[Any, Any], ExtensionArray]\"; expected type\u001b[39;00m\n\u001b[0;32m   1013\u001b[0m     \u001b[38;5;66;03m# \"Union[int, integer[Any]]\"\u001b[39;00m\n\u001b[0;32m   1014\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvalues[i]  \u001b[38;5;66;03m# type: ignore[index]\u001b[39;00m\n\u001b[0;32m   1016\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_slice\u001b[39m(\n\u001b[0;32m   1017\u001b[0m     \u001b[38;5;28mself\u001b[39m, slicer: \u001b[38;5;28mslice\u001b[39m \u001b[38;5;241m|\u001b[39m npt\u001b[38;5;241m.\u001b[39mNDArray[np\u001b[38;5;241m.\u001b[39mbool_] \u001b[38;5;241m|\u001b[39m npt\u001b[38;5;241m.\u001b[39mNDArray[np\u001b[38;5;241m.\u001b[39mintp]\n\u001b[0;32m   1018\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ArrayLike:\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# test function to identify sgRNAs within a given window around the start/stop codon of a transcript\n",
    "# this cell takes 32 min to run, you can avoid it by running the next cell and loading the output from this cell from MockMaterials\n",
    "from sgRNAutils import gRNA_stringencyIterator\n",
    "import os\n",
    "\n",
    "for index, row in TFsdf.iterrows():\n",
    "    result_df = gRNA_stringencyIterator(row, window, refSeqPerChromosome)\n",
    "    display(result_df)\n",
    "    # test whether every sgRNA identified is within the window of 21 bp around the start/stop codon\n",
    "    # you can manually do this by comparing the genomic coordinates of the sgRNA (fmin and fmax) with the coordinates of the start/stop codon (Start and Stop)\n",
    "    for i in range(len(result_df)):\n",
    "        if result_df['fmin'].iloc[i] >= result_df[\"Start\"].iloc[i] - window & result_df['fmax'].iloc[i] <= result_df[\"Stop\"].iloc[i] + window:\n",
    "            print(\"sgRNA within window\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9792987</td>\n",
       "      <td>9793009</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CACAGTTTACGCAGGTCCATGGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  9792987  9793009   2L      +  CACAGTTTACGCAGGTCCATGGG  FBgn0032150   \n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "1   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "1  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "1  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...  \n",
       "1  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26005878</td>\n",
       "      <td>26005900</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATTTCCAGAACCATTTTGTTG</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  26005878  26005900   3R      -  CCATTTCCAGAACCATTTTGTTG  FBgn0002733   \n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "1   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "1  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "1  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...  \n",
       "1  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "      <td>AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...</td>\n",
       "      <td>CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  524040      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "      <td>TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...</td>\n",
       "      <td>GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0070072          X  start_codon  370094  370096      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30852536</td>\n",
       "      <td>30852558</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCACATCGGTATGCAGTCGTCG</td>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>30852549</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...</td>\n",
       "      <td>GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...</td>\n",
       "      <td>CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  30852536  30852558   3R      -  CCCACATCGGTATGCAGTCGTCG  FBgn0003720   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0085709         3R  start_codon  30852547  30852549      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "      <td>GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...</td>\n",
       "      <td>CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  11278453      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "      <td>GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...</td>\n",
       "      <td>ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  8600584      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# to avoid the cell above but still check the results of the gRNA_stringencyIterator function, run this cell\n",
    "# the gRNA_stringencyIterator() function is supposed to identify sgRNAs within a given window around the start/stop codon of a transcript\n",
    "# to test the downstream functions, we need to identifiy representatives for conditions A-G and the sgRNA/start/stop codon orientations \n",
    "# (see MutationLogic_ML_20240124.pptx)\n",
    "# there are 8 different combinations of sgRNA/start/stop codon orientations\n",
    "# however, for two combinations each, the sgRNA and the start/stop codon have effectively the same relative position to each other \n",
    "# (e.g. the two combinations start codon - sgRNA strand (+) - gene strand (+) and stop codon - sgRNA strand (+) - gene strand (-))\n",
    "# therefore, we have identified 4 combinations of sgRNA/start/stop codon orientations with unique relative positions to test the downstream pipeline\n",
    "# further, we have identified representatives to test conditions A-D, which are relevant for the find_best_sgRNA function\n",
    "# further, we have identified representatives to test conditions E-G, which are relevant for the find_best_mutation function\n",
    "sgRNAdf1 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/window_21_bp/FBgn0032150_FBtr0079917.xlsx') # PAM in start/stop, start codon - gene (-) - gRNA (+)\n",
    "sgRNAdf2 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/window_21_bp/FBgn0002733_FBtr0084982.xlsx') # max 15 bp overlap, start codon - gene (-) - gRNA (-)\n",
    "sgRNAdf3 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/window_21_bp/FBgn0003963_FBtr0078063.xlsx') # more than 15 bp overlap, PAM in CDS\n",
    "sgRNAdf4 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/window_21_bp/FBgn0000022_FBtr0070072.xlsx') # start codon - gene (+) - gRNA (+)\n",
    "sgRNAdf5 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/window_21_bp/FBgn0003720_FBtr0085709.xlsx') # start codon - gene (+) - gRNA (-)\n",
    "sgRNAdf6 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/window_21_bp/FBgn0052831_FBtr0091683.xlsx') # SRS in CDS (need to manually change \"PAM in CDS\" to False)\n",
    "sgRNAdf7 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/window_21_bp/FBgn0267792_FBtr0088692.xlsx') # PAM and SRS bot outside CDS\n",
    "\n",
    "for i in [sgRNAdf1, sgRNAdf2, sgRNAdf3, sgRNAdf4, sgRNAdf5, sgRNAdf6, sgRNAdf7]:\n",
    "    display(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9792987</td>\n",
       "      <td>9793009</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CACAGTTTACGCAGGTCCATGGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  9792987  9793009   2L      +  CACAGTTTACGCAGGTCCATGGG  FBgn0032150   \n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0079917         2L  start_codon  9793004  ...              False   \n",
       "1   FBtr0079917         2L  start_codon  9793004  ...               True   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True      False            True       True   \n",
       "1                 False      False           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -2  \n",
       "1  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -3  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CACAGTTTACGCAGGTC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GCACAGTTTACGCAGGTC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26005878</td>\n",
       "      <td>26005900</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATTTCCAGAACCATTTTGTTG</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  26005878  26005900   3R      -  CCATTTCCAGAACCATTTTGTTG  FBgn0002733   \n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0084982         3R  start_codon  26005891  ...              False   \n",
       "1   FBtr0084982         3R  start_codon  26005891  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True       True           False       True   \n",
       "1                  True       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                     SRS_boundary  \\\n",
       "0  [-19, -18, -17, -16, -15, -14]   \n",
       "1  [-19, -18, -17, -16, -15, -14]   \n",
       "\n",
       "                                         mutable_PAM cut_site  \n",
       "0  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...      -10  \n",
       "1  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...       -4  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TTTGTTG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CCATTTCCAGAAC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TTTGTTGTAGTTT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CCAGAAC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                 False       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       15  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n",
      "sgRNA only in HAR\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0070072          X  start_codon  370094  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]        3  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GCTTTGGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CTATCTCTTAAA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30852536</td>\n",
       "      <td>30852558</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCACATCGGTATGCAGTCGTCG</td>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  30852536  30852558   3R      -  CCCACATCGGTATGCAGTCGTCG  FBgn0003720   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0085709         3R  start_codon  30852547  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True      False            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]       -8  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGTCGTCG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CCCACATCGGT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                 False       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       14  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCATCCCAATGCGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'C'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                 False      False            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]      -16  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'A'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CCAGTAAAGCGCAGCAGCA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test the sgRNApositionCheck function which checks how the sgRNA is positioned relative to the start/stop codon\n",
    "from sgRNAutils import sgRNApositionCheck\n",
    "\n",
    "#Create list to store results in\n",
    "positionCheckedList = []\n",
    "\n",
    "#Derive min and max distance from start/stop codon from window size\n",
    "# since the stop of the start/stop codon is zero, the minDistance has to be negative (left of the start/stop codon) \n",
    "# and shifted 3 nt to the left to account for the start/stop codon (see MutationLogic_ML_20240124.pptx)\n",
    "minDistance = (window + 3) * -1\n",
    "maxDistance = window\n",
    "\n",
    "for sgRNAdf in [sgRNAdf1, sgRNAdf2, sgRNAdf3, sgRNAdf4, sgRNAdf5, sgRNAdf6, sgRNAdf7]:\n",
    "    result_df = sgRNApositionCheck(sgRNAdf, minDistance, maxDistance)\n",
    "    display(result_df)\n",
    "    # check whether the sgRNA is positioned within the HA as expected\n",
    "    for i in range(len(result_df)):\n",
    "        positionScore = result_df.iloc[i][\"positionScore\"]\n",
    "        # print positionScore which represents the distance between the end of the start/stop codon and the end of the sgRNA (see MutationLogic_ML_20240124.pptx)\n",
    "        # note that the positionScore consequently represents the number of nts in the HAR that are part of the sgRNA\n",
    "        print(positionScore)\n",
    "        # check if sRNA is only in the HAR\n",
    "        if positionScore >=20:\n",
    "            print(\"sgRNA only in HAR\")\n",
    "            display(result_df.iloc[i][\"HAR\"][positionScore-23:positionScore])\n",
    "            # print the part of the sgRNA that is the start/stop codon\n",
    "            print(refSeqPerChromosome[result_df.iloc[i][\"Chromosome\"]][result_df.iloc[i]['Start']-1:result_df.iloc[i]['Stop']])\n",
    "        elif positionScore <= 0:\n",
    "            print(\"sgRNA only in HAL\")\n",
    "            # print the part of the sgRNA that is the start/stop codon\n",
    "            print(refSeqPerChromosome[result_df.iloc[i][\"Chromosome\"]][result_df.iloc[i]['Start']-1:result_df.iloc[i]['Stop']])\n",
    "            positionScore = positionScore + 2\n",
    "            positionScore = positionScore  -22\n",
    "            # print the only part of the sgRNA that is located in HAL (homology arm left)\n",
    "            display(result_df.iloc[i][\"HAL\"][positionScore:positionScore+23])\n",
    "        else: # print both HAL and HAR\n",
    "            # print the part of the sgRNA that is located in HAR (homology arm right)\n",
    "            display(result_df.iloc[i][\"HAR\"][:positionScore])\n",
    "            # print the part of the sgRNA that is the start/stop codon\n",
    "            print(refSeqPerChromosome[result_df.iloc[i][\"Chromosome\"]][result_df.iloc[i]['Start']-1:result_df.iloc[i]['Stop']])\n",
    "            positionScore = positionScore + 2 \n",
    "            positionScore = positionScore  -22\n",
    "            display(result_df.iloc[i][\"HAL\"][positionScore:])\n",
    "    positionCheckedList.append(result_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9792987</td>\n",
       "      <td>9793009</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CACAGTTTACGCAGGTCCATGGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  9792987  9793009   2L      +  CACAGTTTACGCAGGTCCATGGG  FBgn0032150   \n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0079917         2L  start_codon  9793004  ...              False   \n",
       "1   FBtr0079917         2L  start_codon  9793004  ...               True   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True      False            True       True   \n",
       "1                 False      False           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -2  \n",
       "1  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -3  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26005878</td>\n",
       "      <td>26005900</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATTTCCAGAACCATTTTGTTG</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  26005878  26005900   3R      -  CCATTTCCAGAACCATTTTGTTG  FBgn0002733   \n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0084982         3R  start_codon  26005891  ...              False   \n",
       "1   FBtr0084982         3R  start_codon  26005891  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True       True           False       True   \n",
       "1                  True       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                     SRS_boundary  \\\n",
       "0  [-19, -18, -17, -16, -15, -14]   \n",
       "1  [-19, -18, -17, -16, -15, -14]   \n",
       "\n",
       "                                         mutable_PAM cut_site  \n",
       "0  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...      -10  \n",
       "1  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...       -4  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'GCACAGTTTACGCAGGTCCATGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CCAGAACCATTTTGTTGTAGTTT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test the checkCDSCutandOrder function which selects a sgRNA that cuts in the CDS \n",
    "# and/or as a lower priority an sgRNA that cuts the closest to the start/stop codon (condition D, see MutationLogic_ML_20240124.pptx)\n",
    "from sgRNAutils import checkCDSCutandOrder\n",
    "# test this for some of the TF transcripts with multiple sgRNAs\n",
    "display(positionCheckedList[0])\n",
    "display(positionCheckedList[1])\n",
    "# we expect the second sgRNA in first dataframe to be selected since it cuts inside the CDS (cut_site value is among the CDS_boundary values) and the first one cuts outside\n",
    "# we expect the second sgRNA in the second dataframe to be selected since it is closer to the start/stop codon (4 is smaller than 10)\n",
    "sgRNAdf1_checked = checkCDSCutandOrder(positionCheckedList[0])\n",
    "sgRNAdf2_checked = checkCDSCutandOrder(positionCheckedList[1])\n",
    "display(sgRNAdf1_checked)\n",
    "display(sgRNAdf2_checked)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>cutsite-CDSbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9792987</td>\n",
       "      <td>9793009</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CACAGTTTACGCAGGTCCATGGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "      <td>3</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "      <td>2</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  9792987  9793009   2L      +  CACAGTTTACGCAGGTCCATGGG  FBgn0032150   \n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "1   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "1  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "1  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...              3      HAL   \n",
       "1  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...              2      HAL   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAR             False                  True      False   \n",
       "1          HAR              True                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True       True   \n",
       "1           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site cutsite-CDSbound  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -2                2  \n",
       "1  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -3                3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  False\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>cutsite-CDSbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "      <td>2</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "1   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "1  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "1  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "1  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...              2      HAL   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "1          HAR              True                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "1           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site cutsite-CDSbound  \n",
       "1  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -3                3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>cutsite-CDSbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26005878</td>\n",
       "      <td>26005900</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATTTCCAGAACCATTTTGTTG</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "      <td>7</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "      <td>13</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  26005878  26005900   3R      -  CCATTTCCAGAACCATTTTGTTG  FBgn0002733   \n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "1   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "1  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "1  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...              7      HAL   \n",
       "1  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...             13      HAL   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAR             False                  True       True   \n",
       "1          HAR             False                  True       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "1           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                     SRS_boundary  \\\n",
       "0  [-19, -18, -17, -16, -15, -14]   \n",
       "1  [-19, -18, -17, -16, -15, -14]   \n",
       "\n",
       "                                         mutable_PAM cut_site cutsite-CDSbound  \n",
       "0  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...      -10               10  \n",
       "1  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...       -4                4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  False\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>cutsite-CDSbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "      <td>13</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "1   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "1  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "1  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "1  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...             13      HAL   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "1          HAR             False                  True       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "1           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                     SRS_boundary  \\\n",
       "1  [-19, -18, -17, -16, -15, -14]   \n",
       "\n",
       "                                         mutable_PAM cut_site cutsite-CDSbound  \n",
       "1  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...       -4                4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "      <td>AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...</td>\n",
       "      <td>CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...</td>\n",
       "      <td>20</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  524040      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...             20      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       15  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "      <td>AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...</td>\n",
       "      <td>CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...</td>\n",
       "      <td>20</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  524040      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...             20      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       15  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "      <td>TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...</td>\n",
       "      <td>GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...</td>\n",
       "      <td>8</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0070072          X  start_codon  370094  370096      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...              8      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                  True       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]        3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  False\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "      <td>TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...</td>\n",
       "      <td>GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...</td>\n",
       "      <td>8</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0070072          X  start_codon  370094  370096      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...              8      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                  True       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]        3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30852536</td>\n",
       "      <td>30852558</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCACATCGGTATGCAGTCGTCG</td>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>30852549</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...</td>\n",
       "      <td>GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...</td>\n",
       "      <td>CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...</td>\n",
       "      <td>9</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  30852536  30852558   3R      -  CCCACATCGGTATGCAGTCGTCG  FBgn0003720   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0085709         3R  start_codon  30852547  30852549      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...              9      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                  True      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]       -8  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  False\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30852536</td>\n",
       "      <td>30852558</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCACATCGGTATGCAGTCGTCG</td>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>30852549</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...</td>\n",
       "      <td>GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...</td>\n",
       "      <td>CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...</td>\n",
       "      <td>9</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  30852536  30852558   3R      -  CCCACATCGGTATGCAGTCGTCG  FBgn0003720   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0085709         3R  start_codon  30852547  30852549      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...              9      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                  True      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]       -8  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "      <td>GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...</td>\n",
       "      <td>CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...</td>\n",
       "      <td>19</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  11278453      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...             19      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       14  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "      <td>GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...</td>\n",
       "      <td>CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...</td>\n",
       "      <td>19</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  11278453      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...             19      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       14  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "      <td>GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...</td>\n",
       "      <td>ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...</td>\n",
       "      <td>1</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  8600584      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...              1      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]      -16  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "      <td>GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...</td>\n",
       "      <td>ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...</td>\n",
       "      <td>1</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  8600584      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...              1      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]      -16  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test condition A-C (see MutationLogic_ML_20240124.pptx) for identifying the best sgRNA (e.g. PAM in start/stop, max 15 bp overlap, none of the above)\n",
    "# we expect that for the first dataframe the sgRNA is selected that has the PAM in the start/stop codon, no mutation needed\n",
    "# we expect that for the second dataframe the sgRNA is selected that cuts the closest to the start/stop codon, \n",
    "# since both candidate sgRNAs exhibit max 15bp overlap and cut within the CDS, no mutation needed\n",
    "# we expect that for the third dataframe a mutation is needed since the PAM is not located in the start/stop codon and the 3' overlap is more than 15 bp\n",
    "\n",
    "from sgRNAutils import find_best_gRNA  \n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "bestsgRNAList = []\n",
    "\n",
    "for sgRNAdf in positionCheckedList:\n",
    "    # first display all sgRNAs within the given windown around the start/stop codon\n",
    "    display(sgRNAdf)\n",
    "    sgRNA, mutationNeeded = find_best_gRNA(sgRNAdf)\n",
    "    print(\"mutation needed: \", mutationNeeded)\n",
    "    winnerdf = sgRNAdf[sgRNAdf[\"sgRNA_sequence\"] == sgRNA] \n",
    "    # next display the sgRNA that is identified as the best sgRNA\n",
    "    display(winnerdf)\n",
    "    bestsgRNAList.append(winnerdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CGC', 'TGC', 'CTG', 'AAA', 'CGT', 'CTG', 'GAC']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['TCC', 'ATG', 'GAG', 'ATG', 'GAA', 'CTG', 'GTT']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['CTT', 'AGC', 'AGC', 'AAT', 'ACA', 'AGA', 'GGT']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['GCT', 'TTG', 'GGC', 'AGC', 'GAA', 'AAT', 'CAC']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['CAG', 'TCG', 'TCG', 'GAG', 'GGT', 'TCA', 'CCA']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['CAG', 'CTG', 'CAT', 'CCC', 'AAT', 'GCG', 'GAG']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['ATG', 'CCG', 'CTG', 'CAG', 'GAG', 'TCG', 'CCG']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test the function that extracts the sequence of the drosophila genome that is in the CDS of a given transcript and within the given window and chops it into codons \n",
    "from sgRNAutils import codonFragmenter\n",
    "pd.set_option('display.max_column', None)\n",
    "\n",
    "# create an empty list to store the chopped CDSs and the adjusted dataframes\n",
    "CDS_list = []\n",
    "df_adjusted_list = []\n",
    "\n",
    "for bestsgRNA in bestsgRNAList:\n",
    "    CDS, df_adjusted = codonFragmenter(bestsgRNA.iloc[0], bestsgRNA.iloc[0][\"CDS_side\"], bestsgRNA.iloc[0][\"CDS_boundary\"])\n",
    "    CDS_list.append(CDS)\n",
    "    df_adjusted_list.append(df_adjusted)\n",
    "    # now display the chopped up CDS\n",
    "    display(CDS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTATGTTCTGTTGAATATTATACGTATTCACTTATGTTTAAGTCATTTCGAAGGTCTTACCCAGCCGAACATGGCTTCTATCAGCTGGAGGACTGTCTGATCCGAACTGCCACCGCTCCTGTGCTCCCGTATATCCACCGAAGCGCTGTCCACGAAGCAAATGCGGCACAGTTTACGCAGGTC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTATGTTCTGTTGAATATTATACGTATTCACTTATGTTTAAGTCATTTCGAAGGTCTTACCCAGCCGAACATGGCTTCTATCAGCTGGAGGACTGTCTGATCCGAACTGCCACCGCTCCTGTGCTCCCGTATATCCACCGAAGCGCTGTCCACGAAGCAAATGCGGCACAGTTTACGCAGGTC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "HA before chopping and HA after recombining are the same. All is well\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGCTCCAGGATGTCGGCCTTCTCCAGCCGGGTGATGTGCTCGCCCTCCTGGGTCAGGCACTCCACCATGATGTCCTTGAGCTCGTCCAGGCACTTGTTAATCCTGGCACGGCGCTTACGCTCCAGCATGGGCTTCATCACCTTGCGGTACTGATAGGTCTTGGACATCTCCATTTCCAGAAC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGCTCCAGGATGTCGGCCTTCTCCAGCCGGGTGATGTGCTCGCCCTCCTGGGTCAGGCACTCCACCATGATGTCCTTGAGCTCGTCCAGGCACTTGTTAATCCTGGCACGGCGCTTACGCTCCAGCATGGGCTTCATCACCTTGCGGTACTGATAGGTCTTGGACATCTCCATTTCCAGAAC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "HA before chopping and HA after recombining are the same. All is well\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCTATGAAACACAATATGCTTGAATATGCTTTTAATAAGAAAAGCAAACTAGCAATTCTATCTGATTTAGATGTAAACAAATTATTCATTCTATTTGGAATGTTTGGATAATTATGTTATATTAAATTTGAGATCTAAAGTGAATGTATCATTTGTTGTTGATATTGATTGTTGAATATATG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCTATGAAACACAATATGCTTGAATATGCTTTTAATAAGAAAAGCAAACTAGCAATTCTATCTGATTTAGATGTAAACAAATTATTCATTCTATTTGGAATGTTTGGATAATTATGTTATATTAAATTTGAGATCTAAAGTGAATGTATCATTTGTTGTTGATATTGATTGTTGAATATATG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "HA before chopping and HA after recombining are the same. All is well\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGTCATCTTCGGCCTTTAATGGACCCTCTGTTATCCGGAGAAATGCCCGGGAACGCAACCGCGTAAAGCAGGTCAACAATGGCTTCAGCCAACTACGACAACATATCCCTGCGGCCGTAATAGCCGATTTAAGCAATGGTCGCCGGGGAATTGGTCCCGGCGCCAATAAAAAACTGAGCAAA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGTCATCTTCGGCCTTTAATGGACCCTCTGTTATCCGGAGAAATGCCCGGGAACGCAACCGCGTAAAGCAGGTCAACAATGGCTTCAGCCAACTACGACAACATATCCCTGCGGCCGTAATAGCCGATTTAAGCAATGGTCGCCGGGGAATTGGTCCCGGCGCCAATAAAAAACTGAGCAAA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "HA before chopping and HA after recombining are the same. All is well\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACAGCGTGCGTCTTTCGCCAGCGGCATCGAGTAAGTATAGCAAAAAAAAAAATCAGATAGCAATGAAAAAAGAAGAAAACAATCTTCCTAATATATAAAATTAAGACAATCCTTTACTTATACCAACTTTCCCCCCTTCTTGAATTTCCAGGTCGCATTCTATACCATGTGCCCTGCAAAGT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACAGCGTGCGTCTTTCGCCAGCGGCATCGAGTAAGTATAGCAAAAAAAAAAATCAGATAGCAATGAAAAAAGAAGAAAACAATCTTCCTAATATATAAAATTAAGACAATCCTTTACTTATACCAACTTTCCCCCCTTCTTGAATTTCCAGGTCGCATTCTATACCATGTGCCCTGCAAAGT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "HA before chopping and HA after recombining are the same. All is well\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTATGCGGTGGAGTCGCGAGTTTTGTGGACACTAATCGGAGTGTTTCTGTGTGCCTTGTTCCCTCGCCACGGCTGCAGGAAGTGGACGGACGAGGAGATCGACATGCTGCACTCGTCCATAATGCGCTTCTCCGACGACCTGACCAAGATCAGCCTAAGCATCAAGAACCGCACCGTGTAAG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTATGCGGTGGAGTCGCGAGTTTTGTGGACACTAATCGGAGTGTTTCTGTGTGCCTTGTTCCCTCGCCACGGCTGCAGGAAGTGGACGGACGAGGAGATCGACATGCTGCACTCGTCCATAATGCGCTTCTCCGACGACCTGACCAAGATCAGCCTAAGCATCAAGAACCGCACCGTGTAAG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "HA before chopping and HA after recombining are the same. All is well\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACTACGCTTTCTTCCGCAAGTGCGGGGAAATCACCGTCTCGCCGGACGTGCAGAGCTTCGGCTTCAACTGCGCCTTCTGTCCGGCGATCTGCCTGCAGTTCTCCGTGTTCATGGACCACATCCGGGTGCAGCACACGCAGGACGTGAGCCGGCGCTACGAGACTGAGGAGGAGAGGTGTCCG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACTACGCTTTCTTCCGCAAGTGCGGGGAAATCACCGTCTCGCCGGACGTGCAGAGCTTCGGCTTCAACTGCGCCTTCTGTCCGGCGATCTGCCTGCAGTTCTCCGTGTTCATGGACCACATCCGGGTGCAGCACACGCAGGACGTGAGCCGGCGCTACGAGACTGAGGAGGAGAGGTGTCCG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "HA before chopping and HA after recombining are the same. All is well\n"
     ]
    }
   ],
   "source": [
    "# test the function that recombines the codons again and fuses the CDS with the rest of the HA\n",
    "from sgRNAutils import codonReverseFragmenter\n",
    "\n",
    "for i, sgRNA in enumerate(bestsgRNAList):\n",
    "    # retrieve the HA which contains the CDS before the chopping \n",
    "    HA_before_chopping = sgRNA[f\"{sgRNA[\"CDS_side\"].iloc[0]}\"].iloc[0]\n",
    "    df_after_recombining = codonReverseFragmenter(CDS_list[i], df_adjusted_list[i], df_adjusted_list[i][\"CDS_side\"], df_adjusted_list[i][\"CDS_boundary\"])\n",
    "    HA_after_recombining = df_after_recombining[df_after_recombining[\"CDS_side\"]]\n",
    "    # display HA before chopping and after chopping and test whether they are the same\n",
    "    # you can also check manually whether they are the same\n",
    "    display(HA_before_chopping)\n",
    "    display(HA_after_recombining)\n",
    "    # display the length of the HA after recombining, should be again 225 bp\n",
    "    print(len(HA_after_recombining))\n",
    "    if HA_after_recombining == HA_before_chopping:\n",
    "        print(\"HA before chopping and HA after recombining are the same. All is well\")\n",
    "    else: print(\"we have a problem\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n",
      "[]\n",
      "['GCC', 'GCA', 'GCG']\n"
     ]
    }
   ],
   "source": [
    "# test function to identify synonymous codons that changes a given nucleotide but does not change the encoded amino acid\n",
    "from sgRNAutils import find_synonymous_codons\n",
    "# try to mutate first, second and third nucleotide of the codon \"ATG\"\n",
    "# note, its unlikely to find synonymous mutations if the nucleotide you wish to mutate is NOT located in the wobble position (e.g. nucleotide 3 of the codon)\n",
    "codon = \"GCT\"\n",
    "synonymous_codon_1 = find_synonymous_codons(codon, 1)\n",
    "synonymous_codon_2 = find_synonymous_codons(codon, 2)\n",
    "synonymous_codon_3 = find_synonymous_codons(codon, 3)\n",
    "print(synonymous_codon_1)\n",
    "print(synonymous_codon_2)\n",
    "print(synonymous_codon_3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "      <td>GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...</td>\n",
       "      <td>CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...</td>\n",
       "      <td>19</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  11278453      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...             19      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       14  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HAR\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTATGCGGTGGAGTCGCGAGTTTTGTGGACACTAATCGGAGTGTTTCTGTGTGCCTTGTTCCCTCGCCACGGCTGCAGGAAGTGGACGGACGAGGAGATCGACATGCTGCACTCGTCCATAATGCGCTTCTCCGACGACCTGACCAAGATCAGCCTAAGCATCAAGAACCGCACCGTGTAAG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCATCCCAATGCTGAGTCGCCATCGGGGTAAGTTGATTCTATGCGGTGGAGTCGCGAGTTTTGTGGACACTAATCGGAGTGTTTCTGTGTGCCTTGTTCCCTCGCCACGGCTGCAGGAAGTGGACGGACGAGGAGATCGACATGCTGCACTCGTCCATAATGCGCTTCTCCGACGACCTGACCAAGATCAGCCTAAGCATCAAGAACCGCACCGTGTAAG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CATGCAGCTGCATCCCAATGCGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test find_best_mutation function which mutates the HA in a way that it cannot be cut by the sgRNA\n",
    "# test case where PAM is in CDS\n",
    "from sgRNAutils import find_best_mutation\n",
    "pd.set_option('display.max_columns', None)\n",
    "display(bestsgRNAList[5])\n",
    "bestsgRNAList[5][\"mutated?\"] = False\n",
    "CDSside = bestsgRNAList[5][\"CDS_side\"]\n",
    "mutated_sgRNAdf6 = find_best_mutation(bestsgRNAList[5].iloc[0])\n",
    "# compare the HA wich contains the CDS before and after mutation\n",
    "# check whether position of mutation is in the PAM by comparing the HA with the sgRNA\n",
    "print(CDSside.iloc[0])\n",
    "display(bestsgRNAList[5].iloc[0][CDSside].iloc[0])\n",
    "display(mutated_sgRNAdf6[CDSside].iloc[0])\n",
    "display(mutated_sgRNAdf6[\"sgRNA_sequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "      <td>AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...</td>\n",
       "      <td>CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...</td>\n",
       "      <td>20</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  524040      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...             20      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       15  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HAR\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCTATGAAACACAATATGCTTGAATATGCTTTTAATAAGAAAAGCAAACTAGCAATTCTATCTGATTTAGATGTAAACAAATTATTCATTCTATTTGGAATGTTTGGATAATTATGTTATATTAAATTTGAGATCTAAAGTGAATGTATCATTTGTTGTTGATATTGATTGTTGAATATATG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CTTAGCAGCAACACTAGAGGTGAGAATTATTCATACACAGTTTTCTATGAAACACAATATGCTTGAATATGCTTTTAATAAGAAAAGCAAACTAGCAATTCTATCTGATTTAGATGTAAACAAATTATTCATTCTATTTGGAATGTTTGGATAATTATGTTATATTAAATTTGAGATCTAAAGTGAATGTATCATTTGTTGTTGATATTGATTGTTGAATATATG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ATGCTTAGCAGCAATACAAGAGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test find_best_mutation function which mutates the HA in a way that it cannot be cut by the sgRNA\n",
    "# test case where PAM is in CDS\n",
    "pd.set_option('display.max_columns', None)\n",
    "display(bestsgRNAList[2])\n",
    "bestsgRNAList[2][\"mutated?\"] = False\n",
    "CDSside = bestsgRNAList[2][\"CDS_side\"]\n",
    "mutated_sgRNAdf3 = find_best_mutation(bestsgRNAList[2].iloc[0])\n",
    "# compare the HA wich contains the CDS before and after mutation\n",
    "# check position of mutation by comparing the HA with the sgRNA\n",
    "# no synoynmous mutations in the PAM were possible (likely because not in wobble)\n",
    "# therefore the SRS was mutated\n",
    "print(CDSside.iloc[0])\n",
    "display(bestsgRNAList[2].iloc[0][CDSside].iloc[0])\n",
    "display(mutated_sgRNAdf3[CDSside].iloc[0])\n",
    "display(mutated_sgRNAdf3[\"sgRNA_sequence\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "      <td>GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...</td>\n",
       "      <td>ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...</td>\n",
       "      <td>1</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  8600584      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...              1      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]      -16  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HAL\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCACTGCCTCGGCCACACTGCAGTTCGGATTTGCCGGTTGTTTGGATCGCAGATTGAATCGGAAAGCCAGCGGCGTTGGTTCGGAAAAGTGCAGTCTCCGCGTCGCGAGTGCGCCAATAACGAGAAATAACGAGAATTGAAAAAGTTCCTTAAGAGGTGGAGTCCAGTAAAGCGCAGCAGCA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCACTGCCTCGGCCACACTGCAGTTCGGATTTGCCGGTTGTTTGGATCGCAGATTGAATCGGAAAGCCAGCGGCGTTGGTTCGGAAAAGTGCAGTCTCCGCGTCGCGAGTGCGCCAATAACGAGAAATAACGAGAATTGAAAAAGTTCCTTAAGAGGTGGAGTCAAGTAAAGCGCAGCAGCA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CCAGTAAAGCGCAGCAGCAATGA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test find_best_mutation function which mutates the HA in a way that it cannot be cut by the sgRNA\n",
    "# test case where neither PAM nor SRS is in CDS\n",
    "# we expect that the PAM is simply mutated without respecting synonymous mutations\n",
    "pd.set_option('display.max_columns', None)\n",
    "display(bestsgRNAList[6])\n",
    "bestsgRNAList[6][\"mutated?\"] = False\n",
    "nonCDSside = bestsgRNAList[6][\"non_CDS_side\"]\n",
    "mutated_sgRNAdf7 = find_best_mutation(bestsgRNAList[6].iloc[0])\n",
    "# compare the HA wich does NOT contain the CDS before and after mutation\n",
    "# check position of mutation by comparing the HA with the sgRNA\n",
    "print(nonCDSside.iloc[0])\n",
    "display(bestsgRNAList[6].iloc[0][nonCDSside].iloc[0])\n",
    "display(mutated_sgRNAdf7[nonCDSside].iloc[0])\n",
    "display(mutated_sgRNAdf7[\"sgRNA_sequence\"])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
