{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CAT'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test function to form the reverse complement of a DNA sequence\n",
    "from sgRNAutils import revComp\n",
    "revComp('ATG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('TGCTAGAGGG')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test function to extract the DNA sequence at particular genomic coordinates of a particular chromosome\n",
    "from sgRNAutils import refSeq\n",
    "refSeqPerChromosome= refSeq()\n",
    "refSeqPerChromosome['X'][20000080:20000090]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>370697</td>\n",
       "      <td>370699</td>\n",
       "      <td>+</td>\n",
       "      <td>GACATTGTGTCGTTCGTATGTCGCCCATTGAGACCGCCAATGGAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>539954</td>\n",
       "      <td>539956</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0310027</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>491316</td>\n",
       "      <td>491318</td>\n",
       "      <td>+</td>\n",
       "      <td>CACTGATAAAAAATACCCTAAATTATCGCAACGCAAATGGTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0310027</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>539954</td>\n",
       "      <td>539956</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0329895</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0329895</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>540017</td>\n",
       "      <td>540019</td>\n",
       "      <td>+</td>\n",
       "      <td>TTTTTTAATGTTCTGCACTTGCTTAATTACCAGCATGTACCATCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0333012</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>525418</td>\n",
       "      <td>525420</td>\n",
       "      <td>+</td>\n",
       "      <td>ATCATTGGCCAAAAAAGTAGGAATTCTGGGAGGAGGGCATGCCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0333012</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>539954</td>\n",
       "      <td>539956</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0333013</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>491316</td>\n",
       "      <td>491318</td>\n",
       "      <td>+</td>\n",
       "      <td>CACTGATAAAAAATACCCTAAATTATCGCAACGCAAATGGTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0333013</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>539954</td>\n",
       "      <td>539956</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>9791982</td>\n",
       "      <td>9791984</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0331285</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9792792</td>\n",
       "      <td>9792794</td>\n",
       "      <td>-</td>\n",
       "      <td>TAAAATGCGGTGACCTGCAAAAGGAGCCATAAACTAAGCGTCCTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0331285</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>9791982</td>\n",
       "      <td>9791984</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0331286</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9792792</td>\n",
       "      <td>9792794</td>\n",
       "      <td>-</td>\n",
       "      <td>TAAAATGCGGTGACCTGCAAAAGGAGCCATAAACTAAGCGTCCTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0331286</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>9791982</td>\n",
       "      <td>9791984</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FBgn0037584</td>\n",
       "      <td>FBtr0301759</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8362880</td>\n",
       "      <td>8362882</td>\n",
       "      <td>+</td>\n",
       "      <td>GATATACATCTATACATTATACGGATATTTATAGCATATACCTGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>FBgn0037584</td>\n",
       "      <td>FBtr0301759</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>8363892</td>\n",
       "      <td>8363894</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACGCCCCAGTTAAACCACAACAAGTATACGGCTCCATCGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083648</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18506517</td>\n",
       "      <td>18506519</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTAATGCAACTTATTGAATTTTCGAGTGCTTGGCTTAGTTGATTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083648</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18426145</td>\n",
       "      <td>18426147</td>\n",
       "      <td>-</td>\n",
       "      <td>CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083649</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18503977</td>\n",
       "      <td>18503979</td>\n",
       "      <td>-</td>\n",
       "      <td>TATTATATAGTTATTGTAGTTATTGTTTATAGTTATAACCATTAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083649</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18426145</td>\n",
       "      <td>18426147</td>\n",
       "      <td>-</td>\n",
       "      <td>CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083650</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18478069</td>\n",
       "      <td>18478071</td>\n",
       "      <td>-</td>\n",
       "      <td>TGTGTCTAACATGTGAGGAGCATTTCATAATGGTTCTTTTTCGGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083650</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18422638</td>\n",
       "      <td>18422640</td>\n",
       "      <td>-</td>\n",
       "      <td>TCACTCTCTGTCTCCCTCCAACTTATTGTACATTTGCCAAACAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083651</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18435766</td>\n",
       "      <td>18435768</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083651</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18426145</td>\n",
       "      <td>18426147</td>\n",
       "      <td>-</td>\n",
       "      <td>CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083652</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18435766</td>\n",
       "      <td>18435768</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083652</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18426145</td>\n",
       "      <td>18426147</td>\n",
       "      <td>-</td>\n",
       "      <td>CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083640</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18435766</td>\n",
       "      <td>18435768</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083640</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18422638</td>\n",
       "      <td>18422640</td>\n",
       "      <td>-</td>\n",
       "      <td>TCACTCTCTGTCTCCCTCCAACTTATTGTACATTTGCCAAACAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083641</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18515330</td>\n",
       "      <td>18515332</td>\n",
       "      <td>-</td>\n",
       "      <td>GTAAAGTCGGCTCCACATGAGTGGGCGTAAGGCGTTTCGTGGGCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083641</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18416777</td>\n",
       "      <td>18416779</td>\n",
       "      <td>-</td>\n",
       "      <td>ACGAGTACCCTACTTTTATCTATTTGAATATATATATATTATGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083642</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18435766</td>\n",
       "      <td>18435768</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083642</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18426145</td>\n",
       "      <td>18426147</td>\n",
       "      <td>-</td>\n",
       "      <td>CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083643</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18515330</td>\n",
       "      <td>18515332</td>\n",
       "      <td>-</td>\n",
       "      <td>GTAAAGTCGGCTCCACATGAGTGGGCGTAAGGCGTTTCGTGGGCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083643</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18426145</td>\n",
       "      <td>18426147</td>\n",
       "      <td>-</td>\n",
       "      <td>CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083644</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18435766</td>\n",
       "      <td>18435768</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083644</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18416777</td>\n",
       "      <td>18416779</td>\n",
       "      <td>-</td>\n",
       "      <td>ACGAGTACCCTACTTTTATCTATTTGAATATATATATATTATGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083645</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18515330</td>\n",
       "      <td>18515332</td>\n",
       "      <td>-</td>\n",
       "      <td>GTAAAGTCGGCTCCACATGAGTGGGCGTAAGGCGTTTCGTGGGCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083645</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18422638</td>\n",
       "      <td>18422640</td>\n",
       "      <td>-</td>\n",
       "      <td>TCACTCTCTGTCTCCCTCCAACTTATTGTACATTTGCCAAACAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083646</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18435766</td>\n",
       "      <td>18435768</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083646</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18430824</td>\n",
       "      <td>18430826</td>\n",
       "      <td>-</td>\n",
       "      <td>CACTGTACAGAACATTCTTGTAAAGCCTTTTTGTTTAAATCAGCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083647</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18435766</td>\n",
       "      <td>18435768</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083647</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18431447</td>\n",
       "      <td>18431449</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGCTATTTCGGGTTTCGCGTTGTTATCAGTGGGCATGAAGTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0301284</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18515330</td>\n",
       "      <td>18515332</td>\n",
       "      <td>-</td>\n",
       "      <td>GTAAAGTCGGCTCCACATGAGTGGGCGTAAGGCGTTTCGTGGGCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0301284</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18426145</td>\n",
       "      <td>18426147</td>\n",
       "      <td>-</td>\n",
       "      <td>CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0330040</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18435766</td>\n",
       "      <td>18435768</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0330040</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18422077</td>\n",
       "      <td>18422079</td>\n",
       "      <td>-</td>\n",
       "      <td>CAAAAATCCCGAACCGAGCAGAGGTCTCTGCCCCATATAACACCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>26005306</td>\n",
       "      <td>26005308</td>\n",
       "      <td>-</td>\n",
       "      <td>TCCCCAAGCGAACCCAAAACCCCCAACCGAACTGCAGCTGATCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>FBgn0039509</td>\n",
       "      <td>FBtr0085173</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>27288467</td>\n",
       "      <td>27288469</td>\n",
       "      <td>-</td>\n",
       "      <td>GTTCGTCCAGCAGCACCTCGCTAACATTCTTCGAGCGATTGTATTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>FBgn0039509</td>\n",
       "      <td>FBtr0085173</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>27287498</td>\n",
       "      <td>27287500</td>\n",
       "      <td>-</td>\n",
       "      <td>GATTTCTCCAAGGATTAACACGGGAATTACGTCAATTGAAACGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>FBgn0039712</td>\n",
       "      <td>FBtr0085523</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>29861352</td>\n",
       "      <td>29861354</td>\n",
       "      <td>-</td>\n",
       "      <td>TTGGTTCGCACCAGCTCGTTGTTGGAGGCGTTGTACACAACATCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>FBgn0039712</td>\n",
       "      <td>FBtr0085523</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>29860354</td>\n",
       "      <td>29860356</td>\n",
       "      <td>-</td>\n",
       "      <td>CAATTTGCGAGAGTCGACTTACCCATGTTCGAGTAAAAAGCGAGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>FBgn0039712</td>\n",
       "      <td>FBtr0335005</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>29861352</td>\n",
       "      <td>29861354</td>\n",
       "      <td>-</td>\n",
       "      <td>TTGGTTCGCACCAGCTCGTTGTTGGAGGCGTTGTACACAACATCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>FBgn0039712</td>\n",
       "      <td>FBtr0335005</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>29860354</td>\n",
       "      <td>29860356</td>\n",
       "      <td>-</td>\n",
       "      <td>CAATTTGCGAGAGTCGACTTACCCATGTTCGAGTAAAAAGCGAGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>FBgn0030008</td>\n",
       "      <td>FBtr0071155</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8145949</td>\n",
       "      <td>8145951</td>\n",
       "      <td>+</td>\n",
       "      <td>GTGCGTGTCAAATAAATTCGATGGAAACCCGTGAAACGAGGACCGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>FBgn0030008</td>\n",
       "      <td>FBtr0071155</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>8147344</td>\n",
       "      <td>8147346</td>\n",
       "      <td>+</td>\n",
       "      <td>GGGCGTGCAGGCGGACGTAGTCGACGTTGGCGAGGCAGTGTCTCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0111006</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0111006</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>15143449</td>\n",
       "      <td>15143451</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0111007</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0111007</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>15140679</td>\n",
       "      <td>15140681</td>\n",
       "      <td>+</td>\n",
       "      <td>CGCCTCTAGAAATGCCGTTGCCATCATCGCCTTGGTGGGTCCCATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0087415</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0087415</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>15143449</td>\n",
       "      <td>15143451</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0087416</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0087416</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>15143449</td>\n",
       "      <td>15143451</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305213</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305213</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>15143449</td>\n",
       "      <td>15143451</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305214</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305214</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>15143449</td>\n",
       "      <td>15143451</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305215</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305215</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>15143449</td>\n",
       "      <td>15143451</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305216</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305216</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>15143449</td>\n",
       "      <td>15143451</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305217</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0305217</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>15140382</td>\n",
       "      <td>15140384</td>\n",
       "      <td>+</td>\n",
       "      <td>CTCTCGTGTCGTCCATTGCGAAGGCAGTTTATTGTTAAACAGGAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>FBgn0022740</td>\n",
       "      <td>FBtr0089547</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>17757575</td>\n",
       "      <td>17757577</td>\n",
       "      <td>+</td>\n",
       "      <td>TTCTAGTGGGGGGCCCGGCTGAATGGCTGACTCCGGTCATGGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>FBgn0022740</td>\n",
       "      <td>FBtr0089547</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>17760336</td>\n",
       "      <td>17760338</td>\n",
       "      <td>+</td>\n",
       "      <td>CTGTCAGATAAACCTATCAGCTAAACTAATTTCAAGTGGACCTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>FBgn0004895</td>\n",
       "      <td>FBtr0073192</td>\n",
       "      <td>3L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4019094</td>\n",
       "      <td>4019096</td>\n",
       "      <td>+</td>\n",
       "      <td>TCATTTATTCGATTTGTTTATGAGTTTTTAGATGTGCATTATACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>FBgn0004895</td>\n",
       "      <td>FBtr0073192</td>\n",
       "      <td>3L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>4021177</td>\n",
       "      <td>4021179</td>\n",
       "      <td>+</td>\n",
       "      <td>ATATAAACATTACATCACTTATATTTAGGGTTTATGGTCATGGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>FBgn0004895</td>\n",
       "      <td>FBtr0309789</td>\n",
       "      <td>3L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4019094</td>\n",
       "      <td>4019096</td>\n",
       "      <td>+</td>\n",
       "      <td>TCATTTATTCGATTTGTTTATGAGTTTTTAGATGTGCATTATACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>FBgn0004895</td>\n",
       "      <td>FBtr0309789</td>\n",
       "      <td>3L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>4021177</td>\n",
       "      <td>4021179</td>\n",
       "      <td>+</td>\n",
       "      <td>ATATAAACATTACATCACTTATATTTAGGGTTTATGGTCATGGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>30852549</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>30854024</td>\n",
       "      <td>30854026</td>\n",
       "      <td>+</td>\n",
       "      <td>ACAGAATGATGAGGATTAGGCCAACACACACATCCATCGAAAGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11279461</td>\n",
       "      <td>11279463</td>\n",
       "      <td>+</td>\n",
       "      <td>TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091680</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278113</td>\n",
       "      <td>11278115</td>\n",
       "      <td>+</td>\n",
       "      <td>CTGGCGATCCACAGACTCGTCCTCCACATCGCACAATTCATTTGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091680</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11279461</td>\n",
       "      <td>11279463</td>\n",
       "      <td>+</td>\n",
       "      <td>TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091681</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278370</td>\n",
       "      <td>11278372</td>\n",
       "      <td>+</td>\n",
       "      <td>CTGATCAAAATAAAAGTATGAAGTACAAATTACAGCAATGAGATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091681</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11279461</td>\n",
       "      <td>11279463</td>\n",
       "      <td>+</td>\n",
       "      <td>TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091682</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278113</td>\n",
       "      <td>11278115</td>\n",
       "      <td>+</td>\n",
       "      <td>CTGGCGATCCACAGACTCGTCCTCCACATCGCACAATTCATTTGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091682</td>\n",
       "      <td>2L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11279461</td>\n",
       "      <td>11279463</td>\n",
       "      <td>+</td>\n",
       "      <td>TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>FBgn0263240</td>\n",
       "      <td>FBtr0088984</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>7492187</td>\n",
       "      <td>7492189</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>FBgn0263240</td>\n",
       "      <td>FBtr0088984</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>7491049</td>\n",
       "      <td>7491051</td>\n",
       "      <td>-</td>\n",
       "      <td>GACACAGAGTTGCGGCCAGCGCAGCGGAAGCTTAACATCCGGATAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>FBgn0263240</td>\n",
       "      <td>FBtr0329876</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>7492187</td>\n",
       "      <td>7492189</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>FBgn0263240</td>\n",
       "      <td>FBtr0329876</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>7491049</td>\n",
       "      <td>7491051</td>\n",
       "      <td>-</td>\n",
       "      <td>GACACAGAGTTGCGGCCAGCGCAGCGGAAGCTTAACATCCGGATAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>FBgn0035137</td>\n",
       "      <td>FBtr0072548</td>\n",
       "      <td>3L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>361658</td>\n",
       "      <td>361660</td>\n",
       "      <td>+</td>\n",
       "      <td>CACCAAAATCAAGGTCCAAAGCACTAAAACGTGTGTTTTCCACATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>FBgn0035137</td>\n",
       "      <td>FBtr0072548</td>\n",
       "      <td>3L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>365244</td>\n",
       "      <td>365246</td>\n",
       "      <td>+</td>\n",
       "      <td>TATATATTTAAACATTTTTACAACATATCAAATGCTTCCTCCTTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>FBgn0035137</td>\n",
       "      <td>FBtr0072549</td>\n",
       "      <td>3L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>362275</td>\n",
       "      <td>362277</td>\n",
       "      <td>+</td>\n",
       "      <td>CCCTCGTCCTCCCGCTGCAGTTCCTTCTTCTGCGTTTTGTGTGTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>FBgn0035137</td>\n",
       "      <td>FBtr0072549</td>\n",
       "      <td>3L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>365244</td>\n",
       "      <td>365246</td>\n",
       "      <td>+</td>\n",
       "      <td>TATATATTTAAACATTTTTACAACATATCAAATGCTTCCTCCTTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>8610886</td>\n",
       "      <td>8610888</td>\n",
       "      <td>+</td>\n",
       "      <td>AATATTTTTCTTTTAAAATGAAACTTTCCCAAACTTTTAGTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0339354</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0339354</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>8610886</td>\n",
       "      <td>8610888</td>\n",
       "      <td>+</td>\n",
       "      <td>AATATTTTTCTTTTAAAATGAAACTTTCCCAAACTTTTAGTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0343132</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0343132</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10508347</td>\n",
       "      <td>10508349</td>\n",
       "      <td>-</td>\n",
       "      <td>GGGCCAAGCGGCAGTGACTGTCACCGGGCCGGCTGGACAACCCACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089365</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089365</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10487981</td>\n",
       "      <td>10487983</td>\n",
       "      <td>-</td>\n",
       "      <td>TAAAAACATTACCCAACCAATGCGTATTTACCCTATATGTGAAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089347</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089347</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10482940</td>\n",
       "      <td>10482942</td>\n",
       "      <td>-</td>\n",
       "      <td>ACAATGTGTACTTTTATTTGTTTAATTTTAAAAAAATGCGAAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089350</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089350</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10508347</td>\n",
       "      <td>10508349</td>\n",
       "      <td>-</td>\n",
       "      <td>GGGCCAAGCGGCAGTGACTGTCACCGGGCCGGCTGGACAACCCACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089348</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089348</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10515541</td>\n",
       "      <td>10515543</td>\n",
       "      <td>-</td>\n",
       "      <td>CTAGCTAGGATTATGACTAACTAACTGATAGACAAGGATCTCAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089361</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089361</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10491942</td>\n",
       "      <td>10491944</td>\n",
       "      <td>-</td>\n",
       "      <td>CGAATTCAGAGTAAATTTAATTATGTAATTGTGTATAAAGCCGAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089360</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089360</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10497179</td>\n",
       "      <td>10497181</td>\n",
       "      <td>-</td>\n",
       "      <td>TATGATTTCCTTTACATTTAAATATTGCATGCGTATTTATAAAGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089352</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089352</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10525399</td>\n",
       "      <td>10525401</td>\n",
       "      <td>-</td>\n",
       "      <td>CGATGATGAAATTTCCGTTTCCTCTAGAAATGCCAGCAGCGGTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089357</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089357</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10482940</td>\n",
       "      <td>10482942</td>\n",
       "      <td>-</td>\n",
       "      <td>ACAATGTGTACTTTTATTTGTTTAATTTTAAAAAAATGCGAAATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089366</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089366</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10488773</td>\n",
       "      <td>10488775</td>\n",
       "      <td>-</td>\n",
       "      <td>GCCGCAAGAAAAGCTGTACATAAACGGAATGTCCAATATCATACAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089355</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089355</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10500636</td>\n",
       "      <td>10500638</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTTTACAGTCTTACATATTGTTTTGTTGCTGAGCACACACAGACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089353</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089353</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10517925</td>\n",
       "      <td>10517927</td>\n",
       "      <td>-</td>\n",
       "      <td>ACCCCACATGATGATCCATCGATGTTTTCGTTTTCGTGTATGTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089356</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089356</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10526680</td>\n",
       "      <td>10526682</td>\n",
       "      <td>-</td>\n",
       "      <td>TGTTGGGTTACTGGAAACTTCTTTTGCTTAAATGCTTCGCACTATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089359</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089359</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10510640</td>\n",
       "      <td>10510642</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAACGTTTTAGTAAAAGTGTATTTGGAACAAAGAAATTGCCACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089344</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089344</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10523876</td>\n",
       "      <td>10523878</td>\n",
       "      <td>-</td>\n",
       "      <td>GCGATGCTGGATTTAAGCGCAAGGACTCGCTTAATCGGCATTGCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089362</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089362</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10532004</td>\n",
       "      <td>10532006</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTGCTGTCAGCATGCAGTCCCTACTTTGCTGTAAGTGTTACCAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089346</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089346</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10502141</td>\n",
       "      <td>10502143</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTACTTAGTGAGGGCTAGACCCAAATTCCAAAACTAAAAAAAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089349</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089349</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10520394</td>\n",
       "      <td>10520396</td>\n",
       "      <td>-</td>\n",
       "      <td>TTCAATTGTATAATTTTTAAAACAAAGTTTTAAACTTTTAGGTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089345</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089345</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10505073</td>\n",
       "      <td>10505075</td>\n",
       "      <td>-</td>\n",
       "      <td>TAAAATAAATATGTTTACACCCTGCACTAATTGGCCTTAATTTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089342</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089342</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10520394</td>\n",
       "      <td>10520396</td>\n",
       "      <td>-</td>\n",
       "      <td>TTCAATTGTATAATTTTTAAAACAAAGTTTTAAACTTTTAGGTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089364</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089364</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10490099</td>\n",
       "      <td>10490101</td>\n",
       "      <td>-</td>\n",
       "      <td>AAACTTTTCGTTTCATTTGAACTGTCTTTAAGTTAGCCGCAAAACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089351</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089351</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10510640</td>\n",
       "      <td>10510642</td>\n",
       "      <td>-</td>\n",
       "      <td>AAAACGTTTTAGTAAAAGTGTATTTGGAACAAAGAAATTGCCACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089363</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089363</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10519089</td>\n",
       "      <td>10519091</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCGAGGAGCAAGCAGCGGCAGCAGCAGCAGCTGCGGCCATCGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089343</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0089343</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10505073</td>\n",
       "      <td>10505075</td>\n",
       "      <td>-</td>\n",
       "      <td>TAAAATAAATATGTTTACACCCTGCACTAATTGGCCTTAATTTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0100286</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0100286</td>\n",
       "      <td>2R</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>10522335</td>\n",
       "      <td>10522337</td>\n",
       "      <td>-</td>\n",
       "      <td>TGGTGGCCAGCGGCAACTCCACTACCTCCACCACTGAACTGCTTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>FBgn0030933</td>\n",
       "      <td>FBtr0074621</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18514707</td>\n",
       "      <td>18514709</td>\n",
       "      <td>-</td>\n",
       "      <td>CATCAGTAGACTTACATCAGCAAGAGATTAATATAAACAATTGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>FBgn0030933</td>\n",
       "      <td>FBtr0074621</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18513189</td>\n",
       "      <td>18513191</td>\n",
       "      <td>-</td>\n",
       "      <td>CAAACTGAAATTAAAAAAATAGCATTGAAAATTAATAAGACACGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>FBgn0030933</td>\n",
       "      <td>FBtr0346083</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18514707</td>\n",
       "      <td>18514709</td>\n",
       "      <td>-</td>\n",
       "      <td>CATCAGTAGACTTACATCAGCAAGAGATTAATATAAACAATTGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>FBgn0030933</td>\n",
       "      <td>FBtr0346083</td>\n",
       "      <td>X</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18513189</td>\n",
       "      <td>18513191</td>\n",
       "      <td>-</td>\n",
       "      <td>CAAACTGAAATTAAAAAAATAGCATTGAAAATTAATAAGACACGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>FBgn0036791</td>\n",
       "      <td>FBtr0075101</td>\n",
       "      <td>3L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18588859</td>\n",
       "      <td>18588861</td>\n",
       "      <td>+</td>\n",
       "      <td>ACTACTAACCTAAGGGCGTAATCAAAACTTTGTATTACGGACTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>FBgn0036791</td>\n",
       "      <td>FBtr0075101</td>\n",
       "      <td>3L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>18590140</td>\n",
       "      <td>18590142</td>\n",
       "      <td>+</td>\n",
       "      <td>TCGTCGATCGATTTAATGCTATAAACGGTCTCAATTTTAAGCCAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>FBgn0036126</td>\n",
       "      <td>FBtr0076234</td>\n",
       "      <td>3L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11098108</td>\n",
       "      <td>11098110</td>\n",
       "      <td>-</td>\n",
       "      <td>AAGTATACGTATAAATTCAGATTATATGGGCAACCTTCTTGTTTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>FBgn0036126</td>\n",
       "      <td>FBtr0076234</td>\n",
       "      <td>3L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11097642</td>\n",
       "      <td>11097644</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATAACCGTTCGGCCAGGAGAAGCAAAGTCTGCTCATTGAGATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>FBgn0036126</td>\n",
       "      <td>FBtr0332706</td>\n",
       "      <td>3L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11098108</td>\n",
       "      <td>11098110</td>\n",
       "      <td>-</td>\n",
       "      <td>AAGTATACGTATAAATTCAGATTATATGGGCAACCTTCTTGTTTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>FBgn0036126</td>\n",
       "      <td>FBtr0332706</td>\n",
       "      <td>3L</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>11097642</td>\n",
       "      <td>11097644</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATAACCGTTCGGCCAGGAGAAGCAAAGTCTGCTCATTGAGATTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene_ID Transcript_ID Chromosome  Gene_Region     Start      Stop  \\\n",
       "0    FBgn0000022   FBtr0070072          X  start_codon    370094    370096   \n",
       "1    FBgn0000022   FBtr0070072          X   stop_codon    370697    370699   \n",
       "2    FBgn0003963   FBtr0078063         2L  start_codon    524038    524040   \n",
       "3    FBgn0003963   FBtr0078063         2L   stop_codon    539954    539956   \n",
       "4    FBgn0003963   FBtr0310027         2L  start_codon    491316    491318   \n",
       "5    FBgn0003963   FBtr0310027         2L   stop_codon    539954    539956   \n",
       "6    FBgn0003963   FBtr0329895         2L  start_codon    524038    524040   \n",
       "7    FBgn0003963   FBtr0329895         2L   stop_codon    540017    540019   \n",
       "8    FBgn0003963   FBtr0333012         2L  start_codon    525418    525420   \n",
       "9    FBgn0003963   FBtr0333012         2L   stop_codon    539954    539956   \n",
       "10   FBgn0003963   FBtr0333013         2L  start_codon    491316    491318   \n",
       "11   FBgn0003963   FBtr0333013         2L   stop_codon    539954    539956   \n",
       "12   FBgn0032150   FBtr0079917         2L  start_codon   9793004   9793006   \n",
       "13   FBgn0032150   FBtr0079917         2L   stop_codon   9791982   9791984   \n",
       "14   FBgn0032150   FBtr0331285         2L  start_codon   9792792   9792794   \n",
       "15   FBgn0032150   FBtr0331285         2L   stop_codon   9791982   9791984   \n",
       "16   FBgn0032150   FBtr0331286         2L  start_codon   9792792   9792794   \n",
       "17   FBgn0032150   FBtr0331286         2L   stop_codon   9791982   9791984   \n",
       "18   FBgn0037584   FBtr0301759         3R  start_codon   8362880   8362882   \n",
       "19   FBgn0037584   FBtr0301759         3R   stop_codon   8363892   8363894   \n",
       "20   FBgn0004652   FBtr0083648         3R  start_codon  18506517  18506519   \n",
       "21   FBgn0004652   FBtr0083648         3R   stop_codon  18426145  18426147   \n",
       "22   FBgn0004652   FBtr0083649         3R  start_codon  18503977  18503979   \n",
       "23   FBgn0004652   FBtr0083649         3R   stop_codon  18426145  18426147   \n",
       "24   FBgn0004652   FBtr0083650         3R  start_codon  18478069  18478071   \n",
       "25   FBgn0004652   FBtr0083650         3R   stop_codon  18422638  18422640   \n",
       "26   FBgn0004652   FBtr0083651         3R  start_codon  18435766  18435768   \n",
       "27   FBgn0004652   FBtr0083651         3R   stop_codon  18426145  18426147   \n",
       "28   FBgn0004652   FBtr0083652         3R  start_codon  18435766  18435768   \n",
       "29   FBgn0004652   FBtr0083652         3R   stop_codon  18426145  18426147   \n",
       "30   FBgn0004652   FBtr0083640         3R  start_codon  18435766  18435768   \n",
       "31   FBgn0004652   FBtr0083640         3R   stop_codon  18422638  18422640   \n",
       "32   FBgn0004652   FBtr0083641         3R  start_codon  18515330  18515332   \n",
       "33   FBgn0004652   FBtr0083641         3R   stop_codon  18416777  18416779   \n",
       "34   FBgn0004652   FBtr0083642         3R  start_codon  18435766  18435768   \n",
       "35   FBgn0004652   FBtr0083642         3R   stop_codon  18426145  18426147   \n",
       "36   FBgn0004652   FBtr0083643         3R  start_codon  18515330  18515332   \n",
       "37   FBgn0004652   FBtr0083643         3R   stop_codon  18426145  18426147   \n",
       "38   FBgn0004652   FBtr0083644         3R  start_codon  18435766  18435768   \n",
       "39   FBgn0004652   FBtr0083644         3R   stop_codon  18416777  18416779   \n",
       "40   FBgn0004652   FBtr0083645         3R  start_codon  18515330  18515332   \n",
       "41   FBgn0004652   FBtr0083645         3R   stop_codon  18422638  18422640   \n",
       "42   FBgn0004652   FBtr0083646         3R  start_codon  18435766  18435768   \n",
       "43   FBgn0004652   FBtr0083646         3R   stop_codon  18430824  18430826   \n",
       "44   FBgn0004652   FBtr0083647         3R  start_codon  18435766  18435768   \n",
       "45   FBgn0004652   FBtr0083647         3R   stop_codon  18431447  18431449   \n",
       "46   FBgn0004652   FBtr0301284         3R  start_codon  18515330  18515332   \n",
       "47   FBgn0004652   FBtr0301284         3R   stop_codon  18426145  18426147   \n",
       "48   FBgn0004652   FBtr0330040         3R  start_codon  18435766  18435768   \n",
       "49   FBgn0004652   FBtr0330040         3R   stop_codon  18422077  18422079   \n",
       "50   FBgn0002733   FBtr0084982         3R  start_codon  26005891  26005893   \n",
       "51   FBgn0002733   FBtr0084982         3R   stop_codon  26005306  26005308   \n",
       "52   FBgn0039509   FBtr0085173         3R  start_codon  27288467  27288469   \n",
       "53   FBgn0039509   FBtr0085173         3R   stop_codon  27287498  27287500   \n",
       "54   FBgn0039712   FBtr0085523         3R  start_codon  29861352  29861354   \n",
       "55   FBgn0039712   FBtr0085523         3R   stop_codon  29860354  29860356   \n",
       "56   FBgn0039712   FBtr0335005         3R  start_codon  29861352  29861354   \n",
       "57   FBgn0039712   FBtr0335005         3R   stop_codon  29860354  29860356   \n",
       "58   FBgn0030008   FBtr0071155          X  start_codon   8145949   8145951   \n",
       "59   FBgn0030008   FBtr0071155          X   stop_codon   8147344   8147346   \n",
       "60   FBgn0015371   FBtr0111006         2R  start_codon  15138215  15138217   \n",
       "61   FBgn0015371   FBtr0111006         2R   stop_codon  15143449  15143451   \n",
       "62   FBgn0015371   FBtr0111007         2R  start_codon  15138215  15138217   \n",
       "63   FBgn0015371   FBtr0111007         2R   stop_codon  15140679  15140681   \n",
       "64   FBgn0015371   FBtr0087415         2R  start_codon  15138215  15138217   \n",
       "65   FBgn0015371   FBtr0087415         2R   stop_codon  15143449  15143451   \n",
       "66   FBgn0015371   FBtr0087416         2R  start_codon  15138215  15138217   \n",
       "67   FBgn0015371   FBtr0087416         2R   stop_codon  15143449  15143451   \n",
       "68   FBgn0015371   FBtr0305213         2R  start_codon  15138215  15138217   \n",
       "69   FBgn0015371   FBtr0305213         2R   stop_codon  15143449  15143451   \n",
       "70   FBgn0015371   FBtr0305214         2R  start_codon  15138215  15138217   \n",
       "71   FBgn0015371   FBtr0305214         2R   stop_codon  15143449  15143451   \n",
       "72   FBgn0015371   FBtr0305215         2R  start_codon  15138215  15138217   \n",
       "73   FBgn0015371   FBtr0305215         2R   stop_codon  15143449  15143451   \n",
       "74   FBgn0015371   FBtr0305216         2R  start_codon  15138215  15138217   \n",
       "75   FBgn0015371   FBtr0305216         2R   stop_codon  15143449  15143451   \n",
       "76   FBgn0015371   FBtr0305217         2R  start_codon  15138215  15138217   \n",
       "77   FBgn0015371   FBtr0305217         2R   stop_codon  15140382  15140384   \n",
       "78   FBgn0022740   FBtr0089547         2R  start_codon  17757575  17757577   \n",
       "79   FBgn0022740   FBtr0089547         2R   stop_codon  17760336  17760338   \n",
       "80   FBgn0004895   FBtr0073192         3L  start_codon   4019094   4019096   \n",
       "81   FBgn0004895   FBtr0073192         3L   stop_codon   4021177   4021179   \n",
       "82   FBgn0004895   FBtr0309789         3L  start_codon   4019094   4019096   \n",
       "83   FBgn0004895   FBtr0309789         3L   stop_codon   4021177   4021179   \n",
       "84   FBgn0003720   FBtr0085709         3R  start_codon  30852547  30852549   \n",
       "85   FBgn0003720   FBtr0085709         3R   stop_codon  30854024  30854026   \n",
       "86   FBgn0052831   FBtr0091683         2L  start_codon  11278451  11278453   \n",
       "87   FBgn0052831   FBtr0091683         2L   stop_codon  11279461  11279463   \n",
       "88   FBgn0052831   FBtr0091680         2L  start_codon  11278113  11278115   \n",
       "89   FBgn0052831   FBtr0091680         2L   stop_codon  11279461  11279463   \n",
       "90   FBgn0052831   FBtr0091681         2L  start_codon  11278370  11278372   \n",
       "91   FBgn0052831   FBtr0091681         2L   stop_codon  11279461  11279463   \n",
       "92   FBgn0052831   FBtr0091682         2L  start_codon  11278113  11278115   \n",
       "93   FBgn0052831   FBtr0091682         2L   stop_codon  11279461  11279463   \n",
       "94   FBgn0263240   FBtr0088984         2R  start_codon   7492187   7492189   \n",
       "95   FBgn0263240   FBtr0088984         2R   stop_codon   7491049   7491051   \n",
       "96   FBgn0263240   FBtr0329876         2R  start_codon   7492187   7492189   \n",
       "97   FBgn0263240   FBtr0329876         2R   stop_codon   7491049   7491051   \n",
       "98   FBgn0035137   FBtr0072548         3L  start_codon    361658    361660   \n",
       "99   FBgn0035137   FBtr0072548         3L   stop_codon    365244    365246   \n",
       "100  FBgn0035137   FBtr0072549         3L  start_codon    362275    362277   \n",
       "101  FBgn0035137   FBtr0072549         3L   stop_codon    365244    365246   \n",
       "102  FBgn0267792   FBtr0088692         2R  start_codon   8600582   8600584   \n",
       "103  FBgn0267792   FBtr0088692         2R   stop_codon   8610886   8610888   \n",
       "104  FBgn0267792   FBtr0339354         2R  start_codon   8600582   8600584   \n",
       "105  FBgn0267792   FBtr0339354         2R   stop_codon   8610886   8610888   \n",
       "106  FBgn0283521   FBtr0343132         2R  start_codon  10533747  10533749   \n",
       "107  FBgn0283521   FBtr0343132         2R   stop_codon  10508347  10508349   \n",
       "108  FBgn0283521   FBtr0089365         2R  start_codon  10533747  10533749   \n",
       "109  FBgn0283521   FBtr0089365         2R   stop_codon  10487981  10487983   \n",
       "110  FBgn0283521   FBtr0089347         2R  start_codon  10533747  10533749   \n",
       "111  FBgn0283521   FBtr0089347         2R   stop_codon  10482940  10482942   \n",
       "112  FBgn0283521   FBtr0089350         2R  start_codon  10533747  10533749   \n",
       "113  FBgn0283521   FBtr0089350         2R   stop_codon  10508347  10508349   \n",
       "114  FBgn0283521   FBtr0089348         2R  start_codon  10533747  10533749   \n",
       "115  FBgn0283521   FBtr0089348         2R   stop_codon  10515541  10515543   \n",
       "116  FBgn0283521   FBtr0089361         2R  start_codon  10533747  10533749   \n",
       "117  FBgn0283521   FBtr0089361         2R   stop_codon  10491942  10491944   \n",
       "118  FBgn0283521   FBtr0089360         2R  start_codon  10533747  10533749   \n",
       "119  FBgn0283521   FBtr0089360         2R   stop_codon  10497179  10497181   \n",
       "120  FBgn0283521   FBtr0089352         2R  start_codon  10533747  10533749   \n",
       "121  FBgn0283521   FBtr0089352         2R   stop_codon  10525399  10525401   \n",
       "122  FBgn0283521   FBtr0089357         2R  start_codon  10533747  10533749   \n",
       "123  FBgn0283521   FBtr0089357         2R   stop_codon  10482940  10482942   \n",
       "124  FBgn0283521   FBtr0089366         2R  start_codon  10533747  10533749   \n",
       "125  FBgn0283521   FBtr0089366         2R   stop_codon  10488773  10488775   \n",
       "126  FBgn0283521   FBtr0089355         2R  start_codon  10533747  10533749   \n",
       "127  FBgn0283521   FBtr0089355         2R   stop_codon  10500636  10500638   \n",
       "128  FBgn0283521   FBtr0089353         2R  start_codon  10533747  10533749   \n",
       "129  FBgn0283521   FBtr0089353         2R   stop_codon  10517925  10517927   \n",
       "130  FBgn0283521   FBtr0089356         2R  start_codon  10533747  10533749   \n",
       "131  FBgn0283521   FBtr0089356         2R   stop_codon  10526680  10526682   \n",
       "132  FBgn0283521   FBtr0089359         2R  start_codon  10533747  10533749   \n",
       "133  FBgn0283521   FBtr0089359         2R   stop_codon  10510640  10510642   \n",
       "134  FBgn0283521   FBtr0089344         2R  start_codon  10533747  10533749   \n",
       "135  FBgn0283521   FBtr0089344         2R   stop_codon  10523876  10523878   \n",
       "136  FBgn0283521   FBtr0089362         2R  start_codon  10533747  10533749   \n",
       "137  FBgn0283521   FBtr0089362         2R   stop_codon  10532004  10532006   \n",
       "138  FBgn0283521   FBtr0089346         2R  start_codon  10533747  10533749   \n",
       "139  FBgn0283521   FBtr0089346         2R   stop_codon  10502141  10502143   \n",
       "140  FBgn0283521   FBtr0089349         2R  start_codon  10533747  10533749   \n",
       "141  FBgn0283521   FBtr0089349         2R   stop_codon  10520394  10520396   \n",
       "142  FBgn0283521   FBtr0089345         2R  start_codon  10533747  10533749   \n",
       "143  FBgn0283521   FBtr0089345         2R   stop_codon  10505073  10505075   \n",
       "144  FBgn0283521   FBtr0089342         2R  start_codon  10533747  10533749   \n",
       "145  FBgn0283521   FBtr0089342         2R   stop_codon  10520394  10520396   \n",
       "146  FBgn0283521   FBtr0089364         2R  start_codon  10533747  10533749   \n",
       "147  FBgn0283521   FBtr0089364         2R   stop_codon  10490099  10490101   \n",
       "148  FBgn0283521   FBtr0089351         2R  start_codon  10533747  10533749   \n",
       "149  FBgn0283521   FBtr0089351         2R   stop_codon  10510640  10510642   \n",
       "150  FBgn0283521   FBtr0089363         2R  start_codon  10533747  10533749   \n",
       "151  FBgn0283521   FBtr0089363         2R   stop_codon  10519089  10519091   \n",
       "152  FBgn0283521   FBtr0089343         2R  start_codon  10533747  10533749   \n",
       "153  FBgn0283521   FBtr0089343         2R   stop_codon  10505073  10505075   \n",
       "154  FBgn0283521   FBtr0100286         2R  start_codon  10533747  10533749   \n",
       "155  FBgn0283521   FBtr0100286         2R   stop_codon  10522335  10522337   \n",
       "156  FBgn0030933   FBtr0074621          X  start_codon  18514707  18514709   \n",
       "157  FBgn0030933   FBtr0074621          X   stop_codon  18513189  18513191   \n",
       "158  FBgn0030933   FBtr0346083          X  start_codon  18514707  18514709   \n",
       "159  FBgn0030933   FBtr0346083          X   stop_codon  18513189  18513191   \n",
       "160  FBgn0036791   FBtr0075101         3L  start_codon  18588859  18588861   \n",
       "161  FBgn0036791   FBtr0075101         3L   stop_codon  18590140  18590142   \n",
       "162  FBgn0036126   FBtr0076234         3L  start_codon  11098108  11098110   \n",
       "163  FBgn0036126   FBtr0076234         3L   stop_codon  11097642  11097644   \n",
       "164  FBgn0036126   FBtr0332706         3L  start_codon  11098108  11098110   \n",
       "165  FBgn0036126   FBtr0332706         3L   stop_codon  11097642  11097644   \n",
       "\n",
       "    Strand                                      Reference_Seq  \n",
       "0        +  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...  \n",
       "1        +  GACATTGTGTCGTTCGTATGTCGCCCATTGAGACCGCCAATGGAGG...  \n",
       "2        +  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...  \n",
       "3        +  CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...  \n",
       "4        +  CACTGATAAAAAATACCCTAAATTATCGCAACGCAAATGGTAAAGT...  \n",
       "5        +  CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...  \n",
       "6        +  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...  \n",
       "7        +  TTTTTTAATGTTCTGCACTTGCTTAATTACCAGCATGTACCATCTG...  \n",
       "8        +  ATCATTGGCCAAAAAAGTAGGAATTCTGGGAGGAGGGCATGCCAAA...  \n",
       "9        +  CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...  \n",
       "10       +  CACTGATAAAAAATACCCTAAATTATCGCAACGCAAATGGTAAAGT...  \n",
       "11       +  CCTGTAATTTTTCGATGAGGCCCAAAGTGGCCACGTGGCACCCAGA...  \n",
       "12       -  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...  \n",
       "13       -  AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...  \n",
       "14       -  TAAAATGCGGTGACCTGCAAAAGGAGCCATAAACTAAGCGTCCTAC...  \n",
       "15       -  AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...  \n",
       "16       -  TAAAATGCGGTGACCTGCAAAAGGAGCCATAAACTAAGCGTCCTAC...  \n",
       "17       -  AAAATCTTCCCGAATATTATGCATGTAAAAAAGGAAATGGCTAAGT...  \n",
       "18       +  GATATACATCTATACATTATACGGATATTTATAGCATATACCTGCG...  \n",
       "19       +  GCACGCCCCAGTTAAACCACAACAAGTATACGGCTCCATCGCCACC...  \n",
       "20       -  CCTAATGCAACTTATTGAATTTTCGAGTGCTTGGCTTAGTTGATTT...  \n",
       "21       -  CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...  \n",
       "22       -  TATTATATAGTTATTGTAGTTATTGTTTATAGTTATAACCATTAAT...  \n",
       "23       -  CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...  \n",
       "24       -  TGTGTCTAACATGTGAGGAGCATTTCATAATGGTTCTTTTTCGGTC...  \n",
       "25       -  TCACTCTCTGTCTCCCTCCAACTTATTGTACATTTGCCAAACAAGA...  \n",
       "26       -  AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...  \n",
       "27       -  CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...  \n",
       "28       -  AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...  \n",
       "29       -  CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...  \n",
       "30       -  AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...  \n",
       "31       -  TCACTCTCTGTCTCCCTCCAACTTATTGTACATTTGCCAAACAAGA...  \n",
       "32       -  GTAAAGTCGGCTCCACATGAGTGGGCGTAAGGCGTTTCGTGGGCGG...  \n",
       "33       -  ACGAGTACCCTACTTTTATCTATTTGAATATATATATATTATGTAT...  \n",
       "34       -  AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...  \n",
       "35       -  CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...  \n",
       "36       -  GTAAAGTCGGCTCCACATGAGTGGGCGTAAGGCGTTTCGTGGGCGG...  \n",
       "37       -  CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...  \n",
       "38       -  AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...  \n",
       "39       -  ACGAGTACCCTACTTTTATCTATTTGAATATATATATATTATGTAT...  \n",
       "40       -  GTAAAGTCGGCTCCACATGAGTGGGCGTAAGGCGTTTCGTGGGCGG...  \n",
       "41       -  TCACTCTCTGTCTCCCTCCAACTTATTGTACATTTGCCAAACAAGA...  \n",
       "42       -  AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...  \n",
       "43       -  CACTGTACAGAACATTCTTGTAAAGCCTTTTTGTTTAAATCAGCTA...  \n",
       "44       -  AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...  \n",
       "45       -  CCAGCTATTTCGGGTTTCGCGTTGTTATCAGTGGGCATGAAGTTTT...  \n",
       "46       -  GTAAAGTCGGCTCCACATGAGTGGGCGTAAGGCGTTTCGTGGGCGG...  \n",
       "47       -  CTGTTCTGCATAATCGAGAAAATAAATGACGTCATATGGAATACTC...  \n",
       "48       -  AAAAAGTTGTGGGCGGCGAACGAAAACGAAAGTTTACGTTTAAAGT...  \n",
       "49       -  CAAAAATCCCGAACCGAGCAGAGGTCTCTGCCCCATATAACACCAT...  \n",
       "50       -  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...  \n",
       "51       -  TCCCCAAGCGAACCCAAAACCCCCAACCGAACTGCAGCTGATCCCC...  \n",
       "52       -  GTTCGTCCAGCAGCACCTCGCTAACATTCTTCGAGCGATTGTATTT...  \n",
       "53       -  GATTTCTCCAAGGATTAACACGGGAATTACGTCAATTGAAACGGGC...  \n",
       "54       -  TTGGTTCGCACCAGCTCGTTGTTGGAGGCGTTGTACACAACATCGG...  \n",
       "55       -  CAATTTGCGAGAGTCGACTTACCCATGTTCGAGTAAAAAGCGAGAA...  \n",
       "56       -  TTGGTTCGCACCAGCTCGTTGTTGGAGGCGTTGTACACAACATCGG...  \n",
       "57       -  CAATTTGCGAGAGTCGACTTACCCATGTTCGAGTAAAAAGCGAGAA...  \n",
       "58       +  GTGCGTGTCAAATAAATTCGATGGAAACCCGTGAAACGAGGACCGA...  \n",
       "59       +  GGGCGTGCAGGCGGACGTAGTCGACGTTGGCGAGGCAGTGTCTCAA...  \n",
       "60       +  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...  \n",
       "61       +  ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...  \n",
       "62       +  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...  \n",
       "63       +  CGCCTCTAGAAATGCCGTTGCCATCATCGCCTTGGTGGGTCCCATA...  \n",
       "64       +  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...  \n",
       "65       +  ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...  \n",
       "66       +  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...  \n",
       "67       +  ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...  \n",
       "68       +  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...  \n",
       "69       +  ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...  \n",
       "70       +  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...  \n",
       "71       +  ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...  \n",
       "72       +  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...  \n",
       "73       +  ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...  \n",
       "74       +  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...  \n",
       "75       +  ATAGACAACCGTCTAAACAATATATTTAAGCAGTTAGATTGTAAAA...  \n",
       "76       +  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...  \n",
       "77       +  CTCTCGTGTCGTCCATTGCGAAGGCAGTTTATTGTTAAACAGGAGT...  \n",
       "78       +  TTCTAGTGGGGGGCCCGGCTGAATGGCTGACTCCGGTCATGGGATG...  \n",
       "79       +  CTGTCAGATAAACCTATCAGCTAAACTAATTTCAAGTGGACCTGGA...  \n",
       "80       +  TCATTTATTCGATTTGTTTATGAGTTTTTAGATGTGCATTATACCT...  \n",
       "81       +  ATATAAACATTACATCACTTATATTTAGGGTTTATGGTCATGGTAT...  \n",
       "82       +  TCATTTATTCGATTTGTTTATGAGTTTTTAGATGTGCATTATACCT...  \n",
       "83       +  ATATAAACATTACATCACTTATATTTAGGGTTTATGGTCATGGTAT...  \n",
       "84       +  TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...  \n",
       "85       +  ACAGAATGATGAGGATTAGGCCAACACACACATCCATCGAAAGGAC...  \n",
       "86       +  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...  \n",
       "87       +  TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...  \n",
       "88       +  CTGGCGATCCACAGACTCGTCCTCCACATCGCACAATTCATTTGTC...  \n",
       "89       +  TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...  \n",
       "90       +  CTGATCAAAATAAAAGTATGAAGTACAAATTACAGCAATGAGATTG...  \n",
       "91       +  TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...  \n",
       "92       +  CTGGCGATCCACAGACTCGTCCTCCACATCGCACAATTCATTTGTC...  \n",
       "93       +  TAATTTTTGAACACAGTTACATAATATATTTGTTTCTTGTAAGCTT...  \n",
       "94       -  TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...  \n",
       "95       -  GACACAGAGTTGCGGCCAGCGCAGCGGAAGCTTAACATCCGGATAA...  \n",
       "96       -  TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...  \n",
       "97       -  GACACAGAGTTGCGGCCAGCGCAGCGGAAGCTTAACATCCGGATAA...  \n",
       "98       +  CACCAAAATCAAGGTCCAAAGCACTAAAACGTGTGTTTTCCACATG...  \n",
       "99       +  TATATATTTAAACATTTTTACAACATATCAAATGCTTCCTCCTTTA...  \n",
       "100      +  CCCTCGTCCTCCCGCTGCAGTTCCTTCTTCTGCGTTTTGTGTGTCC...  \n",
       "101      +  TATATATTTAAACATTTTTACAACATATCAAATGCTTCCTCCTTTA...  \n",
       "102      +  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...  \n",
       "103      +  AATATTTTTCTTTTAAAATGAAACTTTCCCAAACTTTTAGTTTTTT...  \n",
       "104      +  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...  \n",
       "105      +  AATATTTTTCTTTTAAAATGAAACTTTCCCAAACTTTTAGTTTTTT...  \n",
       "106      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "107      -  GGGCCAAGCGGCAGTGACTGTCACCGGGCCGGCTGGACAACCCACG...  \n",
       "108      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "109      -  TAAAAACATTACCCAACCAATGCGTATTTACCCTATATGTGAAAGT...  \n",
       "110      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "111      -  ACAATGTGTACTTTTATTTGTTTAATTTTAAAAAAATGCGAAATTA...  \n",
       "112      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "113      -  GGGCCAAGCGGCAGTGACTGTCACCGGGCCGGCTGGACAACCCACG...  \n",
       "114      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "115      -  CTAGCTAGGATTATGACTAACTAACTGATAGACAAGGATCTCAAAT...  \n",
       "116      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "117      -  CGAATTCAGAGTAAATTTAATTATGTAATTGTGTATAAAGCCGAGA...  \n",
       "118      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "119      -  TATGATTTCCTTTACATTTAAATATTGCATGCGTATTTATAAAGCG...  \n",
       "120      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "121      -  CGATGATGAAATTTCCGTTTCCTCTAGAAATGCCAGCAGCGGTGGA...  \n",
       "122      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "123      -  ACAATGTGTACTTTTATTTGTTTAATTTTAAAAAAATGCGAAATTA...  \n",
       "124      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "125      -  GCCGCAAGAAAAGCTGTACATAAACGGAATGTCCAATATCATACAT...  \n",
       "126      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "127      -  TTTTTACAGTCTTACATATTGTTTTGTTGCTGAGCACACACAGACA...  \n",
       "128      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "129      -  ACCCCACATGATGATCCATCGATGTTTTCGTTTTCGTGTATGTTTT...  \n",
       "130      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "131      -  TGTTGGGTTACTGGAAACTTCTTTTGCTTAAATGCTTCGCACTATT...  \n",
       "132      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "133      -  AAAACGTTTTAGTAAAAGTGTATTTGGAACAAAGAAATTGCCACAA...  \n",
       "134      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "135      -  GCGATGCTGGATTTAAGCGCAAGGACTCGCTTAATCGGCATTGCAA...  \n",
       "136      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "137      -  GGTGCTGTCAGCATGCAGTCCCTACTTTGCTGTAAGTGTTACCAAA...  \n",
       "138      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "139      -  GGTACTTAGTGAGGGCTAGACCCAAATTCCAAAACTAAAAAAAAAA...  \n",
       "140      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "141      -  TTCAATTGTATAATTTTTAAAACAAAGTTTTAAACTTTTAGGTTTT...  \n",
       "142      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "143      -  TAAAATAAATATGTTTACACCCTGCACTAATTGGCCTTAATTTGCT...  \n",
       "144      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "145      -  TTCAATTGTATAATTTTTAAAACAAAGTTTTAAACTTTTAGGTTTT...  \n",
       "146      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "147      -  AAACTTTTCGTTTCATTTGAACTGTCTTTAAGTTAGCCGCAAAACG...  \n",
       "148      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "149      -  AAAACGTTTTAGTAAAAGTGTATTTGGAACAAAGAAATTGCCACAA...  \n",
       "150      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "151      -  GGTCGAGGAGCAAGCAGCGGCAGCAGCAGCAGCTGCGGCCATCGCT...  \n",
       "152      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "153      -  TAAAATAAATATGTTTACACCCTGCACTAATTGGCCTTAATTTGCT...  \n",
       "154      -  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...  \n",
       "155      -  TGGTGGCCAGCGGCAACTCCACTACCTCCACCACTGAACTGCTTAA...  \n",
       "156      -  CATCAGTAGACTTACATCAGCAAGAGATTAATATAAACAATTGCTT...  \n",
       "157      -  CAAACTGAAATTAAAAAAATAGCATTGAAAATTAATAAGACACGCT...  \n",
       "158      -  CATCAGTAGACTTACATCAGCAAGAGATTAATATAAACAATTGCTT...  \n",
       "159      -  CAAACTGAAATTAAAAAAATAGCATTGAAAATTAATAAGACACGCT...  \n",
       "160      +  ACTACTAACCTAAGGGCGTAATCAAAACTTTGTATTACGGACTTTT...  \n",
       "161      +  TCGTCGATCGATTTAATGCTATAAACGGTCTCAATTTTAAGCCAAT...  \n",
       "162      -  AAGTATACGTATAAATTCAGATTATATGGGCAACCTTCTTGTTTAA...  \n",
       "163      -  CCATAACCGTTCGGCCAGGAGAAGCAAAGTCTGCTCATTGAGATTG...  \n",
       "164      -  AAGTATACGTATAAATTCAGATTATATGGGCAACCTTCTTGTTTAA...  \n",
       "165      -  CCATAACCGTTCGGCCAGGAGAAGCAAAGTCTGCTCATTGAGATTG...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# testing function to identify known transcripts of transcription factors and extract their genomic information\n",
    "# this cell takes about 50s to run but you can avoid it by running the next cell and loading the output from this cell from MockMaterials\n",
    "from sgRNAutils import make_dataframe_from_TFs_list\n",
    "TFsdf = make_dataframe_from_TFs_list('inputfiles/mockMaterials/TFsTruncatedLong.xlsx', refSeqPerChromosome)\n",
    "display(TFsdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG 225 225\n",
      "ACGTTATACTATCTCTTAAAATGGCTTTGGGCAGCGAAAATCA\n",
      "ACGTTATACTATCTCTTAAAXXXGCTTTGGGCAGCGAAAATCA\n",
      "ATG 225 225\n",
      "ATACACATATAGGGAAAACAATGCTTAGCAGCAATACAAGAGG\n",
      "ATACACATATAGGGAAAACAXXXCTTAGCAGCAATACAAGAGG\n",
      "CAT 225 225\n",
      "CGGCACAGTTTACGCAGGTCCATGGGTTTTATTATTTAATTAA\n",
      "CGGCACAGTTTACGCAGGTCXXXGGGTTTTATTATTTAATTAA\n",
      "ATG 225 225\n",
      "TTGCAAGCTTTTAACTGAAAATGTCGCCACCATCTGGTGAATT\n",
      "TTGCAAGCTTTTAACTGAAAXXXTCGCCACCATCTGGTGAATT\n",
      "CAT 225 225\n",
      "CGGACTGAGTTCTTTTGCAGCATTTATGGGAAACTAAAACCGA\n",
      "CGGACTGAGTTCTTTTGCAGXXXTTATGGGAAACTAAAACCGA\n",
      "CAT 225 225\n",
      "GACATCTCCATTTCCAGAACCATTTTGTTGTAGTTTGGTTTTG\n",
      "GACATCTCCATTTCCAGAACXXXTTTGTTGTAGTTTGGTTTTG\n",
      "CAT 225 225\n",
      "AACGCGTTGTTATTATCGCTCATCGTCAACTGCGACGTTTTGG\n",
      "AACGCGTTGTTATTATCGCTXXXCGTCAACTGCGACGTTTTGG\n",
      "ATG 225 225\n",
      "GCCAGGTGGGAGCCGCAATCATGCTCAAGTCCCTCAAGCCGGA\n",
      "GCCAGGTGGGAGCCGCAATCXXXCTCAAGTCCCTCAAGCCGGA\n",
      "ATG 225 225\n",
      "GGCAAAGTGAAGCACCTGCAATGGCCACACTAATACCAGTGAA\n",
      "GGCAAAGTGAAGCACCTGCAXXXGCCACACTAATACCAGTGAA\n",
      "ATG 225 225\n",
      "CACCAACTTGTTTTTGAACAATGCCACGCAAGAAGCGCAGTTC\n",
      "CACCAACTTGTTTTTGAACAXXXCCACGCAAGAAGCGCAGTTC\n",
      "ATG 225 225\n",
      "CTGGGATTTCGTTTTGAAAAATGTTACCCTCTTGCTATGCCAA\n",
      "CTGGGATTTCGTTTTGAAAAXXXTTACCCTCTTGCTATGCCAA\n",
      "ATG 225 225\n",
      "AAATTACGCCCCACATCGGTATGCAGTCGTCGGAGGGTTCACC\n",
      "AAATTACGCCCCACATCGGTXXXCAGTCGTCGGAGGGTTCACC\n",
      "ATG 225 225\n",
      "GCAGACTGGGCGATCTAACCATGCAGCTGCATCCCAATGCGGA\n",
      "GCAGACTGGGCGATCTAACCXXXCAGCTGCATCCCAATGCGGA\n",
      "CAT 225 225\n",
      "GACGCATTGTAGGCCATCTCCATGTCCAGATCCGAGTGGCCAA\n",
      "GACGCATTGTAGGCCATCTCXXXGTCCAGATCCGAGTGGCCAA\n",
      "ATG 225 225\n",
      "GACCAAGGAAGGCACCCACGATGAGCGATGCGACCTATCCGGG\n",
      "GACCAAGGAAGGCACCCACGXXXAGCGATGCGACCTATCCGGG\n",
      "ATG 225 225\n",
      "TCCAGTAAAGCGCAGCAGCAATGATGCCGCTGCAGGAGTCGCC\n",
      "TCCAGTAAAGCGCAGCAGCAXXXATGCCGCTGCAGGAGTCGCC\n",
      "CAT 225 225\n",
      "CAAAACTGCTGATCGTCATCCATTCTGGTTTCTCGGGTGGTTG\n",
      "CAAAACTGCTGATCGTCATCXXXTCTGGTTTCTCGGGTGGTTG\n",
      "CAT 225 225\n",
      "ATGGGCTTTTGCGGCTTGTTCATCTCGTGTGTGGTGTGCGTGT\n",
      "ATGGGCTTTTGCGGCTTGTTXXXCTCGTGTGTGGTGTGCGTGT\n",
      "ATG 225 225\n",
      "AAAGTTTCCATATCTCAATCATGTTCGACGAGCCCATATCCAG\n",
      "AAAGTTTCCATATCTCAATCXXXTTCGACGAGCCCATATCCAG\n"
     ]
    }
   ],
   "source": [
    "# test function to extract 225 bp of the genome left and right of the start/stop codon of a transcript\n",
    "from sgRNAutils import make_homology_arm_fragments\n",
    "import pandas as pd\n",
    "TFsdf = pd.read_excel('inputfiles/mockMaterials/TFsdfTruncatedLong.xlsx')\n",
    "TFsdf = make_homology_arm_fragments(TFsdf, refSeqPerChromosome)\n",
    "TFsdf = TFsdf.reset_index()\n",
    "del TFsdf[\"index\"]\n",
    "\n",
    "# check that the homology arms are 225 bp long\n",
    "# check that the sequence between the homology arms is a start or a stop codon in forward or reverse complement\n",
    "for index, row in TFsdf.iterrows():\n",
    "    print(refSeqPerChromosome[row[\"Chromosome\"]][row['Start']-1:row['Stop']], len(TFsdf.iloc[0][\"HAL\"]), len(TFsdf.iloc[0][\"HAR\"]))\n",
    "    # compare the region around the start/stop to the end of HAL and the beginning of HAR\n",
    "    print(refSeqPerChromosome[row[\"Chromosome\"]][row['Start']-21:row['Stop']+20])\n",
    "    print(TFsdf.iloc[index][\"HAL\"][-20:] + \"XXX\" + TFsdf.iloc[index][\"HAR\"][:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stringency 0\n",
      "stringency 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "      <td>TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...</td>\n",
       "      <td>GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0070072          X  start_codon  370094  370096      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "stringency 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "      <td>AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...</td>\n",
       "      <td>CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  524040      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9792987</td>\n",
       "      <td>9793009</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CACAGTTTACGCAGGTCCATGGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  9792987  9793009   2L      +  CACAGTTTACGCAGGTCCATGGG  FBgn0032150   \n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "1   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "1  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "1  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...  \n",
       "1  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n",
      "stringency 3\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8362874</td>\n",
       "      <td>8362896</td>\n",
       "      <td>3R</td>\n",
       "      <td>+</td>\n",
       "      <td>CTGAAAATGTCGCCACCATCTGG</td>\n",
       "      <td>FBgn0037584</td>\n",
       "      <td>FBtr0301759</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8362880</td>\n",
       "      <td>8362882</td>\n",
       "      <td>+</td>\n",
       "      <td>GATATACATCTATACATTATACGGATATTTATAGCATATACCTGCG...</td>\n",
       "      <td>AATATTTGTTGGCTGTTAGTATTAAATATATTCATATATTTATATT...</td>\n",
       "      <td>TCGCCACCATCTGGTGAATTCCGTTGCCGAGTTTGTCTTAAACAGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8362874  8362896   3R      +  CTGAAAATGTCGCCACCATCTGG  FBgn0037584   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0301759         3R  start_codon  8362880  8362882      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GATATACATCTATACATTATACGGATATTTATAGCATATACCTGCG...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AATATTTGTTGGCTGTTAGTATTAAATATATTCATATATTTATATT...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  TCGCCACCATCTGGTGAATTCCGTTGCCGAGTTTGTCTTAAACAGG...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18506503</td>\n",
       "      <td>18506525</td>\n",
       "      <td>3R</td>\n",
       "      <td>+</td>\n",
       "      <td>GAGTTCTTTTGCAGCATTTATGG</td>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083648</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18506517</td>\n",
       "      <td>18506519</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTAATGCAACTTATTGAATTTTCGAGTGCTTGGCTTAGTTGATTT...</td>\n",
       "      <td>GAAAAAAAAAAATCGAAATATGTCCAGTCGTACTCACGTATGAATG...</td>\n",
       "      <td>TTATGGGAAACTAAAACCGAGCGATTTTATGACCGCGACGATTCGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18506504</td>\n",
       "      <td>18506526</td>\n",
       "      <td>3R</td>\n",
       "      <td>+</td>\n",
       "      <td>AGTTCTTTTGCAGCATTTATGGG</td>\n",
       "      <td>FBgn0004652</td>\n",
       "      <td>FBtr0083648</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18506517</td>\n",
       "      <td>18506519</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTAATGCAACTTATTGAATTTTCGAGTGCTTGGCTTAGTTGATTT...</td>\n",
       "      <td>GAAAAAAAAAAATCGAAATATGTCCAGTCGTACTCACGTATGAATG...</td>\n",
       "      <td>TTATGGGAAACTAAAACCGAGCGATTTTATGACCGCGACGATTCGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  18506503  18506525   3R      +  GAGTTCTTTTGCAGCATTTATGG  FBgn0004652   \n",
       "1  18506504  18506526   3R      +  AGTTCTTTTGCAGCATTTATGGG  FBgn0004652   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0083648         3R  start_codon  18506517  18506519      -   \n",
       "1   FBtr0083648         3R  start_codon  18506517  18506519      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CCTAATGCAACTTATTGAATTTTCGAGTGCTTGGCTTAGTTGATTT...   \n",
       "1  CCTAATGCAACTTATTGAATTTTCGAGTGCTTGGCTTAGTTGATTT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAAAAAAAAAAATCGAAATATGTCCAGTCGTACTCACGTATGAATG...   \n",
       "1  GAAAAAAAAAAATCGAAATATGTCCAGTCGTACTCACGTATGAATG...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  TTATGGGAAACTAAAACCGAGCGATTTTATGACCGCGACGATTCGA...  \n",
       "1  TTATGGGAAACTAAAACCGAGCGATTTTATGACCGCGACGATTCGA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n",
      "stringency 3\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26005878</td>\n",
       "      <td>26005900</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATTTCCAGAACCATTTTGTTG</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  26005878  26005900   3R      -  CCATTTCCAGAACCATTTTGTTG  FBgn0002733   \n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "1   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "1  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "1  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...  \n",
       "1  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n",
      "stringency 3\n",
      "stringency 4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [fmin, fmax, #chr, strand, sgRNA_sequence, Gene_ID, Transcript_ID, Chromosome, Gene_Region, Start, Stop, Strand, Reference_Seq, HAL, HAR]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stringency 0\n",
      "stringency 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8145930</td>\n",
       "      <td>8145952</td>\n",
       "      <td>X</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGGTGGGAGCCGCAATCATGC</td>\n",
       "      <td>FBgn0030008</td>\n",
       "      <td>FBtr0071155</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8145949</td>\n",
       "      <td>8145951</td>\n",
       "      <td>+</td>\n",
       "      <td>GTGCGTGTCAAATAAATTCGATGGAAACCCGTGAAACGAGGACCGA...</td>\n",
       "      <td>TTTTGTCACTAACATTTCAAAAGTAAATGAAATAATAATAAAGATG...</td>\n",
       "      <td>CTCAAGTCCCTCAAGCCGGACACGTACGCCGGAATGGCGAACGCCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8145930  8145952    X      -  CCAGGTGGGAGCCGCAATCATGC  FBgn0030008   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0071155          X  start_codon  8145949  8145951      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GTGCGTGTCAAATAAATTCGATGGAAACCCGTGAAACGAGGACCGA...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TTTTGTCACTAACATTTCAAAAGTAAATGAAATAATAATAAAGATG...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  CTCAAGTCCCTCAAGCCGGACACGTACGCCGGAATGGCGAACGCCA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15138209</td>\n",
       "      <td>15138231</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCTGCAATGGCCACACTAATACC</td>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0111006</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "      <td>TCGTTAATTCAATTAATCAGTCTCTATTCTTTAAGTCTTTACAATT...</td>\n",
       "      <td>GCCACACTAATACCAGTGAACGGCGGCCATCCAGCAGCGAGCGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15138196</td>\n",
       "      <td>15138218</td>\n",
       "      <td>2R</td>\n",
       "      <td>+</td>\n",
       "      <td>GCAAAGTGAAGCACCTGCAATGG</td>\n",
       "      <td>FBgn0015371</td>\n",
       "      <td>FBtr0111006</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>15138215</td>\n",
       "      <td>15138217</td>\n",
       "      <td>+</td>\n",
       "      <td>GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...</td>\n",
       "      <td>TCGTTAATTCAATTAATCAGTCTCTATTCTTTAAGTCTTTACAATT...</td>\n",
       "      <td>GCCACACTAATACCAGTGAACGGCGGCCATCCAGCAGCGAGCGGAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  15138209  15138231   2R      -  CCTGCAATGGCCACACTAATACC  FBgn0015371   \n",
       "1  15138196  15138218   2R      +  GCAAAGTGAAGCACCTGCAATGG  FBgn0015371   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0111006         2R  start_codon  15138215  15138217      +   \n",
       "1   FBtr0111006         2R  start_codon  15138215  15138217      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...   \n",
       "1  GTTGCCACCATGACTGCCACTGCTGCCGGAACTCGAGCTGGCCACC...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TCGTTAATTCAATTAATCAGTCTCTATTCTTTAAGTCTTTACAATT...   \n",
       "1  TCGTTAATTCAATTAATCAGTCTCTATTCTTTAAGTCTTTACAATT...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  GCCACACTAATACCAGTGAACGGCGGCCATCCAGCAGCGAGCGGAC...  \n",
       "1  GCCACACTAATACCAGTGAACGGCGGCCATCCAGCAGCGAGCGGAC...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17757557</td>\n",
       "      <td>17757579</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAACTTGTTTTTGAACAATGCC</td>\n",
       "      <td>FBgn0022740</td>\n",
       "      <td>FBtr0089547</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>17757575</td>\n",
       "      <td>17757577</td>\n",
       "      <td>+</td>\n",
       "      <td>TTCTAGTGGGGGGCCCGGCTGAATGGCTGACTCCGGTCATGGGATG...</td>\n",
       "      <td>GAGAGCTGCAAACTGCGGAGCAGAATGCTCAAAAAGGAGGATCTCC...</td>\n",
       "      <td>CCACGCAAGAAGCGCAGTTCCAGCCCCACGGAGTTCTTTGATGACG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  17757557  17757579   2R      -  CCAACTTGTTTTTGAACAATGCC  FBgn0022740   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0089547         2R  start_codon  17757575  17757577      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTCTAGTGGGGGGCCCGGCTGAATGGCTGACTCCGGTCATGGGATG...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAGAGCTGCAAACTGCGGAGCAGAATGCTCAAAAAGGAGGATCTCC...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  CCACGCAAGAAGCGCAGTTCCAGCCCCACGGAGTTCTTTGATGACG...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n",
      "stringency 3\n",
      "stringency 4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [fmin, fmax, #chr, strand, sgRNA_sequence, Gene_ID, Transcript_ID, Chromosome, Gene_Region, Start, Stop, Strand, Reference_Seq, HAL, HAR]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "stringency 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30852536</td>\n",
       "      <td>30852558</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCACATCGGTATGCAGTCGTCG</td>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>30852549</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...</td>\n",
       "      <td>GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...</td>\n",
       "      <td>CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  30852536  30852558   3R      -  CCCACATCGGTATGCAGTCGTCG  FBgn0003720   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0085709         3R  start_codon  30852547  30852549      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "      <td>GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...</td>\n",
       "      <td>CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  11278453      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7492183</td>\n",
       "      <td>7492205</td>\n",
       "      <td>2R</td>\n",
       "      <td>+</td>\n",
       "      <td>TCTCCATGTCCAGATCCGAGTGG</td>\n",
       "      <td>FBgn0263240</td>\n",
       "      <td>FBtr0088984</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>7492187</td>\n",
       "      <td>7492189</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...</td>\n",
       "      <td>CTGACCCACCTCGAACCAAACAGGCGTGATGTCGCTGCGCTTCTGG...</td>\n",
       "      <td>GTCCAGATCCGAGTGGCCAATTTCCGATAGCCCGATAGCTAGTTAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7492180</td>\n",
       "      <td>7492202</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATCTCCATGTCCAGATCCGAG</td>\n",
       "      <td>FBgn0263240</td>\n",
       "      <td>FBtr0088984</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>7492187</td>\n",
       "      <td>7492189</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...</td>\n",
       "      <td>CTGACCCACCTCGAACCAAACAGGCGTGATGTCGCTGCGCTTCTGG...</td>\n",
       "      <td>GTCCAGATCCGAGTGGCCAATTTCCGATAGCCCGATAGCTAGTTAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7492186</td>\n",
       "      <td>7492208</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATGTCCAGATCCGAGTGGCCA</td>\n",
       "      <td>FBgn0263240</td>\n",
       "      <td>FBtr0088984</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>7492187</td>\n",
       "      <td>7492189</td>\n",
       "      <td>-</td>\n",
       "      <td>TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...</td>\n",
       "      <td>CTGACCCACCTCGAACCAAACAGGCGTGATGTCGCTGCGCTTCTGG...</td>\n",
       "      <td>GTCCAGATCCGAGTGGCCAATTTCCGATAGCCCGATAGCTAGTTAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  7492183  7492205   2R      +  TCTCCATGTCCAGATCCGAGTGG  FBgn0263240   \n",
       "1  7492180  7492202   2R      -  CCATCTCCATGTCCAGATCCGAG  FBgn0263240   \n",
       "2  7492186  7492208   2R      -  CCATGTCCAGATCCGAGTGGCCA  FBgn0263240   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0088984         2R  start_codon  7492187  7492189      -   \n",
       "1   FBtr0088984         2R  start_codon  7492187  7492189      -   \n",
       "2   FBtr0088984         2R  start_codon  7492187  7492189      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...   \n",
       "1  TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...   \n",
       "2  TTTTATTTGGTTCAAAACAAAGTATGCATCATATTATTAATTTATT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  CTGACCCACCTCGAACCAAACAGGCGTGATGTCGCTGCGCTTCTGG...   \n",
       "1  CTGACCCACCTCGAACCAAACAGGCGTGATGTCGCTGCGCTTCTGG...   \n",
       "2  CTGACCCACCTCGAACCAAACAGGCGTGATGTCGCTGCGCTTCTGG...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  GTCCAGATCCGAGTGGCCAATTTCCGATAGCCCGATAGCTAGTTAG...  \n",
       "1  GTCCAGATCCGAGTGGCCAATTTCCGATAGCCCGATAGCTAGTTAG...  \n",
       "2  GTCCAGATCCGAGTGGCCAATTTCCGATAGCCCGATAGCTAGTTAG...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "sgRNA within window\n",
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>361640</td>\n",
       "      <td>361662</td>\n",
       "      <td>3L</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAAGGAAGGCACCCACGATGAG</td>\n",
       "      <td>FBgn0035137</td>\n",
       "      <td>FBtr0072548</td>\n",
       "      <td>3L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>361658</td>\n",
       "      <td>361660</td>\n",
       "      <td>+</td>\n",
       "      <td>CACCAAAATCAAGGTCCAAAGCACTAAAACGTGTGTTTTCCACATG...</td>\n",
       "      <td>TAACGCGACAGTGTGACCGAATATAGTTGTGCAAACACCGCTGCCA...</td>\n",
       "      <td>AGCGATGCGACCTATCCGGGATATTACGGCGGCAACCAGGTGAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>361653</td>\n",
       "      <td>361675</td>\n",
       "      <td>3L</td>\n",
       "      <td>-</td>\n",
       "      <td>CCACGATGAGCGATGCGACCTAT</td>\n",
       "      <td>FBgn0035137</td>\n",
       "      <td>FBtr0072548</td>\n",
       "      <td>3L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>361658</td>\n",
       "      <td>361660</td>\n",
       "      <td>+</td>\n",
       "      <td>CACCAAAATCAAGGTCCAAAGCACTAAAACGTGTGTTTTCCACATG...</td>\n",
       "      <td>TAACGCGACAGTGTGACCGAATATAGTTGTGCAAACACCGCTGCCA...</td>\n",
       "      <td>AGCGATGCGACCTATCCGGGATATTACGGCGGCAACCAGGTGAGCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  361640  361662   3L      -  CCAAGGAAGGCACCCACGATGAG  FBgn0035137   \n",
       "1  361653  361675   3L      -  CCACGATGAGCGATGCGACCTAT  FBgn0035137   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0072548         3L  start_codon  361658  361660      +   \n",
       "1   FBtr0072548         3L  start_codon  361658  361660      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CACCAAAATCAAGGTCCAAAGCACTAAAACGTGTGTTTTCCACATG...   \n",
       "1  CACCAAAATCAAGGTCCAAAGCACTAAAACGTGTGTTTTCCACATG...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TAACGCGACAGTGTGACCGAATATAGTTGTGCAAACACCGCTGCCA...   \n",
       "1  TAACGCGACAGTGTGACCGAATATAGTTGTGCAAACACCGCTGCCA...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  AGCGATGCGACCTATCCGGGATATTACGGCGGCAACCAGGTGAGCA...  \n",
       "1  AGCGATGCGACCTATCCGGGATATTACGGCGGCAACCAGGTGAGCA...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "      <td>GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...</td>\n",
       "      <td>ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  8600584      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10533746</td>\n",
       "      <td>10533768</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATTCTGGTTTCTCGGGTGGTT</td>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0343132</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "      <td>TGTATTTTTAGTGTTAACAGAGTTTTCCGAAACTATGAATTTGGTA...</td>\n",
       "      <td>TCTGGTTTCTCGGGTGGTTGGATGGGGTTGTTGGAAATGGATCCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10533744</td>\n",
       "      <td>10533766</td>\n",
       "      <td>2R</td>\n",
       "      <td>+</td>\n",
       "      <td>ATCCATTCTGGTTTCTCGGGTGG</td>\n",
       "      <td>FBgn0283521</td>\n",
       "      <td>FBtr0343132</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>10533747</td>\n",
       "      <td>10533749</td>\n",
       "      <td>-</td>\n",
       "      <td>GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...</td>\n",
       "      <td>TGTATTTTTAGTGTTAACAGAGTTTTCCGAAACTATGAATTTGGTA...</td>\n",
       "      <td>TCTGGTTTCTCGGGTGGTTGGATGGGGTTGTTGGAAATGGATCCTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  10533746  10533768   2R      -  CCATTCTGGTTTCTCGGGTGGTT  FBgn0283521   \n",
       "1  10533744  10533766   2R      +  ATCCATTCTGGTTTCTCGGGTGG  FBgn0283521   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0343132         2R  start_codon  10533747  10533749      -   \n",
       "1   FBtr0343132         2R  start_codon  10533747  10533749      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...   \n",
       "1  GGCACGATTATTACCGCTGCAGTTCTCCTTTTTCCTTTGAGCTCCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TGTATTTTTAGTGTTAACAGAGTTTTCCGAAACTATGAATTTGGTA...   \n",
       "1  TGTATTTTTAGTGTTAACAGAGTTTTCCGAAACTATGAATTTGGTA...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  TCTGGTTTCTCGGGTGGTTGGATGGGGTTGTTGGAAATGGATCCTG...  \n",
       "1  TCTGGTTTCTCGGGTGGTTGGATGGGGTTGTTGGAAATGGATCCTG...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18514698</td>\n",
       "      <td>18514720</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CGGCTTGTTCATCTCGTGTGTGG</td>\n",
       "      <td>FBgn0030933</td>\n",
       "      <td>FBtr0074621</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>18514707</td>\n",
       "      <td>18514709</td>\n",
       "      <td>-</td>\n",
       "      <td>CATCAGTAGACTTACATCAGCAAGAGATTAATATAAACAATTGCTT...</td>\n",
       "      <td>CATGCTCGATCCTTCCCGCCCTGCTCCTCCGCCTGCGCCAGCCGTG...</td>\n",
       "      <td>CTCGTGTGTGGTGTGCGTGTGTGTTTGTGTATGTGTAGCGTGTCTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  18514698  18514720    X      +  CGGCTTGTTCATCTCGTGTGTGG  FBgn0030933   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0074621          X  start_codon  18514707  18514709      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CATCAGTAGACTTACATCAGCAAGAGATTAATATAAACAATTGCTT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  CATGCTCGATCCTTCCCGCCCTGCTCCTCCGCCTGCGCCAGCCGTG...   \n",
       "\n",
       "                                                 HAR  \n",
       "0  CTCGTGTGTGGTGTGCGTGTGTGTTTGTGTATGTGTAGCGTGTCTT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sgRNA within window\n",
      "stringency 0\n",
      "stringency 1\n",
      "stringency 2\n",
      "stringency 3\n",
      "stringency 4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [fmin, fmax, #chr, strand, sgRNA_sequence, Gene_ID, Transcript_ID, Chromosome, Gene_Region, Start, Stop, Strand, Reference_Seq, HAL, HAR]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test function to identify sgRNAs within a window of 21 bp around the start/stop codon of a transcript\n",
    "# this cell takes 32 min to run, you can avoid it by running the next cell and loading the output from this cell from MockMaterials\n",
    "from sgRNAutils import gRNA_stringencyIterator\n",
    "import os\n",
    "\n",
    "folder_name = \"inputfiles/mockMaterials/sgRNAsdfs\"\n",
    "window = 21\n",
    "for index, row in TFsdf.iterrows():\n",
    "    result_df = gRNA_stringencyIterator(row, window, refSeqPerChromosome)\n",
    "    display(result_df)\n",
    "    if not result_df.empty:\n",
    "        file_name = os.path.join(folder_name, f'{result_df[\"Gene_ID\"].iloc[0]}_{result_df[\"Transcript_ID\"].iloc[0]}.xlsx')\n",
    "        # Save the DataFrame to an Excel file\n",
    "        result_df.to_excel(file_name, index=False)\n",
    "    # test whether every sgRNA identified is within the window of 21 bp around the start/stop codon\n",
    "    for i in range(len(result_df)):\n",
    "        if result_df['fmin'].iloc[i] >= result_df[\"Start\"].iloc[i] - window & result_df['fmax'].iloc[i] <= result_df[\"Stop\"].iloc[i] + window:\n",
    "            print(\"sgRNA within window\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9792987</td>\n",
       "      <td>9793009</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CACAGTTTACGCAGGTCCATGGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  9792987  9793009   2L      +  CACAGTTTACGCAGGTCCATGGG  FBgn0032150   \n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0079917         2L  start_codon  9793004  ...              False   \n",
       "1   FBtr0079917         2L  start_codon  9793004  ...               True   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True      False            True       True   \n",
       "1                 False      False           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -2  \n",
       "1  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -3  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GCGGCACAGTTTACGCAGGTC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TGCGGCACAGTTTACGCAGGTC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26005878</td>\n",
       "      <td>26005900</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATTTCCAGAACCATTTTGTTG</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  26005878  26005900   3R      -  CCATTTCCAGAACCATTTTGTTG  FBgn0002733   \n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0084982         3R  start_codon  26005891  ...              False   \n",
       "1   FBtr0084982         3R  start_codon  26005891  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True       True           False       True   \n",
       "1                  True       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                     SRS_boundary  \\\n",
       "0  [-19, -18, -17, -16, -15, -14]   \n",
       "1  [-19, -18, -17, -16, -15, -14]   \n",
       "\n",
       "                                         mutable_PAM cut_site  \n",
       "0  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...      -10  \n",
       "1  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...       -4  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TTTGTTG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'ATCTCCATTTCCAGAAC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TTTGTTGTAGTTT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'ATTTCCAGAAC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                 False       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       15  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CTTAGCAGCAATACAAGAGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'AACA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0070072          X  start_codon  370094  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]        3  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GCTTTGGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TATACTATCTCTTAAA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30852536</td>\n",
       "      <td>30852558</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCACATCGGTATGCAGTCGTCG</td>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  30852536  30852558   3R      -  CCCACATCGGTATGCAGTCGTCG  FBgn0003720   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0085709         3R  start_codon  30852547  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True      False            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]       -8  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGTCGTCG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'ACGCCCCACATCGGT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                 False       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       14  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCATCCCAATGCGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TAACC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                 False      False            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]      -16  \n",
       "\n",
       "[1 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'A'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GAGTCCAGTAAAGCGCAGCAGCA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# there are 8 different combinations of start/stop - sgRNA strand and gene strand (see MutationLogic_ML_20231001.pptx)\n",
    "# however, for two combinations each, gene and sgRNA have effectively the same relative position to each other \n",
    "# (e.g. the two combinations start codon - sgRNA strand (+) - gene strand (+) and stop codon - sgRNA strand (+) - gene strand (-))\n",
    "# therefore, we have identified 4 combinations with unique relative positions to test the downstream pipeline with\n",
    "# further, we have identified representatives for the 3 possible sgRNA position cases relevant for the find_best_sgRNA function\n",
    "# further, we have identified representatives for the 3 possible sgRNA position cases relevant for the find_best_mutation function\n",
    "from sgRNAutils import sgRNApositionCheck\n",
    "import pandas as pd\n",
    "\n",
    "sgRNAdf1 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/FBgn0032150_FBtr0079917.xlsx') # PAM in start/stop, start codon - gene (-) - gRNA (+)\n",
    "sgRNAdf2 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/FBgn0002733_FBtr0084982.xlsx') # max 15 bp overlap, start codon - gene (-) - gRNA (-)\n",
    "sgRNAdf3 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/FBgn0003963_FBtr0078063.xlsx') # more than 15 bp overlap, PAM in CDS\n",
    "sgRNAdf4 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/FBgn0000022_FBtr0070072.xlsx') # start codon - gene (+) - gRNA (+)\n",
    "sgRNAdf5 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/FBgn0003720_FBtr0085709.xlsx') # start codon - gene (+) - gRNA (-)\n",
    "sgRNAdf6 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/FBgn0052831_FBtr0091683.xlsx') # SRS in CDS (need to manually change \"PAM in CDS\" to False)\n",
    "sgRNAdf7 = pd.read_excel('inputfiles/mockMaterials/sgRNAsdfs/FBgn0267792_FBtr0088692.xlsx') # PAM and SRS bot outside CDS\n",
    "\n",
    "#Create list to store results in\n",
    "positionCheckedList = []\n",
    "\n",
    "#Derive min and max distance from stop position from window size\n",
    "window = 21\n",
    "minDistance = (window + 3) * -1 # leave a 3bp gap to account for start/stop site\n",
    "maxDistance = window\n",
    "\n",
    "for sgRNAdf in [sgRNAdf1, sgRNAdf2, sgRNAdf3, sgRNAdf4, sgRNAdf5, sgRNAdf6, sgRNAdf7]:\n",
    "    result_df = sgRNApositionCheck(sgRNAdf, minDistance, maxDistance)\n",
    "    display(result_df)\n",
    "    # check whether the sgRNA is positioned within the HA as expected\n",
    "    for i in range(len(result_df)):\n",
    "        positionScore = result_df.iloc[i][\"positionScore\"]\n",
    "        # print positionScore which represents the distance between the end of the start/stop codon and the end of the sgRNA\n",
    "        print(positionScore)\n",
    "        # print the part of the sgRNA that is located in HAR\n",
    "        display(result_df.iloc[i][\"HAR\"][:positionScore])\n",
    "        # print the part of the sgRNA that is the start/stop codon\n",
    "        print(refSeqPerChromosome[result_df.iloc[i][\"Chromosome\"]][result_df.iloc[i]['Start']-1:result_df.iloc[i]['Stop']])\n",
    "        positionScore = positionScore -2 # currently the stop of the start/stop codon is zero, we want the stop of the HAL to be -1, 3 nucleotides from the start/stop codon are left out\n",
    "        positionScore = positionScore  -22 # currently we have the stop of the sgRNA, we want the start of the sgRNA, sgRNA is 23 nucleotides long and we want position inclusively\n",
    "        # print the part of the sgRNA that is located in HAL\n",
    "        display(result_df.iloc[i][\"HAL\"][positionScore:])\n",
    "    positionCheckedList.append(result_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9792987</td>\n",
       "      <td>9793009</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CACAGTTTACGCAGGTCCATGGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  9792987  9793009   2L      +  CACAGTTTACGCAGGTCCATGGG  FBgn0032150   \n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0079917         2L  start_codon  9793004  ...              False   \n",
       "1   FBtr0079917         2L  start_codon  9793004  ...               True   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True      False            True       True   \n",
       "1                 False      False           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -2  \n",
       "1  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -3  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26005878</td>\n",
       "      <td>26005900</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATTTCCAGAACCATTTTGTTG</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  26005878  26005900   3R      -  CCATTTCCAGAACCATTTTGTTG  FBgn0002733   \n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start  ...  PAM_in_start/stop  \\\n",
       "0   FBtr0084982         3R  start_codon  26005891  ...              False   \n",
       "1   FBtr0084982         3R  start_codon  26005891  ...              False   \n",
       "\n",
       "  max_15_bp_3’_overhang PAM_in_CDS PAM_outside_CDS SRS_in_CDS  \\\n",
       "0                  True       True           False       True   \n",
       "1                  True       True           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                     SRS_boundary  \\\n",
       "0  [-19, -18, -17, -16, -15, -14]   \n",
       "1  [-19, -18, -17, -16, -15, -14]   \n",
       "\n",
       "                                         mutable_PAM cut_site  \n",
       "0  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...      -10  \n",
       "1  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...       -4  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'GCACAGTTTACGCAGGTCCATGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CCAGAACCATTTTGTTGTAGTTT'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from sgRNAutils import checkCDSCutandOrder\n",
    "import pandas as pd\n",
    "# test this for some of the TF transcripts with multiple sgRNAs\n",
    "# prefers sgRNAs that cut in the CDS and sgRNAs that are closer to the start/stop codon\n",
    "display(positionCheckedList[0])\n",
    "display(positionCheckedList[1])\n",
    "# we expect the second sgRNA in first dataframe to be selected since cuts inside the CDS and the first one cuts outside\n",
    "# we expect the second sgRNA in the second dataframe to be selected since it is closer to the start/stop codon\n",
    "sgRNAdf1_checked = checkCDSCutandOrder(positionCheckedList[0])\n",
    "sgRNAdf2_checked = checkCDSCutandOrder(positionCheckedList[1])\n",
    "display(sgRNAdf1_checked)\n",
    "display(sgRNAdf2_checked)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>cutsite-CDSbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9792987</td>\n",
       "      <td>9793009</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CACAGTTTACGCAGGTCCATGGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "      <td>3</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "      <td>2</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  9792987  9793009   2L      +  CACAGTTTACGCAGGTCCATGGG  FBgn0032150   \n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "1   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "1  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "1  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...              3      HAL   \n",
       "1  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...              2      HAL   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAR             False                  True      False   \n",
       "1          HAR              True                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True       True   \n",
       "1           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site cutsite-CDSbound  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -2                2  \n",
       "1  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -3                3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  False\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>cutsite-CDSbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9792986</td>\n",
       "      <td>9793008</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>GCACAGTTTACGCAGGTCCATGG</td>\n",
       "      <td>FBgn0032150</td>\n",
       "      <td>FBtr0079917</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>9793004</td>\n",
       "      <td>9793006</td>\n",
       "      <td>-</td>\n",
       "      <td>AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...</td>\n",
       "      <td>AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...</td>\n",
       "      <td>GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...</td>\n",
       "      <td>2</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>-3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "1  9792986  9793008   2L      +  GCACAGTTTACGCAGGTCCATGG  FBgn0032150   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "1   FBtr0079917         2L  start_codon  9793004  9793006      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "1  AATTTAATTTTTTTTATAATTAATTTAGTGCTAATCTTTGAGCAGC...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "1  AATCATTACAGATCATGGGCAGCTCCTCAGTAAGATTAAGTGCTAT...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "1  GGGTTTTATTATTTAATTAATGTAAATAAACTGTAATGTTAATGTT...              2      HAL   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "1          HAR              True                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "1           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site cutsite-CDSbound  \n",
       "1  [-8, -7, -6, -5, -4, -3]     [-1, 0]       -3                3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>cutsite-CDSbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>26005878</td>\n",
       "      <td>26005900</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCATTTCCAGAACCATTTTGTTG</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "      <td>7</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "      <td>13</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  26005878  26005900   3R      -  CCATTTCCAGAACCATTTTGTTG  FBgn0002733   \n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "1   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "1  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "1  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...              7      HAL   \n",
       "1  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...             13      HAL   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAR             False                  True       True   \n",
       "1          HAR             False                  True       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "1           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                     SRS_boundary  \\\n",
       "0  [-19, -18, -17, -16, -15, -14]   \n",
       "1  [-19, -18, -17, -16, -15, -14]   \n",
       "\n",
       "                                         mutable_PAM cut_site cutsite-CDSbound  \n",
       "0  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...      -10               10  \n",
       "1  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...       -4                4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  False\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>cutsite-CDSbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26005884</td>\n",
       "      <td>26005906</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGAACCATTTTGTTGTAGTTT</td>\n",
       "      <td>FBgn0002733</td>\n",
       "      <td>FBtr0084982</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>26005891</td>\n",
       "      <td>26005893</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...</td>\n",
       "      <td>CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...</td>\n",
       "      <td>TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...</td>\n",
       "      <td>13</td>\n",
       "      <td>HAL</td>\n",
       "      <td>HAR</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21, -20, -19, -18, -17, -16, -15, -14, ...</td>\n",
       "      <td>-4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "1  26005884  26005906   3R      -  CCAGAACCATTTTGTTGTAGTTT  FBgn0002733   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "1   FBtr0084982         3R  start_codon  26005891  26005893      -   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "1  GGTCCAGCGGACAAAACGTTTCGGTTAGATAAGTTGTTAGCCGTTT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "1  CAGCTGCTTCTGGGCACGCAGCTTCTTCATGTGCTCCACGGTCAGC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "1  TTTGTTGTAGTTTGGTTTTGTTTTTTTTTTATTGGGGGTAAAGTAA...             13      HAL   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "1          HAR             False                  True       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "1           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "1  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                     SRS_boundary  \\\n",
       "1  [-19, -18, -17, -16, -15, -14]   \n",
       "\n",
       "                                         mutable_PAM cut_site cutsite-CDSbound  \n",
       "1  [-22, -21, -20, -19, -18, -17, -16, -15, -14, ...       -4                4  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "      <td>AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...</td>\n",
       "      <td>CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...</td>\n",
       "      <td>20</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  524040      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...             20      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       15  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "      <td>AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...</td>\n",
       "      <td>CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...</td>\n",
       "      <td>20</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  524040      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...             20      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       15  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "      <td>TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...</td>\n",
       "      <td>GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...</td>\n",
       "      <td>8</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0070072          X  start_codon  370094  370096      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...              8      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                  True       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]        3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  False\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>370082</td>\n",
       "      <td>370104</td>\n",
       "      <td>X</td>\n",
       "      <td>+</td>\n",
       "      <td>CTATCTCTTAAAATGGCTTTGGG</td>\n",
       "      <td>FBgn0000022</td>\n",
       "      <td>FBtr0070072</td>\n",
       "      <td>X</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>370094</td>\n",
       "      <td>370096</td>\n",
       "      <td>+</td>\n",
       "      <td>CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...</td>\n",
       "      <td>TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...</td>\n",
       "      <td>GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...</td>\n",
       "      <td>8</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  370082  370104    X      +  CTATCTCTTAAAATGGCTTTGGG  FBgn0000022   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0070072          X  start_codon  370094  370096      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CTAATGAATAGATTGGTGTGTGATGTAGTGATCTAATATGGTGAAG...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  TACTACCTCTCTATTAAAATCAGAGAAAACACTCATCTCAAGAGAC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  GCTTTGGGCAGCGAAAATCACTCTGTTTTCAACGACGACGAGGAGT...              8      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                  True       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]        3  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30852536</td>\n",
       "      <td>30852558</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCACATCGGTATGCAGTCGTCG</td>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>30852549</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...</td>\n",
       "      <td>GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...</td>\n",
       "      <td>CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...</td>\n",
       "      <td>9</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  30852536  30852558   3R      -  CCCACATCGGTATGCAGTCGTCG  FBgn0003720   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0085709         3R  start_codon  30852547  30852549      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...              9      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                  True      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]       -8  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  False\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>30852536</td>\n",
       "      <td>30852558</td>\n",
       "      <td>3R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCACATCGGTATGCAGTCGTCG</td>\n",
       "      <td>FBgn0003720</td>\n",
       "      <td>FBtr0085709</td>\n",
       "      <td>3R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>30852547</td>\n",
       "      <td>30852549</td>\n",
       "      <td>+</td>\n",
       "      <td>TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...</td>\n",
       "      <td>GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...</td>\n",
       "      <td>CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...</td>\n",
       "      <td>9</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  30852536  30852558   3R      -  CCCACATCGGTATGCAGTCGTCG  FBgn0003720   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0085709         3R  start_codon  30852547  30852549      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TATGAGAAATCGTTGATTTCACTGATCGAGTATTGTCCAAAGCGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCCGGCGGTCCTCACAGCAGACAACACAACCCATCGTGATCTCAGC...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGTCGTCGGAGGGTTCACCAGACATGATGGATCAGAAATACAACA...              9      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                  True      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]       -8  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "      <td>GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...</td>\n",
       "      <td>CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...</td>\n",
       "      <td>19</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  11278453      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...             19      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       14  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "      <td>GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...</td>\n",
       "      <td>CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...</td>\n",
       "      <td>19</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  11278453      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...             19      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       14  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "      <td>GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...</td>\n",
       "      <td>ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...</td>\n",
       "      <td>1</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  8600584      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...              1      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]      -16  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutation needed:  True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "      <td>GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...</td>\n",
       "      <td>ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...</td>\n",
       "      <td>1</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  8600584      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...              1      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]      -16  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test for the three different cases for identifying best sgRNA (e.g. PAM in start/stop, max 15 bp overlap, none of the above)\n",
    "# test for 4 unique position combinations (e.g. start/stop codon - sgRNA strand - gene strand\n",
    "# test for three different  cased for identifying best mutation (e.g. PAM in CDS, SRS in CDS, none of the above)\n",
    "# we expect that for the first dataframe the sgRNA is selected that has the PAM in the start/stop codon\n",
    "# we expect that for the second dataframe the sgRNA is selected that has the max 15 bp overlap\n",
    "# we expect that for the third dataframe a mutation is needed\n",
    "\n",
    "from sgRNAutils import find_best_gRNA  \n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "bestsgRNAList = []\n",
    "\n",
    "for sgRNAdf in positionCheckedList:\n",
    "    display(sgRNAdf)\n",
    "    sgRNA, mutationNeeded = find_best_gRNA(sgRNAdf)\n",
    "    print(\"mutation needed: \", mutationNeeded)\n",
    "    winnerdf = sgRNAdf[sgRNAdf[\"sgRNA_sequence\"] == sgRNA] #This is the dataframe row for the winning sgRNA from the original sgRNA dataframe\n",
    "    display(winnerdf)\n",
    "    bestsgRNAList.append(winnerdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TCAGATGGGAAAAGCGATCGTGATTAGTGGAGAGTTACGAGATAGCTTTGTAATGACAAAGTGTAATTTACAAAAGTGCATTTGTTCGTTGAGGCAAAATAATGGTTGCAGATAACTGCACAGGTGCTATAAAAAGAAGGTCGTCGACTAAACCCCGACTTGATAAAACTAAACTTGTAGCTCACCTTTATTCAACGGCTTCGGGTTGGACTGTTTGCGTCGTGA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGCACGTGAGCGCTGCCAGCTCGGTGCCATCAGTAGTAACTCCTGTTGTGACCACTGGTGGGACAACGATCACATTGGGCGGCCCACCACCGCTCCCTAAATCCGAGCACAAAGAGGATGGCAAGCCGCCGCACGGCATCGAAATGTACAAGGTGAACATTGAGGACATTTCGCAGCTCTTCACCTACCACGAGGTCTTTGGCAAAATCCACGGCGATGTGGTC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n"
     ]
    }
   ],
   "source": [
    "# for testing print the HA in CDS\n",
    "# TODO: only adjust this if you think you need to test more in detail\n",
    "HA_0 = checked_0.iloc[0][checked_0.iloc[0][\"CDS_side\"]]\n",
    "display(HA_0)\n",
    "print(len(HA_0))\n",
    "HA_1 = checked_1.iloc[0][checked_1.iloc[0][\"CDS_side\"]]\n",
    "display(HA_1)\n",
    "print(len(HA_1))    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TCA', 'GAT', 'GGG', 'AAA', 'AGC', 'GAT', 'CGT']\n",
      "['TTT', 'GAT', 'GTG', 'GCC', 'ATC', 'CAC', 'GAC']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:496: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) - 1 #HA is 1bp to the right of stop\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:496: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) - 1 #HA is 1bp to the right of stop\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:500: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) + 2 #HA is 3pb to the left of stop but because we index backwards we calculate - 1\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:500: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) + 2 #HA is 3pb to the left of stop but because we index backwards we calculate - 1\n"
     ]
    }
   ],
   "source": [
    "# TODO: only adjust this if you think you need to test more in detail\n",
    "from sgRNAutils import codonFragmenter\n",
    "pd.set_option('display.max_column', None)\n",
    "CDS_0, df_CDS_adjusted_0 = codonFragmenter(checked_0.iloc[0], checked_0.iloc[0][\"CDS_side\"], checked_0.iloc[0][\"CDS_boundary\"])\n",
    "print(CDS_0)\n",
    "CDS_1, df_CDS_adjusted_1 = codonFragmenter(checked_1.iloc[0], checked_1.iloc[0][\"CDS_side\"], checked_1.iloc[0][\"CDS_boundary\"])\n",
    "print(CDS_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:586: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf.at[\"HAR\"] = mutable_region + winnerdf[\"HAR\"][winnerdf[f\"{side}_HA\"][-1]+1:]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'TCAGATGGGAAAAGCGATCGTGATTAGTGGAGAGTTACGAGATAGCTTTGTAATGACAAAGTGTAATTTACAAAAGTGCATTTGTTCGTTGAGGCAAAATAATGGTTGCAGATAACTGCACAGGTGCTATAAAAAGAAGGTCGTCGACTAAACCCCGACTTGATAAAACTAAACTTGTAGCTCACCTTTATTCAACGGCTTCGGGTTGGACTGTTTGCGTCGTGA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'TCAGATGGGAAAAGCGATCGTGATTAGTGGAGAGTTACGAGATAGCTTTGTAATGACAAAGTGTAATTTACAAAAGTGCATTTGTTCGTTGAGGCAAAATAATGGTTGCAGATAACTGCACAGGTGCTATAAAAAGAAGGTCGTCGACTAAACCCCGACTTGATAAAACTAAACTTGTAGCTCACCTTTATTCAACGGCTTCGGGTTGGACTGTTTGCGTCGTGA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "all working\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:588: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf.at[\"HAL\"] =  winnerdf[\"HAL\"][:winnerdf[f\"{side}_HA\"][0]+1] + mutable_region\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGCACGTGAGCGCTGCCAGCTCGGTGCCATCAGTAGTAACTCCTGTTGTGACCACTGGTGGGACAACGATCACATTGGGCGGCCCACCACCGCTCCCTAAATCCGAGCACAAAGAGGATGGCAAGCCGCCGCACGGCATCGAAATGTACAAGGTGAACATTGAGGACATTTCGCAGCTCTTCACCTACCACGAGGTCTTTGGCAAAATCCACGGCGATGTGGTC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CAGCACGTGAGCGCTGCCAGCTCGGTGCCATCAGTAGTAACTCCTGTTGTGACCACTGGTGGGACAACGATCACATTGGGCGGCCCACCACCGCTCCCTAAATCCGAGCACAAAGAGGATGGCAAGCCGCCGCACGGCATCGAAATGTACAAGGTGAACATTGAGGACATTTCGCAGCTCTTCACCTACCACGAGGTCTTTGGCAAAATCCACGGCGATGTGGTC'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "all working\n"
     ]
    }
   ],
   "source": [
    "# TODO: only adjust this if you think you need to test more in detail\n",
    "from sgRNAutils import codonReverseFragmenter\n",
    "df_HA_0_after = codonReverseFragmenter(CDS_0, df_CDS_adjusted_0, checked_0.iloc[0][\"CDS_side\"], checked_0.iloc[0][\"CDS_boundary\"])\n",
    "HA_0_after = df_HA_0_after[df_HA_0_after[\"CDS_side\"]]\n",
    "display(HA_0_after)\n",
    "display(HA_0)\n",
    "print(len(HA_0_after))\n",
    "if HA_0_after == HA_0:\n",
    "    print(\"all working\")\n",
    "else: print(\"not working\")\n",
    "\n",
    "df_HA_1_after = codonReverseFragmenter(CDS_1, df_CDS_adjusted_1, checked_1.iloc[0][\"CDS_side\"], checked_1.iloc[0][\"CDS_boundary\"])\n",
    "HA_1_after = df_HA_1_after[df_HA_1_after[\"CDS_side\"]]\n",
    "display(HA_1_after)\n",
    "display(HA_1)\n",
    "print(len(HA_1_after))\n",
    "if HA_1_after == HA_1:\n",
    "    print(\"all working\")\n",
    "else: print(\"not working\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: only adjust this if you think you need to test more in detail\n",
    "# test function to identify a synonymous codons that changes a given nucleotide but does not change the encoded amino acid\n",
    "from sgRNAutils import find_synonymous_codons\n",
    "# try to mutate first, second and third nucleotide of the codon \"ATG\"\n",
    "codon = \"GCT\"\n",
    "synonymous_codon_1 = find_synonymous_codons(codon, 1)\n",
    "synonymous_codon_2 = find_synonymous_codons(codon, 2)\n",
    "synonymous_codon_3 = find_synonymous_codons(codon, 3)\n",
    "print(synonymous_codon_1)\n",
    "print(synonymous_codon_2)\n",
    "print(synonymous_codon_3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plus\n",
      "['TCA', 'GAT', 'GGG', 'AAA', 'AGC', 'GAT', 'CGT']\n",
      "positionScore: 2\n",
      "CDS start: 1\n",
      "['TCA', 'GAC', 'GGG', 'AAA', 'AGC', 'GAT', 'CGT']\n"
     ]
    }
   ],
   "source": [
    "# TODO: only adjust this if you think you need to test more in detail\n",
    "from sgRNAutils import mutator\n",
    "# formula to calculate base position to be mutated = base position relative to fmax + positionScore - CDS start\n",
    "# to test try different base positions relative to fmax e.g. [-1, 0, 1, 2, 3, 4, 5, 6, 7, 8]\n",
    "base_position_relative_to_fmax = 4\n",
    "if df_CDS_adjusted_0[\"Strand\"] == '+':\n",
    "    codonCoordinates = pd.read_excel(\"inputfiles/fmaxStopScoreML.xlsx\", sheet_name=\"CodonCoordinatePlus\", index_col= 0)\n",
    "    print(\"plus\")\n",
    "elif df_CDS_adjusted_0[\"Strand\"] == '-':\n",
    "    codonCoordinates = pd.read_excel(\"inputfiles/fmaxStopScoreML.xlsx\", sheet_name=\"CodonCoordinateMinus\", index_col= 0)\n",
    "    print(\"minus\")\n",
    "print(CDS_0)\n",
    "print(\"positionScore:\", df_CDS_adjusted_0[\"positionScore\"])\n",
    "print(\"CDS start:\", df_CDS_adjusted_0[\"CDS_boundary\"][0])\n",
    "fragmentedDCS = mutator(base_position_relative_to_fmax,CDS_0,df_CDS_adjusted_0,codonCoordinates)\n",
    "print(fragmentedDCS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>mutated?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11278450</td>\n",
       "      <td>11278472</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>CATGCAGCTGCATCCCAATGCGG</td>\n",
       "      <td>FBgn0052831</td>\n",
       "      <td>FBtr0091683</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>11278451</td>\n",
       "      <td>11278453</td>\n",
       "      <td>+</td>\n",
       "      <td>TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...</td>\n",
       "      <td>GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...</td>\n",
       "      <td>CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...</td>\n",
       "      <td>19</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       fmin      fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  11278450  11278472   2L      +  CATGCAGCTGCATCCCAATGCGG  FBgn0052831   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region     Start      Stop Strand  \\\n",
       "0   FBtr0091683         2L  start_codon  11278451  11278453      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  TTAGCGGACCATTAGAAACACAATTGAGTTTGCATTGGGTGCTTTA...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GCGGGAGAGAGTCGAGAGTCATTCGTTTGTCTCACTCCTGCGAGTG...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTA...             19      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  mutated?  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       14     False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HAR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:497: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) - 1 #HA is 1bp to the right of stop\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:497: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) - 1 #HA is 1bp to the right of stop\n",
      "C:\\Users\\magd5549\\AppData\\Local\\Temp\\ipykernel_10360\\3952113674.py:51: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[\"mutated?\"] = True\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:587: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf.at[\"HAR\"] = mutable_region + winnerdf[\"HAR\"][winnerdf[f\"{side}_HA\"][-1]+1:]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCATCCCAATGCGGAGTCGCCATCGGGGTAAGTTGATTCTATGCGGTGGAGTCGCGAGTTTTGTGGACACTAATCGGAGTGTTTCTGTGTGCCTTGTTCCCTCGCCACGGCTGCAGGAAGTGGACGGACGAGGAGATCGACATGCTGCACTCGTCCATAATGCGCTTCTCCGACGACCTGACCAAGATCAGCCTAAGCATCAAGAACCGCACCGTGTAAG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CAGCTGCATCCCAATGCTGAGTCGCCATCGGGGTAAGTTGATTCTATGCGGTGGAGTCGCGAGTTTTGTGGACACTAATCGGAGTGTTTCTGTGTGCCTTGTTCCCTCGCCACGGCTGCAGGAAGTGGACGGACGAGGAGATCGACATGCTGCACTCGTCCATAATGCGCTTCTCCGACGACCTGACCAAGATCAGCCTAAGCATCAAGAACCGCACCGTGTAAG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CATGCAGCTGCATCCCAATGCGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test find_best_mutation function which mutates the HA in a way that it cannot be cut by the sgRNA\n",
    "# test case where PAM is in CDS, sixth dataframe in bestsgRNAList\n",
    "from sgRNAutils import find_best_mutation\n",
    "pd.set_option('display.max_columns', None)\n",
    "display(bestsgRNAList[5])\n",
    "bestsgRNAList[5][\"mutated?\"] = False\n",
    "CDSside = bestsgRNAList[5][\"CDS_side\"]\n",
    "mutated_sgRNAdf6 = find_best_mutation(bestsgRNAList[5].iloc[0])\n",
    "# compare the HA wich contains the CDS before and after mutation\n",
    "# check position of mutation by comparing the HA with the sgRNA\n",
    "print(CDSside.iloc[0])\n",
    "display(bestsgRNAList[5].iloc[0][CDSside].iloc[0])\n",
    "display(mutated_sgRNAdf6[CDSside].iloc[0])\n",
    "display(mutated_sgRNAdf6[\"sgRNA_sequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>mutated?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>524038</td>\n",
       "      <td>524060</td>\n",
       "      <td>2L</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCTTAGCAGCAATACAAGAGG</td>\n",
       "      <td>FBgn0003963</td>\n",
       "      <td>FBtr0078063</td>\n",
       "      <td>2L</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>524038</td>\n",
       "      <td>524040</td>\n",
       "      <td>+</td>\n",
       "      <td>CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...</td>\n",
       "      <td>AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...</td>\n",
       "      <td>CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...</td>\n",
       "      <td>20</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-8, -7, -6, -5, -4, -3]</td>\n",
       "      <td>[-1, 0]</td>\n",
       "      <td>15</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     fmin    fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  524038  524060   2L      +  ATGCTTAGCAGCAATACAAGAGG  FBgn0003963   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region   Start    Stop Strand  \\\n",
       "0   FBtr0078063         2L  start_codon  524038  524040      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  CAGAAAGGGATGGGCACACTCGCCTGTTTCCTGCAGTCGAACAGAT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  AAGAGCTACAATACGATACCGCACGTTCGTCATAACGAGAATTTGA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCT...             20      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False       True   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0           False       True   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "               SRS_boundary mutable_PAM cut_site  mutated?  \n",
       "0  [-8, -7, -6, -5, -4, -3]     [-1, 0]       15     False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HAR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:497: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) - 1 #HA is 1bp to the right of stop\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:497: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) - 1 #HA is 1bp to the right of stop\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:587: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf.at[\"HAR\"] = mutable_region + winnerdf[\"HAR\"][winnerdf[f\"{side}_HA\"][-1]+1:]\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:497: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) - 1 #HA is 1bp to the right of stop\n",
      "C:\\Users\\magd5549\\AppData\\Local\\Temp\\ipykernel_10360\\3952113674.py:67: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[\"mutated?\"] = True\n",
      "C:\\Users\\magd5549\\AppData\\Local\\Temp\\ipykernel_10360\\3952113674.py:67: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[\"mutated?\"] = True\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:587: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf.at[\"HAR\"] = mutable_region + winnerdf[\"HAR\"][winnerdf[f\"{side}_HA\"][-1]+1:]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'CTTAGCAGCAATACAAGAGGTGAGAATTATTCATACACAGTTTTCTATGAAACACAATATGCTTGAATATGCTTTTAATAAGAAAAGCAAACTAGCAATTCTATCTGATTTAGATGTAAACAAATTATTCATTCTATTTGGAATGTTTGGATAATTATGTTATATTAAATTTGAGATCTAAAGTGAATGTATCATTTGTTGTTGATATTGATTGTTGAATATATG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CTTAGCAGCAACACTAGAGGTGAGAATTATTCATACACAGTTTTCTATGAAACACAATATGCTTGAATATGCTTTTAATAAGAAAAGCAAACTAGCAATTCTATCTGATTTAGATGTAAACAAATTATTCATTCTATTTGGAATGTTTGGATAATTATGTTATATTAAATTTGAGATCTAAAGTGAATGTATCATTTGTTGTTGATATTGATTGTTGAATATATG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ATGCTTAGCAGCAATACAAGAGG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test find_best_mutation function which mutates the HA in a way that it cannot be cut by the sgRNA\n",
    "# test case where PAM is in CDS, third dataframe in bestsgRNAList\n",
    "pd.set_option('display.max_columns', None)\n",
    "display(bestsgRNAList[2])\n",
    "bestsgRNAList[2][\"mutated?\"] = False\n",
    "CDSside = bestsgRNAList[2][\"CDS_side\"]\n",
    "mutated_sgRNAdf3 = find_best_mutation(bestsgRNAList[2].iloc[0])\n",
    "# compare the HA wich contains the CDS before and after mutation\n",
    "# check position of mutation by comparing the HA with the sgRNA\n",
    "# no synoynmous mutations in the PAM were possible (likely because not in wobble)\n",
    "# therefore the SRS was mutated\n",
    "print(CDSside.iloc[0])\n",
    "display(bestsgRNAList[2].iloc[0][CDSside].iloc[0])\n",
    "display(mutated_sgRNAdf3[CDSside].iloc[0])\n",
    "display(mutated_sgRNAdf3[\"sgRNA_sequence\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fmin</th>\n",
       "      <th>fmax</th>\n",
       "      <th>#chr</th>\n",
       "      <th>strand</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Gene_Region</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Reference_Seq</th>\n",
       "      <th>HAL</th>\n",
       "      <th>HAR</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>CDS_side</th>\n",
       "      <th>non_CDS_side</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>max_15_bp_3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>SRS_in_CDS</th>\n",
       "      <th>CDS_boundary</th>\n",
       "      <th>non_CDS_boundary</th>\n",
       "      <th>SRS_boundary</th>\n",
       "      <th>mutable_PAM</th>\n",
       "      <th>cut_site</th>\n",
       "      <th>mutated?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8600563</td>\n",
       "      <td>8600585</td>\n",
       "      <td>2R</td>\n",
       "      <td>-</td>\n",
       "      <td>CCAGTAAAGCGCAGCAGCAATGA</td>\n",
       "      <td>FBgn0267792</td>\n",
       "      <td>FBtr0088692</td>\n",
       "      <td>2R</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>8600582</td>\n",
       "      <td>8600584</td>\n",
       "      <td>+</td>\n",
       "      <td>GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...</td>\n",
       "      <td>GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...</td>\n",
       "      <td>ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...</td>\n",
       "      <td>1</td>\n",
       "      <td>HAR</td>\n",
       "      <td>HAL</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...</td>\n",
       "      <td>[-24, -23, -22, -21, -20, -19, -18, -17, -16, ...</td>\n",
       "      <td>[-19, -18, -17, -16, -15, -14]</td>\n",
       "      <td>[-22, -21]</td>\n",
       "      <td>-16</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      fmin     fmax #chr strand           sgRNA_sequence      Gene_ID  \\\n",
       "0  8600563  8600585   2R      -  CCAGTAAAGCGCAGCAGCAATGA  FBgn0267792   \n",
       "\n",
       "  Transcript_ID Chromosome  Gene_Region    Start     Stop Strand  \\\n",
       "0   FBtr0088692         2R  start_codon  8600582  8600584      +   \n",
       "\n",
       "                                       Reference_Seq  \\\n",
       "0  GCCGCGGGTAAAGGCCTCTCACCAAACAAAGAAAATGTGACGAGCT...   \n",
       "\n",
       "                                                 HAL  \\\n",
       "0  GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCA...   \n",
       "\n",
       "                                                 HAR  positionScore CDS_side  \\\n",
       "0  ATGCCGCTGCAGGAGTCGCCGTCGCCCAGTTGGCAGCTGGAGGACT...              1      HAR   \n",
       "\n",
       "  non_CDS_side PAM_in_start/stop max_15_bp_3’_overhang PAM_in_CDS  \\\n",
       "0          HAL             False                 False      False   \n",
       "\n",
       "  PAM_outside_CDS SRS_in_CDS  \\\n",
       "0            True      False   \n",
       "\n",
       "                                        CDS_boundary  \\\n",
       "0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   \n",
       "\n",
       "                                    non_CDS_boundary  \\\n",
       "0  [-24, -23, -22, -21, -20, -19, -18, -17, -16, ...   \n",
       "\n",
       "                     SRS_boundary mutable_PAM cut_site  mutated?  \n",
       "0  [-19, -18, -17, -16, -15, -14]  [-22, -21]      -16     False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HAL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:501: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) + 2 #HA is 3pb to the left of stop but because we index backwards we calculate - 1\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:501: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[f\"{side}_HA\"] = np.array(side_boundary) + 2 #HA is 3pb to the left of stop but because we index backwards we calculate - 1\n",
      "c:\\Users\\magd5549\\OneDrive - Nexus365\\Bioinformatics\\AutoTagsCRISPR\\sgRNAutils.py:589: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf.at[\"HAL\"] =  winnerdf[\"HAL\"][:winnerdf[f\"{side}_HA\"][0]+1] + mutable_region\n",
      "C:\\Users\\magd5549\\AppData\\Local\\Temp\\ipykernel_10360\\3952113674.py:88: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  winnerdf[\"mutated?\"] = True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCACTGCCTCGGCCACACTGCAGTTCGGATTTGCCGGTTGTTTGGATCGCAGATTGAATCGGAAAGCCAGCGGCGTTGGTTCGGAAAAGTGCAGTCTCCGCGTCGCGAGTGCGCCAATAACGAGAAATAACGAGAATTGAAAAAGTTCCTTAAGAGGTGGAGTCCAGTAAAGCGCAGCAGCA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'GAAGCGCATTGTATTTGGCCACCATCGAGCGATGGCCACACTGCCACTGCCTCGGCCACACTGCAGTTCGGATTTGCCGGTTGTTTGGATCGCAGATTGAATCGGAAAGCCAGCGGCGTTGGTTCGGAAAAGTGCAGTCTCCGCGTCGCGAGTGCGCCAATAACGAGAAATAACGAGAATTGAAAAAGTTCCTTAAGAGGTGGAGTCAAGTAAAGCGCAGCAGCA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'CCAGTAAAGCGCAGCAGCAATGA'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# test find_best_mutation function which mutates the HA in a way that it cannot be cut by the sgRNA\n",
    "# test case where neither PAM nor SRS is in CDS, seventh dataframe in bestsgRNAList\n",
    "# we expect that the PAM is simply mutated without respecting synonymous mutations\n",
    "pd.set_option('display.max_columns', None)\n",
    "display(bestsgRNAList[6])\n",
    "bestsgRNAList[6][\"mutated?\"] = False\n",
    "nonCDSside = bestsgRNAList[6][\"non_CDS_side\"]\n",
    "mutated_sgRNAdf7 = find_best_mutation(bestsgRNAList[6].iloc[0])\n",
    "# compare the HA wich does NOT contain the CDS before and after mutation\n",
    "# check position of mutation by comparing the HA with the sgRNA\n",
    "print(nonCDSside.iloc[0])\n",
    "display(bestsgRNAList[6].iloc[0][nonCDSside].iloc[0])\n",
    "display(mutated_sgRNAdf7[nonCDSside].iloc[0])\n",
    "display(mutated_sgRNAdf7[\"sgRNA_sequence\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# testing the find_best_mutation function for the case that the PAM is inside of the CDS\n",
    "# therefore I will just pretend that the PAM is in the CDS because it will not make a difference for this particular function\n",
    "checked_0[\"mutated?\"] = False\n",
    "checked_0[\"PAM_in_CDS\"] = True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT\n",
      "TGCATTCTTGTGTCCAGATCCGG\n",
      "CAT\n",
      "GCATTCTTGTGTCCAGATCCGGG\n"
     ]
    }
   ],
   "source": [
    "# check whether the start/stop codon and sgRNA is correct\n",
    "for ind, sgRNA in sgRNAdf.iterrows():\n",
    "    print(refSeqPerChromosome[sgRNA[\"Chromosome\"]][sgRNA[\"Start\"]-1:sgRNA[\"Stop\"]])\n",
    "    print(refSeqPerChromosome[sgRNA[\"Chromosome\"]][sgRNA[\"fmin\"]-1:sgRNA[\"fmax\"]])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
