{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (2689803032.py, line 13)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[29], line 13\u001b[0;36m\u001b[0m\n\u001b[0;31m    \"start/stop\":\"start\",#is it N or C termini -> do we need to look at start or stop codon for teh cut\u001b[0m\n\u001b[0m                ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "#Example params:\n",
    "import pandas as pd\n",
    "\n",
    "tfsdict = {\n",
    "        'Gene_ID': 'FBgn0004652',\n",
    "        'Transcript_ID': 'FBtr0083651', \n",
    "        'Chromosome': '3R', \n",
    "        'Gene_Region': 'stop_codon', \n",
    "        'Start': 18426145, \n",
    "        'Stop': 18426147, \n",
    "        'Strand': '-', \n",
    "        'Reference_Seq': 'TTGATCGTAGGACAC',\n",
    "        'upstreamHA': 'CTTGATCGTAGGACACCTGCACAG',\n",
    "        'downstreamHA': 'CCTTGATCGTAGGACACCTGCACA',\n",
    "        \"start/stop\":\"start\",#is it N or C termini -> do we need to look at start or stop codon for teh cut \n",
    "        'genome_start_codon_pos':400, \n",
    "        'genome_stop_codon_pos':700, # or only 1 of those \n",
    "        'strand_type':'+',\n",
    "        \"sgRNA_list_positions\":[[401,425],[456,467],[478,489],[395,415]],#those wil be as genome positions -assumptions - the coordinates correspond to the 1st and last bp of the strand to which the gsRNA will be complementary to\n",
    "        \"sgRNA_list_values\":[\"AAGCGACTA\",\"AAAAAAAATAAAAA\",\"ATATATTTTTTTTTTAAAAA\",\"AGCGCGAAATAATA\"],\n",
    "        \"sgRNA_strand\" = ['-', '+', '-']\n",
    "}\n",
    "\n",
    "tfsDF = pd.DataFrame.to_dict(tfsdict)\n",
    "\n",
    "tfsDF = {'Gene_ID': 'FBgn0004652',\n",
    "                    'Transcript_ID': 'FBtr0083651', \n",
    "                    'Chromosome': '3R', \n",
    "                    'Gene_Region': 'stop_codon', \n",
    "                    'Start': 18426145, \n",
    "                    'Stop': 18426147, \n",
    "                    'Strand': '-', \n",
    "                    'Reference_Seq': 'TTGATCGTAGGACACCTGCACAGATGCTTGATCGTAGGACACCTGCACAGATGCCTTGATCGTAGGACACCTGCACAGATGCTTGATCGTAGGACACCTGCACAGATGCTTGATCGTAGGACACCTGCACAGATGC', \n",
    "                    'upstreamHA': 'CTTGATCGTAGGACACCTGCACAG',\n",
    "                    'downstreamHA': 'CCTTGATCGTAGGACACCTGCACA'\n",
    "                }\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def revComp(inputSeq):\n",
    "  \"\"\"\n",
    "  This function takes an input sequence and returns the reverse complement.\n",
    "\n",
    "  Input: inputSeq in str format\n",
    "  Output: revComp in str format\n",
    "\n",
    "  \"\"\"\n",
    "  complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "  \n",
    "  revComp = \"\"\n",
    "  for base in inputSeq[::-1]:\n",
    "    revComp += complement[(base.upper())]\n",
    "\n",
    "  return revComp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sgRNACataloguer(df):\n",
    "    \"\"\"\n",
    "    Given filtered sgRNAs for a start/stop site in the 'df' format, will create a 'catalogue' dataframe containing \n",
    "    information about the last G position and whether the sgRNAs in the list meet certain conditions.\n",
    "\n",
    "    params: df: a dataframe for one start/stop site of the format: \n",
    "            df= {\"start/stop\":\"start\", #is it N or C termini -> do we need to look at start or stop codon for teh cut \n",
    "            'genome_start_codon_pos':403, \n",
    "            'genome_stop_codon_pos':406, # only one of these - the other is n/a\n",
    "            'strand_type':'+',\n",
    "            \"sgRNA_list_positions\":[[401,425],[403, 427]],#those wil be as genome positions -assumptions - the coordinates correspond to the 1st and last bp of the strand to which the gsRNA will be complementary to\n",
    "            \"sgRNA_list_values\":[\"AAGCGACTA\", \"CCTGTAA\"],\n",
    "            \"sgRNA_strand\" : ['-', '+']}\n",
    "    \n",
    "    output: dataframe with columns \"start/stop\", \"genome_start_codon_pos\", \"genome_stop_codon_pos\",\n",
    "        \"strand_type\", \"sgRNA_list_values\", \"sgRNA_strand\", \"sgRNA_fmax\", \"positionScore\", \"PAM_in_start/stop\",\n",
    "        \"<15_bp3’_overhang\", \"PAM_in_CDS\", \"PAM_outside_CDS\", \"Cutsite_in_CDS\", \"relative_position_score\", \"sgRNA_mutable_bases\"\n",
    "    \n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "\n",
    "    #Making 'df' into an sgRNA catalogue dataframe. This makes it easier to subset sgRNAs by condition.\n",
    "    sgRNACatalogue = pd.DataFrame.from_dict(df)\n",
    "    #keep only fmax to simplify indexing later\n",
    "    for ind, sgRNA in sgRNACatalogue.iterrows():\n",
    "        sgRNACatalogue.at[ind, \"sgRNA_fmax\"] = sgRNACatalogue[\"sgRNA_list_positions\"][ind][1] #split to fmax and fmin, delete []\n",
    "    sgRNACatalogue = sgRNACatalogue.drop(columns=[\"sgRNA_list_positions\"]) #delete the old position value to simplify dataframe (optional)\n",
    "    sgRNACatalogue[\"sgRNA_fmax\"] = sgRNACatalogue[\"sgRNA_fmax\"].astype(int) #convert to int\n",
    "\n",
    "    #Adding position scores (fmax - stop)\n",
    "    sgRNACatalogue[\"positionScore\"] = sgRNACatalogue[\"sgRNA_fmax\"] - sgRNACatalogue[\"genome_stop_codon_pos\"]\n",
    "                                                                         \n",
    "    #Dataframe containing parameter ranges to interpret the positon score, based on gene strand, sgRNA strand, and start/stop\n",
    "    positionScoreParameters = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\")\n",
    "\n",
    "    #Per parameter, append the sgRNACatalogue with a TRUE/FALSE value per condition.\n",
    "    #Also translate the column of the positionScoreParameters df indicating how many sgRNA bases within 6bp of PAM ccan be mutated inside CDS.\n",
    "    #Lastly, translate positionScore into the relative position of the last 'G' or 'C' of PAM.\n",
    "    booleanColumns = [\"PAM_in_start/stop\", \"<15_bp3’_overhang\", \"PAM_in_CDS\", \"PAM_outside_CDS\", \"Cutsite_in_CDS\"]\n",
    "    positionColumns = [\"relative_position_score\", \"sgRNA_mutable_bases\"]\n",
    "    sgRNACatalogue = sgRNACatalogue.reindex(columns = sgRNACatalogue.columns.tolist() + booleanColumns + positionColumns)\n",
    "    \n",
    "    for ind, sgRNA in sgRNACatalogue.iterrows():\n",
    "\n",
    "        #Extract the appropriate parameter row per sgRNA\n",
    "        conditions = positionScoreParameters.loc[(positionScoreParameters[\"start/stop\"] == sgRNA['start/stop']) & (positionScoreParameters[\"strand_type\"] == sgRNA['strand_type']) & (positionScoreParameters[\"sgRNA_strand\"] == sgRNA['sgRNA_strand'])]\n",
    "        conditions = conditions.reset_index(drop = True)\n",
    "\n",
    "        #Per column, input true/false as to whether the position score meets that condition\n",
    "        for col in booleanColumns:\n",
    "            colValue = conditions.at[0,col] #extract parameter range values from dataframe\n",
    "            #Process the value into a range (in format list [min, max])\n",
    "            #If the values should be 'more than' or 'less than', 25 is used as a max or -25 as min because distances cannot be more than 20\n",
    "            if \">\" in colValue: #could simplify this further by just defining all as ranges in excel\n",
    "                minMax = [int(colValue[1:]), 25]\n",
    "            elif \"<\" in colValue:\n",
    "                minMax = [-25,int(colValue[1:])]\n",
    "            elif \":\" in colValue:\n",
    "                min, max = colValue.split(\":\")\n",
    "                minMax = [int(min), int(max)]\n",
    "            else:\n",
    "                print(\"Incorrect format of range value. Verify inputs.\")\n",
    "            \n",
    "            #Into the output dataframe, print true/false as to whether the positionScore has met the condition for that column\n",
    "            sgRNACatalogue.at[ind, col] = bool(sgRNACatalogue.at[ind, \"positionScore\"] in range(minMax[0], minMax[1]))\n",
    "       \n",
    "        #Calculate PAM last G or C position using positionScore. Add this to the sgRNA catalogue.\n",
    "        sgRNACatalogue.at[ind, \"relative_position_score\"] = sgRNACatalogue.at[ind, \"positionScore\"] + int(conditions.at[0, \"relative_position_score\"])\n",
    "\n",
    "        #Calculate how many bases of the sgRNA recognition site are in the CDS and can therefore be mutated.\n",
    "        #This is given in the column \"sgRNA_mutable_bases\", in a list of the format [x,y,z,...] indicating the position score for which 1-6 bases can be mutated.\n",
    "        #Accounting for the case where there is more than one positionScore for which 6bp can be mutated, this translates to:\n",
    "        mutateBases = [1,2,3,4,5,6,6,6,6,6,6,6,6,6,6,6,6]\n",
    "        if str(sgRNACatalogue.at[ind, \"positionScore\"]) in conditions.at[0, \"sgRNA_mutable_bases\"]:\n",
    "            index = conditions.at[0, \"sgRNA_mutable_bases\"].index(str(sgRNACatalogue.at[ind, \"positionScore\"]))\n",
    "            sgRNACatalogue.at[ind, \"sgRNA_mutable_bases\"] = str(mutateBases[index])\n",
    "        else:\n",
    "            sgRNACatalogue.at[ind, \"sgRNA_mutable_bases\"] = False\n",
    "    \n",
    "    sgRNACatalogue[\"relative_position_score\"] = sgRNACatalogue[\"relative_position_score\"].astype(int) #convert to int\n",
    "\n",
    "    return sgRNACatalogue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start/stop</th>\n",
       "      <th>genome_start_codon_pos</th>\n",
       "      <th>genome_stop_codon_pos</th>\n",
       "      <th>strand_type</th>\n",
       "      <th>sgRNA_list_values</th>\n",
       "      <th>sgRNA_strand</th>\n",
       "      <th>sgRNA_fmax</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>&lt;15_bp3’_overhang</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>Cutsite_in_CDS</th>\n",
       "      <th>relative_position_score</th>\n",
       "      <th>sgRNA_mutable_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>start</td>\n",
       "      <td>403</td>\n",
       "      <td>406</td>\n",
       "      <td>+</td>\n",
       "      <td>AAGCGACTA</td>\n",
       "      <td>-</td>\n",
       "      <td>425</td>\n",
       "      <td>19</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>start</td>\n",
       "      <td>403</td>\n",
       "      <td>406</td>\n",
       "      <td>+</td>\n",
       "      <td>CCTGTAA</td>\n",
       "      <td>+</td>\n",
       "      <td>427</td>\n",
       "      <td>21</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>44</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start/stop  genome_start_codon_pos  genome_stop_codon_pos strand_type  \\\n",
       "0      start                     403                    406           +   \n",
       "1      start                     403                    406           +   \n",
       "\n",
       "  sgRNA_list_values sgRNA_strand  sgRNA_fmax  positionScore PAM_in_start/stop  \\\n",
       "0         AAGCGACTA            -         425             19              True   \n",
       "1           CCTGTAA            +         427             21             False   \n",
       "\n",
       "  <15_bp3’_overhang PAM_in_CDS PAM_outside_CDS Cutsite_in_CDS  \\\n",
       "0              True      False           False           True   \n",
       "1             False       True           False           True   \n",
       "\n",
       "   relative_position_score sgRNA_mutable_bases  \n",
       "0                       19                   6  \n",
       "1                       44               False  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sgRNACatalogue = sgRNACataloguer(df)\n",
    "display(sgRNACatalogue)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "def checkCDSCutandOrder(sgRNACatalogue):\n",
    "    \"\"\"\n",
    "    Given an sgRNACatalogue or subset, will calculate firstly rows where the sgRNA cuts inside CDS. If multiple, selects that which cuts closest to start/stop.\n",
    "    If none, selects closest cut sgRNA that is outside CDS.\n",
    "    \n",
    "    \"\"\"\n",
    "    #C. Check cut site in CDS\n",
    "    conditionC = sgRNACatalogue[sgRNACatalogue[\"Cutsite_in_CDS\"] == True]\n",
    "    conditionCclosestCut = conditionC.sort_values('positionScore') #This value is the absolute value of positionScore, so indicates those that cut closest to start/stop\n",
    "\n",
    "    if len(conditionCclosestCut) > 0: #cuts in CDS, closest cut (C1, C2)\n",
    "        conditionCclosestCut = conditionCclosestCut.reset_index(drop = True) #reset index\n",
    "        winnersgRNA = conditionCclosestCut.at[0, \"sgRNA_sequence\"]\n",
    "        winnerFound = True\n",
    "    else: #no sgRNAs cut in CDS, select closest that still met condition B (C3)\n",
    "        nonCDSclosestCut = sgRNACatalogue.sort_values('PAMRelativeEnd')\n",
    "        winnersgRNA = nonCDSclosestCut.at[0, \"sgRNA_sequence\"]\n",
    "        winnerFound = True\n",
    "\n",
    "    return winnersgRNA, winnerFound\n",
    "\n",
    "def find_best_gRNA(df):\n",
    "    \"\"\"\n",
    "    params: df={\n",
    "            \"start/stop\":\"start\",#is it N or C termini -> do we need to look at start or stop codon for teh cut \n",
    "            'genome_start_codon_pos':400, \n",
    "            'genome_stop_codon_pos':700, # or only 1 of those \n",
    "            'strand_type':'+',\n",
    "            \"sgRNA_list_positions\":[[401,425],[456,467],[478,489],[395,415]],#those wil be as genome positions -assumptions - the coordinates correspond to the 1st and last bp of the strand to which the gsRNA will be complementary to\n",
    "            \"sgRNA_list_values\":[\"AAGCGACTA\",\"AAAAAAAATAAAAA\",\"ATATATTTTTTTTTTAAAAA\",\"AGCGCGAAATAATA\"]\n",
    "        }\n",
    "    \n",
    "    \"\"\"\n",
    "    #Score the sgRNAs for this site\n",
    "    sgRNACatalogue = sgRNACataloguer(df)\n",
    "\n",
    "    #Set up the winning guide RNA\n",
    "    winnerFound = False\n",
    "    mutationNeeded = False\n",
    "    winnersgRNA = \"\" #Could update this to be more information about the guide e.g. strand, position\n",
    "\n",
    "    #If there is no guide RNAs at this site, flag this. The resulting sgRNA will be an empty string.\n",
    "    if len(sgRNACatalogue) == 0:\n",
    "        print(f\"No guide RNAs were found for the start/stop site beginning at position {df['genome_start_codon_pos']}\")\n",
    "        winnerFound = True #To avoid looping through other conditions if there are no sgRNAs\n",
    "\n",
    "    #A. Ideal condition - PAM in start/stop\n",
    "    if winnerFound == False:\n",
    "        conditionA = sgRNACatalogue[sgRNACatalogue[\"PAM_in_start/stop\"] == True] #This is the subset df for which PAM is in the start/stop\n",
    "        if len(conditionA) > 0: #if there is one or more sgRNAs for this condition, select the first as the winner\n",
    "            conditionA = conditionA.reset_index(drop = True) #reset index\n",
    "            winnersgRNA = conditionA.at[0, \"sgRNA_list_values\"]\n",
    "            winnerFound = True\n",
    "\n",
    "    #B. sgRNA overhang is less than 15bp\n",
    "    if winnerFound == False:\n",
    "        conditionB = sgRNACatalogue[sgRNACatalogue[\"<15_bp3’_overhang\"] == True]\n",
    "        if len(conditionB) == 1:\n",
    "            winnersgRNA = conditionB\n",
    "            winnerFound = True\n",
    "        elif len(conditionB) > 1: #more than one, select in CDS preferencially, and closest cut\n",
    "            winnersgRNA, winnerFound = checkCDSCutandOrder(conditionB)    \n",
    "\n",
    "    #D. sgRNA overhang is more than 15bp, need to mutate\n",
    "    if winnerFound == False:\n",
    "        if len(sgRNACatalogue) == 1:\n",
    "            winnersgRNA = sgRNACatalogue.at[0, \"sgRNA_list_values\"]\n",
    "        else: #more than 1 sgRNA\n",
    "            winnersgRNA, winnerFound = checkCDSCutandOrder(sgRNACatalogue)            \n",
    "\n",
    "        mutationNeeded = True\n",
    "\n",
    "    return winnersgRNA, mutationNeeded, sgRNACatalogue[sgRNACatalogue[\"sgRNA_list_values\"] == winnersgRNA] #could adapt return to be just the true/false for mutation and the catalogue\n",
    "    #output dictionary with one set of values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('AAGCGACTA',\n",
       " False,\n",
       "   start/stop  genome_start_codon_pos  genome_stop_codon_pos strand_type  \\\n",
       " 0      start                     403                    406           +   \n",
       " \n",
       "   sgRNA_list_values sgRNA_strand  sgRNA_fmax  positionScore PAM_in_start/stop  \\\n",
       " 0         AAGCGACTA            -         425             19              True   \n",
       " \n",
       "   <15_bp3’_overhang PAM_in_CDS PAM_outside_CDS Cutsite_in_CDS  \\\n",
       " 0              True      False           False           True   \n",
       " \n",
       "    relative_position_score sgRNA_mutable_bases  \n",
       " 0                       19                   6  )"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "winnersgRNA, mutationNeeded, sgRNACatalogue = find_best_gRNA(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Before mutating, iterate through stringency files\n",
    "\n",
    "#This should be specified in the runner. Also in the runner, pass winner sgRNA catalogue to mutator only where mutationNeeded == True.\n",
    "#Make sure to re-index the dataframe so that the one row can be indexed as '0'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codonFragmenter(winner_sgRNACatalogue, tfsDF):\n",
    "    \"\"\"\n",
    "    For the start/stop site, will create a list of codons in the appropriate range where sgRNAs might be found (start/stop ±20 bp).\n",
    "    If gene is on - strand, these will be revComp codons.\n",
    "    \"\"\"\n",
    "    \n",
    "    #Extract HA sequences - just the 21bp near start/stop\n",
    "    HAL = tfsDF[\"upstreamHA\"][-21:]\n",
    "    HAR = tfsDF[\"downstreamHA\"][0:21]\n",
    "\n",
    "    #Start codon list\n",
    "    codonList = []\n",
    "\n",
    "    #Add HAL codons\n",
    "    for codonBase1 in range(0, len(HAL), 3):\n",
    "        codonList.append(HAL[codonBase1:codonBase1+3])\n",
    "\n",
    "    #add start/stop - this is added as 'ATG' (even if stop) but will never be mutated.\n",
    "    codonList.append('ATG')\n",
    "\n",
    "    #Add HAR codons\n",
    "    for codonBase1 in range(0, len(HAR), 3):\n",
    "        codonList.append(HAR[codonBase1:codonBase1+3])\n",
    "    \n",
    "    #Define gene strand:\n",
    "    if winner_sgRNACatalogue[\"strand_type\"] == '-': #If on the minus strand, take revComp per codon\n",
    "        for ind, codon in enumerate(codonList):\n",
    "            codonList[ind] = revComp(codon)\n",
    "\n",
    "    return codonList\n",
    "\n",
    "def codonReverseFragmenter(mutableCodons, tfsDF, winner_sgRNACatalogue):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    HAL = \"\"\n",
    "    HAR = \"\"\n",
    "\n",
    "    #return to + strand if the gene is on -\n",
    "    if winner_sgRNACatalogue[\"strand_type\"] == '-': #If on the minus strand, take revComp per codon\n",
    "        for ind, codon in enumerate(mutableCodons):\n",
    "            mutableCodons[ind] = revComp(codon)\n",
    "    \n",
    "    #replace HAL\n",
    "    for codon in range (0, 7):\n",
    "        HAL += codon\n",
    "\n",
    "    #replace HAR\n",
    "    for codon in range (8, 15):\n",
    "        HAR += codon\n",
    "\n",
    "    #check this indexing\n",
    "    tfsDF[\"upstreamHA\"] = tfsDF[\"upstreamHA\"][:-21] + HAL\n",
    "    tfsDF[\"downstreamHA\"] = HAR + tfsDF[\"downstreamHA\"][21:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "unmatched ')' (203083978.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[74], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    def mutator()):\u001b[0m\n\u001b[0m                 ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unmatched ')'\n"
     ]
    }
   ],
   "source": [
    "def mutator(winner_sgRNACatalogue, tfsDF):\n",
    "    \"\"\"\n",
    "    In the case where a fragment or primer needs to be mutated, will mutate in CDS (preferably PAM, if not in the sgRNA). \n",
    "    If not possible, will mutate PAM outside of CDS to NTG/CTN.\n",
    "    \n",
    "    params:\n",
    "        sequenceToMutate: the homology arm or primer fragment that will be mutated.\n",
    "        winner_sgRNACatalogue: sgRNACatalogue in format as above, with only the row for the winner sgRNA selected.\n",
    "    \"\"\"\n",
    "\n",
    "    import pandas as pd\n",
    "\n",
    "    #Mutated starts as false\n",
    "    mutated = False\n",
    "\n",
    "    #Codon fragmenter - codons are now in mutable format in a list, from leftmost of HAL to right-most of HAR. RevComp if gene is -\n",
    "    mutableCodons = codonFragmenter(winner_sgRNACatalogue, tfsDF)\n",
    "\n",
    "    ## Coordinate information per codon\n",
    "    if winner_sgRNACatalogue.at[0, \"strand_type\"] == '+':\n",
    "        codonCoordinates = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\", sheet_name=\"CodonCoordinatePlus\", index_col= 0)\n",
    "    elif winner_sgRNACatalogue.at[0, \"strand_type\"] == '-':\n",
    "        codonCoordinates = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\", sheet_name=\"CodonCoordinateMinus\", index_col= 0)\n",
    "\n",
    "    #PAM last G coordinate\n",
    "    PAMcoordinate = winner_sgRNACatalogue.at[0, 'relative_position_score']\n",
    "\n",
    "    #Direction to move away from PAM\n",
    "    if winner_sgRNACatalogue.at[0, 'sgRNA_strand'] == '+':\n",
    "        direction = 'decreasing'\n",
    "    else:\n",
    "        direction = 'increasing'\n",
    "\n",
    "    #Coordinates for bases: PAM1, PAM2, PAM3, sgRNA1, sgRNA2, sgRNA3, sgRNA4, sgRNA5, sgRNA6\n",
    "    mutationCoordinates = []\n",
    "    if direction == 'decreasing':\n",
    "        for shift in range(0, 8):\n",
    "            mutationCoordinates.append(PAMcoordinate - shift)\n",
    "    elif direction == 'increasing':\n",
    "        for shift in range(0, 8):\n",
    "            mutationCoordinates.append(PAMcoordinate + shift)\n",
    "\n",
    "    #A. Mutate PAM in CDS\n",
    "    if sgRNACatalogue[\"PAM_in_CDS\"] == True:\n",
    "        mutationIndex = 0\n",
    "        while mutated == False:\n",
    "            PAMcoordinate = mutationCoordinates[mutationIndex]\n",
    "            codon = mutableCodons[codonCoordinates.at[PAMcoordinate, 'codon']] #This is the codon that this base is in\n",
    "            base = codonCoordinates.at[PAMcoordinate, 'base']\n",
    "            potentialCodons = find_synonymous_codons(querycodon = codon, baseToChange = base) #base is the base number within the codon, 1-3\n",
    "            if potentialCodons > 0:\n",
    "                mutableCodons[codon] = potentialCodons[0]\n",
    "                mutated = True\n",
    "            else:\n",
    "                mutationIndex +=1\n",
    "            if mutationIndex == 3:\n",
    "                break\n",
    "    \n",
    "    #B. Mutate sgRNA in CDS\n",
    "    #Check if mutation was successful in A, then check there are mutable bases of the sgRNA in CDS \n",
    "    if mutated == False and sgRNACatalogue[\"PAM_in_CDS\"] != False:\n",
    "        mutate = sgRNACatalogue[\"sgRNA_mutable_bases\"] #This is the number of bases (1-6) of the sgRNA that we can mutate\n",
    "        sgMutatedBases = 0\n",
    "        baseIndex = 3 #this is the first index in mutationCoordinates that is part of the sgRNA rather than PAM\n",
    "        moveIndex = 0 #this is to keep track of how many bases we try to mutate, as this should be capped at 'mutate' distance\n",
    "        while moveIndex < mutate: #mutate a max of 2 bases, and can only cover bases within the 'mutate' range\n",
    "            coordinate = mutableCodons[baseIndex + moveIndex]\n",
    "            codon = mutableCodons[codonCoordinates.at[coordinate, 'codon']] #This is the codon that this base is in\n",
    "            base = codonCoordinates.at[coordinate, 'base']\n",
    "            potentialCodons = find_synonymous_codons(querycodon = codon, baseToChange = base) #base is the base number within the codon, 1-3\n",
    "            if potentialCodons > 0: #if mutation is possible, accept this mutation\n",
    "                mutableCodons[codon] = potentialCodons[0]\n",
    "                sgMutatedBases += 1 #add 1 to the mutated bases count\n",
    "            moveIndex += 1 #move to the next base\n",
    "            if sgMutatedBases == 2: #if we've made two mutations, set mutation check to true and exit loop\n",
    "                mutated == True\n",
    "                break\n",
    "\n",
    "    #Mutate PAM outside CDS - to NTG or CTN\n",
    "    if mutated == False:\n",
    "        PAM2pos = mutationCoordinates[1] #this is the relative position of the second G/C of PAM\n",
    "        codon = mutableCodons[codonCoordinates.at[PAM2pos, 'codon']] #This is the codon that this base is in\n",
    "        base = codonCoordinates.at[PAMcoordinate, 'base']\n",
    "        newCodon = codon[:base-1] + 'T' + codon[base:] #add a 'T' where the old base was \n",
    "        mutableCodons[PAM2pos] = newCodon\n",
    "\n",
    "    #Replace the potentially mutated homology arms\n",
    "    mutableCodons = codonReverseFragmenter(mutableCodons, tfsDF, winner_sgRNACatalogue)\n",
    "\n",
    "    #Indicate if mutation has occurred\n",
    "    tfsDF['mutated'] = mutated\n",
    "\n",
    "    return tfsDF\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Codon</th>\n",
       "      <th>Base</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Relative_coordinate_in_mutable_area</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     Codon  Base\n",
       "Relative_coordinate_in_mutable_area             \n",
       "0                                        0     2\n",
       "1                                        0     1\n",
       "2                                        1     3\n",
       "3                                        1     2\n",
       "4                                        1     1\n",
       "5                                        2     3\n",
       "6                                        2     2\n",
       "7                                        2     1\n",
       "8                                        3     3\n",
       "9                                        3     2\n",
       "10                                       3     1\n",
       "11                                       4     3\n",
       "12                                       4     2\n",
       "13                                       4     1\n",
       "14                                       5     3\n",
       "15                                       5     2\n",
       "16                                       5     1\n",
       "17                                       6     3\n",
       "18                                       6     2\n",
       "19                                       6     1\n",
       "20                                       7     3\n",
       "21                                       7     2\n",
       "22                                       7     1\n",
       "23                                       8     3\n",
       "24                                       8     2\n",
       "25                                       8     1\n",
       "26                                       9     3\n",
       "27                                       9     2\n",
       "28                                       9     1\n",
       "29                                      10     3\n",
       "30                                      10     2\n",
       "31                                      10     1\n",
       "32                                      11     3\n",
       "33                                      11     2\n",
       "34                                      11     1\n",
       "35                                      12     3\n",
       "36                                      12     2\n",
       "37                                      12     1\n",
       "38                                      13     3\n",
       "39                                      13     2\n",
       "40                                      13     1\n",
       "41                                      14     3\n",
       "42                                      14     2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "codonCoordinates = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\", sheet_name=\"CodonCoordinateMinus\", index_col= 0)\n",
    "display(codonCoordinates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_synonymous_codons(query_codon, base_to_change, codon_table_excel = \"inputfiles/codon_table.xlsx\"):\n",
    "\n",
    "    '''\n",
    "    Uses the amino acids table to select codons that encode for the same amino acid as the query codon. Will mutate at the specified base only.\n",
    "\n",
    "    Params:\n",
    "        query_codon: string, codon to select synonymous codons for\n",
    "        base_to_change: the base within the codon that needs to change. Value from 1-3.\n",
    "        codon_table_excel: string, path to an excel file that lists per codon which amino acid that codon encodes.\n",
    "    \n",
    "    Returns:\n",
    "        synonymous_codons: list of strings, each string is a codon that encodes for the same amino acid as the query codon.\n",
    "    \n",
    "    '''\n",
    "\n",
    "    import pandas as pd\n",
    "\n",
    "    synonymous_codons = []\n",
    "\n",
    "    codon_table = pd.read_excel(codon_table_excel)\n",
    "\n",
    "    #Extract amino acid given codon\n",
    "    amino_acid_query = codon_table[codon_table['codon'] == query_codon].iloc[0][\"amino_acid\"]\n",
    "\n",
    "    #Subset df for other rows corresponding to this amino acid\n",
    "    same_aa_df=codon_table[codon_table[\"amino_acid\"] == amino_acid_query]\n",
    "\n",
    "    #Convert the codons to a list\n",
    "    codon_list = same_aa_df[\"codon\"].values.tolist()\n",
    "\n",
    "    #Keep only codons where the indicated base has changed\n",
    "    for codon in codon_list:\n",
    "        if codon[base_to_change-1] == query_codon[base_to_change-1]:\n",
    "            codon_list.remove(codon)\n",
    "\n",
    "    return codon_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sgRNACataloguer(df):\n",
    "    \"\"\"\n",
    "    Given filtered sgRNAs for a start/stop site in the 'df' format, will create a 'catalogue' dataframe containing \n",
    "    information about the last G position and whether the sgRNAs in the list meet certain conditions.\n",
    "\n",
    "    params: df: a dataframe for one start/stop site of the format: \n",
    "            df= {\"start/stop\":\"start\", #is it N or C termini -> do we need to look at start or stop codon for teh cut \n",
    "            'genome_start_codon_pos':403, \n",
    "            'genome_stop_codon_pos':406, # only one of these - the other is n/a\n",
    "            'strand_type':'+',\n",
    "            \"sgRNA_list_positions\":[[401,425],[403, 427]],#those wil be as genome positions -assumptions - the coordinates correspond to the 1st and last bp of the strand to which the gsRNA will be complementary to\n",
    "            \"sgRNA_list_values\":[\"AAGCGACTA\", \"CCTGTAA\"],\n",
    "            \"sgRNA_strand\" : ['-', '+']}\n",
    "    \n",
    "    output: dataframe with columns \"start/stop\", \"genome_start_codon_pos\", \"genome_stop_codon_pos\",\n",
    "        \"strand_type\", \"sgRNA_list_values\", \"sgRNA_strand\", \"sgRNA_fmax\", \"positionScore\", \"PAM_in_start/stop\",\n",
    "        \"<15_bp3’_overhang\", \"PAM_in_CDS\", \"PAM_outside_CDS\", \"Cutsite_in_CDS\", \"relative_position_score\", \"sgRNA_mutable_bases\"\n",
    "    \n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "\n",
    "    #Making 'df' into an sgRNA catalogue dataframe. This makes it easier to subset sgRNAs by condition.\n",
    "    sgRNACatalogue = pd.DataFrame.from_dict(df)\n",
    "    #keep only fmax to simplify indexing later\n",
    "    for ind, sgRNA in sgRNACatalogue.iterrows():\n",
    "        sgRNACatalogue.at[ind, \"sgRNA_fmax\"] = sgRNACatalogue[\"sgRNA_list_positions\"][ind][1] #split to fmax and fmin, delete []\n",
    "    sgRNACatalogue = sgRNACatalogue.drop(columns=[\"sgRNA_list_positions\"]) #delete the old position value to simplify dataframe (optional)\n",
    "    sgRNACatalogue[\"sgRNA_fmax\"] = sgRNACatalogue[\"sgRNA_fmax\"].astype(int) #convert to int\n",
    "\n",
    "    #Adding position scores (fmax - stop)\n",
    "    sgRNACatalogue[\"positionScore\"] = sgRNACatalogue[\"sgRNA_fmax\"] - sgRNACatalogue[\"genome_stop_codon_pos\"]\n",
    "                                                                         \n",
    "    #Dataframe containing parameter ranges to interpret the positon score, based on gene strand, sgRNA strand, and start/stop\n",
    "    positionScoreParameters = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\")\n",
    "\n",
    "    #Per parameter, append the sgRNACatalogue with a TRUE/FALSE value per condition.\n",
    "    #Also translate the column of the positionScoreParameters df indicating how many sgRNA bases within 6bp of PAM ccan be mutated inside CDS.\n",
    "    #Lastly, translate positionScore into the relative position of the last 'G' or 'C' of PAM.\n",
    "    booleanColumns = [\"PAM_in_start/stop\", \"<15_bp3’_overhang\", \"PAM_in_CDS\", \"PAM_outside_CDS\", \"Cutsite_in_CDS\"]\n",
    "    positionColumns = [\"relative_position_score\", \"sgRNA_mutable_bases\"]\n",
    "    sgRNACatalogue = sgRNACatalogue.reindex(columns = sgRNACatalogue.columns.tolist() + booleanColumns + positionColumns)\n",
    "    \n",
    "    for ind, sgRNA in sgRNACatalogue.iterrows():\n",
    "\n",
    "        #Extract the appropriate parameter row per sgRNA\n",
    "        conditions = positionScoreParameters.loc[(positionScoreParameters[\"start/stop\"] == sgRNA['start/stop']) & (positionScoreParameters[\"strand_type\"] == sgRNA['strand_type']) & (positionScoreParameters[\"sgRNA_strand\"] == sgRNA['sgRNA_strand'])]\n",
    "        conditions = conditions.reset_index(drop = True)\n",
    "\n",
    "        #Per column, input true/false as to whether the position score meets that condition\n",
    "        for col in booleanColumns:\n",
    "            colValue = conditions.at[0,col] #extract parameter range values from dataframe\n",
    "            #Process the value into a range (in format list [min, max])\n",
    "            #If the values should be 'more than' or 'less than', 25 is used as a max or -25 as min because distances cannot be more than 20\n",
    "            if \">\" in colValue: #could simplify this further by just defining all as ranges in excel\n",
    "                minMax = [int(colValue[1:]), 25]\n",
    "            elif \"<\" in colValue:\n",
    "                minMax = [-25,int(colValue[1:])]\n",
    "            elif \":\" in colValue:\n",
    "                min, max = colValue.split(\":\")\n",
    "                minMax = [int(min), int(max)]\n",
    "            else:\n",
    "                print(\"Incorrect format of range value. Verify inputs.\")\n",
    "            \n",
    "            #Into the output dataframe, print true/false as to whether the positionScore has met the condition for that column\n",
    "            sgRNACatalogue.at[ind, col] = bool(sgRNACatalogue.at[ind, \"positionScore\"] in range(minMax[0], minMax[1]))\n",
    "       \n",
    "        #Calculate PAM last G or C position using positionScore. Add this to the sgRNA catalogue.\n",
    "        sgRNACatalogue.at[ind, \"relative_position_score\"] = sgRNACatalogue.at[ind, \"positionScore\"] + int(conditions.at[0, \"relative_position_score\"])\n",
    "\n",
    "        #Calculate how many bases of the sgRNA recognition site are in the CDS and can therefore be mutated.\n",
    "        #This is given in the column \"sgRNA_mutable_bases\", in a list of the format [x,y,z,...] indicating the position score for which 1-6 bases can be mutated.\n",
    "        #Accounting for the case where there is more than one positionScore for which 6bp can be mutated, this translates to:\n",
    "        mutateBases = [1,2,3,4,5,6,6,6,6,6,6,6,6,6,6,6,6]\n",
    "        if str(sgRNACatalogue.at[ind, \"positionScore\"]) in conditions.at[0, \"sgRNA_mutable_bases\"]:\n",
    "            index = conditions.at[0, \"sgRNA_mutable_bases\"].index(str(sgRNACatalogue.at[ind, \"positionScore\"]))\n",
    "            sgRNACatalogue.at[ind, \"sgRNA_mutable_bases\"] = str(mutateBases[index])\n",
    "        else:\n",
    "            sgRNACatalogue.at[ind, \"sgRNA_mutable_bases\"] = False\n",
    "    \n",
    "    sgRNACatalogue[\"relative_position_score\"] = sgRNACatalogue[\"relative_position_score\"].astype(int) #convert to int\n",
    "\n",
    "    return sgRNACatalogue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute 'any'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m a \u001b[39m=\u001b[39m [\u001b[39m\"\u001b[39m\u001b[39mATG\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mATA\u001b[39m\u001b[39m\"\u001b[39m]\n\u001b[0;32m----> 3\u001b[0m \u001b[39mif\u001b[39;00m (x[:\u001b[39m-\u001b[39m\u001b[39m1\u001b[39m] \u001b[39mfor\u001b[39;00m x \u001b[39min\u001b[39;00m a\u001b[39m.\u001b[39;49many()) \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m      4\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m      5\u001b[0m     \u001b[39m#a.remove[x]\u001b[39;00m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'list' object has no attribute 'any'"
     ]
    }
   ],
   "source": [
    "a = [\"ATG\", \"ATA\"]\n",
    "\n",
    "if (x[:-1] for x in a.any()) == \"A\":\n",
    "    print(\"A\")\n",
    "    #a.remove[x]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "codon = \"ATA\"\n",
    "a.remove(codon)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATG']\n"
     ]
    }
   ],
   "source": [
    "print(a)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Free variables\n",
    "\n",
    "#refSeqPerChromosome\n",
    "from Bio import SeqIO\n",
    "\n",
    "refSeqPerChromosome = {}\n",
    "\n",
    "for seq in SeqIO.parse(open(\"inputfiles/dmel-all-chromosome-r6.48.fasta\"), 'fasta'):\n",
    "    refSeqPerChromosome[seq.id] = seq.seq \n",
    "\n",
    "#TFsdf\n",
    "TFsdf = make_dataframe_from_TFs_list(\"inputfiles/TFs.xlsx\", refSeqPerChromosome)\n",
    "display(TFsdf)\n",
    "\n",
    "\"\"\"\n",
    "#singletfsDF = pandas series for one row extracted from tfsDF\n",
    "for index, row in TFsdf.iterrows():\n",
    "    if index == 0:\n",
    "        singletfsDF = row\n",
    "        print(singletfsDF)\n",
    "        \n",
    "#gRNAFileAnnotation = dataframe of gRNAs as extracted from original gff file\n",
    "import pandas as pd\n",
    "gRNAFileAnnotation = pd.read_csv(\"inputfiles/Hu.2019.8.28.sgRNA_designs/1to3NonCdsOffTarget_low_stringency.gff\", sep = \"\\t\", index_col = False)\n",
    "gRNAFileAnnotation = gRNAFileAnnotation.assign(target_site_variation= \"\")\n",
    "index = 0 #TODO@Marina improve this code\n",
    "for attribute in gRNAFileAnnotation['attributes']:\n",
    "    fullatt = (attribute).split(\";\")\n",
    "    gRNAFileAnnotation.at[index,\"target_site_variation\"] = fullatt[8]\n",
    "    index+=1\n",
    "\n",
    "print(type(gRNAFileAnnotation))\n",
    "\n",
    "#GenomeCoordinatesFiltered = filtered gRNAs for 20bp region, no target site variation, and with start/stop info from tfsDF\n",
    "GenomeCoordinatesFiltered = gRNAFileAnnotation[(gRNAFileAnnotation['target_site_variation'] == \"target_site_variation=none observed\") & \n",
    "                                                (gRNAFileAnnotation['#chr'] == singletfsDF[\"Chromosome\"]) & \n",
    "                                                (gRNAFileAnnotation['fmin']-1 >= singletfsDF[\"Start\"] - 20)]\n",
    "for (columnName, columnData) in singletfsDF.items():\n",
    "       GenomeCoordinatesFiltered = GenomeCoordinatesFiltered.assign(**{columnName: columnData})\n",
    "display(GenomeCoordinatesFiltered)\n",
    "\"\"\"\n",
    "#filteredgRNAs - theoretical sgRNAs for one site\n",
    "filteredgRNAs = pd.read_excel(\"inputfiles/mockMaterials/filtered_gRNAs.xlsx\", index_col= 0)\n",
    "\n",
    "display(filteredgRNAs)\n",
    "\n",
    "sgRNAdf = sgRNApositionCheck(filteredgRNAs)\n",
    "\n",
    "display(sgRNAdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def workingsgRNArunner(refSeqPerChromosome = refSeqPerChrom, TFsdf = TFsDF):\n",
    "    \"\"\"\n",
    "    Check functions so far work with the proposed changes.\n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "    \n",
    "    #set up the output DF - this is the winning sgRNA per site in TFsdf\n",
    "    TFsdfWinnersandMutated = pd.DataFrame(columns=[\"fmin\", \"fmax\", \"#chr\", \"strand\", \"sgRNA_sequence\", \"Gene_ID\",\n",
    "                                                   \"Transcript_ID\", \"Chromosome\", \"Gene_Region\", \"Start\", \"Stop\",\n",
    "                                                   \"Strand\", \"Reference_Seq\", \"upstreamHA\", \"downstreamHA\", \"positionScore\",\n",
    "                                                    \"PAM_in_start/stop\", \"<15_bp3’_overhang\", \"PAM_in_CDS\", \"PAM_outside_CDS\",\n",
    "                                                    \"CDS_boundary\", \"lastG\", \"cutSite\", \"mutated\"])\n",
    "    \n",
    "    #Per row of this dataframe, select a guideRNA\n",
    "    for ind, row in TFsdf.iterrows():\n",
    "        print(f\"Selecting guide RNA for TFsdf row {ind}\")\n",
    "\n",
    "        filtered_sgRNA = gRNA_stringencyIterator(row, refSeqPerChromosome)\n",
    "\n",
    "        #Score the sgRNAs for this site\n",
    "        sgRNAdf = sgRNApositionCheck(filtered_sgRNA)\n",
    "\n",
    "        #Add a column to establish whether mutation has occurred. This will be set to 'True' in the mutator function and starts as False by default.\n",
    "        sgRNAdf['mutated'] = False\n",
    "\n",
    "        #If no sgRNAs were found at any stringency\n",
    "        if len(sgRNAdf) == 0:\n",
    "            print(f\"No sgRNAs found at all stringencies for {ind}.\")\n",
    "            print(sgRNAdf)\n",
    "\n",
    "            #Just input the information about this site into the final DF. The columns about the guideRNA will be filled with 'NaN', indicating no guideRNA could be found.\n",
    "            for col in list(row.index):\n",
    "                TFsdfWinnersandMutated.at[ind, col] = row[col]\n",
    "            display(TFsdfWinnersandMutated)\n",
    "        else: #If we have at least one sgRNA\n",
    "            #Select winner\n",
    "            winnersgRNA, mutationNeeded = find_best_gRNA(sgRNAdf) #winnersgRNA is a string of the winning sequence, mutationNeeded is a bool variable\n",
    "            winnerdf = sgRNAdf[sgRNAdf[\"sgRNA_sequence\"] == winnersgRNA] #This is the dataframe row for the winning sgRNA from the original sgRNA dataframe\n",
    "            winnerdf = winnerdf.loc[0] #This is the pandas series for the same information\n",
    "            display(winnerdf)\n",
    "            if mutationNeeded == True: #run mutator if indicated\n",
    "                winnerdf = mutator(winnerdf) #this will mutate HAL or HAR as needed and return the original DF with mutated sequences, and the mutated column set to True\n",
    "        \n",
    "            #Add the winning sgRNA into the output df for this TF start/stop site\n",
    "            TFsdfWinnersandMutated.loc[ind] = winnerdf\n",
    "            display(TFsdfWinnersandMutated)\n",
    "    \n",
    "        #Return dataframe as an excel file (This should be unindented one, but while testing I'd like to see the output file updated after each row of the TFsdf)\n",
    "        TFsdfWinnersandMutated.to_excel(\"outputFiles/winningsgRNAs.xlsx\")\n",
    "\n",
    "    return TFsdfWinnersandMutated\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
