{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df= {\"start/stop\":\"start\", #is it N or C termini -> do we need to look at start or stop codon for teh cut \n",
    "            'genome_start_codon_pos':403, \n",
    "            'genome_stop_codon_pos':406, # or only 1 of those \n",
    "            'strand_type':'+',\n",
    "            \"sgRNA_list_positions\":[[401,425], [403, 427], [381, 405]],#those wil be as genome positions -assumptions - the coordinates correspond to the 1st and last bp of the strand to which the gsRNA will be complementary to\n",
    "            \"sgRNA_list_values\":[\"AAGCGACTA\", \"CCGTC\", \"CTGACT\"],\n",
    "            \"sgRNA_strand\" : ['-', '+', '+']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_dataframe_from_TFs_list(TF_list, refSeqPerChromosome, annotation):\n",
    "    '''\n",
    "    Creating a dataframe from sequence information and genes of interest. Depends on function revComp(). \n",
    "\n",
    "    params:\n",
    "            TF_list: xlsx file listing the genes of interest\n",
    "            ref_genome: fasta file containing the gene sequence information for the reference genome\n",
    "            annotation: gtf file containing the gene annotation information for the reference genome\n",
    "\n",
    "\n",
    "    returns: dataframe with dictionaries for every isotype for each gene of interest\n",
    "        {..., \n",
    "        108843:{    'Gene_ID': 'FBgn0004652',\n",
    "                    'Transcript_ID': 'FBtr0083651', \n",
    "                    'Chromosome': '3R', \n",
    "                    'Gene_Region': 'stop_codon', \n",
    "                    'Start': 18426145, \n",
    "                    'Stop': 18426147, \n",
    "                    'Strand': '-', \n",
    "                    'Reference_Seq': 'TTGATCGTAGGACAC', \n",
    "                }\n",
    "        ...,\n",
    "        }\n",
    "\n",
    "    '''\n",
    "    import pandas as pd\n",
    "    \n",
    "    #This is the input file containing the TFs we want to query\n",
    "    #Imported as a pandas dataframe\n",
    "    queryTFsdf = pd.read_excel(TF_list)\n",
    "    #Ths contains 765 TFs total\n",
    "\n",
    "    #This is the .gtf file with annotations for each gene on the reference genome\n",
    "    #Note that I'll use these for the info categories of the final pandas df (rather than the transgenic genome annotations)\n",
    "    refAnnotationsHeaders = [\"Chromosome\", \"Source\", \"Gene_Region\", \"Start\", \"Stop\", \"Score\", \"Strand\", \"Frame\", \"Attribute\"]\n",
    "    refGenomeAnnotation = pd.read_csv(annotation, sep = \"\\t\", header = None, index_col = False, names = refAnnotationsHeaders)\n",
    "    \n",
    "    #This is to reformat the \"Attribute\" category in refGenomeAnnotation, to extract Gene_ID, Gene_Symbol, and Transcript ID\n",
    "    index = 0\n",
    "\n",
    "    #Add new categories to the dataframe\n",
    "    refGenomeAnnotation = refGenomeAnnotation.assign(Gene_ID = \"\", Gene_Symbol = \"\", Transcript_ID = \"\")\n",
    "\n",
    "    #For each attribute value, extract the gene ID and symbol and add this to the new categories\n",
    "    for attribute in refGenomeAnnotation['Attribute']:\n",
    "        fullatt = (refGenomeAnnotation.loc[index][\"Attribute\"]).replace(\";\", \"\")\n",
    "        fullatt = fullatt.replace('\"', \"\")\n",
    "        fullattsplit = fullatt.split(\" \")\n",
    "        refGenomeAnnotation.at[index,\"Gene_ID\"] = fullattsplit[1]\n",
    "        refGenomeAnnotation.at[index,\"Gene_Symbol\"] = fullattsplit[3]\n",
    "        if len(fullattsplit) > 4:\n",
    "            refGenomeAnnotation.at[index,\"Transcript_ID\"] = fullattsplit[5]\n",
    "            refGenomeAnnotation.at[index,\"Transcript_Symbol\"] = fullattsplit[7]\n",
    "        index+=1\n",
    "\n",
    "    #Delete Attributes category\n",
    "    del refGenomeAnnotation[\"Attribute\"]\n",
    "\n",
    "    #Select only rows that TFs are in, and keep only the start and stop codon gene regions\n",
    "\n",
    "    refGenomeAnnotation = refGenomeAnnotation.loc[refGenomeAnnotation[\"Gene_Region\"].isin([\"start_codon\", \"stop_codon\"])]\n",
    "\n",
    "    TFsdf = refGenomeAnnotation[[\"Gene_ID\", \"Transcript_ID\", \"Chromosome\", \"Gene_Region\", \"Start\", \"Stop\", \"Strand\"]].loc[refGenomeAnnotation[\"Gene_ID\"].isin(queryTFsdf[\"Flybase_ID\"])]\n",
    "\n",
    "    #Add reference genome sequence per gene region\n",
    "    #This will correspond to 1.6kb upstream and downstream of ATG/stop codon \n",
    "    TFsdf = TFsdf.assign(Reference_Seq = \"\")\n",
    "\n",
    "    for index, rowcontents in TFsdf.iterrows():\n",
    "            #Define 3.2kb gene region\n",
    "            regionStart = rowcontents[\"Start\"] - 1601\n",
    "            regionStop = rowcontents[\"Stop\"] + 1600\n",
    "\n",
    "            #Add reference sequence\n",
    "            refPosStrandSeq = str(refSeqPerChromosome[rowcontents[\"Chromosome\"]][regionStart:regionStop]) #This is the + strand seq, so goes from end to beginning\n",
    "            TFsdf.at[index,\"Reference_Seq\"] = revComp(refPosStrandSeq)\n",
    "\n",
    "    #Create fragments for the HDR-arm\n",
    "    TFsdf = make_homology_arm_fragments(TFsdf)\n",
    "\n",
    "    #re-index\n",
    "    TFsdf = TFsdf.reset_index()\n",
    "    del TFsdf[\"index\"]\n",
    "           \n",
    "    #Create a dictionary of dictionaries for the df\n",
    "    TFsdict_of_dict = {}\n",
    "    for index, rowcontents in TFsdf.iterrows():\n",
    "        TFsdict_of_dict[index] = rowcontents.to_dict()\n",
    "\n",
    "    return TFsdf, TFsdict_of_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def filter_gRNA(gRNA_file, TF_dict, refSeqPerChromosome):\n",
    "    '''\n",
    "    params:\n",
    "        gRNA_file: gff file\n",
    "        TF_dict: dataframe with the following format:\n",
    "\n",
    "            {    'Gene_ID': 'FBgn0004652',\n",
    "                'Transcript_ID': 'FBtr0083651', \n",
    "                'Chromosome': '3R', \n",
    "                'Gene_Region': 'stop_codon', \n",
    "                'Start': 18426145, \n",
    "                'Stop': 18426147, \n",
    "                'Strand': '-', \n",
    "                'Reference_Seq': 'TTGATCGTAGGACAC', \n",
    "                'Transgenic_Seq': 'GACCCTAGGACCGG'\n",
    "            }\n",
    "        returns: dictionary with the following format\n",
    "\n",
    "            df={\n",
    "            \"start/stop\":\"start\",#is it N or C termini -> do we need to look at start or stop codon for teh cut \n",
    "            'genome_start_codon_pos':400, \n",
    "            'genome_stop_codon_pos':700, # or only 1 of those \n",
    "            'strand_type':'+',\n",
    "            \"sgRNA_list_positions\":[[401,425],[456,467],[478,489],[395,415]],#those wil be as genome positions -assumptions - the coordinates correspond to the 1st and last bp of the strand to which the gsRNA will be complementary to\n",
    "            \"sgRNA_list_values\":[\"AAGCGACTA\",\"AAAAAAAATAAAAA\",\"ATATATTTTTTTTTTAAAAA\",\"AGCGCGAAATAATA\"]\n",
    "            \"sgRNA_strand\" = '-'\n",
    "        }\n",
    "    '''\n",
    "    import pandas as pd\n",
    "\n",
    "    # create a dataframe from the gRNA files\n",
    "    gRNAFileAnnotation = pd.read_csv(gRNA_file, sep = \"\\t\", index_col = False)\n",
    "\n",
    "    # add a new category to the dataframe that provides information about whether the sequence deviates from the transgenic strain \n",
    "    gRNAFileAnnotation = gRNAFileAnnotation.assign(target_site_variation= \"\")\n",
    "\n",
    "    # reformat the \"Attribute\" category in refGenomeAnnotation, to extract Gene_ID, Gene_Symbol, and Transcript ID\n",
    "    index = 0 #TODO@Marina improve this code\n",
    "\n",
    "    # for each attribute value, extract the gene ID and symbol and add this to the new categories\n",
    "    for attribute in gRNAFileAnnotation['attributes']:\n",
    "\n",
    "        fullatt = (attribute).split(\";\")\n",
    "        gRNAFileAnnotation.at[index,\"target_site_variation\"] = fullatt[8]\n",
    "        index+=1\n",
    "    \n",
    "    # shorten file to essential information\n",
    "    GenomeCoordinates = gRNAFileAnnotation.loc[:,[\"target_site_variation\", \"fmin\", \"fmax\", \"#chr\", \"strand\"]]\n",
    "\n",
    "    # initialize a list that stores the positions of the gRNAs\n",
    "    sgRNA_list_positions = []\n",
    "    sgRNA_list_values = []\n",
    "\n",
    "    # check whether the sgRNAs match the transgeneic genome, whether the sgRNAs match to the same chromosome as the transcription factor\n",
    "    # select sgRNA that are located maximally 20 bp upstream of the start/stop codon of the transcription factors\n",
    "    # select sgRNA that are located maximally 20 bp downstream of the start/stop codon of the transcription factors\n",
    "\n",
    "    for gRNA in range(len(GenomeCoordinates)):\n",
    "        \n",
    "        if GenomeCoordinates['target_site_variation'].iloc[gRNA] == \"target_site_variation=none observed\" and \\\n",
    "        GenomeCoordinates['#chr'].iloc[gRNA] == TF_dict[\"Chromosome\"] and \\\n",
    "        int(GenomeCoordinates['fmin'].iloc[gRNA])-1 >= TF_dict[\"Start\"] - 20 and \\\n",
    "        int(GenomeCoordinates['fmax'].iloc[gRNA]) <= TF_dict[\"Stop\"] + 20:\n",
    "                \n",
    "            sgRNA_list_positions.append([int(GenomeCoordinates['fmin'].iloc[gRNA])-1, int(GenomeCoordinates['fmax'].iloc[gRNA])])\n",
    "        \n",
    "            # to retrieve the nucleotide sequence of the sgRNA, load the fruit fly reference genome and extract the sequence at the genomic coordinates provided in the sgRNA file\n",
    "\n",
    "            gRNA_string = str(refSeqPerChromosome[GenomeCoordinates['#chr'].iloc[gRNA]][int(GenomeCoordinates['fmin'].iloc[gRNA])-1:int(GenomeCoordinates['fmax'].iloc[gRNA])])\n",
    "\n",
    "            # if the gRNA lies on the minus strand than provide reverse complement sequence\n",
    "\n",
    "            if GenomeCoordinates['strand'].iloc[gRNA] == '-':\n",
    "\n",
    "                gRNA_string = revComp(gRNA_string)\n",
    "\n",
    "            sgRNA_list_values.append(gRNA_string)\n",
    "\n",
    "    TF_dict[\"sgRNA_list_positions\"] = sgRNA_list_positions\n",
    "    TF_dict[\"sgRNA_list_values\"] = sgRNA_list_values\n",
    "    TF_dict[\"sgRNA_strand\"] = GenomeCoordinates['strand'].iloc[gRNA]\n",
    "\n",
    "\n",
    "    return(TF_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def revComp(inputSeq):\n",
    "  \"\"\"\n",
    "  This function takes an input sequence and returns the reverse complement.\n",
    "\n",
    "  Input: inputSeq in str format\n",
    "  Output: revComp in str format\n",
    "\n",
    "  \"\"\"\n",
    "  complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "  \n",
    "  revComp = \"\"\n",
    "  for base in inputSeq[::-1]:\n",
    "    revComp += complement[(base.upper())]\n",
    "\n",
    "  return revComp\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>sgRNA_stop</th>\n",
       "      <th>sgRNA_strand</th>\n",
       "      <th>positionScore</th>\n",
       "      <th>P6inCDS</th>\n",
       "      <th>Mutate2</th>\n",
       "      <th>Mutate1</th>\n",
       "      <th>OutsideCDS</th>\n",
       "      <th>CutSiteInCDS</th>\n",
       "      <th>PAMinCDS</th>\n",
       "      <th>15bpOverhangL</th>\n",
       "      <th>15bpOverhangR</th>\n",
       "      <th>PAMRelativeEnd</th>\n",
       "      <th>PAMinStartStop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAGCGACTA</td>\n",
       "      <td>425</td>\n",
       "      <td>-</td>\n",
       "      <td>19</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>19.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCGTC</td>\n",
       "      <td>427</td>\n",
       "      <td>+</td>\n",
       "      <td>21</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>21.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTGACT</td>\n",
       "      <td>405</td>\n",
       "      <td>+</td>\n",
       "      <td>-1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sgRNA_sequence sgRNA_stop sgRNA_strand positionScore P6inCDS Mutate2  \\\n",
       "0      AAGCGACTA        425            -            19   False   False   \n",
       "1          CCGTC        427            +            21    True   False   \n",
       "2         CTGACT        405            +            -1   False   False   \n",
       "\n",
       "  Mutate1 OutsideCDS CutSiteInCDS PAMinCDS 15bpOverhangL 15bpOverhangR  \\\n",
       "0   False       True        False    False          True         False   \n",
       "1   False      False         True     True         False          True   \n",
       "2   False       True        False    False          True         False   \n",
       "\n",
       "   PAMRelativeEnd PAMinStartStop  \n",
       "0            19.0          False  \n",
       "1            21.0          False  \n",
       "2             1.0           True  "
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positionScore(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [],
   "source": [
    "def checkCDSCutandOrder(sgRNACatalogue):\n",
    "    \"\"\"\n",
    "    Given an sgRNACatalogue or subset, will calculate firstly rows where the sgRNA cuts inside CDS. If multiple, selects that which cuts closest to start/stop.\n",
    "    If none, selects closest cut sgRNA that is outside CDS.\n",
    "    \n",
    "    \"\"\"\n",
    "    #C. Check cut site in CDS\n",
    "    conditionC = sgRNACatalogue[sgRNACatalogue[\"CutSiteInCDS\"] == True]\n",
    "    conditionCclosestCut = conditionC.sort_values('PAMRelativeEnd') #This value is the absolute value of positionScore, so indicates those that cut closest to start/stop\n",
    "\n",
    "    if len(conditionCclosestCut) >1: #cuts in CDS, closest cut (C1, C2)\n",
    "        conditionCclosestCut = conditionCclosestCut.reset_index(drop = True) #reset index\n",
    "        winnersgRNA = conditionCclosestCut.at[0, \"sgRNA_sequence\"]\n",
    "        winnerFound = True\n",
    "    else: #not sgRNAs cut in CDS, select closest that still met condition B (C3)\n",
    "        nonCDSclosestCut = sgRNACatalogue.sort_values('PAMRelativeEnd')\n",
    "        winnersgRNA = nonCDSclosestCut.at[0, \"sgRNA_sequence\"]\n",
    "        winnerFound = True\n",
    "\n",
    "    return winnersgRNA, winnerFound\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_best_gRNA(df):\n",
    "    \"\"\"\n",
    "    params: df={\n",
    "            \"start/stop\":\"start\",#is it N or C termini -> do we need to look at start or stop codon for teh cut \n",
    "            'genome_start_codon_pos':400, \n",
    "            'genome_stop_codon_pos':700, # or only 1 of those \n",
    "            'strand_type':'+',\n",
    "            \"sgRNA_list_positions\":[[401,425],[456,467],[478,489],[395,415]],#those wil be as genome positions -assumptions - the coordinates correspond to the 1st and last bp of the strand to which the gsRNA will be complementary to\n",
    "            \"sgRNA_list_values\":[\"AAGCGACTA\",\"AAAAAAAATAAAAA\",\"ATATATTTTTTTTTTAAAAA\",\"AGCGCGAAATAATA\"]\n",
    "        }\n",
    "    \n",
    "    \"\"\"\n",
    "    #Score the sgRNAs for this site\n",
    "    sgRNACatalogue = positionScore(df)\n",
    "\n",
    "    #Set up the winning guide RNA\n",
    "    winnerFound = False\n",
    "    mutationNeeded = False\n",
    "    winnersgRNA = \"\" #Could update this to be more information about the guide e.g. strand, position\n",
    "\n",
    "    #If there is no guide RNAs at this site, flag this. The resulting sgRNA will be an empty string.\n",
    "    if len(sgRNACatalogue) == 0:\n",
    "        print(f\"No guide RNAs were found for the start/stop site beginning at position {df['genome_start_codon_pos']}\")\n",
    "        winnerFound = True #To avoid looping through other conditions if there are no sgRNAs\n",
    "\n",
    "    #A. Ideal condition - PAM in start/stop\n",
    "    if winnerFound == False:\n",
    "        conditionA = sgRNACatalogue[sgRNACatalogue[\"PAMinStartStop\"] == True] #This is the subset df for which PAM is in the start/stop\n",
    "        if len(conditionA) > 0: #if there is one or more sgRNAs for this condition, select the first as the winner\n",
    "            conditionA = conditionA.reset_index(drop = True) #reset index\n",
    "            winnersgRNA = conditionA.at[0, \"sgRNA_sequence\"]\n",
    "            winnerFound = True\n",
    "\n",
    "    #B. sgRNA overhang is less than 15bp\n",
    "    if winnerFound == False:\n",
    "        conditionB = sgRNACatalogue[sgRNACatalogue[\"15bpOverhangL\"] == False & sgRNACatalogue[\"15bpOverhangR\"] == False]\n",
    "        if len(conditionB) == 1:\n",
    "            winnersgRNA = conditionB\n",
    "            winnerFound = True\n",
    "        elif len(conditionB) > 1: #more than one, select in CDS preferencially, and closest cut\n",
    "            winnersgRNA, winnerFound = checkCDSCutandOrder(conditionB)    \n",
    "\n",
    "    #D. sgRNA overhang is more than 15bp, need to mutate\n",
    "    if winnerFound == False:\n",
    "        if len(sgRNACatalogue) == 1:\n",
    "            winnersgRNA = sgRNACatalogue.at[0, \"sgRNA_sequence\"]\n",
    "        else: #more than 1 sgRNA\n",
    "            winnersgRNA, winnerFound = checkCDSCutandOrder(sgRNACatalogue)            \n",
    "\n",
    "        mutationNeeded = True\n",
    "\n",
    "\n",
    "    return winnersgRNA, mutationNeeded, sgRNACatalogue[sgRNACatalogue[\"sgRNA_sequence\"] == winnersgRNA]\n",
    "#could adapt return to be just the true/false for mutation and the catalogue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('CTGACT',\n",
       " False,\n",
       "   sgRNA_sequence sgRNA_stop sgRNA_strand positionScore P6inCDS Mutate2  \\\n",
       " 2         CTGACT        405            +            -1   False   False   \n",
       " \n",
       "   Mutate1 OutsideCDS CutSiteInCDS PAMinCDS 15bpOverhangL 15bpOverhangR  \\\n",
       " 2   False       True        False    False          True         False   \n",
       " \n",
       "    PAMRelativeEnd PAMinStartStop  \n",
       " 2             1.0           True  )"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_best_gRNA(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codonFragmenter(sequenceString, type = \"homologyArm\", direction = 'HAL', geneStrand = '+'):\n",
    "    \"\"\"\n",
    "    Fragments codons in order moving away from the start or stop site. If gene on - strand, will take the reverse complement codon.\n",
    "\n",
    "    params:\n",
    "        sequence: sequence to fragment into codons as string.\n",
    "        type: one of \"homologyArm\" or \"primer\". If primer, will truncate at far end into a fragment divisible by 3 before fragmenting into codons.\n",
    "        direction: 'HAL' or 'HAR'. If on left homology arm, codon order is from right to left. If on right, codon order is from left to right.\n",
    "        geneStrand: if +, codons are the original + strand sequence. If -, each individual codon has been reverse complemented.\n",
    "    \"\"\"\n",
    "\n",
    "    #If sequence is for primer HAL-R, will take the reverse complement so that all fragments/primers used in code below are in the same orientation.\n",
    "    if type ==\"primer\" & direction == 'HAL':\n",
    "        sequenceDirected = revComp(sequenceString)\n",
    "    else:\n",
    "        sequenceDirected = sequenceString\n",
    "\n",
    "    #Define the sequence string to use/mutate. If primer, truncate until divisible by 3.\n",
    "    if type == \"primer\":\n",
    "        remainder = len(sequenceDirected) % 3\n",
    "        if remainder == 1:\n",
    "            sequence = sequenceDirected[:-1]\n",
    "        elif remainder == 2:\n",
    "            sequence = sequenceDirected[:-2]\n",
    "        else:\n",
    "            sequence = sequenceDirected #if already divisible by 3\n",
    "    elif type == \"homologyArm\":\n",
    "        sequence = sequenceDirected #no modification needed, always 225bp so divisible by 3\n",
    "    else:\n",
    "        print(\"Sequence type invalid.\")\n",
    "\n",
    "    orderedCodons = []\n",
    "\n",
    "    #Define codons in threes along to sequence.\n",
    "    for codonBase1 in range(0, len(sequence), 3):\n",
    "        orderedCodons.append(sequence[codonBase1:codonBase1+3]) #Append this to orderedCodons\n",
    "    if direction == 'HAL': #If on the left, reverse the ordered codons list to be moving away from the start/stop\n",
    "        orderedCodons.reverse()\n",
    "    if geneStrand == '-': #If on the minus strand, take revComp per codon\n",
    "        for ind, codon in enumerate(orderedCodons):\n",
    "            orderedCodons[ind] = revComp(codon)\n",
    "    \n",
    "    return(orderedCodons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutator(sequenceToMutate, winner_sgRNACatalogue, sequenceType = \"homologyArm\"):\n",
    "    \"\"\"\n",
    "    In the case where a fragment or primer needs to be mutated, will mutate in CDS (preferably PAM, if not in the sgRNA). If not possible, will mutate PAM outside of CDS to NGT.\n",
    "\n",
    "    params:\n",
    "        sequenceToMutate: the homology arm or primer fragment that will be mutated.\n",
    "        winner_sgRNACatalogue: sgRNACatalogue in format as above, with only the row for the winner sgRNA selected.\n",
    "        sequenceType: one of 'homologyArm' or 'primer'.\n",
    "    \"\"\"\n",
    "\n",
    "    #Check mutable conditions and posiions\n",
    "    if positionBoolean.at[]\n",
    "\n",
    "    #1. Mutate PAM in CDS\n",
    "    if positionBoolean.at[1,\"PAMinCDS\"] is True:\n",
    "        \n",
    "    #after mutating, revComp back"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def runnersgRNAandPrimer(TF_list, type = \"homologyArms\"):\n",
    "\n",
    "    \"\"\"\n",
    "    TF_list: .xlsx file of TFs. Either the full TFs list in the first instance, or a shorter list for which to calculate primers.\n",
    "    Type: One of \"HomologyArm\" or \"primers\" - homology arms will be returned in the normal instance.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    #make TFsdf\n",
    "    TFsdf, TFsdict_of_dic = make_dataframe_from_TFs_list(TF_list)\n",
    "\n",
    "    #sgRNAs\n",
    "    sgRNAfiles = [\"\"] #sgRNA files in stringency order (low to high)\n",
    "    stringency = 0\n",
    "    sgRNAselected = False\n",
    "\n",
    "    #Loop through the different files and stop if an sgRNA is found\n",
    "    while sgRNAselected == False:\n",
    "        #From this, filter guideRNAs nearby\n",
    "        df = filter_gRNA(sgRNAfiles[stringency], TFsdict_of_dic)\n",
    "\n",
    "        #Extract necessary boolean and position values\n",
    "        sgRNACatalogue = positionScore(df)\n",
    "\n",
    "        #find best guideRNA\n",
    "        winnersgRNA, mutationNeeded, winnerCatalogue = find_best_gRNA(sgRNACatalogue)\n",
    "\n",
    "        #Check if a winner was found, go to next stringency level\n",
    "        if winnersgRNA != \"\":\n",
    "            sgRNAselected == True\n",
    "        else:\n",
    "            stringency +=1\n",
    "            if stringency > 3:\n",
    "                print(\"no sgRNAs could be found for any stringency file.\")\n",
    "                sgRNAselected = True #To get out of the while loop\n",
    "\n",
    "    #Find homology arms or primers\n",
    "\n",
    "    #Find homology arms\n",
    "    if type == \"homologyArms\":\n",
    "        tfsDf = make_homology_arm_fragments(tfsDF)\n",
    "\n",
    "    #Find primers\n",
    "    elif type == \"primers\":\n",
    "        TFsdf = primersRunner(TFs = 'inputFiles/truncatedTFsinput.xlsx') #replace this with the TFs list of only those where you need primers\n",
    "\n",
    "    else:\n",
    "        print(\"invalid type. Should be one of 'homologyArms' or 'primers')\n",
    "              \n",
    "    #Mutate as needed\n",
    "    if mutationNeeded == True:\n",
    "        mutator(sequenceToMutate, winnerCatalogue, sequenceType = \"homologyArm\")\n",
    "\n",
    "    #return sgRNA selected, HAR fragments or primers, and I assume metrics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def mutate_sgRNA_recognition_site_in_HDR_plasmid(director, df):\n",
    "    \"\"\"\n",
    "    Note that for now, this does not indicate the number of mutations (for simplicity), but this will be either one or two if df[\"mutated\"] = \"yes\"\n",
    "    \"\"\"\n",
    "    #Define all codons in the homology arm fragment\n",
    "    fragmentedCodons = codonFragmenter(df[\"sgRNA_list_values\"][0], direction = director, geneStrand = df[\"strand_type\"])\n",
    "    mutationIndex = [] #codon number from the list of the 1 or 2 codons to mutate\n",
    "    #for 1 mutation, this will be one index, for 2, there will be 2\n",
    "\n",
    "    for ind in mutationIndex:\n",
    "        codontoMutate = fragmentedCodons[mutationIndex[0]] #extract a codon to mutate\n",
    "        synonCodons = find_synonymous_codons(query_codon)\n",
    "        if len(synonCodons) != 0: #at least one synonymous codon is found\n",
    "            df[\"mutated\"] == \"yes\" #indicate a mutation has been made\n",
    "            fragmentedCodons[mutationIndex[0]] = synonCodons[0] #replace the mutated codon in the list\n",
    "            reverseFragmenter(fragmentedCodons, direction = director, geneStrand = df[\"strand_type\"])\n",
    "\n",
    "    return df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverseFragmenter(mutatedOrderedCodons, direction = 'HAL', geneStrand = '+'):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "\n",
    "    fullString = \"\"\n",
    "\n",
    "    if geneStrand == \"-\":\n",
    "        for ind, codon in enumerate(mutatedOrderedCodons): #Take + strand sequences of the codons\n",
    "            mutatedOrderedCodons[ind] = revComp(codon)\n",
    "    if direction == 'HAL':\n",
    "        mutatedOrderedCodons.reverse()\n",
    "    for codon in mutatedOrderedCodons:\n",
    "        fullString.append(codon)\n",
    "    \n",
    "    return fullString"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def positionScore(df):\n",
    "    \"\"\"\n",
    "    Boolean evaluations of whether sgRNAs for a site meets required best-pick conditions and positional information about these sgRNAs.\n",
    "\n",
    "    params:\n",
    "        df: a dataframe for one start/stop site of the format: \n",
    "            df= {\"start/stop\":\"start\", #is it N or C termini -> do we need to look at start or stop codon for teh cut \n",
    "            'genome_start_codon_pos':403, \n",
    "            'genome_stop_codon_pos':406, # only one of these - the other is n/a\n",
    "            'strand_type':'+',\n",
    "            \"sgRNA_list_positions\":[[401,425],[403, 427]],#those wil be as genome positions -assumptions - the coordinates correspond to the 1st and last bp of the strand to which the gsRNA will be complementary to\n",
    "            \"sgRNA_list_values\":[\"AAGCGACTA\", \"CCTGTAA\"],\n",
    "            \"sgRNA_strand\" : ['-', '+']}\n",
    "    \"\"\"\n",
    "\n",
    "    import pandas as pd\n",
    "\n",
    "    #Set up a new dataframe where boolean values will be stored per sgRNA in the input df\n",
    "    sgRNACatalogue = pd.DataFrame(columns = [\"sgRNA_sequence\", \"sgRNA_stop\", \"sgRNA_strand\"], index = range(1, len(df[\"sgRNA_list_values\"])))\n",
    "\n",
    "    #Iterating through sgRNAs in df, add to the sgRNACatalogue\n",
    "    for ind, sgRNA in enumerate(df[\"sgRNA_list_values\"]):\n",
    "        newCatalogueRow = [sgRNA, df[\"sgRNA_list_positions\"][ind][1], df[\"sgRNA_strand\"][ind]]\n",
    "        sgRNACatalogue.loc[ind] = newCatalogueRow\n",
    "\n",
    "    #Re-order the index as we've been appending to the top\n",
    "    sgRNACatalogue = sgRNACatalogue.sort_index()\n",
    "\n",
    "    #Calculate fmax of sgRNA - stop of genome start/stop position - this will be used for position scoring throughout\n",
    "    #only have fmin in case of start codon - recalc fmax from this for homogeneity:\n",
    "    if df[\"start/stop\"] == \"start\":\n",
    "        df[\"genome_stop_codon_pos\"] == df[\"genome_stop_codon_pos\"] + 2\n",
    "    \n",
    "    sgRNACatalogue[\"positionScore\"] = sgRNACatalogue[\"sgRNA_stop\"] - df[\"genome_stop_codon_pos\"] #This is fmax - stop\n",
    "\n",
    "    #Dataframe containing parameter ranges to interpret the positon score, based on gene strand, sgRNA strand, and start/stop\n",
    "    positionScoreParameters = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\")\n",
    "\n",
    "    #Add the boolean columns to our output dataframe\n",
    "    booleanColumns = [\"P6inCDS\", \"Mutate2\", \"Mutate1\", \"OutsideCDS\", \"CutSiteInCDS\", \"PAMinCDS\", \"15bpOverhangL\", \"15bpOverhangR\", \"PAMoutsideCDS\"]\n",
    "    PAMColumns = [\"PAMRelativeEnd\", \"PAMinStartStop\"]\n",
    "    sgRNACatalogue = sgRNACatalogue.reindex(columns = sgRNACatalogue.columns.tolist() + booleanColumns + PAMColumns)\n",
    "\n",
    "    for ind, sgRNA in sgRNACatalogue.iterrows():\n",
    "\n",
    "        #Extract the appropriate parameter row per sgRNA\n",
    "        conditions = positionScoreParameters.loc[(positionScoreParameters[\"start/stop\"] == df[\"start/stop\"]) & (positionScoreParameters[\"strand_type\"] == df[\"strand_type\"]) & (positionScoreParameters[\"sgRNA_strand\"] == sgRNACatalogue.at[ind, 'sgRNA_strand'])]\n",
    "        conditions = conditions.reset_index(drop = True)\n",
    "\n",
    "        #Per column, input true/false as to whether the position score meets that condition\n",
    "        for col in booleanColumns:\n",
    "            colValue = conditions.at[0,col] #extract parameter range values from dataframe\n",
    "            #Process the value into a range (in format list [min, max])\n",
    "            #If the values should be 'more than' or 'less than', 250 is used as a max or -250 as min because fragments are 225bp, so distances cannot be more than this\n",
    "            if \">\" in colValue: \n",
    "                minMax = [int(colValue[1:]), 250]\n",
    "            elif \"<\" in colValue:\n",
    "                minMax = [-250,int(colValue[1:])]\n",
    "            elif \":\" in colValue:\n",
    "                min, max = colValue.split(\":\")\n",
    "                minMax = [int(min), int(max)]\n",
    "            else:\n",
    "                print(\"Incorrect format of range value. Verify inputs.\")\n",
    "            \n",
    "            #Into the output dataframe, print true/false as to whether the positionScore has met the condition for that column\n",
    "            sgRNACatalogue.at[ind, col] = bool(sgRNACatalogue.at[ind, \"positionScore\"] in range(minMax[0], minMax[1]))\n",
    "\n",
    "        #Calculate PAM position as a relative distance to the left end of start/stop if PAM is on the left, or right if on right (this can be calculated from the position score with the same equation for all cases)\n",
    "        sgRNACatalogue.at[ind, \"PAMRelativeEnd\"] = int(abs(sgRNACatalogue.at[ind, \"positionScore\"]))\n",
    "\n",
    "        #Calculate if PAM is in start/stop - where positionScore is 2 or less\n",
    "        sgRNACatalogue.at[ind, \"PAMinStartStop\"] = bool(sgRNACatalogue.at[ind,\"PAMRelativeEnd\"] < 3)\n",
    "\n",
    "    return sgRNACatalogue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "def conditionCheck(df):\n",
    "    \"\"\"\n",
    "    Given filtered sgRNAs for a start/stop site in the 'df' format, will filter these for those that meet ideal sgRNA conditions.\n",
    "    \n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "\n",
    "    #Making 'df' into an sgRNA catalogue dataframe. This makes it easier to subset sgRNAs by condition.\n",
    "    sgRNACatalogue = pd.DataFrame.from_dict(df)\n",
    "    #keep only fmax to simplify indexing later\n",
    "    for ind, sgRNA in sgRNACatalogue.iterrows():\n",
    "        sgRNACatalogue.at[ind, \"sgRNA_fmax\"] = sgRNACatalogue[\"sgRNA_list_positions\"][ind][1]\n",
    "    sgRNACatalogue = sgRNACatalogue.drop(columns=[\"sgRNA_list_positions\"]) #delete the old position value to simplify dataframe (optional)\n",
    "\n",
    "    #Adding position scores (fmax - stop)\n",
    "    sgRNACatalogue[\"positionScore\"] = sgRNACatalogue[\"sgRNA_fmax\"]- sgRNACatalogue[\"genome_stop_codon_pos\"]\n",
    "    \n",
    "    #Dataframe containing parameter ranges to interpret the positon score, based on gene strand, sgRNA strand, and start/stop\n",
    "    positionScoreParameters = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\")\n",
    "\n",
    "    #Per parameter, append the sgRNACatalogue with a TRUE/FALSE value per condition.\n",
    "    #Also translate the column of the positionScoreParameters df indicating how many sgRNA bases within 6bp of PAM ccan be mutated inside CDS.\n",
    "    #Lastly, translate positionScore into the relative position of the last 'G' or 'C' of PAM.\n",
    "    booleanColumns = [\"PAM_in_start/stop\", \"<15_bp3’_overhang\", \"PAM_in_CDS\", \"PAM_outside_CDS\", \"Cutsite_in_CDS\"]\n",
    "    positionColumns = [\"relative_position_score\", \"sgRNA_mutable_bases\"]\n",
    "    sgRNACatalogue = sgRNACatalogue.reindex(columns = sgRNACatalogue.columns.tolist() + booleanColumns + positionColumns)\n",
    "    \n",
    "    for ind, sgRNA in sgRNACatalogue.iterrows():\n",
    "\n",
    "        #Extract the appropriate parameter row per sgRNA\n",
    "        conditions = positionScoreParameters.loc[(positionScoreParameters[\"start/stop\"] == sgRNA['start/stop']) & (positionScoreParameters[\"strand_type\"] == sgRNA['strand_type']) & (positionScoreParameters[\"sgRNA_strand\"] == sgRNA['sgRNA_strand'])]\n",
    "        conditions = conditions.reset_index(drop = True)\n",
    "\n",
    "        #Per column, input true/false as to whether the position score meets that condition\n",
    "        for col in booleanColumns:\n",
    "            colValue = conditions.at[0,col] #extract parameter range values from dataframe\n",
    "            #Process the value into a range (in format list [min, max])\n",
    "            #If the values should be 'more than' or 'less than', 25 is used as a max or -25 as min because distances cannot be more than 20\n",
    "            if \">\" in colValue: \n",
    "                minMax = [int(colValue[1:]), 25]\n",
    "            elif \"<\" in colValue:\n",
    "                minMax = [-25,int(colValue[1:])]\n",
    "            elif \":\" in colValue:\n",
    "                min, max = colValue.split(\":\")\n",
    "                minMax = [int(min), int(max)]\n",
    "            else:\n",
    "                print(\"Incorrect format of range value. Verify inputs.\")\n",
    "            \n",
    "            #Into the output dataframe, print true/false as to whether the positionScore has met the condition for that column\n",
    "            sgRNACatalogue.at[ind, col] = bool(sgRNACatalogue.at[ind, \"positionScore\"] in range(minMax[0], minMax[1]))\n",
    "       \n",
    "        #Calculate PAM last G or C position using positionScore. Add this to the sgRNA catalogue.\n",
    "        sgRNACatalogue.at[ind, \"relative_position_score\"] = sgRNACatalogue.at[ind, \"positionScore\"] + int(conditions.at[0, \"relative_position_score\"])\n",
    "\n",
    "        #Calculate how many bases of the sgRNA recognition site are in the CDS and can therefore be mutated.\n",
    "        #This is given in the column \"sgRNA_mutable_bases\", in a list of the format [x,y,z,...] indicating the position score for which 1-6 bases can be mutated.\n",
    "        #Accounting for the case where there is more than one positionScore for which 6bp can be mutated, this translates to:\n",
    "        mutateBases = [1,2,3,4,5,6,6,6,6,6,6,6,6,6,6,6,6]\n",
    "        if sgRNACatalogue.at[ind, \"positionScore\"] in conditions.at[0, \"sgRNA_mutable_bases\"]:\n",
    "            index = conditions.at[0, \"sgRNA_mutable_bases\"].index(sgRNACatalogue.at[ind, \"positionScore\"])\n",
    "            sgRNACatalogue.at[ind, \"sgRNA_mutable_bases\"] = mutateBases[index]\n",
    "        else:\n",
    "            sgRNACatalogue.at[ind, \"sgRNA_mutable_bases\"] = False\n",
    "\n",
    "    #A. Check if PAM in CDS\n",
    "\n",
    "    #C. Check cut site in CDS\n",
    "    conditionC = sgRNACatalogue[sgRNACatalogue[\"CutSiteInCDS\"] == True]\n",
    "    conditionCclosestCut = conditionC.sort_values('PAMRelativeEnd') #This value is the absolute value of positionScore, so indicates those that cut closest to start/stop\n",
    "\n",
    "    if len(conditionCclosestCut) >1: #cuts in CDS, closest cut (C1, C2)\n",
    "        conditionCclosestCut = conditionCclosestCut.reset_index(drop = True) #reset index\n",
    "        winnersgRNA = conditionCclosestCut.at[0, \"sgRNA_sequence\"]\n",
    "        winnerFound = True\n",
    "    else: #not sgRNAs cut in CDS, select closest that still met condition B (C3)\n",
    "        nonCDSclosestCut = sgRNACatalogue.sort_values('PAMRelativeEnd')\n",
    "        winnersgRNA = nonCDSclosestCut.at[0, \"sgRNA_sequence\"]\n",
    "        winnerFound = True\n",
    "\n",
    "    return winnersgRNA, winnerFound\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "list = [4,5,6,7,8,9]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "10 is not in list",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[43], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[39mlist\u001b[39;49m\u001b[39m.\u001b[39;49mindex(\u001b[39m10\u001b[39;49m)\n",
      "\u001b[0;31mValueError\u001b[0m: 10 is not in list"
     ]
    }
   ],
   "source": [
    "list.index(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "df={\n",
    "            \"start/stop\":\"start\",#is it N or C termini -> do we need to look at start or stop codon for teh cut \n",
    "            'genome_start_codon_pos':400, \n",
    "            'genome_stop_codon_pos':700, # or only 1 of those \n",
    "            'strand_type':'+',\n",
    "            \"sgRNA_list_positions\":[[401,425],[456,467],[478,489],[395,415]],#those wil be as genome positions -assumptions - the coordinates correspond to the 1st and last bp of the strand to which the gsRNA will be complementary to\n",
    "            \"sgRNA_list_values\":[\"AAGCGACTA\",\"AAAAAAAATAAAAA\",\"ATATATTTTTTTTTTAAAAA\",\"AGCGCGAAATAATA\"],\n",
    "            \"sgRNA_strand\" : ['-', '+', '-', '+'],\n",
    "            \"positionScores\": [0, 5, 20, 18]\n",
    " }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "sgRNACatalogue = pd.DataFrame.from_dict(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    " #keep only fmax to simplify indexing later\n",
    "for ind, sgRNA in sgRNACatalogue.iterrows():\n",
    "    sgRNACatalogue.at[ind, \"sgRNA_fmax\"] = sgRNACatalogue[\"sgRNA_list_positions\"][ind][1]\n",
    "\n",
    "sgRNACatalogue = sgRNACatalogue.drop(columns=[\"sgRNA_list_positions\"])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start/stop</th>\n",
       "      <th>genome_start_codon_pos</th>\n",
       "      <th>genome_stop_codon_pos</th>\n",
       "      <th>strand_type</th>\n",
       "      <th>sgRNA_list_values</th>\n",
       "      <th>sgRNA_strand</th>\n",
       "      <th>positionScores</th>\n",
       "      <th>sgRNA_fmax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>start</td>\n",
       "      <td>400</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>AAGCGACTA</td>\n",
       "      <td>-</td>\n",
       "      <td>-275.0</td>\n",
       "      <td>425.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>start</td>\n",
       "      <td>400</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>AAAAAAAATAAAAA</td>\n",
       "      <td>+</td>\n",
       "      <td>-233.0</td>\n",
       "      <td>467.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>start</td>\n",
       "      <td>400</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>ATATATTTTTTTTTTAAAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>-211.0</td>\n",
       "      <td>489.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>start</td>\n",
       "      <td>400</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>AGCGCGAAATAATA</td>\n",
       "      <td>+</td>\n",
       "      <td>-285.0</td>\n",
       "      <td>415.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start/stop  genome_start_codon_pos  genome_stop_codon_pos strand_type  \\\n",
       "0      start                     400                    700           +   \n",
       "1      start                     400                    700           +   \n",
       "2      start                     400                    700           +   \n",
       "3      start                     400                    700           +   \n",
       "\n",
       "      sgRNA_list_values sgRNA_strand  positionScores  sgRNA_fmax  \n",
       "0             AAGCGACTA            -          -275.0       425.0  \n",
       "1        AAAAAAAATAAAAA            +          -233.0       467.0  \n",
       "2  ATATATTTTTTTTTTAAAAA            -          -211.0       489.0  \n",
       "3        AGCGCGAAATAATA            +          -285.0       415.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(sgRNACatalogue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start/stop</th>\n",
       "      <th>strand_type</th>\n",
       "      <th>sgRNA_strand</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_in_CDS2</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>Cutsite_in_CDS</th>\n",
       "      <th>relative_position_score</th>\n",
       "      <th>sgRNA_mutable_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>start</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>0:3</td>\n",
       "      <td>&lt;15</td>\n",
       "      <td>&gt;2</td>\n",
       "      <td>-1</td>\n",
       "      <td>&gt;5</td>\n",
       "      <td>23</td>\n",
       "      <td>[4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>start</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>18:21</td>\n",
       "      <td>&gt;5</td>\n",
       "      <td>-1</td>\n",
       "      <td>&lt;18</td>\n",
       "      <td>&gt;17</td>\n",
       "      <td>0</td>\n",
       "      <td>[15,16,17,18,19,20]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>start</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>0:3</td>\n",
       "      <td>&lt;15</td>\n",
       "      <td>-1</td>\n",
       "      <td>&gt;2</td>\n",
       "      <td>&lt;5</td>\n",
       "      <td>23</td>\n",
       "      <td>[5,4,3,2,1,0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>start</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>18:21</td>\n",
       "      <td>&gt;5</td>\n",
       "      <td>&lt;18</td>\n",
       "      <td>-1</td>\n",
       "      <td>&lt;15</td>\n",
       "      <td>0</td>\n",
       "      <td>[16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>stop</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>0:3</td>\n",
       "      <td>&lt;15</td>\n",
       "      <td>-1</td>\n",
       "      <td>&gt;2</td>\n",
       "      <td>&lt;5</td>\n",
       "      <td>23</td>\n",
       "      <td>[5,4,3,2,1,0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>stop</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>18:21</td>\n",
       "      <td>&lt;5</td>\n",
       "      <td>&lt;17</td>\n",
       "      <td>01</td>\n",
       "      <td>&lt;15</td>\n",
       "      <td>0</td>\n",
       "      <td>[16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>stop</td>\n",
       "      <td>-</td>\n",
       "      <td>+</td>\n",
       "      <td>0:3</td>\n",
       "      <td>&lt;15</td>\n",
       "      <td>&gt;2</td>\n",
       "      <td>-1</td>\n",
       "      <td>&gt;5</td>\n",
       "      <td>23</td>\n",
       "      <td>[4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>stop</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>18:21</td>\n",
       "      <td>&gt;5</td>\n",
       "      <td>-1</td>\n",
       "      <td>&lt;18</td>\n",
       "      <td>&gt;17</td>\n",
       "      <td>0</td>\n",
       "      <td>[15,16,17,18,19,20]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start/stop strand_type sgRNA_strand PAM_in_start/stop PAM_in_CDS  \\\n",
       "0      start           +            +               0:3        <15   \n",
       "1      start           +            -             18:21         >5   \n",
       "2      start           -            +               0:3        <15   \n",
       "3      start           -            -             18:21         >5   \n",
       "4       stop           +            +               0:3        <15   \n",
       "5       stop           +            -             18:21         <5   \n",
       "6       stop           -            +               0:3        <15   \n",
       "7       stop           -            -             18:21         >5   \n",
       "\n",
       "  PAM_in_CDS2 PAM_outside_CDS Cutsite_in_CDS  relative_position_score  \\\n",
       "0          >2              -1             >5                       23   \n",
       "1          -1             <18            >17                        0   \n",
       "2          -1              >2             <5                       23   \n",
       "3         <18              -1            <15                        0   \n",
       "4          -1              >2             <5                       23   \n",
       "5         <17              01            <15                        0   \n",
       "6          >2              -1             >5                       23   \n",
       "7          -1             <18            >17                        0   \n",
       "\n",
       "                              sgRNA_mutable_bases  \n",
       "0  [4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]  \n",
       "1                             [15,16,17,18,19,20]  \n",
       "2                                   [5,4,3,2,1,0]  \n",
       "3        [16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1]  \n",
       "4                                   [5,4,3,2,1,0]  \n",
       "5        [16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1]  \n",
       "6  [4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]  \n",
       "7                             [15,16,17,18,19,20]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sgRNACatalogue[\"positionScore\"] = sgRNACatalogue[\"sgRNA_fmax\"]- sgRNACatalogue[\"genome_stop_codon_pos\"]\n",
    "\n",
    "positionScoreParameters = pd.read_excel(\"inputfiles/fmaxStopScore.xlsx\")\n",
    "display(positionScoreParameters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start/stop</th>\n",
       "      <th>strand_type</th>\n",
       "      <th>sgRNA_strand</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_in_CDS2</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>Cutsite_in_CDS</th>\n",
       "      <th>relative_position_score</th>\n",
       "      <th>sgRNA_mutable_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>start</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>18:21</td>\n",
       "      <td>&gt;5</td>\n",
       "      <td>-1</td>\n",
       "      <td>&lt;18</td>\n",
       "      <td>&gt;17</td>\n",
       "      <td>0</td>\n",
       "      <td>[15,16,17,18,19,20]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start/stop strand_type sgRNA_strand PAM_in_start/stop PAM_in_CDS  \\\n",
       "0      start           +            -             18:21         >5   \n",
       "\n",
       "  PAM_in_CDS2 PAM_outside_CDS Cutsite_in_CDS  relative_position_score  \\\n",
       "0          -1             <18            >17                        0   \n",
       "\n",
       "   sgRNA_mutable_bases  \n",
       "0  [15,16,17,18,19,20]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start/stop</th>\n",
       "      <th>strand_type</th>\n",
       "      <th>sgRNA_strand</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_in_CDS2</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>Cutsite_in_CDS</th>\n",
       "      <th>relative_position_score</th>\n",
       "      <th>sgRNA_mutable_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>start</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>0:3</td>\n",
       "      <td>&lt;15</td>\n",
       "      <td>&gt;2</td>\n",
       "      <td>-1</td>\n",
       "      <td>&gt;5</td>\n",
       "      <td>23</td>\n",
       "      <td>[4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start/stop strand_type sgRNA_strand PAM_in_start/stop PAM_in_CDS  \\\n",
       "0      start           +            +               0:3        <15   \n",
       "\n",
       "  PAM_in_CDS2 PAM_outside_CDS Cutsite_in_CDS  relative_position_score  \\\n",
       "0          >2              -1             >5                       23   \n",
       "\n",
       "                              sgRNA_mutable_bases  \n",
       "0  [4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start/stop</th>\n",
       "      <th>strand_type</th>\n",
       "      <th>sgRNA_strand</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_in_CDS2</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>Cutsite_in_CDS</th>\n",
       "      <th>relative_position_score</th>\n",
       "      <th>sgRNA_mutable_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>start</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>18:21</td>\n",
       "      <td>&gt;5</td>\n",
       "      <td>-1</td>\n",
       "      <td>&lt;18</td>\n",
       "      <td>&gt;17</td>\n",
       "      <td>0</td>\n",
       "      <td>[15,16,17,18,19,20]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start/stop strand_type sgRNA_strand PAM_in_start/stop PAM_in_CDS  \\\n",
       "0      start           +            -             18:21         >5   \n",
       "\n",
       "  PAM_in_CDS2 PAM_outside_CDS Cutsite_in_CDS  relative_position_score  \\\n",
       "0          -1             <18            >17                        0   \n",
       "\n",
       "   sgRNA_mutable_bases  \n",
       "0  [15,16,17,18,19,20]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start/stop</th>\n",
       "      <th>strand_type</th>\n",
       "      <th>sgRNA_strand</th>\n",
       "      <th>PAM_in_start/stop</th>\n",
       "      <th>PAM_in_CDS</th>\n",
       "      <th>PAM_in_CDS2</th>\n",
       "      <th>PAM_outside_CDS</th>\n",
       "      <th>Cutsite_in_CDS</th>\n",
       "      <th>relative_position_score</th>\n",
       "      <th>sgRNA_mutable_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>start</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>0:3</td>\n",
       "      <td>&lt;15</td>\n",
       "      <td>&gt;2</td>\n",
       "      <td>-1</td>\n",
       "      <td>&gt;5</td>\n",
       "      <td>23</td>\n",
       "      <td>[4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start/stop strand_type sgRNA_strand PAM_in_start/stop PAM_in_CDS  \\\n",
       "0      start           +            +               0:3        <15   \n",
       "\n",
       "  PAM_in_CDS2 PAM_outside_CDS Cutsite_in_CDS  relative_position_score  \\\n",
       "0          >2              -1             >5                       23   \n",
       "\n",
       "                              sgRNA_mutable_bases  \n",
       "0  [4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Extract the appropriate parameter row per sgRNA\n",
    "for ind, sgRNA in sgRNACatalogue.iterrows():\n",
    "    conditions = positionScoreParameters.loc[(positionScoreParameters[\"start/stop\"] == sgRNA['start/stop']) & (positionScoreParameters[\"strand_type\"] == sgRNA['strand_type']) & (positionScoreParameters[\"sgRNA_strand\"] == sgRNA['sgRNA_strand'])]\n",
    "    conditions = conditions.reset_index(drop = True)\n",
    "    display(conditions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "        mutateBases = [1,2,3,4,5,6,6,6,6,6,6,6,6,6,6,6,6]\n",
    "        if sgRNACatalogue.at[ind, \"positionScore\"] in conditions.at[0, \"sgRNA_mutable_bases\"]:\n",
    "            index = conditions.at[0, \"sgRNA_mutable_bases\"].index(sgRNACatalogue.at[ind, \"positionScore\"])\n",
    "            sgRNACatalogue.at[ind, \"sgRNA_mutable_bases\"] = mutateBases[index]\n",
    "        else:\n",
    "            sgRNACatalogue.at[ind, \"sgRNA_mutable_bases\"] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-275.0\n",
      "-233.0\n",
      "-211.0\n",
      "-285.0\n"
     ]
    }
   ],
   "source": [
    "for ind, row in sgRNACatalogue.iterrows():\n",
    "    print(row[\"positionScore\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start/stop</th>\n",
       "      <th>genome_start_codon_pos</th>\n",
       "      <th>genome_stop_codon_pos</th>\n",
       "      <th>strand_type</th>\n",
       "      <th>sgRNA_list_values</th>\n",
       "      <th>sgRNA_strand</th>\n",
       "      <th>positionScores</th>\n",
       "      <th>sgRNA_fmax</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>start</td>\n",
       "      <td>400</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>AAGCGACTA</td>\n",
       "      <td>-</td>\n",
       "      <td>-275.0</td>\n",
       "      <td>425.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>start</td>\n",
       "      <td>400</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>AAAAAAAATAAAAA</td>\n",
       "      <td>+</td>\n",
       "      <td>-233.0</td>\n",
       "      <td>467.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>start</td>\n",
       "      <td>400</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>ATATATTTTTTTTTTAAAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>-211.0</td>\n",
       "      <td>489.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>start</td>\n",
       "      <td>400</td>\n",
       "      <td>700</td>\n",
       "      <td>+</td>\n",
       "      <td>AGCGCGAAATAATA</td>\n",
       "      <td>+</td>\n",
       "      <td>-285.0</td>\n",
       "      <td>415.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  start/stop  genome_start_codon_pos  genome_stop_codon_pos strand_type  \\\n",
       "0      start                     400                    700           +   \n",
       "1      start                     400                    700           +   \n",
       "2      start                     400                    700           +   \n",
       "3      start                     400                    700           +   \n",
       "\n",
       "      sgRNA_list_values sgRNA_strand  positionScores  sgRNA_fmax  \n",
       "0             AAGCGACTA            -          -275.0       425.0  \n",
       "1        AAAAAAAATAAAAA            +          -233.0       467.0  \n",
       "2  ATATATTTTTTTTTTAAAAA            -          -211.0       489.0  \n",
       "3        AGCGCGAAATAATA            +          -285.0       415.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(sgRNACatalogue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
