{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "E133olOcNouR"
      },
      "outputs": [],
      "source": [
        "### This code will take in a TF list as an excel file and return a pd dataframe of the relevant features and gene region sequences for each start and stop codon of interest.\n",
        "# Contact scro4473@ox.ac.uk for questions and queries"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Import packages\n",
        "!pip install biopython\n",
        "!pip install primer3-py\n",
        "\n",
        "\n",
        "import pandas as pd\n",
        "from Bio import SeqIO\n",
        "import primer3 as p3\n",
        "from tqdm import tqdm"
      ],
      "metadata": {
        "id": "ksj8F76d2isQ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ec8e7263-fc90-488c-dcb5-2ff9e5d022ab"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: biopython in /usr/local/lib/python3.8/dist-packages (1.80)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.8/dist-packages (from biopython) (1.21.6)\n",
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: primer3-py in /usr/local/lib/python3.8/dist-packages (0.6.1)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "id": "SX1ay0pOmnDx",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "6f6c42cd-fa6e-4bcf-9191-7a9614b79afb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Functions\n",
        "\n",
        "#ReverseComplement\n",
        "def revComp(inputSeq):\n",
        "  \"\"\"\n",
        "  This function takes an input sequence and returns the reverse complement.\n",
        "\n",
        "  Input: inputSeq in str format\n",
        "  Output: revComp in str format\n",
        "\n",
        "  \"\"\"\n",
        "  complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
        "  \n",
        "  revComp = \"\"\n",
        "  for base in inputSeq[::-1]:\n",
        "    revComp += complement[(base.upper())]\n",
        "\n",
        "  return revComp"
      ],
      "metadata": {
        "id": "32LUgzhHR1qh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#This is the input file containing the TFs we want to query\n",
        "#Imported as a pandas dataframe\n",
        "queryTFsdf = pd.read_excel('/content/drive/My Drive/bioinformatics project/TFs.xlsx')\n",
        "#Ths contains 765 TFs total\n",
        "\n",
        "#This is the .gtf file with annotations for each gene on the reference genome\n",
        "#Note that I'll use these for the info categories of the final pandas df (rather than the transgenic genome annotations)\n",
        "refAnnotationsHeaders = [\"Chromosome\", \"Source\", \"Gene_Region\", \"Start\", \"Stop\", \"Score\", \"Strand\", \"Frame\", \"Attribute\"]\n",
        "refGenomeAnnotation = pd.read_csv(r'/content/drive/My Drive/bioinformatics project/dmel-all-r6.48.gtf', sep = \"\\t\", header = None, index_col = False, names = refAnnotationsHeaders)\n",
        "\n",
        "#This is the FASTA file of the reference genome sequence\n",
        "refSeqPerChromosome = {}\n",
        "for seq in SeqIO.parse(open('/content/drive/My Drive/bioinformatics project/dmel-all-chromosome-r6.48.fasta'), 'fasta'):\n",
        "  refSeqPerChromosome[seq.id] = seq.seq\n",
        "\n",
        "#This is the nos-Cas9 on 2 sequence - use for chromosome 3 only\n",
        "on2SeqPerChromosome = {}\n",
        "for seq in SeqIO.parse(open('/content/drive/My Drive/bioinformatics project/dmel6-nos-Cas9_on_2.fasta'), 'fasta'):\n",
        "  on2SeqPerChromosome[seq.id] = seq.seq\n",
        "\n",
        "#This is the nos-Cas9 on 3 sequence - use for all other chromosomes\n",
        "on3SeqPerChromosome = {}\n",
        "for seq in SeqIO.parse(open('/content/drive/My Drive/bioinformatics project/dmel6-nos-Cas9_on_3.fasta'), 'fasta'):\n",
        "  on3SeqPerChromosome[seq.id] = seq.seq"
      ],
      "metadata": {
        "id": "VkzTkR7yuYWJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#This is to reformat the \"Attribute\" category in refGenomeAnnotation, to extract Gene_ID, Gene_Symbol, and Transcript ID\n",
        "index = 0\n",
        "\n",
        "#Add new categories to the dataframe\n",
        "refGenomeAnnotation = refGenomeAnnotation.assign(Gene_ID = \"\", Gene_Symbol = \"\", Transcript_ID = \"\")\n",
        "\n",
        "#For each attribute value, extract the gene ID and symbol and add this to the new categories\n",
        "for attribute in refGenomeAnnotation['Attribute']:\n",
        "  fullatt = (refGenomeAnnotation.loc[index][\"Attribute\"]).replace(\";\", \"\")\n",
        "  fullatt = fullatt.replace('\"', \"\")\n",
        "  fullattsplit = fullatt.split(\" \")\n",
        "  refGenomeAnnotation.at[index,\"Gene_ID\"] = fullattsplit[1]\n",
        "  refGenomeAnnotation.at[index,\"Gene_Symbol\"] = fullattsplit[3]\n",
        "  if len(fullattsplit) == 8:\n",
        "    refGenomeAnnotation.at[index,\"Transcript_ID\"] = fullattsplit[5]\n",
        "  index+=1\n",
        "\n",
        "#Delete Attributes category\n",
        "del refGenomeAnnotation[\"Attribute\"]\n",
        "\n",
        "display(refGenomeAnnotation)"
      ],
      "metadata": {
        "id": "IOoB5z9UaE6K",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "outputId": "0a45428e-b3b1-4302-fcd9-5e523a7ed86e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "       Chromosome   Source Gene_Region     Start      Stop Score Strand Frame  \\\n",
              "0               X  FlyBase        gene  19961297  19969323     .      +     .   \n",
              "1               X  FlyBase        mRNA  19961689  19968479     .      +     .   \n",
              "2               X  FlyBase        5UTR  19961689  19961845     .      +     .   \n",
              "3               X  FlyBase        exon  19961689  19961845     .      +     .   \n",
              "4               X  FlyBase        exon  19963955  19964071     .      +     .   \n",
              "...           ...      ...         ...       ...       ...   ...    ...   ...   \n",
              "549156         2L  FlyBase         CDS    814243    815221     .      -     0   \n",
              "549157         2L  FlyBase         CDS    813951    814184     .      -     2   \n",
              "549158         2L  FlyBase         CDS    813460    813893     .      -     2   \n",
              "549159         2L  FlyBase  stop_codon    813457    813459     .      -     0   \n",
              "549160         2L  FlyBase        3UTR    813314    813456     .      -     .   \n",
              "\n",
              "            Gene_ID Gene_Symbol Transcript_ID  \n",
              "0       FBgn0031081        Nep3                \n",
              "1       FBgn0031081        Nep3   FBtr0070000  \n",
              "2       FBgn0031081        Nep3   FBtr0070000  \n",
              "3       FBgn0031081        Nep3   FBtr0070000  \n",
              "4       FBgn0031081        Nep3   FBtr0070000  \n",
              "...             ...         ...           ...  \n",
              "549156  FBgn0031281        Saf6   FBtr0332974  \n",
              "549157  FBgn0031281        Saf6   FBtr0332974  \n",
              "549158  FBgn0031281        Saf6   FBtr0332974  \n",
              "549159  FBgn0031281        Saf6   FBtr0332974  \n",
              "549160  FBgn0031281        Saf6   FBtr0332974  \n",
              "\n",
              "[549161 rows x 11 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-5e3071f1-4bb4-4ce5-8dd0-74baefb3829c\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Source</th>\n",
              "      <th>Gene_Region</th>\n",
              "      <th>Start</th>\n",
              "      <th>Stop</th>\n",
              "      <th>Score</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Frame</th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>Gene_Symbol</th>\n",
              "      <th>Transcript_ID</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>X</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>gene</td>\n",
              "      <td>19961297</td>\n",
              "      <td>19969323</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>FBgn0031081</td>\n",
              "      <td>Nep3</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>X</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>mRNA</td>\n",
              "      <td>19961689</td>\n",
              "      <td>19968479</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>FBgn0031081</td>\n",
              "      <td>Nep3</td>\n",
              "      <td>FBtr0070000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>X</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>5UTR</td>\n",
              "      <td>19961689</td>\n",
              "      <td>19961845</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>FBgn0031081</td>\n",
              "      <td>Nep3</td>\n",
              "      <td>FBtr0070000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>X</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>exon</td>\n",
              "      <td>19961689</td>\n",
              "      <td>19961845</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>FBgn0031081</td>\n",
              "      <td>Nep3</td>\n",
              "      <td>FBtr0070000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>X</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>exon</td>\n",
              "      <td>19963955</td>\n",
              "      <td>19964071</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>FBgn0031081</td>\n",
              "      <td>Nep3</td>\n",
              "      <td>FBtr0070000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>549156</th>\n",
              "      <td>2L</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>CDS</td>\n",
              "      <td>814243</td>\n",
              "      <td>815221</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>FBgn0031281</td>\n",
              "      <td>Saf6</td>\n",
              "      <td>FBtr0332974</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>549157</th>\n",
              "      <td>2L</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>CDS</td>\n",
              "      <td>813951</td>\n",
              "      <td>814184</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>2</td>\n",
              "      <td>FBgn0031281</td>\n",
              "      <td>Saf6</td>\n",
              "      <td>FBtr0332974</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>549158</th>\n",
              "      <td>2L</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>CDS</td>\n",
              "      <td>813460</td>\n",
              "      <td>813893</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>2</td>\n",
              "      <td>FBgn0031281</td>\n",
              "      <td>Saf6</td>\n",
              "      <td>FBtr0332974</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>549159</th>\n",
              "      <td>2L</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>813457</td>\n",
              "      <td>813459</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>FBgn0031281</td>\n",
              "      <td>Saf6</td>\n",
              "      <td>FBtr0332974</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>549160</th>\n",
              "      <td>2L</td>\n",
              "      <td>FlyBase</td>\n",
              "      <td>3UTR</td>\n",
              "      <td>813314</td>\n",
              "      <td>813456</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>.</td>\n",
              "      <td>FBgn0031281</td>\n",
              "      <td>Saf6</td>\n",
              "      <td>FBtr0332974</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>549161 rows × 11 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-5e3071f1-4bb4-4ce5-8dd0-74baefb3829c')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-5e3071f1-4bb4-4ce5-8dd0-74baefb3829c button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-5e3071f1-4bb4-4ce5-8dd0-74baefb3829c');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Select only rows that TFs are in, and keep only the start and stop codon gene regions\n",
        "\n",
        "refGenomeAnnotation = refGenomeAnnotation.loc[refGenomeAnnotation[\"Gene_Region\"].isin([\"start_codon\", \"stop_codon\"])]\n",
        "\n",
        "TFsdf = refGenomeAnnotation[[\"Gene_ID\", \"Transcript_ID\", \"Chromosome\", \"Gene_Region\", \"Start\", \"Stop\", \"Strand\"]].loc[refGenomeAnnotation[\"Gene_ID\"].isin(queryTFsdf[\"Flybase_ID\"])]\n",
        "\n",
        "display(TFsdf)"
      ],
      "metadata": {
        "id": "5KiXnJquKzwo",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "outputId": "34a56c79-f616-43aa-f40b-c2a01bf93b22"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "            Gene_ID Transcript_ID Chromosome  Gene_Region     Start      Stop  \\\n",
              "141     FBgn0031086   FBtr0070032          X  start_codon  20092383  20092385   \n",
              "143     FBgn0031086   FBtr0070032          X   stop_codon  20091603  20091605   \n",
              "453     FBgn0000022   FBtr0070072          X  start_codon    370094    370096   \n",
              "455     FBgn0000022   FBtr0070072          X   stop_codon    370697    370699   \n",
              "461     FBgn0004170   FBtr0070073          X  start_codon    396177    396179   \n",
              "...             ...           ...        ...          ...       ...       ...   \n",
              "546480  FBgn0031375   FBtr0330200         2L   stop_codon   1951056   1951058   \n",
              "546486  FBgn0031375   FBtr0330201         2L  start_codon   1954323   1954325   \n",
              "546489  FBgn0031375   FBtr0330201         2L   stop_codon   1950927   1950929   \n",
              "546495  FBgn0031375   FBtr0330202         2L  start_codon   1954323   1954325   \n",
              "546498  FBgn0031375   FBtr0330202         2L   stop_codon   1950795   1950797   \n",
              "\n",
              "       Strand  \n",
              "141         -  \n",
              "143         -  \n",
              "453         +  \n",
              "455         +  \n",
              "461         +  \n",
              "...       ...  \n",
              "546480      -  \n",
              "546486      -  \n",
              "546489      -  \n",
              "546495      -  \n",
              "546498      -  \n",
              "\n",
              "[4370 rows x 7 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3bc852e8-1fc5-490a-8ad9-49f0363164f2\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>Transcript_ID</th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Gene_Region</th>\n",
              "      <th>Start</th>\n",
              "      <th>Stop</th>\n",
              "      <th>Strand</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>141</th>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>X</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>20092383</td>\n",
              "      <td>20092385</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>143</th>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>X</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>20091603</td>\n",
              "      <td>20091605</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>453</th>\n",
              "      <td>FBgn0000022</td>\n",
              "      <td>FBtr0070072</td>\n",
              "      <td>X</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>370094</td>\n",
              "      <td>370096</td>\n",
              "      <td>+</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>455</th>\n",
              "      <td>FBgn0000022</td>\n",
              "      <td>FBtr0070072</td>\n",
              "      <td>X</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>370697</td>\n",
              "      <td>370699</td>\n",
              "      <td>+</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>461</th>\n",
              "      <td>FBgn0004170</td>\n",
              "      <td>FBtr0070073</td>\n",
              "      <td>X</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>396177</td>\n",
              "      <td>396179</td>\n",
              "      <td>+</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546480</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330200</td>\n",
              "      <td>2L</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>1951056</td>\n",
              "      <td>1951058</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546486</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330201</td>\n",
              "      <td>2L</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>1954323</td>\n",
              "      <td>1954325</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546489</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330201</td>\n",
              "      <td>2L</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>1950927</td>\n",
              "      <td>1950929</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546495</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>2L</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>1954323</td>\n",
              "      <td>1954325</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546498</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>2L</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>1950795</td>\n",
              "      <td>1950797</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4370 rows × 7 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3bc852e8-1fc5-490a-8ad9-49f0363164f2')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-3bc852e8-1fc5-490a-8ad9-49f0363164f2 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-3bc852e8-1fc5-490a-8ad9-49f0363164f2');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Add reference genome sequence per gene region\n",
        "#This will correspond to 1.3kb upstream and downstream of ATG/stop codon \n",
        "TFsdf = TFsdf.assign(Reference_Seq = \"\", Transgenic_Seq = \"\")\n",
        "\n",
        "#FASTAs for reference and transgenic are in format of a dictionary of chromosome:\"sequence\"\n",
        "\n",
        "for index, rowcontents in TFsdf.iterrows():\n",
        "  if rowcontents[\"Strand\"] == \"+\":\n",
        "\n",
        "    #Define 2.6kb gene region\n",
        "    regionStart = rowcontents[\"Start\"] - 1601\n",
        "    regionStop = rowcontents[\"Stop\"] + 1600\n",
        "\n",
        "    #Add reference sequence\n",
        "    TFsdf.at[index,\"Reference_Seq\"] = str(refSeqPerChromosome[rowcontents[\"Chromosome\"]][regionStart:regionStop])\n",
        "    \n",
        "    #Add appropriate transgenic sequence, depending on the chromosome\n",
        "    if rowcontents[\"Chromosome\"].startswith(\"3\"):\n",
        "      TFsdf.at[index,\"Transgenic_Seq\"] = str(on2SeqPerChromosome[rowcontents[\"Chromosome\"]][regionStart:regionStop])\n",
        "    else:\n",
        "      TFsdf.at[index,\"Transgenic_Seq\"] = str(on3SeqPerChromosome[rowcontents[\"Chromosome\"]][regionStart:regionStop])\n",
        "\n",
        "  if rowcontents[\"Strand\"] == \"-\":\n",
        "\n",
        "    #Define 2.6kb gene region\n",
        "    regionStart = rowcontents[\"Start\"] - 1601\n",
        "    regionStop = rowcontents[\"Stop\"] + 1600\n",
        "\n",
        "    #Add reference sequence\n",
        "    refPosStrandSeq = str(refSeqPerChromosome[rowcontents[\"Chromosome\"]][regionStart:regionStop]) #This is the + strand seq, so goes from end to beginning\n",
        "    TFsdf.at[index,\"Reference_Seq\"] = revComp(refPosStrandSeq)\n",
        "\n",
        "    #Add appropriate transgenic sequence, depending on the chromosome\n",
        "    if rowcontents[\"Chromosome\"].startswith(\"3\"):\n",
        "      transgStrandSeq = str(on2SeqPerChromosome[rowcontents[\"Chromosome\"]][regionStart:regionStop])\n",
        "      TFsdf.at[index,\"Transgenic_Seq\"] = revComp(transgStrandSeq)\n",
        "    else:\n",
        "      transgStrandSeq = str(on3SeqPerChromosome[rowcontents[\"Chromosome\"]][regionStart:regionStop])\n",
        "      TFsdf.at[index,\"Transgenic_Seq\"] = revComp(transgStrandSeq)\n",
        "\n",
        "display(TFsdf)\n",
        "\n",
        "TFsdf.to_excel(\"TFsdf.xlsx\")  "
      ],
      "metadata": {
        "id": "rw4qJX5wL7zQ",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 488
        },
        "outputId": "691e5ba9-b753-46f3-8139-67d4e0a678fd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "            Gene_ID Transcript_ID Chromosome  Gene_Region     Start      Stop  \\\n",
              "141     FBgn0031086   FBtr0070032          X  start_codon  20092383  20092385   \n",
              "143     FBgn0031086   FBtr0070032          X   stop_codon  20091603  20091605   \n",
              "453     FBgn0000022   FBtr0070072          X  start_codon    370094    370096   \n",
              "455     FBgn0000022   FBtr0070072          X   stop_codon    370697    370699   \n",
              "461     FBgn0004170   FBtr0070073          X  start_codon    396177    396179   \n",
              "...             ...           ...        ...          ...       ...       ...   \n",
              "546480  FBgn0031375   FBtr0330200         2L   stop_codon   1951056   1951058   \n",
              "546486  FBgn0031375   FBtr0330201         2L  start_codon   1954323   1954325   \n",
              "546489  FBgn0031375   FBtr0330201         2L   stop_codon   1950927   1950929   \n",
              "546495  FBgn0031375   FBtr0330202         2L  start_codon   1954323   1954325   \n",
              "546498  FBgn0031375   FBtr0330202         2L   stop_codon   1950795   1950797   \n",
              "\n",
              "       Strand                                      Reference_Seq  \\\n",
              "141         -  CTGTGTCTTCAAAATTCTGGCGGCTGCGAGACGCGACAAAGGAAGT...   \n",
              "143         -  ATAACGATACCCAGAGCATGTGTATAGTACTCGTAGTATATATCCA...   \n",
              "453         +  ATCGGCTAGAATACCTGTACGATCAACATTGGTACTGAAATGTACG...   \n",
              "455         +  ACAAATCTATTTCGGCACGGTTTCCTAGCGTGCAACCAGAATTCTG...   \n",
              "461         +  TATTCTAGGCATGTCACCGGCATTTCTCCCAACCATTTATTAAAAT...   \n",
              "...       ...                                                ...   \n",
              "546480      -  CGCTCCGGTTTCCTTCGCCACCAACGATGACGACGAGGATGAGGAT...   \n",
              "546486      -  ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...   \n",
              "546489      -  TGACTCCGCTTTCAAGAAGTATGTGCCGGGTCCGTGTTCGGGCGCC...   \n",
              "546495      -  ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...   \n",
              "546498      -  CCTGCTGTACTATGCACCCAATCAGCTGATGTGCGCCGCTGCTGCC...   \n",
              "\n",
              "                                           Transgenic_Seq  \n",
              "141     GCGAAGGACCCAAAGGAATTTTCGAGGGCCATTTACATCTAAGAGT...  \n",
              "143     CAATCCAGCCTAACTGAGTCCGTTGCATTAGCGGCAATCCCGACAG...  \n",
              "453     ACGATCAACATTGGTACTGAAATGTACGGAAGCTGTCCTGTAATCG...  \n",
              "455     TTTCCTAGCGTGCAACCAGAATTCTGACAATGGTTTTAAGAAATAA...  \n",
              "461     ATTTCTCCCAACCATTTATTAAAATGCTTCAATGGAATTTTGGAAA...  \n",
              "...                                                   ...  \n",
              "546480  TTACACCTCCAGTCGAGTGTAAGCGATTTAATGAATCTCACATAAA...  \n",
              "546486  TTTGTTTTATATTTTATTTTTGCACATTTTGCCGGATTTTCAAGTT...  \n",
              "546489  GGTCTTTGTTTGGGTTAAGTGGAGTGTAAGCCAGGGCCAAGCACGG...  \n",
              "546495  TTTGTTTTATATTTTATTTTTGCACATTTTGCCGGATTTTCAAGTT...  \n",
              "546498  CTGGCCCTGAATTATCCTGCGGAAATGTCCTGCCGCTGGGAGCACG...  \n",
              "\n",
              "[4370 rows x 9 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-fccc56c3-30ca-49b3-9fce-edbc351e6618\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>Transcript_ID</th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Gene_Region</th>\n",
              "      <th>Start</th>\n",
              "      <th>Stop</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Reference_Seq</th>\n",
              "      <th>Transgenic_Seq</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>141</th>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>X</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>20092383</td>\n",
              "      <td>20092385</td>\n",
              "      <td>-</td>\n",
              "      <td>CTGTGTCTTCAAAATTCTGGCGGCTGCGAGACGCGACAAAGGAAGT...</td>\n",
              "      <td>GCGAAGGACCCAAAGGAATTTTCGAGGGCCATTTACATCTAAGAGT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>143</th>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>X</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>20091603</td>\n",
              "      <td>20091605</td>\n",
              "      <td>-</td>\n",
              "      <td>ATAACGATACCCAGAGCATGTGTATAGTACTCGTAGTATATATCCA...</td>\n",
              "      <td>CAATCCAGCCTAACTGAGTCCGTTGCATTAGCGGCAATCCCGACAG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>453</th>\n",
              "      <td>FBgn0000022</td>\n",
              "      <td>FBtr0070072</td>\n",
              "      <td>X</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>370094</td>\n",
              "      <td>370096</td>\n",
              "      <td>+</td>\n",
              "      <td>ATCGGCTAGAATACCTGTACGATCAACATTGGTACTGAAATGTACG...</td>\n",
              "      <td>ACGATCAACATTGGTACTGAAATGTACGGAAGCTGTCCTGTAATCG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>455</th>\n",
              "      <td>FBgn0000022</td>\n",
              "      <td>FBtr0070072</td>\n",
              "      <td>X</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>370697</td>\n",
              "      <td>370699</td>\n",
              "      <td>+</td>\n",
              "      <td>ACAAATCTATTTCGGCACGGTTTCCTAGCGTGCAACCAGAATTCTG...</td>\n",
              "      <td>TTTCCTAGCGTGCAACCAGAATTCTGACAATGGTTTTAAGAAATAA...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>461</th>\n",
              "      <td>FBgn0004170</td>\n",
              "      <td>FBtr0070073</td>\n",
              "      <td>X</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>396177</td>\n",
              "      <td>396179</td>\n",
              "      <td>+</td>\n",
              "      <td>TATTCTAGGCATGTCACCGGCATTTCTCCCAACCATTTATTAAAAT...</td>\n",
              "      <td>ATTTCTCCCAACCATTTATTAAAATGCTTCAATGGAATTTTGGAAA...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546480</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330200</td>\n",
              "      <td>2L</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>1951056</td>\n",
              "      <td>1951058</td>\n",
              "      <td>-</td>\n",
              "      <td>CGCTCCGGTTTCCTTCGCCACCAACGATGACGACGAGGATGAGGAT...</td>\n",
              "      <td>TTACACCTCCAGTCGAGTGTAAGCGATTTAATGAATCTCACATAAA...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546486</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330201</td>\n",
              "      <td>2L</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>1954323</td>\n",
              "      <td>1954325</td>\n",
              "      <td>-</td>\n",
              "      <td>ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...</td>\n",
              "      <td>TTTGTTTTATATTTTATTTTTGCACATTTTGCCGGATTTTCAAGTT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546489</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330201</td>\n",
              "      <td>2L</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>1950927</td>\n",
              "      <td>1950929</td>\n",
              "      <td>-</td>\n",
              "      <td>TGACTCCGCTTTCAAGAAGTATGTGCCGGGTCCGTGTTCGGGCGCC...</td>\n",
              "      <td>GGTCTTTGTTTGGGTTAAGTGGAGTGTAAGCCAGGGCCAAGCACGG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546495</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>2L</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>1954323</td>\n",
              "      <td>1954325</td>\n",
              "      <td>-</td>\n",
              "      <td>ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...</td>\n",
              "      <td>TTTGTTTTATATTTTATTTTTGCACATTTTGCCGGATTTTCAAGTT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>546498</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>2L</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>1950795</td>\n",
              "      <td>1950797</td>\n",
              "      <td>-</td>\n",
              "      <td>CCTGCTGTACTATGCACCCAATCAGCTGATGTGCGCCGCTGCTGCC...</td>\n",
              "      <td>CTGGCCCTGAATTATCCTGCGGAAATGTCCTGCCGCTGGGAGCACG...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4370 rows × 9 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-fccc56c3-30ca-49b3-9fce-edbc351e6618')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-fccc56c3-30ca-49b3-9fce-edbc351e6618 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-fccc56c3-30ca-49b3-9fce-edbc351e6618');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def DesignPrimer(template,stringency,GC_content,max_end_GC,size,GC_clamp,TH_max_hairpin,max_polyx,\n",
        "                 primer_region, primer_type,primer_number):\n",
        " \n",
        "  '''\n",
        "  THIS FUNCTION CALCULATES A PRIMER GIVEN THE CONDITIONS AND HOW STRICT\n",
        "   THOSE CONDITIONS SHOULD BE\n",
        "\n",
        "  stringency is a list with n elements (i.e., n different stringency levels). Each condition is a \n",
        "  list (or a list of list). The list contains n elements (one for each striengency level). The \n",
        "  innermost list in a list of list contains [min,opt,max]\n",
        "  \n",
        "  '''\n",
        "  if primer_type == \"F\":\n",
        "    left = 1\n",
        "    right = 0\n",
        "  elif primer_type == \"R\":\n",
        "    left = 0\n",
        "    right = 1\n",
        "\n",
        "  primer = p3.bindings.designPrimers(\n",
        "    {\n",
        "        'SEQUENCE_TEMPLATE': template,\n",
        "        'SEQUENCE_INCLUDED_REGION': primer_region, \n",
        "    },\n",
        "    {\n",
        "        'PRIMER_NUM_RETURN':primer_number,\n",
        "\n",
        "        'PRIMER_TASK': \"generic\",\n",
        "        'PRIMER_PICK_LEFT_PRIMER': left,\n",
        "        'PRIMER_PICK_INTERNAL_OLIGO': 0,\n",
        "        'PRIMER_PICK_RIGHT_PRIMER': right,\n",
        "\n",
        "        'PRIMER_MIN_GC': GC_content[stringency][0],\n",
        "        'PRIMER_OPT_GC_PERCENT': GC_content[stringency][1],\n",
        "        'PRIMER_MAX_GC': GC_content[stringency][2],\n",
        "     \n",
        "        'PRIMER_MIN_SIZE': size[stringency][0],\n",
        "        'PRIMER_OPT_SIZE': size[stringency][1],\n",
        "        'PRIMER_MAX_SIZE': size[stringency][2],\n",
        "     \n",
        "        'PRIMER_MAX_END_GC': max_end_GC[stringency],\n",
        "     \n",
        "        'PRIMER_GC_CLAMP': GC_clamp[stringency],    \n",
        "             \n",
        "        'PRIMER_MAX_HAIRPIN_TH':TH_max_hairpin[stringency],\n",
        "    \n",
        "        'PRIMER_MAX_POLY_X': max_polyx[stringency],\n",
        "    })\n",
        "  \n",
        "  return primer\n",
        "\n",
        "\n",
        "def DoIHaveAPrimer(extension,Gene_ID,Transcript_ID,start_stop,primer_type, primer_name,stringency,primer):\n",
        "  \n",
        "  if extension == 0:\n",
        "    Extended_tag = \"\"\n",
        "  elif extension == 1:\n",
        "    Extended_tag = \"e\"\n",
        "\n",
        "\n",
        "  if primer_type == \"F\":\n",
        "    if primer['PRIMER_LEFT_NUM_RETURNED']>0:\n",
        "      primers_cleanup = {'Gene_ID': Gene_ID,\n",
        "                         'Transcript_ID': Transcript_ID,\n",
        "                         'position':start_stop,\n",
        "                         'primer_type': primer_name,\n",
        "                         'primer_sequence': primer['PRIMER_LEFT_0_SEQUENCE'],\n",
        "                         'stringency_level': f'{Extended_tag}{stringency+1}'}\n",
        "      warning_variable = False\n",
        "    else:\n",
        "      primers_cleanup = ()\n",
        "      warning_variable = True\n",
        "\n",
        "  elif primer_type == \"R\":\n",
        "    if primer['PRIMER_RIGHT_NUM_RETURNED']>0:\n",
        "      primers_cleanup = {'Gene_ID': Gene_ID,\n",
        "                         'Transcript_ID': Transcript_ID,\n",
        "                         'position':start_stop,\n",
        "                         'primer_type': primer_name,\n",
        "                         'primer_sequence': primer['PRIMER_RIGHT_0_SEQUENCE'],\n",
        "                         'stringency_level': f'{Extended_tag}{stringency+1}'}\n",
        "      warning_variable = False\n",
        "    else:\n",
        "      primers_cleanup = ()\n",
        "      warning_variable = True\n",
        "\n",
        "  return primers_cleanup, warning_variable"
      ],
      "metadata": {
        "id": "y4yhzWddpojb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def SixPrimersCalculator(Gene_ID, Transcript_ID,start_stop,template,primer_name,initial_primer_region,enlarged_primer_region,GC_content,size,max_end_GC,GC_clamp,TH_max_hairpin,max_polyx,stringency_levels):\n",
        "  primers_table = pd.DataFrame() # DataFrame used to save the 6 primers\n",
        "  primers_cleanup = ()\n",
        "\n",
        "  for pt in range(0,len(primer_name)): #loop to get as many primers as specified in the primer_name list (here 6)\n",
        "    \n",
        "    # defining what type of primer I am working with \n",
        "    if \"-F\" in primer_name[pt]:\n",
        "      primer_type = \"F\"\n",
        "    elif \"-R\" in primer_name[pt]:\n",
        "      primer_type = \"R\"\n",
        "\n",
        "    # defining how many primer outputs I want to have (one is NOT enough for the special cases)\n",
        "    if ((pt==2) or (pt==3)):  # output is only one primer if we do not care about the exact initial position (i.e., for 4/6 primers)\n",
        "      primer_number = 1000\n",
        "    else:\n",
        "      primer_number = 1\n",
        "\n",
        "    #print('----')\n",
        " \n",
        "   ###############################################################################\n",
        "    # loop to ACTUALLY take care of the primers (one at a time, considering different stringencies)\n",
        "    stringency = 0 #initial condition (most stringent conditions)\n",
        "\n",
        "    safety_net_primer_forward = ()  # empty variables to save HAL-R and HAR-F primers at a random position (BUT still in the right range)\n",
        "    safety_net_primer_reverse = ()  # this is useful in case we can't find any satisfactory primers in the right position\n",
        "    primer_region = initial_primer_region\n",
        "\n",
        "    while stringency<stringency_levels: \n",
        "      #print(primers_cleanup)\n",
        "       \n",
        "      ## taking care of the special cases of validation primers - their position depends on other primers (if those primers do not exist then validation primers are not generated)\n",
        "\n",
        "      # THIS PART TAKES CARE OF DEFINING THE UNKNOWN RANGES FOR VAL-F AND VAL-R\n",
        "      if primer_name[pt] == \"val-F\" and stringency == 0 and primer_region == initial_primer_region:\n",
        "        x = \"PLACEHOLDER\"\n",
        "        if primers_cleanup != ():\n",
        "          #print(primers_cleanup)\n",
        "          x = primers_cleanup[\"primer_sequence\"]\n",
        "          if str(x) in template: #was addressed differently before\n",
        "            primer_region[1][0]=(template.index(primers_cleanup['primer_sequence'])) - 99 - primer_region[1][1]#CHECK WITH EXPANSION  ----- in reality this might be 1bp off, might need further checks\n",
        "            #print(primer_region[1][0])\n",
        "            enlarged_primer_region[1][0] = primer_region[1][0]\n",
        "\n",
        "          elif x not in template:\n",
        "            primers_cleanup = {'Gene_ID': Gene_ID,\n",
        "                               'Transcript_ID': Transcript_ID,\n",
        "                                'position':start_stop,\n",
        "                                'primer_type': primer_name[pt],\n",
        "                                'primer_sequence': \"primer could not be calculated because reference primer (HAL-F) does not exist\",\n",
        "                                'stringency_level': \"NA\"}\n",
        "                \n",
        "            primers_cleanup_table = pd.DataFrame([primers_cleanup]) # coverting to dataframe\n",
        "            primers_table = pd.concat([primers_table,primers_cleanup_table]) # extending existing list with dataframe that was just created\n",
        "            stringency = stringency_levels # leaving the loop\n",
        "\n",
        "      if primer_name[pt] == \"val-R\"and stringency == 0 and primer_region == initial_primer_region:\n",
        "        tempo_reverse = \"PLACEHOLDER\"\n",
        "        if primers_cleanup != ():\n",
        "          tempo_reverse = revComp(primers_cleanup['primer_sequence']) #was addressed differently before\n",
        "          if  tempo_reverse in template:\n",
        "            primer_region[5][0]=template.index(tempo_reverse) + len(tempo_reverse) + 100  #  ----- in reality this might be 1bp off, might need further checks !!!!!!!\n",
        "            enlarged_primer_region[5][0] = primer_region[5][0]\n",
        "\n",
        "\n",
        "          elif x not in template:\n",
        "            primers_cleanup = {'Gene_ID': Gene_ID,\n",
        "                               'Transcript_ID': Transcript_ID,\n",
        "                                    'position':start_stop,\n",
        "                                    'primer_type': primer_name[pt],\n",
        "                                    'primer_sequence': \"primer could not be because reference primer (HAR-R) does not exist\",\n",
        "                                    'stringency_level': \"NA\"}\n",
        "                \n",
        "            primers_cleanup_table = pd.DataFrame([primers_cleanup]) # coverting to dataframe\n",
        "            primers_table = pd.concat([primers_table,primers_cleanup_table]) # extending existing list with dataframe that was just created\n",
        "            stringency = stringency_levels # leaving the loop\n",
        "\n",
        "\n",
        "      if stringency<stringency_levels:    #LAST BIT WAS TO MOVE THIS\n",
        "        primer = DesignPrimer(template,stringency,GC_content,max_end_GC,size,GC_clamp,TH_max_hairpin,max_polyx,    # designing the primer\n",
        "                          primer_region[pt], primer_type, primer_number)\n",
        "        #print(primer)\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "###### now we have to check that the primers are good enough ####\n",
        "\n",
        "\n",
        "      ####### special cases -  HAL-R and HAR -F #############\n",
        "\n",
        "      if primer_type == \"F\": # necessary only for special cases 2 and 3 - the # of calculated primers allows us to loop over them and choose the best\n",
        "        number_of_calculated_primers = primer['PRIMER_LEFT_NUM_RETURNED']  \n",
        "      elif primer_type == \"R\":\n",
        "        number_of_calculated_primers = primer['PRIMER_RIGHT_NUM_RETURNED']\n",
        "      \n",
        "      if (pt == 2 or pt == 3): # TO DO NEXT: when we add the extra level this will need to be skipped --> must add and stirngency<stringency level\n",
        "        primers_cleanup = () # empty variable to save the final primer of choice\n",
        "\n",
        "        for sc in range(0,number_of_calculated_primers): # evaluating all the primers obtained\n",
        "\n",
        "            if (pt==2): # HAL-R\n",
        "\n",
        "              # saving a random \"good\" primer as a \"safety net\" in case we can not find any other in the exact position\n",
        "              if safety_net_primer_reverse == ():   # saving of a \"net\" happens only if we find a primer + no primers can get overwritten\n",
        "                safety_net_primer_reverse = primer['PRIMER_RIGHT_0_SEQUENCE'] # saving the first one - for each primer the safety net is overwritten\n",
        "              \n",
        "              temporary_name_string = f'PRIMER_RIGHT_{sc}_SEQUENCE'  \n",
        "              reverse_complement = revComp(primer[temporary_name_string]) # checking position of the primer\n",
        "              primer_position = template.index(reverse_complement)\n",
        "              if (primer_position + len(reverse_complement)) == 1600: \n",
        "                primers_cleanup = {'Gene_ID': Gene_ID, # if the position is correct we choose it!! --> we can now leave the while loop!\n",
        "                                   'Transcript_ID': Transcript_ID,\n",
        "                                   'position':start_stop,\n",
        "                                   'primer_type': primer_name[pt],\n",
        "                                   'primer_sequence': primer[f'PRIMER_RIGHT_{sc}_SEQUENCE'],\n",
        "                                   'stringency_level': f'{stringency+1}'}\n",
        "                #print(primers_cleanup)\n",
        "\n",
        "            if (pt==3): #HAR-F\n",
        "\n",
        "              # saving a random \"good\" primer as a \"safety net\" in case we can not find any other in the exact position\n",
        "              if safety_net_primer_forward == ():\n",
        "                safety_net_primer_forward = primer['PRIMER_LEFT_0_SEQUENCE']\n",
        "            \n",
        "              temporary_name_string = f'PRIMER_LEFT_{sc}_SEQUENCE'\n",
        "              primer_position = template.index(primer[temporary_name_string])\n",
        "              if primer_position == 1603: # checking primer position\n",
        "                primers_cleanup = {'Gene_ID': Gene_ID,  # if the position is correct we choose it!! --> we can now leave the while loop!\n",
        "                                   'Transcript_ID': Transcript_ID,\n",
        "                                   'position':start_stop,\n",
        "                                   'primer_type': primer_name[pt],\n",
        "                                   'primer_sequence': primer[f'PRIMER_LEFT_{sc}_SEQUENCE'],\n",
        "                                   'stringency_level': f'{stringency+1}'}\n",
        "                #print(primers_cleanup)\n",
        "       \n",
        "        # TO DO: careful with stringency levels after extenstion\n",
        "        if stringency == (stringency_levels-1) and primers_cleanup == (): # i.e. if we are at the last stringency and we still haven't left the loop....\n",
        "          if (pt==2):\n",
        "            if safety_net_primer_reverse == ():\n",
        "              last_resort = template[1575:1600]\n",
        "              primer_last_resort = revComp(last_resort)\n",
        "              primers_cleanup = {'Gene_ID': Gene_ID,\n",
        "                                 'Transcript_ID': Transcript_ID,\n",
        "                                  'position':start_stop,\n",
        "                                  'primer_type': primer_name[pt],\n",
        "                                  'primer_sequence': primer_last_resort,\n",
        "                                  'stringency_level': \"25bp no conditions\"}\n",
        "              #print(primers_cleanup)\n",
        "            else:\n",
        "              safety_net_primer_reverse_t = revComp(safety_net_primer_reverse)\n",
        "              net_position = template.index(safety_net_primer_reverse_t)\n",
        "              extened_safety_net_rev_comp = template[net_position:1600]\n",
        "              extended_safety_net_primer_reverse = revComp(extened_safety_net_rev_comp)\n",
        "              primers_cleanup = {'Gene_ID': Gene_ID,\n",
        "                                 'Transcript_ID': Transcript_ID,\n",
        "                                  'position':start_stop,\n",
        "                                  'primer_type': primer_name[pt],\n",
        "                                  'primer_sequence': extended_safety_net_primer_reverse,\n",
        "                                  'stringency_level': f\"extended from {stringency+1}\"}\n",
        "              #print(primers_cleanup)\n",
        "          if (pt==3):\n",
        "            if safety_net_primer_forward == ():\n",
        "              primers_cleanup = {'Gene_ID': Gene_ID,\n",
        "                                 'Transcript_ID': Transcript_ID,\n",
        "                                  'position':start_stop,\n",
        "                                  'primer_type': primer_name[pt],\n",
        "                                  'primer_sequence': template[1603:1628],\n",
        "                                  'stringency_level': \"25bp no conditions\"}\n",
        "              #print(primers_cleanup)\n",
        "            else:\n",
        "              net_position = template.index(safety_net_primer_forward)\n",
        "              extension = template [1603:net_position]\n",
        "              extended_safety_net_primer_forward = extension+safety_net_primer_forward\n",
        "              primers_cleanup = {'Gene_ID': Gene_ID,\n",
        "                                 'Transcript_ID': Transcript_ID,\n",
        "                                  'position':start_stop,\n",
        "                                  'primer_type': primer_name[pt],\n",
        "                                  'primer_sequence': extended_safety_net_primer_forward,\n",
        "                                  'stringency_level': f\"extended from {stringency+1}\"}\n",
        "              #print(primers_cleanup)\n",
        "\n",
        "        if primers_cleanup == ():\n",
        "          stringency +=1\n",
        "        else:\n",
        "          primers_cleanup_table = pd.DataFrame([primers_cleanup]) # coverting to dataframe\n",
        "          primers_table = pd.concat([primers_table,primers_cleanup_table]) # extending existing list with dataframe that was just created\n",
        "          stringency = stringency_levels # TO DO: SEE HOW TO MODIFY THIS NEXT\n",
        "\n",
        "\n",
        "      ########### normal cases #########\n",
        "      if (pt == 0 or pt == 1 or pt == 4 or pt == 5):\n",
        "        if primer_region == initial_primer_region:\n",
        "          extension = 0\n",
        "        else:\n",
        "          extension = 1\n",
        "\n",
        "        primers_cleanup, warning_variable = DoIHaveAPrimer(extension,Gene_ID,Transcript_ID,start_stop,primer_type, primer_name[pt],stringency,primer)      # function 2 created by me \n",
        "        \n",
        "        if warning_variable == False: # i.e., if I do not get a warning then we are done --> we leave the while loop with \n",
        "          primers_cleanup_table = pd.DataFrame([primers_cleanup]) # coverting to dataframe\n",
        "          primers_table = pd.concat([primers_table,primers_cleanup_table]) # extending existing list with dataframe that was just created\n",
        "          stringency = stringency_levels # leaving the loop\n",
        "        else:\n",
        "          if stringency == (stringency_levels-1):\n",
        "            if primer_region == enlarged_primer_region:\n",
        "              primers_cleanup = {'Gene_ID': Gene_ID,\n",
        "                                 'Transcript_ID': Transcript_ID,\n",
        "                                  'position':start_stop,\n",
        "                                  'primer_type': primer_name[pt],\n",
        "                                  'primer_sequence': \"primer could not be calculated\",\n",
        "                                  'stringency_level': \"NA\"}\n",
        "              #print(primers_cleanup)\n",
        "              \n",
        "              primers_cleanup_table = pd.DataFrame([primers_cleanup]) # coverting to dataframe\n",
        "              primers_table = pd.concat([primers_table,primers_cleanup_table]) # extending existing list with dataframe that was just created\n",
        "              stringency = stringency_levels # leaving the loop\n",
        "\n",
        "            elif primer_region == initial_primer_region:\n",
        "              stringency = 0  \n",
        "              primer_region = enlarged_primer_region\n",
        "          else:\n",
        "              stringency +=1\n",
        "      #print(stringency)\n",
        "\n",
        "  return(primers_table)    \n"
      ],
      "metadata": {
        "id": "G8fnNscLpxsT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "\n",
        "# CONDITIONS VARIABLES - [stringent[min,opt,max], relaxed, desperate]\n",
        "primer_name = [\"HAL-F\",\"val-F\",\"HAL-R\",\"HAR-F\",\"HAR-R\",\"val-R\"]\n",
        "\n",
        "initial_primer_region = [[500,200],[\"unknown\",100],[1570,29],[1603,30],[2500,200],[\"unknown\",100]]\n",
        "extended_primer_region = [[400,400],[\"unknown\",200],[1570,29],[1603,30],[2100,400],[\"unknown\",200]]\n",
        "#extended_primer_region = [[0,10],[200,10],[1270,29],[1303,30],[2200,10],[2500,10]] ### TO BE MODIFIED!!!!\n",
        "GC_content = [[30,50,70],[25,50,75],[20,50,80]]\n",
        "size = [[18,20,25],[18,20,26],[17,20,30]]\n",
        "max_end_GC = [3,4,4]\n",
        "GC_clamp = [1,1,0]\n",
        "TH_max_hairpin = [47.00,48.00,72.00]\n",
        "max_polyx = [5,6,8] \n",
        "stringency_levels = 3\n",
        "\n",
        "\n",
        "\n",
        "primers_table = ()\n",
        "primers_table = pd.DataFrame()\n",
        "#display(TFsdf)\n",
        "TFsdf_i = TFsdf.reset_index()\n",
        "display(TFsdf_i)\n",
        "\n",
        "\n",
        "for x in tqdm(range(0,len(TFsdf[\"Gene_ID\"]))):   #len(TFsdf[\"Gene_ID\"])\n",
        "  single = SixPrimersCalculator(TFsdf_i[\"Gene_ID\"][x],TFsdf_i[\"Transcript_ID\"][x],TFsdf_i[\"Gene_Region\"][x],TFsdf_i[\"Reference_Seq\"][x],primer_name,initial_primer_region,extended_primer_region,GC_content,size,max_end_GC,GC_clamp,TH_max_hairpin,max_polyx,stringency_levels)\n",
        "  primers_table = pd.concat([primers_table,single])\n",
        "\n",
        "display(primers_table)\n",
        "primers_table.to_excel(\"primers_output.xlsx\")  "
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 912
        },
        "id": "94vWJ5VErNIX",
        "outputId": "ba158c17-7398-497b-92f7-c980e67fc877"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "       index      Gene_ID Transcript_ID Chromosome  Gene_Region     Start  \\\n",
              "0        141  FBgn0031086   FBtr0070032          X  start_codon  20092383   \n",
              "1        143  FBgn0031086   FBtr0070032          X   stop_codon  20091603   \n",
              "2        453  FBgn0000022   FBtr0070072          X  start_codon    370094   \n",
              "3        455  FBgn0000022   FBtr0070072          X   stop_codon    370697   \n",
              "4        461  FBgn0004170   FBtr0070073          X  start_codon    396177   \n",
              "...      ...          ...           ...        ...          ...       ...   \n",
              "4365  546480  FBgn0031375   FBtr0330200         2L   stop_codon   1951056   \n",
              "4366  546486  FBgn0031375   FBtr0330201         2L  start_codon   1954323   \n",
              "4367  546489  FBgn0031375   FBtr0330201         2L   stop_codon   1950927   \n",
              "4368  546495  FBgn0031375   FBtr0330202         2L  start_codon   1954323   \n",
              "4369  546498  FBgn0031375   FBtr0330202         2L   stop_codon   1950795   \n",
              "\n",
              "          Stop Strand                                      Reference_Seq  \\\n",
              "0     20092385      -  CTGTGTCTTCAAAATTCTGGCGGCTGCGAGACGCGACAAAGGAAGT...   \n",
              "1     20091605      -  ATAACGATACCCAGAGCATGTGTATAGTACTCGTAGTATATATCCA...   \n",
              "2       370096      +  ATCGGCTAGAATACCTGTACGATCAACATTGGTACTGAAATGTACG...   \n",
              "3       370699      +  ACAAATCTATTTCGGCACGGTTTCCTAGCGTGCAACCAGAATTCTG...   \n",
              "4       396179      +  TATTCTAGGCATGTCACCGGCATTTCTCCCAACCATTTATTAAAAT...   \n",
              "...        ...    ...                                                ...   \n",
              "4365   1951058      -  CGCTCCGGTTTCCTTCGCCACCAACGATGACGACGAGGATGAGGAT...   \n",
              "4366   1954325      -  ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...   \n",
              "4367   1950929      -  TGACTCCGCTTTCAAGAAGTATGTGCCGGGTCCGTGTTCGGGCGCC...   \n",
              "4368   1954325      -  ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...   \n",
              "4369   1950797      -  CCTGCTGTACTATGCACCCAATCAGCTGATGTGCGCCGCTGCTGCC...   \n",
              "\n",
              "                                         Transgenic_Seq  \n",
              "0     GCGAAGGACCCAAAGGAATTTTCGAGGGCCATTTACATCTAAGAGT...  \n",
              "1     CAATCCAGCCTAACTGAGTCCGTTGCATTAGCGGCAATCCCGACAG...  \n",
              "2     ACGATCAACATTGGTACTGAAATGTACGGAAGCTGTCCTGTAATCG...  \n",
              "3     TTTCCTAGCGTGCAACCAGAATTCTGACAATGGTTTTAAGAAATAA...  \n",
              "4     ATTTCTCCCAACCATTTATTAAAATGCTTCAATGGAATTTTGGAAA...  \n",
              "...                                                 ...  \n",
              "4365  TTACACCTCCAGTCGAGTGTAAGCGATTTAATGAATCTCACATAAA...  \n",
              "4366  TTTGTTTTATATTTTATTTTTGCACATTTTGCCGGATTTTCAAGTT...  \n",
              "4367  GGTCTTTGTTTGGGTTAAGTGGAGTGTAAGCCAGGGCCAAGCACGG...  \n",
              "4368  TTTGTTTTATATTTTATTTTTGCACATTTTGCCGGATTTTCAAGTT...  \n",
              "4369  CTGGCCCTGAATTATCCTGCGGAAATGTCCTGCCGCTGGGAGCACG...  \n",
              "\n",
              "[4370 rows x 10 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-0a5107e2-06e7-467d-a1ac-be11d6ec63be\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>index</th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>Transcript_ID</th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Gene_Region</th>\n",
              "      <th>Start</th>\n",
              "      <th>Stop</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Reference_Seq</th>\n",
              "      <th>Transgenic_Seq</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>141</td>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>X</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>20092383</td>\n",
              "      <td>20092385</td>\n",
              "      <td>-</td>\n",
              "      <td>CTGTGTCTTCAAAATTCTGGCGGCTGCGAGACGCGACAAAGGAAGT...</td>\n",
              "      <td>GCGAAGGACCCAAAGGAATTTTCGAGGGCCATTTACATCTAAGAGT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>143</td>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>X</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>20091603</td>\n",
              "      <td>20091605</td>\n",
              "      <td>-</td>\n",
              "      <td>ATAACGATACCCAGAGCATGTGTATAGTACTCGTAGTATATATCCA...</td>\n",
              "      <td>CAATCCAGCCTAACTGAGTCCGTTGCATTAGCGGCAATCCCGACAG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>453</td>\n",
              "      <td>FBgn0000022</td>\n",
              "      <td>FBtr0070072</td>\n",
              "      <td>X</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>370094</td>\n",
              "      <td>370096</td>\n",
              "      <td>+</td>\n",
              "      <td>ATCGGCTAGAATACCTGTACGATCAACATTGGTACTGAAATGTACG...</td>\n",
              "      <td>ACGATCAACATTGGTACTGAAATGTACGGAAGCTGTCCTGTAATCG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>455</td>\n",
              "      <td>FBgn0000022</td>\n",
              "      <td>FBtr0070072</td>\n",
              "      <td>X</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>370697</td>\n",
              "      <td>370699</td>\n",
              "      <td>+</td>\n",
              "      <td>ACAAATCTATTTCGGCACGGTTTCCTAGCGTGCAACCAGAATTCTG...</td>\n",
              "      <td>TTTCCTAGCGTGCAACCAGAATTCTGACAATGGTTTTAAGAAATAA...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>461</td>\n",
              "      <td>FBgn0004170</td>\n",
              "      <td>FBtr0070073</td>\n",
              "      <td>X</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>396177</td>\n",
              "      <td>396179</td>\n",
              "      <td>+</td>\n",
              "      <td>TATTCTAGGCATGTCACCGGCATTTCTCCCAACCATTTATTAAAAT...</td>\n",
              "      <td>ATTTCTCCCAACCATTTATTAAAATGCTTCAATGGAATTTTGGAAA...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4365</th>\n",
              "      <td>546480</td>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330200</td>\n",
              "      <td>2L</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>1951056</td>\n",
              "      <td>1951058</td>\n",
              "      <td>-</td>\n",
              "      <td>CGCTCCGGTTTCCTTCGCCACCAACGATGACGACGAGGATGAGGAT...</td>\n",
              "      <td>TTACACCTCCAGTCGAGTGTAAGCGATTTAATGAATCTCACATAAA...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4366</th>\n",
              "      <td>546486</td>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330201</td>\n",
              "      <td>2L</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>1954323</td>\n",
              "      <td>1954325</td>\n",
              "      <td>-</td>\n",
              "      <td>ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...</td>\n",
              "      <td>TTTGTTTTATATTTTATTTTTGCACATTTTGCCGGATTTTCAAGTT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4367</th>\n",
              "      <td>546489</td>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330201</td>\n",
              "      <td>2L</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>1950927</td>\n",
              "      <td>1950929</td>\n",
              "      <td>-</td>\n",
              "      <td>TGACTCCGCTTTCAAGAAGTATGTGCCGGGTCCGTGTTCGGGCGCC...</td>\n",
              "      <td>GGTCTTTGTTTGGGTTAAGTGGAGTGTAAGCCAGGGCCAAGCACGG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4368</th>\n",
              "      <td>546495</td>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>2L</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>1954323</td>\n",
              "      <td>1954325</td>\n",
              "      <td>-</td>\n",
              "      <td>ACAAGAGCCGCACAACAAGCCAGATTAACGGAAAATGCAGTCGAAA...</td>\n",
              "      <td>TTTGTTTTATATTTTATTTTTGCACATTTTGCCGGATTTTCAAGTT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4369</th>\n",
              "      <td>546498</td>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>2L</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>1950795</td>\n",
              "      <td>1950797</td>\n",
              "      <td>-</td>\n",
              "      <td>CCTGCTGTACTATGCACCCAATCAGCTGATGTGCGCCGCTGCTGCC...</td>\n",
              "      <td>CTGGCCCTGAATTATCCTGCGGAAATGTCCTGCCGCTGGGAGCACG...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>4370 rows × 10 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0a5107e2-06e7-467d-a1ac-be11d6ec63be')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-0a5107e2-06e7-467d-a1ac-be11d6ec63be button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-0a5107e2-06e7-467d-a1ac-be11d6ec63be');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 4370/4370 [20:10<00:00,  3.61it/s]\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "        Gene_ID Transcript_ID     position primer_type  \\\n",
              "0   FBgn0031086   FBtr0070032  start_codon       HAL-F   \n",
              "0   FBgn0031086   FBtr0070032  start_codon       val-F   \n",
              "0   FBgn0031086   FBtr0070032  start_codon       HAL-R   \n",
              "0   FBgn0031086   FBtr0070032  start_codon       HAR-F   \n",
              "0   FBgn0031086   FBtr0070032  start_codon       HAR-R   \n",
              "..          ...           ...          ...         ...   \n",
              "0   FBgn0031375   FBtr0330202   stop_codon       val-F   \n",
              "0   FBgn0031375   FBtr0330202   stop_codon       HAL-R   \n",
              "0   FBgn0031375   FBtr0330202   stop_codon       HAR-F   \n",
              "0   FBgn0031375   FBtr0330202   stop_codon       HAR-R   \n",
              "0   FBgn0031375   FBtr0330202   stop_codon       val-R   \n",
              "\n",
              "                   primer_sequence stringency_level  \n",
              "0         TCTTGGTTTTTGAATTTTGGCATG                1  \n",
              "0             TCGGGTAACTCGACCATAGC                1  \n",
              "0   CGTAATAATATTGCAATATGTGTCACTGTT  extended from 3  \n",
              "0         GATACGACACCAATCTTCCAGTCC                1  \n",
              "0           TCGCACTATTCTTCGAATGTCG                1  \n",
              "..                             ...              ...  \n",
              "0             TCAACGCCCACTACAACCTC                1  \n",
              "0       GAAGAGACCGTTCAGTGCGCGGACAG  extended from 3  \n",
              "0               GCAGAAGGACCCCTGTCC                3  \n",
              "0             TGCTCCAGACCCATTTCCAC                1  \n",
              "0   TGTCAAGTAAAGTTATTCATATTGAGCACT                3  \n",
              "\n",
              "[26220 rows x 6 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6fb23ec6-afb3-4e9f-addc-fe190df19f37\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>Transcript_ID</th>\n",
              "      <th>position</th>\n",
              "      <th>primer_type</th>\n",
              "      <th>primer_sequence</th>\n",
              "      <th>stringency_level</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>HAL-F</td>\n",
              "      <td>TCTTGGTTTTTGAATTTTGGCATG</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>val-F</td>\n",
              "      <td>TCGGGTAACTCGACCATAGC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>HAL-R</td>\n",
              "      <td>CGTAATAATATTGCAATATGTGTCACTGTT</td>\n",
              "      <td>extended from 3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>HAR-F</td>\n",
              "      <td>GATACGACACCAATCTTCCAGTCC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031086</td>\n",
              "      <td>FBtr0070032</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>HAR-R</td>\n",
              "      <td>TCGCACTATTCTTCGAATGTCG</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>val-F</td>\n",
              "      <td>TCAACGCCCACTACAACCTC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>HAL-R</td>\n",
              "      <td>GAAGAGACCGTTCAGTGCGCGGACAG</td>\n",
              "      <td>extended from 3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>HAR-F</td>\n",
              "      <td>GCAGAAGGACCCCTGTCC</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>HAR-R</td>\n",
              "      <td>TGCTCCAGACCCATTTCCAC</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>FBgn0031375</td>\n",
              "      <td>FBtr0330202</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>val-R</td>\n",
              "      <td>TGTCAAGTAAAGTTATTCATATTGAGCACT</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>26220 rows × 6 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6fb23ec6-afb3-4e9f-addc-fe190df19f37')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-6fb23ec6-afb3-4e9f-addc-fe190df19f37 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-6fb23ec6-afb3-4e9f-addc-fe190df19f37');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {}
        }
      ]
    }
  ]
}
